// @optique/core/dist/internal/mode-dispatch.js
//#region src/internal/mode-dispatch.ts
/**
* Dispatches to sync or async implementation based on mode.
*
* This function encapsulates the necessary type assertions when branching
* on runtime mode values. TypeScript cannot narrow `ModeValue<M, T>` based
* on `mode === "async"` checks, so we must use type assertions.
*
* @param mode The execution mode.
* @param syncFn Function to call for sync execution.
* @param asyncFn Function to call for async execution.
* @returns The result with correct mode wrapping.
* @since 0.10.0
*/
function dispatchByMode(mode, syncFn, asyncFn) {
	if (mode === "async") return asyncFn();
	return syncFn();
}
function wrapForMode(mode, value) {
	if (mode === "async") return Promise.resolve(value);
	if (value instanceof Promise) throw new TypeError("Synchronous mode cannot wrap Promise value.");
	return value;
}
/**
* Maps a mode-wrapped value while preserving its execution mode.
*
* @param mode The execution mode.
* @param value The mode-wrapped value to transform.
* @param mapFn Mapping function applied to the unwrapped value.
* @returns The mapped value with correct mode wrapping.
* @since 1.0.0
*/
function mapModeValue(mode, value, mapFn) {
	if (mode === "async") return Promise.resolve(value).then(mapFn);
	if (value instanceof Promise) throw new TypeError("Synchronous mode cannot map Promise value.");
	return mapFn(value);
}
/**
* Dispatches iterable to sync or async implementation based on mode.
*
* @param mode The execution mode.
* @param syncFn Function returning sync iterable.
* @param asyncFn Function returning async iterable.
* @returns The iterable with correct mode wrapping.
* @internal
* @since 0.10.0
*/
function dispatchIterableByMode(mode, syncFn, asyncFn) {
	if (mode === "async") return asyncFn();
	return syncFn();
}

//#endregion
export { dispatchByMode, dispatchIterableByMode, mapModeValue, wrapForMode };
