// @optique/run/dist/print.cjs
const require_rolldown_runtime = require('./_virtual/rolldown_runtime.cjs');
const node_process = require_rolldown_runtime.__toESM(require("node:process"));
const __optique_core_message = require_rolldown_runtime.__toESM(require("@optique/core/message"));

//#region src/print.ts
/**
* Prints a formatted message to stdout with automatic terminal detection.
*
* This function automatically detects terminal capabilities (colors, width)
* and formats the message accordingly. It's ideal for general application
* output that should be visible to users.
*
* @param message The structured message to print.
* @param options Optional formatting options to override defaults.
*
* @example
* ```typescript
* import { print } from "@optique/run";
* import { message, optionName } from "@optique/core/message";
*
* const configFile = "config.json";
* const port = 3000;
*
* print(message`Configuration loaded from ${configFile}`);
* print(message`Using ${optionName("--port")} ${port}`);
* ```
*
* @since 0.3.0
*/
function print(message, options = {}) {
	const printer = createPrinter({
		stream: options.stream ?? "stdout",
		colors: options.colors,
		quotes: options.quotes,
		maxWidth: options.maxWidth
	});
	printer(message);
}
function printError(message, options = {}) {
	const stream = options.stream ?? "stderr";
	const output = node_process.default[stream];
	const quotes = options.quotes ?? !output.isTTY;
	const printer = createPrinter({
		stream,
		colors: options.colors,
		quotes,
		maxWidth: options.maxWidth
	});
	const errorMessage = [{
		type: "text",
		text: "Error: "
	}, ...message];
	printer(errorMessage);
	if (options.exitCode != null) node_process.default.exit(options.exitCode);
}
/**
* Creates a custom printer function with predefined formatting options.
*
* This is useful when you need consistent formatting across multiple print
* operations or when you want to override the automatic terminal detection.
*
* @param options Formatting options for the printer.
* @returns A printer function that can be called with messages.
*
* @example
* ```typescript
* import { createPrinter } from "@optique/run";
* import { message, metavar } from "@optique/core/message";
*
* // Create a printer with forced colors and no quotes
* const printer = createPrinter({
*   colors: true,
*   quotes: false,
*   stream: "stdout",
* });
*
* printer(message`Starting server on ${metavar("PORT")}...`);
* printer(message`Ready to accept connections`);
* ```
*
* @since 0.3.0
*/
function createPrinter(options = {}) {
	const stream = options.stream ?? "stdout";
	const output = node_process.default[stream];
	const formatOptions = {
		colors: options.colors ?? output.isTTY,
		quotes: options.quotes,
		maxWidth: options.maxWidth ?? output.columns
	};
	return (message) => {
		const formatted = (0, __optique_core_message.formatMessage)(message, formatOptions);
		if (stream === "stderr") node_process.default.stderr.write(formatted + "\n");
		else node_process.default.stdout.write(formatted + "\n");
	};
}

//#endregion
exports.createPrinter = createPrinter;
exports.print = print;
exports.printError = printError;
