// @optique/run/dist/run.cjs
const require_rolldown_runtime = require('./_virtual/rolldown_runtime.cjs');
const __optique_core_facade = require_rolldown_runtime.__toESM(require("@optique/core/facade"));
const node_path = require_rolldown_runtime.__toESM(require("node:path"));
const node_process = require_rolldown_runtime.__toESM(require("node:process"));

//#region src/run.ts
function getProgramHelpMetadata(metadata) {
	return {
		brief: metadata.brief,
		description: metadata.description,
		examples: metadata.examples,
		author: metadata.author,
		bugs: metadata.bugs,
		footer: metadata.footer
	};
}
function resolveProgramInput(parserOrProgram, options) {
	if ("parser" in parserOrProgram && "metadata" in parserOrProgram) return {
		parser: parserOrProgram.parser,
		options: options.programName == null ? {
			...options,
			programName: parserOrProgram.metadata.name
		} : options,
		programMetadata: getProgramHelpMetadata(parserOrProgram.metadata)
	};
	return {
		parser: parserOrProgram,
		options
	};
}
function run(parserOrProgram, options = {}) {
	return runImpl(parserOrProgram, options);
}
function runSync(parserOrProgram, options = {}) {
	const parserToCheck = "parser" in parserOrProgram && "metadata" in parserOrProgram ? parserOrProgram.parser : parserOrProgram;
	if (parserToCheck.mode !== "sync") throw new TypeError("Cannot use an async parser with runSync(). Use run() or runAsync() instead.");
	const contexts = options.contexts;
	if (contexts && contexts.length > 0) {
		const resolved = resolveProgramInput(parserOrProgram, options);
		const { parser, programMetadata } = resolved;
		options = resolved.options;
		const { programName, args, coreOptions } = buildCoreOptions(options, programMetadata);
		const runWithOptions = {
			...coreOptions,
			contextOptions: options.contextOptions,
			args
		};
		return (0, __optique_core_facade.runWithSync)(parser, programName, contexts, runWithOptions);
	}
	return runImpl(parserOrProgram, options);
}
function runAsync(parserOrProgram, options = {}) {
	const result = runImpl(parserOrProgram, options);
	return Promise.resolve(result);
}
/**
* Builds core run options from the simplified RunOptions format.
*
* Converts the user-friendly RunOptions (string-based help/version/completion)
* into the verbose CoreRunOptions format expected by `runParser()`, `runWith()`,
* and `runWithSync()` from `@optique/core/facade`.
*
* @internal
*/
function buildCoreOptions(options, programMetadata) {
	const programName = options.programName ?? node_path.default.basename(node_process.default.argv[1] ?? "cli");
	const args = options.args ?? node_process.default.argv.slice(2);
	const stdout = options.stdout ?? ((line) => {
		node_process.default.stdout.write(`${line}\n`);
	});
	const stderr = options.stderr ?? ((line) => {
		node_process.default.stderr.write(`${line}\n`);
	});
	const onExit = options.onExit ?? ((exitCode) => node_process.default.exit(exitCode));
	const colors = options.colors ?? node_process.default.stdout.isTTY;
	const maxWidth = options.maxWidth ?? node_process.default.stdout.columns;
	const showDefault = options.showDefault;
	const showChoices = options.showChoices;
	const sectionOrder = options.sectionOrder;
	const help = options.help;
	const version = options.version;
	const completion = options.completion;
	const aboveError = options.aboveError ?? "usage";
	const errorExitCode = options.errorExitCode ?? 1;
	const brief = options.brief ?? programMetadata?.brief;
	const description = options.description ?? programMetadata?.description;
	const examples = options.examples ?? programMetadata?.examples;
	const author = options.author ?? programMetadata?.author;
	const bugs = options.bugs ?? programMetadata?.bugs;
	const footer = options.footer ?? programMetadata?.footer;
	const onShow = () => onExit(0);
	const helpConfig = (() => {
		if (!help) return void 0;
		if (typeof help === "string") switch (help) {
			case "command": return {
				command: true,
				onShow
			};
			case "option": return {
				option: true,
				onShow
			};
			case "both": return {
				command: true,
				option: true,
				onShow
			};
		}
		return {
			...help,
			onShow
		};
	})();
	const versionConfig = (() => {
		if (!version) return void 0;
		if (typeof version === "string") return {
			value: version,
			option: true,
			onShow
		};
		return {
			...version,
			onShow
		};
	})();
	const completionConfig = (() => {
		if (!completion) return void 0;
		if (typeof completion === "string") switch (completion) {
			case "command": return {
				command: true,
				onShow
			};
			case "option": return {
				option: true,
				onShow
			};
			case "both": return {
				command: true,
				option: true,
				onShow
			};
		}
		return {
			...completion,
			onShow
		};
	})();
	const coreOptions = {
		stderr,
		stdout,
		colors,
		maxWidth,
		showDefault,
		showChoices,
		sectionOrder,
		help: helpConfig,
		version: versionConfig,
		completion: completionConfig,
		aboveError,
		brief,
		description,
		examples,
		author,
		bugs,
		footer,
		onError() {
			return onExit(errorExitCode);
		}
	};
	return {
		programName,
		args,
		coreOptions
	};
}
function runImpl(parserOrProgram, options = {}) {
	const resolved = resolveProgramInput(parserOrProgram, options);
	const { parser, programMetadata } = resolved;
	options = resolved.options;
	const contexts = options.contexts;
	if (contexts && contexts.length > 0) {
		const { programName: programName$1, args: args$1, coreOptions: coreOptions$1 } = buildCoreOptions(options, programMetadata);
		const runWithOptions = {
			...coreOptions$1,
			contextOptions: options.contextOptions,
			args: args$1
		};
		return (0, __optique_core_facade.runWith)(parser, programName$1, contexts, runWithOptions);
	}
	const { programName, args, coreOptions } = buildCoreOptions(options, programMetadata);
	return (0, __optique_core_facade.runParser)(parser, programName, args, coreOptions);
}

//#endregion
exports.run = run;
exports.runAsync = runAsync;
exports.runSync = runSync;
