// @optique/core/dist/annotation-state.cjs
const require_internal_annotations = require('./internal/annotations.cjs');
const require_internal_parser = require('./internal/parser.cjs');

//#region src/annotation-state.ts
/**
* Shared targets for annotation-view proxies.
*
* Maps a proxy returned by {@link withAnnotationView} back to its original
* target object.
*
* @internal
*/
const annotationViewTargets = /* @__PURE__ */ new WeakMap();
const delegatedAnnotationCloneTargets = /* @__PURE__ */ new WeakMap();
/**
* Unwraps an annotation-view proxy to its original target object.
*
* @param value The candidate value that may be an annotation-view proxy.
* @returns The original target object when the input is a tracked
*          annotation-view proxy; otherwise the input value unchanged.
* @internal
*/
function unwrapAnnotationView(value) {
	if (value == null || typeof value !== "object") return value;
	return annotationViewTargets.get(value) ?? value;
}
function unwrapDelegatedAnnotationClone(value) {
	if (value == null || typeof value !== "object") return value;
	return delegatedAnnotationCloneTargets.get(value) ?? value;
}
/**
* Creates an annotation-aware proxy without changing the target shape.
*
* @param state The object state to expose through an annotation-aware view.
* @param annotations The annotations to surface through the proxy.
* @returns A proxy over the unwrapped target object that reports the supplied
*          annotations while preserving the target's structural behavior.
* @since 1.0.0
*/
function withAnnotationView(state, annotations) {
	const target = unwrapAnnotationView(state);
	const view = new Proxy(target, {
		get(target$1, key) {
			if (key === require_internal_annotations.annotationKey) return annotations;
			const value = Reflect.get(target$1, key, target$1);
			return typeof value === "function" ? value.bind(target$1) : value;
		},
		has(target$1, key) {
			return key === require_internal_annotations.annotationKey || Reflect.has(target$1, key);
		}
	});
	annotationViewTargets.set(view, target);
	return view;
}
/**
* Removes Optique's internal primitive-state annotation wrapper when present.
*
* @param state The parser state to normalize.
* @returns The wrapped primitive sentinel when the input is an injected
*          annotation wrapper; otherwise the original input unchanged.
* @internal
*/
function normalizeInjectedAnnotationState(state) {
	return require_internal_annotations.unwrapInjectedAnnotationWrapper(state);
}
function isNonPlainDelegatedObject(state) {
	const proto = Object.getPrototypeOf(state);
	if (proto === Array.prototype || proto === Date.prototype || proto === Map.prototype || proto === Set.prototype || proto === RegExp.prototype) return false;
	return proto !== Object.prototype && proto !== null;
}
function inheritDelegatedAnnotations(parentState, childState) {
	const target = normalizeDelegatedAnnotationState(childState);
	const delegatedState = require_internal_annotations.inheritAnnotations(parentState, target);
	if (delegatedState !== target) delegatedAnnotationCloneTargets.set(delegatedState, target);
	return delegatedState;
}
/**
* Removes Optique's internal annotation carriers from a delegated state.
*
* This unwraps primitive-state annotation wrappers, tracked delegated clones,
* and annotation-view proxies used for object states.
*
* @param state The delegated state to normalize.
* @returns The original underlying state value.
* @internal
*/
function normalizeDelegatedAnnotationState(state) {
	return normalizeInjectedAnnotationState(unwrapDelegatedAnnotationClone(unwrapAnnotationView(state)));
}
/**
* Returns whether the given state uses an internal delegated annotation carrier.
*
* @param state The candidate state to inspect.
* @returns `true` when the state is an injected primitive wrapper, a tracked
*          delegated clone, or an annotation-view proxy.
* @internal
*/
function hasDelegatedAnnotationCarrier(state) {
	return state != null && typeof state === "object" && (require_internal_annotations.isInjectedAnnotationWrapper(state) || delegatedAnnotationCloneTargets.has(state) || annotationViewTargets.has(state));
}
function getOrCreateNestedNormalizationClone(entry) {
	entry.clone ??= entry.createClone();
	return entry.clone;
}
function getPendingNestedNormalizationEntry(originalValue, normalizedValue, seen) {
	if (originalValue == null || typeof originalValue !== "object") return void 0;
	const entry = seen.get(originalValue);
	return entry != null && entry.clone != null && entry.clone === normalizedValue && (!entry.finalized || entry.preferCloneOnRead) ? entry : void 0;
}
function createPendingNestedNormalizationClone(source) {
	if (!Array.isArray(source)) return Object.create(Object.getPrototypeOf(source));
	if (Object.getPrototypeOf(source) === Array.prototype) return [];
	try {
		return source.slice(0, 0);
	} catch {
		return [];
	}
}
function createPendingNestedNormalizationCollectionClone(source) {
	const proto = Object.getPrototypeOf(source);
	if (source instanceof Map && proto === Map.prototype) return /* @__PURE__ */ new Map();
	if (source instanceof Set && proto === Set.prototype) return /* @__PURE__ */ new Set();
	const Constructor = source.constructor;
	if (typeof Constructor === "function") try {
		return new Constructor();
	} catch {}
	return source instanceof Map ? /* @__PURE__ */ new Map() : /* @__PURE__ */ new Set();
}
function normalizeNestedDelegatedStructuredState(source, seen, preferPendingClone) {
	const entry = {
		clone: void 0,
		createClone: () => createPendingNestedNormalizationClone(source),
		finalized: false,
		result: source,
		preferCloneOnRead: false
	};
	seen.set(source, entry);
	const overrides = /* @__PURE__ */ new Map();
	let changed = false;
	let hasPendingAliasOverride = false;
	for (const key of Reflect.ownKeys(source)) {
		const descriptor = Object.getOwnPropertyDescriptor(source, key);
		if (descriptor == null || !("value" in descriptor)) continue;
		const nextValue = normalizeNestedDelegatedAnnotationState(descriptor.value, seen, true);
		if (nextValue === descriptor.value) continue;
		overrides.set(key, nextValue);
		if (getPendingNestedNormalizationEntry(descriptor.value, nextValue, seen) == null) changed = true;
		else hasPendingAliasOverride = true;
	}
	if (!changed && !hasPendingAliasOverride) {
		entry.finalized = true;
		entry.result = source;
		return source;
	}
	const descriptors = Object.getOwnPropertyDescriptors(source);
	for (const [key, nextValue] of overrides) {
		const descriptor = descriptors[key];
		if (descriptor == null || !("value" in descriptor)) continue;
		descriptors[key] = {
			...descriptor,
			value: nextValue
		};
	}
	const clone = getOrCreateNestedNormalizationClone(entry);
	Object.defineProperties(clone, descriptors);
	entry.finalized = true;
	entry.preferCloneOnRead = !changed && hasPendingAliasOverride;
	entry.result = changed ? clone : source;
	return changed || preferPendingClone ? clone : source;
}
function normalizeNestedDelegatedMapState(source, seen, preferPendingClone) {
	const entry = {
		clone: void 0,
		createClone: () => createPendingNestedNormalizationCollectionClone(source),
		finalized: false,
		result: source,
		preferCloneOnRead: false
	};
	seen.set(source, entry);
	const normalizedEntries = [];
	const overrides = /* @__PURE__ */ new Map();
	let changed = false;
	let hasPendingAliasOverride = false;
	for (const [key, value] of source) {
		const nextKey = normalizeNestedDelegatedAnnotationState(key, seen, true);
		const nextValue = normalizeNestedDelegatedAnnotationState(value, seen, true);
		normalizedEntries.push([nextKey, nextValue]);
		if (nextKey !== key) if (getPendingNestedNormalizationEntry(key, nextKey, seen) == null) changed = true;
		else hasPendingAliasOverride = true;
		if (nextValue !== value) if (getPendingNestedNormalizationEntry(value, nextValue, seen) == null) changed = true;
		else hasPendingAliasOverride = true;
	}
	for (const key of Reflect.ownKeys(source)) {
		const descriptor = Object.getOwnPropertyDescriptor(source, key);
		if (descriptor == null || !("value" in descriptor)) continue;
		const nextValue = normalizeNestedDelegatedAnnotationState(descriptor.value, seen, true);
		if (nextValue === descriptor.value) continue;
		overrides.set(key, nextValue);
		if (getPendingNestedNormalizationEntry(descriptor.value, nextValue, seen) == null) changed = true;
		else hasPendingAliasOverride = true;
	}
	if (!changed && !hasPendingAliasOverride) {
		entry.finalized = true;
		entry.result = source;
		return source;
	}
	const clone = getOrCreateNestedNormalizationClone(entry);
	for (const [key, value] of normalizedEntries) clone.set(key, value);
	const descriptors = Object.getOwnPropertyDescriptors(source);
	for (const [key, nextValue] of overrides) {
		const descriptor = descriptors[key];
		if (descriptor == null || !("value" in descriptor)) continue;
		descriptors[key] = {
			...descriptor,
			value: nextValue
		};
	}
	Object.defineProperties(clone, descriptors);
	entry.finalized = true;
	entry.preferCloneOnRead = !changed && hasPendingAliasOverride;
	entry.result = changed ? clone : source;
	return changed || preferPendingClone ? clone : source;
}
function normalizeNestedDelegatedSetState(source, seen, preferPendingClone) {
	const entry = {
		clone: void 0,
		createClone: () => createPendingNestedNormalizationCollectionClone(source),
		finalized: false,
		result: source,
		preferCloneOnRead: false
	};
	seen.set(source, entry);
	const normalizedValues = [];
	const overrides = /* @__PURE__ */ new Map();
	let changed = false;
	let hasPendingAliasOverride = false;
	for (const value of source) {
		const nextValue = normalizeNestedDelegatedAnnotationState(value, seen, true);
		normalizedValues.push(nextValue);
		if (nextValue === value) continue;
		if (getPendingNestedNormalizationEntry(value, nextValue, seen) == null) changed = true;
		else hasPendingAliasOverride = true;
	}
	for (const key of Reflect.ownKeys(source)) {
		const descriptor = Object.getOwnPropertyDescriptor(source, key);
		if (descriptor == null || !("value" in descriptor)) continue;
		const nextValue = normalizeNestedDelegatedAnnotationState(descriptor.value, seen, true);
		if (nextValue === descriptor.value) continue;
		overrides.set(key, nextValue);
		if (getPendingNestedNormalizationEntry(descriptor.value, nextValue, seen) == null) changed = true;
		else hasPendingAliasOverride = true;
	}
	if (!changed && !hasPendingAliasOverride) {
		entry.finalized = true;
		entry.result = source;
		return source;
	}
	const clone = getOrCreateNestedNormalizationClone(entry);
	for (const value of normalizedValues) clone.add(value);
	const descriptors = Object.getOwnPropertyDescriptors(source);
	for (const [key, nextValue] of overrides) {
		const descriptor = descriptors[key];
		if (descriptor == null || !("value" in descriptor)) continue;
		descriptors[key] = {
			...descriptor,
			value: nextValue
		};
	}
	Object.defineProperties(clone, descriptors);
	entry.finalized = true;
	entry.preferCloneOnRead = !changed && hasPendingAliasOverride;
	entry.result = changed ? clone : source;
	return changed || preferPendingClone ? clone : source;
}
/**
* Recursively removes delegated annotation carriers from plain-object, array,
* and built-in collection structures.
*
* Nested plain objects, arrays, Maps, and Sets are shallow-cloned only when a
* delegated carrier is found below them. Other non-plain objects are unwrapped
* at the top level and then preserved as-is to avoid mutating or reconstructing
* class instances.
*
* @param value The candidate value to normalize.
* @param seen Tracks already-normalized objects so cyclic values keep their
*             shape.
* @returns The original value when no delegated carriers are present, or a
*          normalized clone with delegated carriers removed.
* @internal
*/
function normalizeNestedDelegatedAnnotationState(value, seen = /* @__PURE__ */ new WeakMap(), preferPendingClone = false) {
	const normalized = normalizeDelegatedAnnotationState(value);
	if (normalized == null || typeof normalized !== "object") return normalized;
	const source = normalized;
	const existing = seen.get(source);
	if (existing != null) {
		if (!existing.finalized) return getOrCreateNestedNormalizationClone(existing);
		return preferPendingClone && existing.preferCloneOnRead ? getOrCreateNestedNormalizationClone(existing) : existing.result;
	}
	if (Array.isArray(source)) return normalizeNestedDelegatedStructuredState(source, seen, preferPendingClone);
	if (source instanceof Map) return normalizeNestedDelegatedMapState(source, seen, preferPendingClone);
	if (source instanceof Set) return normalizeNestedDelegatedSetState(source, seen, preferPendingClone);
	const proto = Object.getPrototypeOf(source);
	if (proto !== Object.prototype && proto !== null) return normalized;
	return normalizeNestedDelegatedStructuredState(source, seen, preferPendingClone);
}
/**
* Creates a short-lived delegated state that exposes the parent's annotations
* regardless of the child state's runtime shape.
*
* Primitive and nullish states use `injectAnnotations()`, clone-based object
* delegation is tracked so it can be normalized back out later, and non-plain
* objects use an annotation-view proxy so class invariants such as private
* fields remain intact.
*
* @param parentState The state carrying the annotations to delegate.
* @param childState The child state that should observe those annotations.
* @returns A delegated child state that exposes the parent's annotations.
* @internal
*/
function getDelegatedAnnotationState(parentState, childState) {
	const annotations = require_internal_annotations.getAnnotations(parentState);
	if (annotations === void 0) return childState;
	if (require_internal_annotations.isInjectedAnnotationWrapper(childState)) return require_internal_annotations.injectAnnotations(normalizeInjectedAnnotationState(childState), annotations);
	if (childState == null || typeof childState !== "object") return require_internal_annotations.injectAnnotations(childState, annotations);
	if (require_internal_annotations.getAnnotations(childState) === annotations && (delegatedAnnotationCloneTargets.has(childState) || annotationViewTargets.has(childState))) return childState;
	if (isNonPlainDelegatedObject(childState)) return withAnnotationView(childState, annotations);
	return inheritDelegatedAnnotations(parentState, childState);
}
/**
* Returns whether a state is still at the initial sentinel after normalizing
* Optique's injected annotation wrapper.
*
* This treats plain `undefined` and annotation-wrapped `undefined` the same.
*
* @param state The parser state to inspect.
* @returns `true` when the normalized state is still `undefined`;
*          otherwise `false`.
* @internal
*/
function isAnnotationWrappedInitialState(state) {
	return normalizeInjectedAnnotationState(state) === void 0;
}
/**
* Propagates parent annotations into a child parse state when the child parser
* explicitly opts into parent annotation inheritance.
*
* @param parentState The parent parser state that may carry annotations.
* @param childState The child parse state that may receive inherited
*                   annotations.
* @param parser The child parser whose inheritance marker controls whether
*               wrapper injection is allowed.
* @returns The original child state when no injection is needed or possible,
*          or an annotation-injected child state that preserves the original
*          sentinel or object shape when inheritance applies.
* @internal
*/
function getWrappedChildParseState(parentState, childState, parser) {
	const annotations = require_internal_annotations.getAnnotations(parentState);
	const shouldInheritAnnotations = Reflect.get(parser, require_internal_parser.inheritParentAnnotationsKey) === true;
	if (childState == null) {
		if (annotations !== void 0 && shouldInheritAnnotations) return require_internal_annotations.injectAnnotations(childState, annotations);
		return childState;
	}
	if (annotations === void 0 || typeof childState !== "object" || require_internal_annotations.getAnnotations(childState) === annotations || !shouldInheritAnnotations) return childState;
	const injectedState = require_internal_annotations.injectAnnotations(childState, annotations);
	return require_internal_annotations.getAnnotations(injectedState) === annotations ? injectedState : childState;
}
/**
* Propagates parent annotations into a child state while preserving the child
* state's shape for parsers that do not opt into full wrapper injection.
*
* @param parentState The parent parser state that may carry annotations.
* @param childState The child state that may receive inherited annotations.
* @param parser The child parser whose inheritance marker controls whether
*               full wrapper injection is allowed.
* @returns The original child state when no wrapping is needed, an
*          annotation-injected child state when inheritance applies, or an
*          annotation-view proxy that preserves the child's object shape.
* @internal
*/
function getWrappedChildState(parentState, childState, parser) {
	const annotations = require_internal_annotations.getAnnotations(parentState);
	const shouldInheritAnnotations = Reflect.get(parser, require_internal_parser.inheritParentAnnotationsKey) === true;
	if (childState == null) {
		if (annotations !== void 0 && shouldInheritAnnotations) return require_internal_annotations.injectAnnotations(childState, annotations);
		return childState;
	}
	if (annotations === void 0 || typeof childState !== "object" || require_internal_annotations.getAnnotations(childState) === annotations) return childState;
	if (shouldInheritAnnotations) {
		const injectedState = require_internal_annotations.injectAnnotations(childState, annotations);
		if (require_internal_annotations.getAnnotations(injectedState) === annotations) return injectedState;
	}
	return withAnnotationView(childState, annotations);
}
/**
* Reconciles object-owned child state with parent annotations using the same
* shared object-state inheritance rule across parser families.
*
* @param parentState The parent parser state that may carry annotations.
* @param childState The object-owned child state to reconcile.
* @returns The original child state when no reconciliation is needed, or a
*          child state with inherited annotations when the object state should
*          carry the parent's annotations.
* @internal
*/
function reconcileObjectChildState(parentState, childState) {
	const annotations = require_internal_annotations.getAnnotations(parentState);
	if (annotations === void 0 || childState == null || typeof childState !== "object" || require_internal_annotations.getAnnotations(childState) === annotations) return childState;
	return require_internal_annotations.inheritAnnotations(parentState, childState);
}

//#endregion
exports.annotationViewTargets = annotationViewTargets;
exports.getDelegatedAnnotationState = getDelegatedAnnotationState;
exports.getWrappedChildParseState = getWrappedChildParseState;
exports.getWrappedChildState = getWrappedChildState;
exports.hasDelegatedAnnotationCarrier = hasDelegatedAnnotationCarrier;
exports.isAnnotationWrappedInitialState = isAnnotationWrappedInitialState;
exports.normalizeDelegatedAnnotationState = normalizeDelegatedAnnotationState;
exports.normalizeInjectedAnnotationState = normalizeInjectedAnnotationState;
exports.normalizeNestedDelegatedAnnotationState = normalizeNestedDelegatedAnnotationState;
exports.reconcileObjectChildState = reconcileObjectChildState;
exports.unwrapAnnotationView = unwrapAnnotationView;
exports.withAnnotationView = withAnnotationView;
