// @optique/core/dist/dependency-metadata.cjs
const require_internal_dependency = require('./internal/dependency.cjs');

//#region src/dependency-metadata.ts
/**
* Extracts {@link ParserDependencyMetadata} from a value parser by reading
* old-protocol markers (`dependencySourceMarker`, `derivedValueParserMarker`,
* etc.).
*
* Returns `undefined` if the parser has no dependency-related markers.
*
* @param valueParser The value parser to inspect.
* @returns Metadata, or `undefined` for plain parsers.
* @internal
* @since 1.0.0
*/
function extractDependencyMetadata(valueParser) {
	if (require_internal_dependency.isDependencySource(valueParser)) return { source: {
		kind: "source",
		sourceId: valueParser[require_internal_dependency.dependencyId],
		extractSourceValue: extractFromBareState,
		preservesSourceValue: true
	} };
	if (require_internal_dependency.isDerivedValueParser(valueParser)) {
		const isMultiSource = require_internal_dependency.dependencyIds in valueParser && valueParser[require_internal_dependency.dependencyIds] != null;
		const allIds = isMultiSource ? valueParser[require_internal_dependency.dependencyIds] : [valueParser[require_internal_dependency.dependencyId]];
		let defaultValuesFn;
		if (require_internal_dependency.defaultValues in valueParser && valueParser[require_internal_dependency.defaultValues] != null) defaultValuesFn = valueParser[require_internal_dependency.defaultValues];
		else if (require_internal_dependency.singleDefaultValue in valueParser && valueParser[require_internal_dependency.singleDefaultValue] != null) {
			const singleFn = valueParser[require_internal_dependency.singleDefaultValue];
			defaultValuesFn = () => [singleFn()];
		}
		const parser = valueParser;
		const replayParse = (rawInput, depValues) => {
			const depArg = isMultiSource ? depValues : depValues[0];
			return parser[require_internal_dependency.parseWithDependency](rawInput, depArg);
		};
		const suggestFn = require_internal_dependency.suggestWithDependency in parser ? parser[require_internal_dependency.suggestWithDependency] : void 0;
		const replaySuggest = suggestFn != null ? (prefix, depValues) => {
			const depArg = isMultiSource ? depValues : depValues[0];
			return suggestFn(prefix, depArg);
		} : void 0;
		return { derived: {
			kind: "derived",
			dependencyIds: allIds,
			getDefaultDependencyValues: defaultValuesFn,
			replayParse,
			replaySuggest
		} };
	}
	return void 0;
}
/**
* Extracts the source parse result from a plain source state.
*
* Accepts either a `DependencySourceState` or a bare
* `ValueParserResult`-shaped object, and returns `undefined` for
* unrelated states. Used as the base `extractSourceValue` for plain
* dependency sources.
*/
function extractFromBareState(state) {
	if (require_internal_dependency.isDependencySourceState(state)) return state.result;
	if (state != null && typeof state === "object" && Object.hasOwn(state, "success") && typeof state.success === "boolean" && (state.success && Object.hasOwn(state, "value") || !state.success && Object.hasOwn(state, "error"))) return state;
	return void 0;
}
/**
* Wraps an inner `extractSourceValue` to unwrap `[innerState]` first.
* Used by `optional()` and `withDefault()` which wrap state in a
* single-element array.
*/
function unwrapArrayThenExtract(innerExtract) {
	return (state) => {
		if (Array.isArray(state) && state.length === 1) return innerExtract(state[0]);
		return innerExtract(state);
	};
}
/**
* Composes dependency metadata through a modifier wrapper.
*
* - `"optional"`: composes `extractSourceValue` with array unwrapping.
* - `"withDefault"`: adds `getMissingSourceValue` if the inner parser
*   preserves a dependency source.
* - `"map"`: sets `transformsSourceValue` and clears
*   `preservesSourceValue`.
*
* Returns `undefined` if `inner` is `undefined`.
*
* @param inner The inner parser's metadata.
* @param wrapperKind The type of modifier being applied.
* @param options Additional options for certain wrapper kinds.
* @returns Composed metadata, or `undefined`.
* @internal
* @since 1.0.0
*/
function composeDependencyMetadata(inner, wrapperKind, options) {
	if (inner === void 0) return void 0;
	switch (wrapperKind) {
		case "optional": {
			if (inner.source != null && inner.source.extractSourceValue != null) return {
				...inner,
				source: {
					...inner.source,
					extractSourceValue: unwrapArrayThenExtract(inner.source.extractSourceValue)
				}
			};
			return inner;
		}
		case "withDefault": {
			const wrappedExtract = inner.source?.extractSourceValue != null ? unwrapArrayThenExtract(inner.source.extractSourceValue) : void 0;
			const preservesSourceValue = inner.source?.preservesSourceValue !== false;
			if (inner.source != null) return {
				...inner,
				source: {
					...inner.source,
					...wrappedExtract != null && { extractSourceValue: wrappedExtract },
					...preservesSourceValue && options?.defaultValue != null && { getMissingSourceValue: options.defaultValue }
				}
			};
			return inner;
		}
		case "map": {
			const result = {
				...inner,
				transform: { transformsSourceValue: true }
			};
			if (inner.source != null) return {
				...result,
				source: {
					...inner.source,
					preservesSourceValue: false
				}
			};
			return result;
		}
	}
}

//#endregion
exports.composeDependencyMetadata = composeDependencyMetadata;
exports.extractDependencyMetadata = extractDependencyMetadata;
