// @optique/core/dist/execution-context.cjs

//#region src/execution-context.ts
/**
* Appends a child parser segment to the current execution path.
* @internal
*/
function withChildExecPath(exec, segment) {
	if (exec == null) return void 0;
	return {
		...exec,
		path: [...exec.path ?? [], segment]
	};
}
/**
* Merges child-owned execution fields back into the parent execution context.
* @internal
*/
function mergeChildExec(parent, child) {
	if (parent == null) return child;
	if (child == null) return parent;
	return {
		...parent,
		trace: child.trace ?? parent.trace,
		dependencyRuntime: child.dependencyRuntime ?? parent.dependencyRuntime,
		dependencyRegistry: child.dependencyRegistry ?? parent.dependencyRegistry,
		commandPath: child.commandPath ?? parent.commandPath,
		preCompletedByParser: child.preCompletedByParser ?? parent.preCompletedByParser,
		excludedSourceFields: child.excludedSourceFields ?? parent.excludedSourceFields
	};
}
/**
* Creates a child parser context while keeping flat and nested execution data
* aligned.
* @internal
*/
function withChildContext(context, segment, state, usage) {
	const exec = withChildExecPath(context.exec, segment);
	const dependencyRegistry = context.dependencyRegistry ?? exec?.dependencyRegistry;
	return {
		...context,
		state,
		...usage != null ? { usage } : {},
		...exec != null ? {
			exec: dependencyRegistry === exec.dependencyRegistry ? exec : {
				...exec,
				dependencyRegistry
			},
			dependencyRegistry
		} : {}
	};
}

//#endregion
exports.mergeChildExec = mergeChildExec;
exports.withChildContext = withChildContext;
exports.withChildExecPath = withChildExecPath;
