// @optique/core/dist/extension.cjs
const require_internal_annotations = require('./internal/annotations.cjs');
const require_mode_dispatch = require('./internal/mode-dispatch.cjs');
const require_internal_parser = require('./internal/parser.cjs');
const require_annotation_state = require('./annotation-state.cjs');

//#region src/extension.ts
const emptyTraits = Object.freeze({});
/**
* Defines stable extension traits on a parser object.
*
* @param parser The parser object to annotate.
* @param traits Traits to enable.
* @throws {TypeError} If a trait property cannot be defined on `parser`.
* @since 1.0.0
*/
function defineTraits(parser, traits) {
	if (traits.inheritsAnnotations === true) require_internal_parser.defineInheritedAnnotationParser(parser);
	if (traits.completesFromSource === true) Object.defineProperty(parser, require_internal_parser.unmatchedNonCliDependencySourceStateMarker, {
		value: true,
		configurable: true,
		enumerable: true
	});
	if (traits.requiresSourceBinding === true) Object.defineProperty(parser, require_internal_parser.annotationWrapperRequiresSourceBindingKey, {
		value: true,
		configurable: true,
		enumerable: false
	});
}
/**
* Reads the stable extension traits defined on a parser object.
*
* @param parser The parser object to inspect.
* @returns The enabled traits.  Returns an empty object when none are set.
* @since 1.0.0
*/
function getTraits(parser) {
	const traits = {
		...Reflect.get(parser, require_internal_parser.inheritParentAnnotationsKey) === true ? { inheritsAnnotations: true } : {},
		...Reflect.get(parser, require_internal_parser.unmatchedNonCliDependencySourceStateMarker) === true ? { completesFromSource: true } : {},
		...Reflect.get(parser, require_internal_parser.annotationWrapperRequiresSourceBindingKey) === true ? { requiresSourceBinding: true } : {}
	};
	return Object.keys(traits).length > 0 ? traits : emptyTraits;
}
/**
* Delegates suggest-time runtime nodes to an inner parser while preserving an
* outer parser's own source metadata node.
*
* @param innerParser The wrapped parser that owns the underlying nodes.
* @param outerParser The outer parser that may contribute its own source node.
* @param state The outer parser state.
* @param path The parser path within the parse tree.
* @param innerState The state to use when collecting inner nodes.
* @param position Whether the outer node is appended or prepended.
* @returns The composed runtime nodes.
* @since 1.0.0
*/
function delegateSuggestNodes(innerParser, outerParser, state, path, innerState, position = "append") {
	return require_internal_parser.getDelegatingSuggestRuntimeNodes(innerParser, outerParser, state, path, innerState, position);
}
/**
* Maps the source capability of a parser's dependency metadata while
* preserving any derived or transform capabilities unchanged.
*
* @param parser The parser whose source metadata should be transformed.
* @param mapSource Function that transforms the source capability.
* @returns The dependency metadata with its source capability transformed when
*          present; otherwise the original dependency metadata, or
*          `undefined` when the parser has no dependency metadata.
* @since 1.0.0
*/
function mapSourceMetadata(parser, mapSource) {
	return require_internal_parser.composeWrappedSourceMetadata(parser.dependencyMetadata, mapSource);
}

//#endregion
exports.defineTraits = defineTraits;
exports.delegateSuggestNodes = delegateSuggestNodes;
exports.dispatchByMode = require_mode_dispatch.dispatchByMode;
exports.getTraits = getTraits;
exports.inheritAnnotations = require_internal_annotations.inheritAnnotations;
exports.injectAnnotations = require_internal_annotations.injectAnnotations;
exports.isInjectedAnnotationState = require_internal_annotations.isInjectedAnnotationState;
exports.mapModeValue = require_mode_dispatch.mapModeValue;
exports.mapSourceMetadata = mapSourceMetadata;
exports.unwrapInjectedAnnotationState = require_internal_annotations.unwrapInjectedAnnotationState;
exports.withAnnotationView = require_annotation_state.withAnnotationView;
exports.wrapForMode = require_mode_dispatch.wrapForMode;
