// @optique/core/dist/facade.js
import { injectAnnotations, isInjectedAnnotationWrapper, unwrapInjectedAnnotationWrapper } from "./internal/annotations.js";
import { commandLine, formatMessage, lineBreak, message, optionName, text, value } from "./message.js";
import { validateCommandNames, validateContextIds, validateMetaNameCollisions, validateOptionNames, validateProgramName } from "./validate.js";
import { formatUsage, isSuggestionHidden } from "./usage.js";
import { formatDocPage } from "./doc.js";
import { dispatchByMode } from "./internal/mode-dispatch.js";
import { collectExplicitSourceValues, collectExplicitSourceValuesAsync, createDependencyRuntimeContext } from "./dependency-runtime.js";
import { createInputTrace } from "./input-trace.js";
import { createParserContext, getDocPage } from "./internal/parser.js";
import { bash, fish, nu, pwsh, zsh } from "./completion.js";
import { allowDuplicateLeadingCommandNamesKey, hiddenCommandAliasesKey } from "./internal/command-alias.js";
import { completeOrExtractPhase2Seed } from "./phase2-seed.js";
import { group, longestMatch, object } from "./constructs.js";
import { multiple, optional, withDefault } from "./modifiers.js";
import { string } from "./valueparser.js";
import { argument, command, constant, flag, option } from "./primitives.js";

//#region src/facade.ts
/**
* Combines built-in meta commands while intentionally bypassing duplicate
* leading command name validation in `createLongestMatch()`.
*/
function longestMatchForMetaCommands(...parsers) {
	const options = { [allowDuplicateLeadingCommandNamesKey]: true };
	return longestMatch(...parsers, options);
}
const SuppressedErrorCtor = typeof SuppressedError === "function" ? SuppressedError : (() => {
	class SuppressedErrorPolyfill extends Error {
		error;
		suppressed;
		constructor(error, suppressed, message$1) {
			super(message$1);
			this.name = "SuppressedError";
			this.error = error;
			this.suppressed = suppressed;
		}
	}
	return SuppressedErrorPolyfill;
})();
function isPlainObject(value$1) {
	const proto = Object.getPrototypeOf(value$1);
	return proto === Object.prototype || proto === null;
}
function prepareParsedForContexts(parsed, deferred, deferredKeys) {
	if (!deferred) return parsed;
	if (deferredKeys != null && deferredKeys.size === 0 && parsed != null && typeof parsed === "object") return void 0;
	if (deferredKeys != null && deferredKeys.size > 0 && parsed != null && typeof parsed === "object" && !isPlainObject(parsed) && !Array.isArray(parsed)) return void 0;
	if (deferredKeys != null && deferredKeys.size > 0 && parsed != null && typeof parsed === "object" && (isPlainObject(parsed) || Array.isArray(parsed))) {
		const getDeferredEntry = (key) => {
			const entry = deferredKeys.get(key);
			if (entry !== void 0) return entry;
			if (typeof key === "string") {
				const num = Number(key);
				if (Number.isInteger(num)) return deferredKeys.get(num);
			} else if (typeof key === "number") return deferredKeys.get(String(key));
			return void 0;
		};
		const ownKeys = Reflect.ownKeys(parsed);
		let hasMatchingKey = false;
		for (const key of ownKeys) if (getDeferredEntry(key) !== void 0) {
			hasMatchingKey = true;
			break;
		}
		if (hasMatchingKey) {
			const isArray = Array.isArray(parsed);
			let allDeferred = true;
			for (const key of ownKeys) {
				if (isArray && key === "length") continue;
				const desc = Object.getOwnPropertyDescriptor(parsed, key);
				if (desc != null && "value" in desc && getDeferredEntry(key) === void 0) {
					allDeferred = false;
					break;
				}
			}
			if (allDeferred) return void 0;
			const clone = isArray ? new Array(parsed.length) : Object.create(Object.getPrototypeOf(parsed));
			for (const key of ownKeys) {
				const desc = Object.getOwnPropertyDescriptor(parsed, key);
				if (desc == null) continue;
				const entry = getDeferredEntry(key);
				if ("value" in desc && entry !== void 0) if (entry === null) Object.defineProperty(clone, key, {
					...desc,
					value: void 0
				});
				else Object.defineProperty(clone, key, {
					...desc,
					value: prepareParsedForContexts(desc.value, true, entry)
				});
				else Object.defineProperty(clone, key, desc);
			}
			return clone;
		}
		return void 0;
	}
	if (parsed == null || typeof parsed !== "object") return void 0;
	return parsed;
}
function isBufferUnchanged(previous, current) {
	return current.length > 0 && current.length === previous.length && current.every((item, i) => item === previous[i]);
}
function getFailureProgress(args, buffer, consumedInStep) {
	return {
		remainingArgs: buffer.slice(consumedInStep),
		consumedCount: args.length - buffer.length + consumedInStep
	};
}
function createParseExec(parser) {
	return {
		usage: parser.usage,
		phase: "parse",
		path: [],
		commandPath: [],
		trace: createInputTrace()
	};
}
function getCommandPath(exec) {
	return exec?.commandPath ?? [];
}
function createCompleteExec(exec, context) {
	const runtime = createDependencyRuntimeContext();
	return {
		...exec,
		phase: "complete",
		dependencyRuntime: runtime,
		dependencyRegistry: runtime.registry,
		commandPath: getCommandPath(context.exec) ?? exec.commandPath,
		trace: context.exec?.trace ?? context.trace ?? exec.trace
	};
}
function attemptParseSync(parser, args, mode = "complete") {
	const shouldUnwrapAnnotatedValue = isInjectedAnnotationWrapper(parser.initialState);
	const exec = createParseExec(parser);
	let context = createParserContext({
		buffer: args,
		state: parser.initialState,
		optionsTerminated: false
	}, exec);
	do {
		const result = parser.parse(context);
		if (!result.success) {
			const progress = getFailureProgress(args, context.buffer, result.consumed);
			return {
				kind: "failure",
				error: result.error,
				remainingArgs: progress.remainingArgs,
				consumedCount: progress.consumedCount,
				optionsTerminated: context.optionsTerminated,
				commandPath: getCommandPath(context.exec)
			};
		}
		const previousBuffer = context.buffer;
		context = result.next;
		if (isBufferUnchanged(previousBuffer, context.buffer)) {
			const progress = getFailureProgress(args, previousBuffer, result.consumed.length);
			return {
				kind: "failure",
				error: message`Unexpected option or argument: ${context.buffer[0]}.`,
				remainingArgs: progress.remainingArgs,
				consumedCount: progress.consumedCount,
				optionsTerminated: context.optionsTerminated,
				commandPath: getCommandPath(context.exec)
			};
		}
	} while (context.buffer.length > 0);
	if (mode === "parse-only") return {
		kind: "success",
		value: void 0
	};
	const endResult = parser.complete(context.state, createCompleteExec(exec, context));
	if (!endResult.success) return {
		kind: "failure",
		error: endResult.error,
		remainingArgs: [],
		consumedCount: args.length,
		optionsTerminated: context.optionsTerminated,
		commandPath: getCommandPath(context.exec)
	};
	return {
		kind: "success",
		value: shouldUnwrapAnnotatedValue ? unwrapInjectedAnnotationWrapper(endResult.value) : endResult.value
	};
}
async function attemptParseAsync(parser, args, mode = "complete") {
	const shouldUnwrapAnnotatedValue = isInjectedAnnotationWrapper(parser.initialState);
	const exec = createParseExec(parser);
	let context = createParserContext({
		buffer: args,
		state: parser.initialState,
		optionsTerminated: false
	}, exec);
	do {
		const result = await parser.parse(context);
		if (!result.success) {
			const progress = getFailureProgress(args, context.buffer, result.consumed);
			return {
				kind: "failure",
				error: result.error,
				remainingArgs: progress.remainingArgs,
				consumedCount: progress.consumedCount,
				optionsTerminated: context.optionsTerminated,
				commandPath: getCommandPath(context.exec)
			};
		}
		const previousBuffer = context.buffer;
		context = result.next;
		if (isBufferUnchanged(previousBuffer, context.buffer)) {
			const progress = getFailureProgress(args, previousBuffer, result.consumed.length);
			return {
				kind: "failure",
				error: message`Unexpected option or argument: ${context.buffer[0]}.`,
				remainingArgs: progress.remainingArgs,
				consumedCount: progress.consumedCount,
				optionsTerminated: context.optionsTerminated,
				commandPath: getCommandPath(context.exec)
			};
		}
	} while (context.buffer.length > 0);
	if (mode === "parse-only") return {
		kind: "success",
		value: void 0
	};
	const endResult = await parser.complete(context.state, createCompleteExec(exec, context));
	if (!endResult.success) return {
		kind: "failure",
		error: endResult.error,
		remainingArgs: [],
		consumedCount: args.length,
		optionsTerminated: context.optionsTerminated,
		commandPath: getCommandPath(context.exec)
	};
	return {
		kind: "success",
		value: shouldUnwrapAnnotatedValue ? unwrapInjectedAnnotationWrapper(endResult.value) : endResult.value
	};
}
function createPhase2SeedExec(parser, context) {
	const exec = {
		usage: parser.usage,
		phase: "parse",
		path: [],
		commandPath: [],
		trace: createInputTrace()
	};
	const runtime = createDependencyRuntimeContext();
	return {
		...exec,
		phase: "complete",
		dependencyRuntime: runtime,
		dependencyRegistry: runtime.registry,
		commandPath: getCommandPath(context.exec),
		trace: context.exec?.trace ?? context.trace ?? exec.trace
	};
}
function createPhase2SeedContext(parser, args) {
	const exec = {
		usage: parser.usage,
		phase: "parse",
		path: [],
		commandPath: [],
		trace: createInputTrace()
	};
	return createParserContext({
		buffer: args,
		state: parser.initialState,
		optionsTerminated: false
	}, exec);
}
function extractPhase2SeedSync(parser, args) {
	let context = createPhase2SeedContext(parser, args);
	do {
		const result = parser.parse(context);
		if (!result.success) return completeOrExtractPhase2Seed(parser, context.state, createPhase2SeedExec(parser, context));
		const previousBuffer = context.buffer;
		context = result.next;
		if (isBufferUnchanged(previousBuffer, context.buffer)) return completeOrExtractPhase2Seed(parser, context.state, createPhase2SeedExec(parser, context));
	} while (context.buffer.length > 0);
	return completeOrExtractPhase2Seed(parser, context.state, createPhase2SeedExec(parser, context));
}
async function extractPhase2SeedAsync(parser, args) {
	let context = createPhase2SeedContext(parser, args);
	do {
		const result = await parser.parse(context);
		if (!result.success) return await completeOrExtractPhase2Seed(parser, context.state, createPhase2SeedExec(parser, context));
		const previousBuffer = context.buffer;
		context = result.next;
		if (isBufferUnchanged(previousBuffer, context.buffer)) return await completeOrExtractPhase2Seed(parser, context.state, createPhase2SeedExec(parser, context));
	} while (context.buffer.length > 0);
	return await completeOrExtractPhase2Seed(parser, context.state, createPhase2SeedExec(parser, context));
}
function getMetaCommandAliases(names) {
	const [, firstAlias, ...restAliases] = names;
	return firstAlias == null ? void 0 : [firstAlias, ...restAliases];
}
/**
* Creates help parsers based on the sub-config.
*/
function createHelpParser(commandConfig, optionConfig) {
	let helpCommand = null;
	let helpOption = null;
	if (commandConfig) {
		const names = commandConfig.names ?? ["help"];
		const innerParser = multiple(argument(string({ metavar: "COMMAND" }), { description: message`Command name to show help for.` }));
		const aliases = getMetaCommandAliases(names);
		helpCommand = command(names[0], innerParser, {
			description: message`Show help information.`,
			...aliases != null ? { [hiddenCommandAliasesKey]: aliases } : {},
			hidden: commandConfig.hidden
		});
	}
	if (optionConfig) {
		const names = optionConfig.names ?? ["--help"];
		helpOption = flag(...names, {
			description: message`Show help information.`,
			hidden: optionConfig.hidden
		});
	}
	return {
		helpCommand,
		helpOption
	};
}
/**
* Creates version parsers based on the sub-config.
*/
function createVersionParser(commandConfig, optionConfig) {
	let versionCommand = null;
	let versionOption = null;
	if (commandConfig) {
		const names = commandConfig.names ?? ["version"];
		const innerParser = object({});
		const aliases = getMetaCommandAliases(names);
		versionCommand = command(names[0], innerParser, {
			description: message`Show version information.`,
			...aliases != null ? { [hiddenCommandAliasesKey]: aliases } : {},
			hidden: commandConfig.hidden
		});
	}
	if (optionConfig) {
		const names = optionConfig.names ?? ["--version"];
		versionOption = flag(...names, {
			description: message`Show version information.`,
			hidden: optionConfig.hidden
		});
	}
	return {
		versionCommand,
		versionOption
	};
}
const metaResultBrand = Symbol("@optique/core/facade/meta");
/**
* Creates completion parsers based on the sub-config.
*/
function createCompletionParser(programName, availableShells, commandConfig, optionConfig) {
	let completionCommand = null;
	let completionOption = null;
	const shellList = [];
	for (const shell in availableShells) {
		if (shellList.length > 0) shellList.push(text(", "));
		shellList.push(value(shell));
	}
	const completionInner = object({
		shell: optional(argument(string({ metavar: "SHELL" }), { description: message`Shell type (${shellList}). Generate completion script when used alone, or provide completions when followed by arguments.` })),
		args: multiple(argument(string({ metavar: "ARG" }), { description: message`Command line arguments for completion suggestions (used by shell integration; you usually don't need to provide this).` }))
	});
	if (commandConfig) {
		const names = commandConfig.names ?? ["completion"];
		const displayName = names[0];
		const completionCommandConfig = {
			brief: message`Generate shell completion script or provide completions.`,
			description: message`Generate shell completion script or provide completions.`,
			footer: message`Examples:${lineBreak()}  Bash:       ${commandLine(`eval "$(${programName} ${displayName} bash)"`)}${lineBreak()}  zsh:        ${commandLine(`eval "$(${programName} ${displayName} zsh)"`)}${lineBreak()}  fish:       ${commandLine(`eval "$(${programName} ${displayName} fish)"`)}${lineBreak()}  PowerShell: ${commandLine(`${programName} ${displayName} pwsh > ${programName}-completion.ps1; . ./${programName}-completion.ps1`)}${lineBreak()}  Nushell:    ${commandLine(`${programName} ${displayName} nu | save ${programName}-completion.nu; source ./${programName}-completion.nu`)}`
		};
		const aliases = getMetaCommandAliases(names);
		completionCommand = command(names[0], completionInner, {
			...completionCommandConfig,
			...aliases != null ? { [hiddenCommandAliasesKey]: aliases } : {},
			hidden: commandConfig.hidden
		});
	}
	if (optionConfig) {
		const names = optionConfig.names ?? ["--completion"];
		const completionOptions = [];
		for (const name of names) completionOptions.push(option(name, string({ metavar: "SHELL" }), {
			description: message`Generate shell completion script.`,
			hidden: optionConfig.hidden
		}));
		const completionOptionParser = completionOptions.length === 1 ? completionOptions[0] : longestMatch(...completionOptions);
		const argsParser = withDefault(multiple(argument(string({ metavar: "ARG" }), { description: message`Command line arguments for completion suggestions (used by shell integration; you usually don't need to provide this).` })), []);
		completionOption = object({
			shell: completionOptionParser,
			args: argsParser
		});
	}
	return {
		completionCommand,
		completionOption
	};
}
function createMetaOptionDocParser(source) {
	return {
		mode: "sync",
		$valueType: [],
		$stateType: [],
		priority: 200,
		usage: source.usage,
		leadingNames: source.leadingNames,
		acceptingAnyToken: false,
		initialState: null,
		parse: () => ({
			success: false,
			error: message`Documentation-only meta option.`,
			consumed: 0
		}),
		complete: (state) => ({
			success: true,
			value: state
		}),
		suggest: function* () {},
		getDocFragments: (_state, defaultValue) => source.getDocFragments({
			kind: "available",
			state: source.initialState
		}, defaultValue)
	};
}
function combineWithHelpVersion(originalParser, helpParsers, versionParsers, completionParsers, groups) {
	const parsers = [];
	if (helpParsers.helpOption) {
		const helpOptionDocParser = createMetaOptionDocParser(helpParsers.helpOption);
		const wrappedHelp = groups?.helpOptionGroup ? group(groups.helpOptionGroup, helpOptionDocParser) : helpOptionDocParser;
		parsers.push(wrappedHelp);
	}
	if (versionParsers.versionOption) {
		const versionOptionDocParser = createMetaOptionDocParser(versionParsers.versionOption);
		const wrappedVersion = groups?.versionOptionGroup ? group(groups.versionOptionGroup, versionOptionDocParser) : versionOptionDocParser;
		parsers.push(wrappedVersion);
	}
	if (versionParsers.versionCommand) {
		const versionParser = object({
			[metaResultBrand]: constant(true),
			help: constant(false),
			version: constant(true),
			completion: constant(false),
			result: versionParsers.versionCommand,
			helpFlag: helpParsers.helpOption ? optional(helpParsers.helpOption) : constant(false)
		});
		parsers.push(groups?.versionCommandGroup ? group(groups.versionCommandGroup, versionParser) : versionParser);
	}
	if (completionParsers.completionCommand) {
		const completionParser = object({
			[metaResultBrand]: constant(true),
			help: constant(false),
			version: constant(false),
			completion: constant(true),
			completionData: completionParsers.completionCommand,
			helpFlag: helpParsers.helpOption ? optional(helpParsers.helpOption) : constant(false)
		});
		parsers.push(groups?.completionCommandGroup ? group(groups.completionCommandGroup, completionParser) : completionParser);
	}
	if (helpParsers.helpCommand) {
		const helpParser = object({
			[metaResultBrand]: constant(true),
			help: constant(true),
			version: constant(false),
			completion: constant(false),
			commands: helpParsers.helpCommand
		});
		parsers.push(groups?.helpCommandGroup ? group(groups.helpCommandGroup, helpParser) : helpParser);
	}
	parsers.push(object({
		[metaResultBrand]: constant(true),
		help: constant(false),
		version: constant(false),
		completion: constant(false),
		result: originalParser
	}));
	const mainParserIndex = parsers.length - 1;
	if (parsers.length === 1) return parsers[0];
	let combined = longestMatchForMetaCommands(...parsers);
	const topUsage = combined.usage[0];
	if (topUsage?.type === "exclusive" && mainParserIndex > 0) {
		const terms = [...topUsage.terms];
		const [mainTerm] = terms.splice(mainParserIndex, 1);
		const lenientCount = (helpParsers.helpOption ? 1 : 0) + (versionParsers.versionOption ? 1 : 0);
		terms.splice(lenientCount, 0, mainTerm);
		combined = {
			...combined,
			usage: [{
				...topUsage,
				terms
			}]
		};
	}
	return combined;
}
function classifyOptionMeta(scanArgs, fullArgs, helpOptionNames, versionOptionNames, completionOptionNames) {
	let lastHelpVersion;
	for (let i = 0; i < scanArgs.length; i++) {
		const arg = scanArgs[i];
		if (helpOptionNames.includes(arg)) lastHelpVersion = {
			index: i,
			kind: "help"
		};
		else if (versionOptionNames.includes(arg)) lastHelpVersion = {
			index: i,
			kind: "version"
		};
		const equalsMatch = completionOptionNames.find((name) => arg.startsWith(name + "="));
		if (equalsMatch != null) return {
			lastHelpVersion,
			completion: {
				index: i,
				shell: arg.slice(equalsMatch.length + 1),
				args: fullArgs.slice(i + 1)
			}
		};
		if (completionOptionNames.includes(arg)) {
			const shell = scanArgs[i + 1] ?? "";
			return {
				lastHelpVersion,
				completion: {
					index: i,
					shell,
					args: shell === "" ? [] : fullArgs.slice(i + 2)
				}
			};
		}
	}
	return { lastHelpVersion };
}
function getHelpCommandContext(commandPath, args, helpIndex) {
	const commands = [...commandPath];
	for (let i = 0; i < helpIndex; i++) {
		const arg = args[i];
		if (!arg.startsWith("-")) commands.push(arg);
	}
	return commands;
}
function classifyParseFailure(failure, helpOptionNames, helpCommandNames, versionOptionNames, versionCommandNames, completionOptionNames, completionCommandNames) {
	if (failure.remainingArgs.length < 1) return {
		type: "error",
		error: failure.error
	};
	const hasConsumedPrefix = failure.consumedCount > 0;
	const firstArg = failure.remainingArgs[0];
	if (!hasConsumedPrefix && completionCommandNames.includes(firstArg)) {
		const secondArg = failure.remainingArgs[1];
		if (helpOptionNames.includes(secondArg)) return {
			type: "help",
			commands: [firstArg]
		};
		return {
			type: "completion",
			source: "command",
			shell: secondArg ?? "",
			args: failure.remainingArgs.slice(2)
		};
	}
	const optionArgs = failure.optionsTerminated ? [] : (() => {
		const terminatorIndex = failure.remainingArgs.indexOf("--");
		return terminatorIndex >= 0 ? failure.remainingArgs.slice(0, terminatorIndex) : failure.remainingArgs;
	})();
	const { lastHelpVersion, completion } = classifyOptionMeta(optionArgs, failure.remainingArgs, helpOptionNames, versionOptionNames, completionOptionNames);
	if (!hasConsumedPrefix && versionCommandNames.includes(firstArg)) {
		const secondArg = failure.remainingArgs[1];
		if (helpOptionNames.includes(secondArg)) return {
			type: "help",
			commands: [firstArg]
		};
		const isCompletionImmediatelyAfter = completion?.index === 1;
		if (secondArg == null || isCompletionImmediatelyAfter || lastHelpVersion?.index === 1 && lastHelpVersion.kind === "version") return { type: "version" };
		return {
			type: "error",
			error: failure.error
		};
	}
	let commandAction;
	if (!hasConsumedPrefix && helpCommandNames.includes(firstArg)) commandAction = {
		index: 0,
		kind: "help"
	};
	const winner = lastHelpVersion == null ? commandAction : commandAction == null || lastHelpVersion.index >= commandAction.index ? lastHelpVersion : commandAction;
	if (winner?.kind === "help") {
		if (winner === commandAction) return {
			type: "help",
			commands: failure.remainingArgs.slice(1)
		};
		return {
			type: "help",
			commands: getHelpCommandContext(failure.commandPath, failure.remainingArgs, winner.index),
			preferUserCommandDocs: failure.commandPath.length > 0
		};
	}
	if (winner?.kind === "version") return { type: "version" };
	if (completion != null) return {
		type: "completion",
		source: "option",
		shell: completion.shell,
		commandPath: failure.commandPath,
		args: completion.args
	};
	return {
		type: "error",
		error: failure.error
	};
}
/**
* Handles shell completion requests.
* @since 0.6.0
*/
function handleCompletion(completionArgs, programName, parser, completionParser, stdout, stderr, onCompletion, onError, availableShells, colors, maxWidth, completionCommandDisplayName, completionOptionDisplayName, isOptionMode, sectionOrder, rootOptionSuggestions = []) {
	const shellName = completionArgs[0] || "";
	const args = completionArgs.slice(1);
	const callOnError = (code) => onError(code);
	const callOnCompletion = (code) => onCompletion(code);
	if (!shellName) {
		stderr("Error: Missing shell name for completion.\n");
		if (completionParser) {
			const displayName = isOptionMode ? completionOptionDisplayName ?? "--completion" : completionCommandDisplayName ?? "completion";
			const doc = getDocPage(completionParser, [displayName]);
			if (doc) stderr(formatDocPage(programName, doc, {
				colors,
				maxWidth,
				sectionOrder
			}));
		}
		return dispatchByMode(parser.mode, () => {
			const result = callOnError(1);
			if (result instanceof Promise) throw new RunParserError("Synchronous parser returned async result.");
			return result;
		}, async () => callOnError(1));
	}
	const shell = availableShells[shellName];
	if (!shell) {
		const available = [];
		for (const name in availableShells) {
			if (available.length > 0) available.push(text(", "));
			available.push(value(name));
		}
		stderr(formatMessage(message`Error: Unsupported shell ${shellName}. Available shells: ${available}.`, {
			colors,
			quotes: !colors
		}));
		return dispatchByMode(parser.mode, () => {
			const result = callOnError(1);
			if (result instanceof Promise) throw new RunParserError("Synchronous parser returned async result.");
			return result;
		}, async () => callOnError(1));
	}
	if (args.length === 0) {
		const completionArg = isOptionMode ? completionOptionDisplayName ?? "--completion" : completionCommandDisplayName ?? "completion";
		const script = shell.generateScript(programName, [completionArg, shellName]);
		stdout(script);
		return dispatchByMode(parser.mode, () => {
			const result = callOnCompletion(0);
			if (result instanceof Promise) throw new RunParserError("Synchronous parser returned async result.");
			return result;
		}, async () => callOnCompletion(0));
	}
	return dispatchByMode(parser.mode, () => {
		const syncParser = parser;
		const completionSuggestions = getSyncCompletionSuggestions(syncParser, args, rootOptionSuggestions);
		const suggestions = withRootOptionSuggestions(completionSuggestions.suggestions, args, rootOptionSuggestions, completionSuggestions.argumentContext);
		for (const chunk of shell.encodeSuggestions(suggestions)) stdout(chunk);
		const result = callOnCompletion(0);
		if (result instanceof Promise) throw new RunParserError("Synchronous parser returned async result.");
		return result;
	}, async () => {
		const asyncParser = parser;
		const completionSuggestions = await getAsyncCompletionSuggestions(asyncParser, args, rootOptionSuggestions);
		for (const chunk of shell.encodeSuggestions(withRootOptionSuggestions(completionSuggestions.suggestions, args, rootOptionSuggestions, completionSuggestions.argumentContext))) stdout(chunk);
		return callOnCompletion(0);
	});
}
function withRootOptionSuggestions(suggestions, args, extraSuggestions, argumentContext) {
	if (extraSuggestions.length === 0) return suggestions;
	const prefix = args.at(-1) ?? "";
	if (argumentContext.optionsTerminated) return suggestions;
	if (argumentContext.completingOptionValue) return suggestions;
	if (argumentContext.completedRootOption) return suggestions;
	if (!(prefix === "" || prefix.startsWith("--") || prefix.startsWith("-") || prefix.startsWith("/") || prefix.startsWith("+"))) return suggestions;
	const seen = new Set(suggestions.flatMap((suggestion) => suggestion.kind === "literal" ? [suggestion.text] : []));
	const combined = [...suggestions];
	for (const suggestion of extraSuggestions) {
		if (prefix === "-" && suggestion.text.startsWith("--")) continue;
		if (prefix !== "" && !suggestion.text.startsWith(prefix)) continue;
		if (seen.has(suggestion.text)) continue;
		combined.push(suggestion);
		seen.add(suggestion.text);
	}
	return combined;
}
function getSyncCompletionSuggestions(parser, args, rootOptionSuggestions) {
	const prefix = args.at(-1) ?? "";
	let context = createCompletionArgumentParserContext(parser, args.slice(0, -1));
	while (context.buffer.length > 0) {
		const result = parser.parse(context);
		if (result instanceof Promise) throw new RunParserError("Synchronous parser returned async result.");
		if (!result.success) {
			const argumentContext = getFailedCompletionArgumentContext(result, context, parser, rootOptionSuggestions);
			return {
				suggestions: Array.from(parser.suggest(withCompletionSuggestRuntime(parser, context), prefix)),
				argumentContext
			};
		}
		const previousBuffer = context.buffer;
		context = result.next;
		if (isCompletionBufferUnchanged(previousBuffer, context.buffer)) return {
			suggestions: [],
			argumentContext: getCompletedCompletionArgumentContext(context)
		};
	}
	return {
		suggestions: Array.from(parser.suggest(withCompletionSuggestRuntime(parser, context), prefix)),
		argumentContext: getCompletedCompletionArgumentContext(context)
	};
}
async function getAsyncCompletionSuggestions(parser, args, rootOptionSuggestions) {
	const prefix = args.at(-1) ?? "";
	let context = createCompletionArgumentParserContext(parser, args.slice(0, -1));
	while (context.buffer.length > 0) {
		const result = await parser.parse(context);
		if (!result.success) {
			const argumentContext = getFailedCompletionArgumentContext(result, context, parser, rootOptionSuggestions);
			const suggestions$1 = [];
			const suggestContext$1 = await withCompletionSuggestRuntimeAsync(parser, context);
			for await (const suggestion of parser.suggest(suggestContext$1, prefix)) suggestions$1.push(suggestion);
			return {
				suggestions: suggestions$1,
				argumentContext
			};
		}
		const previousBuffer = context.buffer;
		context = result.next;
		if (isCompletionBufferUnchanged(previousBuffer, context.buffer)) return {
			suggestions: [],
			argumentContext: getCompletedCompletionArgumentContext(context)
		};
	}
	const suggestions = [];
	const suggestContext = await withCompletionSuggestRuntimeAsync(parser, context);
	for await (const suggestion of parser.suggest(suggestContext, prefix)) suggestions.push(suggestion);
	return {
		suggestions,
		argumentContext: getCompletedCompletionArgumentContext(context)
	};
}
function getCompletedCompletionArgumentContext(context) {
	return {
		completingOptionValue: false,
		completedRootOption: false,
		optionsTerminated: context.optionsTerminated
	};
}
function withCompletionSuggestRuntime(parser, context) {
	const runtime = createDependencyRuntimeContext();
	const nodes = getCompletionSuggestRuntimeNodes(parser, context.state, context.exec?.path ?? []);
	if (nodes.length > 0) collectExplicitSourceValues(nodes, runtime);
	return addCompletionSuggestRuntime(context, runtime);
}
async function withCompletionSuggestRuntimeAsync(parser, context) {
	const runtime = createDependencyRuntimeContext();
	const nodes = getCompletionSuggestRuntimeNodes(parser, context.state, context.exec?.path ?? []);
	if (nodes.length > 0) await collectExplicitSourceValuesAsync(nodes, runtime);
	return addCompletionSuggestRuntime(context, runtime);
}
function addCompletionSuggestRuntime(context, runtime) {
	return {
		...context,
		dependencyRegistry: runtime.registry,
		exec: context.exec ? {
			...context.exec,
			dependencyRuntime: runtime,
			dependencyRegistry: runtime.registry
		} : void 0
	};
}
function getCompletionSuggestRuntimeNodes(parser, state, path) {
	if (typeof parser.getSuggestRuntimeNodes === "function") return parser.getSuggestRuntimeNodes(state, path);
	if (parser.dependencyMetadata?.source == null) return [];
	return [{
		path,
		parser,
		state
	}];
}
function createCompletionArgumentParserContext(parser, args) {
	return createParserContext({
		buffer: args,
		state: parser.initialState,
		optionsTerminated: false
	}, {
		usage: parser.usage,
		phase: "suggest",
		path: [],
		trace: createInputTrace()
	});
}
function getFailedCompletionArgumentContext(result, context, parser, rootOptionSuggestions) {
	const activeValueOptionNames = collectActiveValueOptionNames(parser.usage, context.exec?.commandPath ?? [], result.consumed > 0, parser.leadingNames);
	const token = context.buffer[0];
	return {
		completingOptionValue: (result.consumed === 0 || result.consumed === 1) && activeValueOptionNames.has(token),
		completedRootOption: result.consumed === 0 && isRootOptionToken(token, rootOptionSuggestions),
		optionsTerminated: context.optionsTerminated
	};
}
function isCompletionBufferUnchanged(before, after) {
	return before.length === after.length && before.every((arg, index) => arg === after[index]);
}
function collectActiveValueOptionNames(usage, commandPath, includeDirectAfterCommandOptions, rootLeadingNames) {
	const optionNames = {
		value: /* @__PURE__ */ new Set(),
		flag: /* @__PURE__ */ new Set()
	};
	if (commandPath.length === 0) {
		collectRootOptionNames(usage, optionNames, rootLeadingNames);
		return optionNames.value;
	} else collectActiveOptionNames(usage, commandPath, optionNames, false, includeDirectAfterCommandOptions);
	const names = new Set(optionNames.value);
	for (const name of optionNames.flag) names.delete(name);
	return names;
}
function collectRootOptionNames(usage, names, rootLeadingNames) {
	for (const term of usage) collectRootOptionNamesFromTerm(term, names, rootLeadingNames);
}
function collectRootOptionNamesFromTerm(term, names, rootLeadingNames) {
	switch (term.type) {
		case "option":
			for (const name of term.names) {
				if (!rootLeadingNames.has(name)) continue;
				if (term.metavar != null) names.value.add(name);
				else names.flag.add(name);
			}
			return;
		case "optional":
		case "multiple":
		case "sequence":
			collectRootOptionNames(term.terms, names, rootLeadingNames);
			return;
		case "exclusive":
			for (const branch of term.terms) collectRootOptionNames(branch, names, rootLeadingNames);
			return;
		case "argument":
		case "command":
		case "literal":
		case "passthrough":
		case "ellipsis": return;
	}
}
function collectActiveOptionNames(usage, commandPath, names, fromExclusive, includeDirectAfterCommandOptions) {
	if (commandPath.length === 0) {
		collectOptionNamesAtCurrentCommandDepth(usage, names, false);
		return;
	}
	const [commandName, ...rest] = commandPath;
	for (let i = 0; i < usage.length; i++) {
		const term = usage[i];
		if (term.type === "command" && term.name === commandName) {
			const remainingUsage = usage.slice(i + 1);
			if (rest.length === 0) collectOptionNamesAtCurrentCommandDepth(remainingUsage, names, !fromExclusive && includeDirectAfterCommandOptions);
			else collectActiveOptionNames(remainingUsage, rest, names, fromExclusive, includeDirectAfterCommandOptions);
		} else if (term.type === "exclusive") for (const branch of term.terms) collectActiveOptionNames(branch, commandPath, names, true, includeDirectAfterCommandOptions);
		else if (term.type === "optional" || term.type === "multiple" || term.type === "sequence") collectActiveOptionNames(term.terms, commandPath, names, fromExclusive, includeDirectAfterCommandOptions);
	}
}
function isRootOptionToken(token, rootOptionSuggestions) {
	return token != null && rootOptionSuggestions.some((suggestion) => token === suggestion.text || token.startsWith(`${suggestion.text}=`));
}
function collectOptionNamesAtCurrentCommandDepth(usage, names, afterMatchedCommand) {
	for (const term of usage) if (collectOptionNamesAtCurrentCommandDepthFromTerm(term, names, afterMatchedCommand)) return;
}
function collectOptionNamesAtCurrentCommandDepthFromTerm(term, names, afterMatchedCommand) {
	switch (term.type) {
		case "command": return !afterMatchedCommand;
		case "option":
			if (term.metavar != null) for (const name of term.names) names.value.add(name);
			else for (const name of term.names) names.flag.add(name);
			return false;
		case "optional":
		case "multiple":
		case "sequence":
			collectOptionNamesAtCurrentCommandDepth(term.terms, names, afterMatchedCommand);
			return false;
		case "exclusive":
			for (const branch of term.terms) collectOptionNamesAtCurrentCommandDepth(branch, names, false);
			return false;
		case "argument":
		case "literal":
		case "passthrough":
		case "ellipsis": return false;
	}
}
function getRootMetaOptionSuggestions(helpOptionConfig, helpOptionNames, versionOptionConfig, versionOptionNames) {
	const suggestions = [];
	if (helpOptionConfig != null && !isSuggestionHidden(helpOptionConfig.hidden)) suggestions.push(...helpOptionNames.map((name) => ({
		kind: "literal",
		text: name
	})));
	if (versionOptionConfig != null && !isSuggestionHidden(versionOptionConfig.hidden)) suggestions.push(...versionOptionNames.map((name) => ({
		kind: "literal",
		text: name
	})));
	return suggestions;
}
/**
* Validates the configured version value.
*
* @param value Runtime version value from configuration.
* @returns The validated version string.
* @throws {TypeError} If the value is not a string, is empty, or contains
*         ASCII control characters.
*/
function validateVersionValue(value$1) {
	if (typeof value$1 !== "string") {
		const type = Array.isArray(value$1) ? "array" : typeof value$1;
		throw new TypeError(`Expected version value to be a string, but got ${type}.`);
	}
	if (value$1 === "") throw new TypeError("Version value must not be empty.");
	if (/[\x00-\x1f\x7f]/.test(value$1)) throw new TypeError("Version value must not contain control characters.");
	return value$1;
}
function runParser(parserOrProgram, programNameOrArgs, argsOrOptions, optionsParam) {
	const isProgram = typeof programNameOrArgs !== "string";
	let parser;
	let programName;
	let args;
	let options;
	if (isProgram) {
		const program = parserOrProgram;
		parser = program.parser;
		programName = program.metadata.name;
		args = programNameOrArgs;
		options = argsOrOptions ?? {};
		options = {
			...options,
			brief: options.brief ?? program.metadata.brief,
			description: options.description ?? program.metadata.description,
			examples: options.examples ?? program.metadata.examples,
			author: options.author ?? program.metadata.author,
			bugs: options.bugs ?? program.metadata.bugs,
			footer: options.footer ?? program.metadata.footer
		};
	} else {
		parser = parserOrProgram;
		programName = programNameOrArgs;
		args = argsOrOptions;
		options = optionsParam ?? {};
	}
	validateProgramName(programName);
	const { colors, maxWidth, showDefault, showChoices, sectionOrder, aboveError = "usage", onError = () => {
		throw new RunParserError("Failed to parse command line arguments.");
	}, stderr = console.error, stdout = console.log, brief, description, examples, author, bugs, footer } = options;
	const norm = (c) => c === true ? {} : c;
	const helpCommandConfig = norm(options.help?.command);
	const helpOptionConfig = norm(options.help?.option);
	const onHelp = options.help?.onShow ?? (() => ({}));
	const versionCommandConfig = norm(options.version?.command);
	const versionOptionConfig = norm(options.version?.option);
	const versionValue = options.version ? validateVersionValue(options.version.value) : void 0;
	const onVersion = options.version?.onShow ?? (() => ({}));
	const completionCommandConfig = norm(options.completion?.command);
	const completionOptionConfig = norm(options.completion?.option);
	const onCompletion = options.completion?.onShow ?? (() => ({}));
	const onCompletionResult = (code) => onCompletion(code);
	const onErrorResult = (code) => onError(code);
	if (helpOptionConfig?.names) validateOptionNames(helpOptionConfig.names, "Help option");
	if (helpCommandConfig?.names) validateCommandNames(helpCommandConfig.names, "Help command");
	if (versionOptionConfig?.names) validateOptionNames(versionOptionConfig.names, "Version option");
	if (versionCommandConfig?.names) validateCommandNames(versionCommandConfig.names, "Version command");
	if (completionOptionConfig?.names) validateOptionNames(completionOptionConfig.names, "Completion option");
	if (completionCommandConfig?.names) validateCommandNames(completionCommandConfig.names, "Completion command");
	const helpOptionNames = helpOptionConfig?.names ?? ["--help"];
	const helpCommandNames = helpCommandConfig?.names ?? ["help"];
	const versionOptionNames = versionOptionConfig?.names ?? ["--version"];
	const versionCommandNames = versionCommandConfig?.names ?? ["version"];
	const completionCommandNames = completionCommandConfig?.names ?? ["completion"];
	const completionOptionNames = completionOptionConfig?.names ?? ["--completion"];
	const rootOptionSuggestions = getRootMetaOptionSuggestions(helpOptionConfig, helpOptionNames, versionOptionConfig, versionOptionNames);
	const activeMetaEntries = [];
	if (options.help && helpOptionConfig) activeMetaEntries.push([
		"option",
		"help option",
		helpOptionNames
	]);
	if (options.help && helpCommandConfig) activeMetaEntries.push([
		"command",
		"help command",
		helpCommandNames
	]);
	if (options.version && versionOptionConfig) activeMetaEntries.push([
		"option",
		"version option",
		versionOptionNames
	]);
	if (options.version && versionCommandConfig) activeMetaEntries.push([
		"command",
		"version command",
		versionCommandNames
	]);
	if (options.completion && completionOptionConfig) activeMetaEntries.push([
		"option",
		"completion option",
		completionOptionNames,
		true
	]);
	if (options.completion && completionCommandConfig) activeMetaEntries.push([
		"command",
		"completion command",
		completionCommandNames
	]);
	validateMetaNameCollisions(activeMetaEntries);
	const defaultShells = {
		bash,
		fish,
		nu,
		pwsh,
		zsh
	};
	const availableShells = options.completion?.shells ? {
		...defaultShells,
		...options.completion.shells
	} : defaultShells;
	const helpParsers = options.help ? createHelpParser(helpCommandConfig, helpOptionConfig) : {
		helpCommand: null,
		helpOption: null
	};
	const versionParsers = options.version ? createVersionParser(versionCommandConfig, versionOptionConfig) : {
		versionCommand: null,
		versionOption: null
	};
	const completionParsers = options.completion ? createCompletionParser(programName, availableShells, completionCommandConfig, completionOptionConfig) : {
		completionCommand: null,
		completionOption: null
	};
	const augmentedParser = !options.help && !options.version && !options.completion ? parser : combineWithHelpVersion(parser, helpParsers, versionParsers, completionParsers, {
		helpCommandGroup: helpCommandConfig?.group,
		helpOptionGroup: helpOptionConfig?.group,
		versionCommandGroup: versionCommandConfig?.group,
		versionOptionGroup: versionOptionConfig?.group,
		completionCommandGroup: completionCommandConfig?.group,
		completionOptionGroup: completionOptionConfig?.group
	});
	const handleResult = (classified) => {
		switch (classified.type) {
			case "success": return classified.value;
			case "version":
				stdout(versionValue);
				return onVersion(0);
			case "completion": return handleCompletion([
				classified.shell,
				...classified.commandPath ?? [],
				...classified.args
			], programName, parser, classified.source === "command" ? completionParsers.completionCommand : completionParsers.completionOption, stdout, stderr, onCompletionResult, onErrorResult, availableShells, colors, maxWidth, completionCommandNames[0], completionOptionNames[0], classified.source === "option", sectionOrder, rootOptionSuggestions);
			case "help": {
				let helpGeneratorParser;
				let docGeneratorParser;
				const helpAsCommand = helpCommandConfig != null;
				const versionAsCommand = versionCommandConfig != null;
				const completionAsCommand = completionCommandConfig != null;
				const helpAsOption = helpOptionConfig != null;
				const versionAsOption = versionOptionConfig != null;
				const completionAsOption = completionOptionConfig != null;
				const requestedCommand = classified.commands[0];
				if (requestedCommand != null && !classified.preferUserCommandDocs && completionCommandNames.includes(requestedCommand) && completionAsCommand && completionParsers.completionCommand) {
					helpGeneratorParser = completionParsers.completionCommand;
					docGeneratorParser = helpGeneratorParser;
				} else if (requestedCommand != null && !classified.preferUserCommandDocs && helpCommandNames.includes(requestedCommand) && helpAsCommand && helpParsers.helpCommand) {
					helpGeneratorParser = helpParsers.helpCommand;
					docGeneratorParser = helpGeneratorParser;
				} else if (requestedCommand != null && !classified.preferUserCommandDocs && versionCommandNames.includes(requestedCommand) && versionAsCommand && versionParsers.versionCommand) {
					helpGeneratorParser = versionParsers.versionCommand;
					docGeneratorParser = helpGeneratorParser;
				} else {
					const commandParsers = [parser];
					const groupedMeta = {};
					const ungroupedMeta = [];
					const addMeta = (p, groupLabel) => {
						if (groupLabel) (groupedMeta[groupLabel] ??= []).push(p);
						else ungroupedMeta.push(p);
					};
					if (helpAsCommand && helpParsers.helpCommand) addMeta(helpParsers.helpCommand, helpCommandConfig?.group);
					if (versionAsCommand && versionParsers.versionCommand) addMeta(versionParsers.versionCommand, versionCommandConfig?.group);
					if (completionAsCommand && completionParsers.completionCommand) addMeta(completionParsers.completionCommand, completionCommandConfig?.group);
					commandParsers.push(...ungroupedMeta);
					for (const [label, parsers] of Object.entries(groupedMeta)) if (parsers.length === 1) commandParsers.push(group(label, parsers[0]));
					else commandParsers.push(group(label, longestMatch(...parsers)));
					const groupedMetaOptions = {};
					const ungroupedMetaOptions = [];
					const addMetaOption = (p, groupLabel) => {
						if (groupLabel) (groupedMetaOptions[groupLabel] ??= []).push(p);
						else ungroupedMetaOptions.push(p);
					};
					if (helpAsOption && helpParsers.helpOption) addMetaOption(helpParsers.helpOption, helpOptionConfig?.group);
					if (versionAsOption && versionParsers.versionOption) addMetaOption(versionParsers.versionOption, versionOptionConfig?.group);
					if (completionAsOption && completionParsers.completionOption) addMetaOption(completionParsers.completionOption, completionOptionConfig?.group);
					commandParsers.push(...ungroupedMetaOptions);
					for (const [label, optParsers] of Object.entries(groupedMetaOptions)) if (optParsers.length === 1) commandParsers.push(group(label, optParsers[0]));
					else commandParsers.push(group(label, longestMatch(...optParsers)));
					if (commandParsers.length === 1) helpGeneratorParser = commandParsers[0];
					else helpGeneratorParser = longestMatchForMetaCommands(...commandParsers);
					docGeneratorParser = classified.commands.length > 0 ? parser : helpGeneratorParser;
				}
				const reportInvalidHelpCommand = (validationError) => {
					stderr(`Usage: ${indentLines(formatUsage(programName, augmentedParser.usage, {
						colors,
						maxWidth: maxWidth == null ? void 0 : maxWidth - 7,
						expandCommands: true
					}), 7)}`);
					const errorMessage = formatMessage(validationError, {
						colors,
						quotes: !colors
					});
					stderr(`Error: ${errorMessage}`);
					return onError(1);
				};
				const displayHelp = (doc) => {
					if (doc != null) {
						const isMetaCommandHelp = requestedCommand != null && completionCommandNames.includes(requestedCommand) || requestedCommand != null && helpCommandNames.includes(requestedCommand) || requestedCommand != null && versionCommandNames.includes(requestedCommand);
						const isSubcommandHelp = classified.commands.length > 0;
						const isTopLevel = !isSubcommandHelp;
						const shouldOverride = !isMetaCommandHelp && !isSubcommandHelp;
						const augmentedDoc = {
							...doc,
							brief: shouldOverride ? brief ?? doc.brief : doc.brief,
							description: shouldOverride ? description ?? doc.description : doc.description,
							examples: isTopLevel && !isMetaCommandHelp ? examples ?? doc.examples : void 0,
							author: isTopLevel && !isMetaCommandHelp ? author ?? doc.author : void 0,
							bugs: isTopLevel && !isMetaCommandHelp ? bugs ?? doc.bugs : void 0,
							footer: shouldOverride ? footer ?? doc.footer : doc.footer ?? footer
						};
						stdout(formatDocPage(programName, augmentedDoc, {
							colors,
							maxWidth,
							showDefault,
							showChoices,
							sectionOrder
						}));
					}
					return onHelp(0);
				};
				if (classified.commands.length > 0) {
					let validationContext = {
						buffer: [...classified.commands],
						optionsTerminated: false,
						state: helpGeneratorParser.initialState,
						usage: helpGeneratorParser.usage
					};
					const processStep = (stepResult) => {
						if (!stepResult.success) return stepResult.error;
						if (stepResult.consumed.length < 1) return message`Unexpected option or subcommand: ${optionName(validationContext.buffer[0])}.`;
						validationContext = {
							...validationContext,
							buffer: stepResult.next.buffer,
							optionsTerminated: stepResult.next.optionsTerminated,
							state: stepResult.next.state,
							usage: stepResult.next.usage ?? validationContext.usage
						};
						return validationContext.buffer.length > 0 ? "continue" : null;
					};
					let validationResult = "continue";
					while (validationResult === "continue") {
						const stepResult = helpGeneratorParser.parse(validationContext);
						if (stepResult instanceof Promise) {
							const asyncValidate = async (result) => {
								let res = processStep(result);
								while (res === "continue") {
									const next = helpGeneratorParser.parse(validationContext);
									const resolved = next instanceof Promise ? await next : next;
									res = processStep(resolved);
								}
								if (res != null) return reportInvalidHelpCommand(res);
								const docOrPromise$1 = getDocPage(docGeneratorParser, classified.commands);
								return docOrPromise$1 instanceof Promise ? docOrPromise$1.then(displayHelp) : displayHelp(docOrPromise$1);
							};
							return stepResult.then(asyncValidate);
						}
						validationResult = processStep(stepResult);
					}
					if (validationResult != null) return reportInvalidHelpCommand(validationResult);
				}
				const docOrPromise = getDocPage(docGeneratorParser, classified.commands);
				if (docOrPromise instanceof Promise) return docOrPromise.then(displayHelp);
				return displayHelp(docOrPromise);
			}
			case "error": {
				const displayError = (doc, currentAboveError) => {
					let effectiveAboveError = currentAboveError;
					if (effectiveAboveError === "help") if (doc == null) effectiveAboveError = "usage";
					else {
						const augmentedDoc = {
							...doc,
							brief: brief ?? doc.brief,
							description: description ?? doc.description,
							examples: examples ?? doc.examples,
							author: author ?? doc.author,
							bugs: bugs ?? doc.bugs,
							footer: footer ?? doc.footer
						};
						stderr(formatDocPage(programName, augmentedDoc, {
							colors,
							maxWidth,
							showDefault,
							showChoices
						}));
					}
					if (effectiveAboveError === "usage") stderr(`Usage: ${indentLines(formatUsage(programName, augmentedParser.usage, {
						colors,
						maxWidth: maxWidth == null ? void 0 : maxWidth - 7,
						expandCommands: true
					}), 7)}`);
					const errorMessage = formatMessage(classified.error, {
						colors,
						quotes: !colors
					});
					stderr(`Error: ${errorMessage}`);
					return onError(1);
				};
				if (aboveError === "help") {
					const parserForDoc = args.length < 1 ? augmentedParser : parser;
					const docOrPromise = getDocPage(parserForDoc, args);
					if (docOrPromise instanceof Promise) return docOrPromise.then((doc) => displayError(doc, aboveError));
					return displayError(docOrPromise, aboveError);
				}
				return displayError(void 0, aboveError);
			}
			default: throw new RunParserError("Unexpected parse result type");
		}
	};
	const parserMode = parser.mode;
	return dispatchByMode(parserMode, () => {
		const attempted = attemptParseSync(parser, args);
		const classified = attempted.kind === "success" ? {
			type: "success",
			value: attempted.value
		} : classifyParseFailure(attempted, helpOptionConfig ? [...helpOptionNames] : [], helpCommandConfig ? [...helpCommandNames] : [], versionOptionConfig ? [...versionOptionNames] : [], versionCommandConfig ? [...versionCommandNames] : [], completionOptionConfig ? [...completionOptionNames] : [], completionCommandConfig ? [...completionCommandNames] : []);
		const handled = handleResult(classified);
		if (handled instanceof Promise) throw new RunParserError("Synchronous parser returned async result.");
		return handled;
	}, async () => {
		const attempted = await attemptParseAsync(parser, args);
		const classified = attempted.kind === "success" ? {
			type: "success",
			value: attempted.value
		} : classifyParseFailure(attempted, helpOptionConfig ? [...helpOptionNames] : [], helpCommandConfig ? [...helpCommandNames] : [], versionOptionConfig ? [...versionOptionNames] : [], versionCommandConfig ? [...versionCommandNames] : [], completionOptionConfig ? [...completionOptionNames] : [], completionCommandConfig ? [...completionCommandNames] : []);
		const handled = handleResult(classified);
		return handled instanceof Promise ? await handled : handled;
	});
}
/**
* Runs a synchronous command-line parser with the given options.
*
* This is a type-safe version of {@link runParser} that only accepts sync
* parsers. Use this when you know your parser is sync-only to get direct
* return values without Promise wrappers.
*
* @template TParser The sync parser type being executed.
* @template THelp The return type of the onHelp callback.
* @template TError The return type of the onError callback.
* @param parser The synchronous command-line parser to execute.
* @param programName The name of the program for help messages.
* @param args The command-line arguments to parse.
* @param options Configuration options for customizing behavior.
* @returns The parsed result if successful.
* @throws {TypeError} If an async parser is passed at runtime.  Use
* {@link runParser} or {@link runParserAsync} for async parsers.
* @since 0.9.0
*/
function runParserSync(parser, programName, args, options) {
	if (parser.mode !== "sync") throw new TypeError("Cannot use an async parser with runParserSync(). Use runParser() or runParserAsync() instead.");
	return runParser(parser, programName, args, options);
}
/**
* Runs any command-line parser asynchronously with the given options.
*
* This function accepts parsers of any mode (sync or async) and always
* returns a Promise. Use this when working with parsers that may contain
* async value parsers.
*
* @template TParser The parser type being executed.
* @template THelp The return type of the onHelp callback.
* @template TError The return type of the onError callback.
* @param parser The command-line parser to execute.
* @param programName The name of the program for help messages.
* @param args The command-line arguments to parse.
* @param options Configuration options for customizing behavior.
* @returns A Promise of the parsed result if successful.
* @since 0.9.0
*/
function runParserAsync(parser, programName, args, options) {
	const result = runParser(parser, programName, args, options);
	return Promise.resolve(result);
}
/**
* An error class used to indicate that the command line arguments
* could not be parsed successfully.
*/
var RunParserError = class extends Error {
	constructor(message$1) {
		super(message$1);
		this.name = "RunParserError";
	}
};
function indentLines(text$1, indent) {
	return text$1.split("\n").join("\n" + " ".repeat(indent));
}
function isMetaEarlyExit(classified) {
	return classified.type === "help" || classified.type === "version" || classified.type === "completion";
}
function classifyEarlyExitFailure(failure, options) {
	const norm = (c) => c === true ? {} : c;
	const helpOptionConfig = norm(options.help?.option);
	const helpCommandConfig = norm(options.help?.command);
	const versionOptionConfig = norm(options.version?.option);
	const versionCommandConfig = norm(options.version?.command);
	const completionOptionConfig = norm(options.completion?.option);
	const completionCommandConfig = norm(options.completion?.command);
	return classifyParseFailure(failure, helpOptionConfig ? [...helpOptionConfig.names ?? ["--help"]] : [], helpCommandConfig ? [...helpCommandConfig.names ?? ["help"]] : [], versionOptionConfig ? [...versionOptionConfig.names ?? ["--version"]] : [], versionCommandConfig ? [...versionCommandConfig.names ?? ["version"]] : [], completionOptionConfig ? [...completionOptionConfig.names ?? ["--completion"]] : [], completionCommandConfig ? [...completionCommandConfig.names ?? ["completion"]] : []);
}
function shouldProbeEarlyExit(options, needsTwoPhase) {
	return needsTwoPhase && (options.help != null || options.version != null || options.completion != null);
}
/**
* Checks if the arguments contain parser-visible meta requests that can be
* handled without collecting source-context annotations first.
*/
function needsEarlyExitSync(parser, args, options) {
	const attempted = attemptParseSync(parser, args, "parse-only");
	if (attempted.kind === "success") return false;
	return isMetaEarlyExit(classifyEarlyExitFailure(attempted, options));
}
/**
* Async variant of {@link needsEarlyExitSync}.
*/
async function needsEarlyExitAsync(parser, args, options) {
	const attempted = await attemptParseAsync(parser, args, "parse-only");
	if (attempted.kind === "success") return false;
	return isMetaEarlyExit(classifyEarlyExitFailure(attempted, options));
}
/**
* Merges multiple annotation objects, with earlier contexts having priority.
*
* When the same symbol key exists in multiple annotations, the value from
* the earlier context (lower index in the array) takes precedence.
*
* @param annotationsList Array of annotations to merge.
* @returns Merged annotations object.
*/
function mergeAnnotations(annotationsList) {
	const result = {};
	for (let i = annotationsList.length - 1; i >= 0; i--) {
		const annotations = annotationsList[i];
		for (const key of Object.getOwnPropertySymbols(annotations)) result[key] = annotations[key];
	}
	return result;
}
function validateContextPhases(contexts) {
	for (const context of contexts) {
		const phase = context.phase;
		if (phase !== "single-pass" && phase !== "two-pass") throw new TypeError(`Context ${String(context.id)} must declare phase as "single-pass" or "two-pass".`);
	}
}
/**
* Collects phase 1 annotations from all contexts and determines whether
* two-phase parsing is needed.
*
* @param contexts Source contexts to collect annotations from.
* @param options Optional context-required options to pass to each context.
* @returns Promise with merged annotations, per-context snapshots, and a
* two-phase hint.
*/
async function collectPhase1Annotations(contexts, options) {
	const annotationsList = [];
	let snapshots;
	for (const context of contexts) {
		const request = { phase: "phase1" };
		const result = context.getAnnotations(request, options);
		const annotations = result instanceof Promise ? await result : result;
		const internalAnnotations = context.getInternalAnnotations?.(request, annotations);
		const snapshot = internalAnnotations == null ? annotations : mergeAnnotations([annotations, internalAnnotations]);
		annotationsList.push(snapshot);
		if (snapshots != null) snapshots.push(snapshot);
		else if (context.phase === "two-pass") snapshots = [...annotationsList];
	}
	return {
		annotations: mergeAnnotations(annotationsList),
		needsTwoPhase: snapshots != null,
		snapshots: snapshots ?? []
	};
}
/**
* Collects final annotations from all contexts.
*
* `single-pass` contexts reuse their phase-1 snapshot. `two-pass` contexts
* are recollected with the parsed value and replace their own phase-1
* snapshot in the final merge.
*
* @param contexts Source contexts to collect annotations from.
* @param phase1Snapshots Per-context snapshots collected during phase 1.
* @param parsed Optional parsed result from a previous parse pass.
* @param options Optional context-required options to pass to each context.
* @returns Promise that resolves to merged annotations.
*/
async function collectFinalAnnotations(contexts, phase1Snapshots, parsed, options, deferred, deferredKeys) {
	const annotationsList = [];
	const preparedParsed = prepareParsedForContexts(parsed, deferred, deferredKeys);
	for (let index = 0; index < contexts.length; index++) {
		const context = contexts[index];
		if (context.phase === "single-pass") {
			annotationsList.push(phase1Snapshots[index]);
			continue;
		}
		const request = {
			phase: "phase2",
			parsed: preparedParsed
		};
		const result = context.getAnnotations(request, options);
		const annotations = result instanceof Promise ? await result : result;
		const internalAnnotations = context.getInternalAnnotations?.(request, annotations);
		const mergedAnnotations = internalAnnotations == null ? annotations : mergeAnnotations([annotations, internalAnnotations]);
		annotationsList.push(mergedAnnotations);
	}
	return { annotations: mergeAnnotations(annotationsList) };
}
/**
* Collects phase 1 annotations from all contexts synchronously and determines
* whether two-phase parsing is needed.
*
* @param contexts Source contexts to collect annotations from.
* @param options Optional context-required options to pass to each context.
* @returns Merged annotations, per-context snapshots, and a two-phase hint.
* @throws {TypeError} If any context returns a Promise.
*/
function collectPhase1AnnotationsSync(contexts, options) {
	const annotationsList = [];
	let snapshots;
	for (const context of contexts) {
		const request = { phase: "phase1" };
		const result = context.getAnnotations(request, options);
		if (result instanceof Promise) throw new TypeError(`Context ${String(context.id)} returned a Promise in sync mode. Use runWith() or runWithAsync() for async contexts.`);
		const internalAnnotations = context.getInternalAnnotations?.(request, result);
		const snapshot = internalAnnotations == null ? result : mergeAnnotations([result, internalAnnotations]);
		annotationsList.push(snapshot);
		if (snapshots != null) snapshots.push(snapshot);
		else if (context.phase === "two-pass") snapshots = [...annotationsList];
	}
	return {
		annotations: mergeAnnotations(annotationsList),
		needsTwoPhase: snapshots != null,
		snapshots: snapshots ?? []
	};
}
/**
* Collects final annotations from all contexts synchronously.
*
* `single-pass` contexts reuse their phase-1 snapshot. `two-pass` contexts
* are recollected with the parsed value and replace their own phase-1
* snapshot in the final merge.
*
* @param contexts Source contexts to collect annotations from.
* @param phase1Snapshots Per-context snapshots collected during phase 1.
* @param parsed Optional parsed result from a previous parse pass.
* @param options Optional context-required options to pass to each context.
* @returns Merged annotations.
* @throws {TypeError} If any context returns a Promise.
*/
function collectFinalAnnotationsSync(contexts, phase1Snapshots, parsed, options, deferred, deferredKeys) {
	const annotationsList = [];
	const preparedParsed = prepareParsedForContexts(parsed, deferred, deferredKeys);
	for (let index = 0; index < contexts.length; index++) {
		const context = contexts[index];
		if (context.phase === "single-pass") {
			annotationsList.push(phase1Snapshots[index]);
			continue;
		}
		const request = {
			phase: "phase2",
			parsed: preparedParsed
		};
		const result = context.getAnnotations(request, options);
		if (result instanceof Promise) throw new TypeError(`Context ${String(context.id)} returned a Promise in sync mode. Use runWith() or runWithAsync() for async contexts.`);
		const internalAnnotations = context.getInternalAnnotations?.(request, result);
		const mergedAnnotations = internalAnnotations == null ? result : mergeAnnotations([result, internalAnnotations]);
		annotationsList.push(mergedAnnotations);
	}
	return { annotations: mergeAnnotations(annotationsList) };
}
/**
* Disposes all contexts that implement `AsyncDisposable` or `Disposable`.
* Prefers `[Symbol.asyncDispose]` over `[Symbol.dispose]`.
*
* @param contexts Source contexts to dispose.
*/
async function disposeContexts(contexts) {
	const errors = [];
	for (const context of contexts) try {
		if (Symbol.asyncDispose in context && typeof context[Symbol.asyncDispose] === "function") await context[Symbol.asyncDispose]();
		else if (Symbol.dispose in context && typeof context[Symbol.dispose] === "function") context[Symbol.dispose]();
	} catch (error) {
		errors.push(error);
	}
	if (errors.length === 1) throw errors[0];
	if (errors.length > 1) throw new AggregateError(errors, "Failed to dispose one or more source contexts.");
}
/**
* Disposes all contexts that implement `Disposable` synchronously.
* Falls back to `[Symbol.asyncDispose]` when it completes synchronously.
*
* @param contexts Source contexts to dispose.
*/
function disposeContextsSync(contexts) {
	const errors = [];
	for (const context of contexts) try {
		if (Symbol.dispose in context && typeof context[Symbol.dispose] === "function") context[Symbol.dispose]();
		else if (Symbol.asyncDispose in context && typeof context[Symbol.asyncDispose] === "function") {
			const result = context[Symbol.asyncDispose]();
			if (typeof result === "object" && result !== null && "then" in result && typeof result.then === "function") throw new TypeError(`Context ${String(context.id)} returned a Promise from Symbol.asyncDispose in sync mode. Use runWith() or runWithAsync() for async disposal.`);
		}
	} catch (error) {
		errors.push(error);
	}
	if (errors.length === 1) throw errors[0];
	if (errors.length > 1) throw new AggregateError(errors, "Failed to dispose one or more source contexts.");
}
/**
* Body of {@link runWith}, extracted so that the caller can handle
* disposal outside a `finally` block (avoiding `no-unsafe-finally` lint).
*/
async function runWithBody(parser, programName, contexts, args, options) {
	validateContextIds(contexts);
	validateContextPhases(contexts);
	const ctxOptions = options.contextOptions;
	const { annotations: phase1Annotations, needsTwoPhase, snapshots: phase1Snapshots } = await collectPhase1Annotations(contexts, ctxOptions);
	if (shouldProbeEarlyExit(options, needsTwoPhase)) {
		const earlyExitParser = injectAnnotationsIntoParser(parser, phase1Annotations);
		if (await needsEarlyExitAsync(earlyExitParser, args, options)) {
			if (parser.mode === "async") return runParser(earlyExitParser, programName, args, options);
			return Promise.resolve(runParser(earlyExitParser, programName, args, options));
		}
	}
	const augmentedParser1 = injectAnnotationsIntoParser(parser, phase1Annotations);
	if (!needsTwoPhase) {
		if (parser.mode === "async") return runParser(augmentedParser1, programName, args, options);
		return Promise.resolve(runParser(augmentedParser1, programName, args, options));
	}
	const firstPassSeed = await dispatchByMode(parser.mode, () => extractPhase2SeedSync(augmentedParser1, args), () => extractPhase2SeedAsync(augmentedParser1, args));
	if (firstPassSeed == null) {
		const fallbackParser = injectAnnotationsIntoParser(parser, phase1Annotations);
		if (parser.mode === "async") return runParser(fallbackParser, programName, args, options);
		return Promise.resolve(runParser(fallbackParser, programName, args, options));
	}
	const { annotations: finalAnnotations } = await collectFinalAnnotations(contexts, phase1Snapshots, firstPassSeed.value, ctxOptions, firstPassSeed.deferred, firstPassSeed.deferredKeys);
	const augmentedParser2 = injectAnnotationsIntoParser(parser, finalAnnotations);
	if (parser.mode === "async") return runParser(augmentedParser2, programName, args, options);
	return Promise.resolve(runParser(augmentedParser2, programName, args, options));
}
/**
* Runs a parser with multiple source contexts.
*
* This function automatically handles single-pass and two-pass contexts with
* proper priority. Earlier contexts in the array override later ones.
*
* The function uses a smart two-phase approach:
*
* 1. *Phase 1*: Collect annotations from all contexts.
* 2. *First parse*: Parse with Phase 1 annotations. If that pass finishes
*    successfully, its value becomes the phase-two input. If the parser
*    reaches a usable intermediate state but still does not complete
*    successfully, the runner extracts a best-effort seed from that state
*    instead.
* 3. *Phase 2*: Call `getAnnotations({ phase: "phase2", parsed })` on all
*    two-pass contexts with the first pass value. Deferred or otherwise
*    unresolved fields in `parsed` may be `undefined`. Each two-pass
*    context's phase-two return
*    value replaces its own phase-one contribution for the final parse, so
*    returning `{}` clears any annotations that context provided during
*    phase 1. Single-pass contexts reuse their phase-one snapshot.
* 4. *Second parse*: Parse again with the merged phase-two annotations.
*
* If all contexts are single-pass, the second parse is skipped for
* optimization. Phase 2 is also skipped when the first pass does not yield
* any usable seed at all.
*
* @template TParser The parser type.
* @template THelp Return type when help is shown.
* @template TError Return type when an error occurs.
* @param parser The parser to execute.
* @param programName Name of the program for help/error output.
* @param contexts Source contexts to use (priority: earlier overrides later).
* @param options Run options including args, help, version, etc.
* @returns Promise that resolves to the parsed result.
* @throws {TypeError} If two or more contexts share the same
* {@link SourceContext.id}.
* @throws {TypeError} If any context omits `phase` or declares an invalid
* phase value.
* @throws {SuppressedError} If the runner throws and a context's disposal
* also throws.  The original error is available via `.suppressed` and the
* disposal error via `.error`.
* @since 0.10.0
*
* @example
* ```typescript
* import { runWith } from "@optique/core/facade";
* import type { SourceContext } from "@optique/core/context";
*
* const envContext: SourceContext = {
*   id: Symbol.for("@myapp/env"),
*   phase: "single-pass",
*   getAnnotations() {
*     return { [Symbol.for("@myapp/env")]: process.env };
*   }
* };
*
* const result = await runWith(
*   parser,
*   "myapp",
*   [envContext],
*   { args: process.argv.slice(2) }
* );
* ```
*/
async function runWith(parser, programName, contexts, options) {
	const args = options?.args ?? [];
	if (contexts.length === 0) {
		if (parser.mode === "async") return runParser(parser, programName, args, options);
		return Promise.resolve(runParser(parser, programName, args, options));
	}
	let result;
	let primaryError;
	let hasPrimaryError = false;
	try {
		result = await runWithBody(parser, programName, contexts, args, options);
	} catch (error) {
		hasPrimaryError = true;
		primaryError = error;
	}
	try {
		await disposeContexts(contexts);
	} catch (disposeError) {
		if (hasPrimaryError) throw new SuppressedErrorCtor(disposeError, primaryError, "An error was suppressed during context disposal.");
		throw disposeError;
	}
	if (hasPrimaryError) throw primaryError;
	return result;
}
/**
* Body of {@link runWithSync}, extracted so that the caller can handle
* disposal outside a `finally` block (avoiding `no-unsafe-finally` lint).
*/
function runWithSyncBody(parser, programName, contexts, args, options) {
	validateContextIds(contexts);
	validateContextPhases(contexts);
	const ctxOptions = options.contextOptions;
	const { annotations: phase1Annotations, needsTwoPhase, snapshots: phase1Snapshots } = collectPhase1AnnotationsSync(contexts, ctxOptions);
	if (shouldProbeEarlyExit(options, needsTwoPhase)) {
		const earlyExitParser = injectAnnotationsIntoParser(parser, phase1Annotations);
		if (needsEarlyExitSync(earlyExitParser, args, options)) return runParser(earlyExitParser, programName, args, options);
	}
	const augmentedParser1 = injectAnnotationsIntoParser(parser, phase1Annotations);
	if (!needsTwoPhase) return runParser(augmentedParser1, programName, args, options);
	const firstPassSeed = extractPhase2SeedSync(augmentedParser1, args);
	if (firstPassSeed == null) {
		const fallbackParser = injectAnnotationsIntoParser(parser, phase1Annotations);
		return runParser(fallbackParser, programName, args, options);
	}
	const { annotations: finalAnnotations } = collectFinalAnnotationsSync(contexts, phase1Snapshots, firstPassSeed.value, ctxOptions, firstPassSeed.deferred, firstPassSeed.deferredKeys);
	const augmentedParser2 = injectAnnotationsIntoParser(parser, finalAnnotations);
	return runParser(augmentedParser2, programName, args, options);
}
/**
* Runs a synchronous parser with multiple source contexts.
*
* This is the sync-only variant of {@link runWith}. All contexts must return
* annotations synchronously (not Promises). It uses the same two-phase
* best-effort seed extraction as {@link runWith} when two-pass contexts are
* present. In two-phase runs, each two-pass context's phase-two return value
* replaces that context's phase-one contribution for the final parse, so
* returning `{}` clears any annotations that context provided during phase 1.
*
* @template TParser The sync parser type.
* @template THelp Return type when help is shown.
* @template TError Return type when an error occurs.
* @param parser The synchronous parser to execute.
* @param programName Name of the program for help/error output.
* @param contexts Source contexts to use (priority: earlier overrides later).
* @param options Run options including args, help, version, etc.
* @returns The parsed result.
* @throws {TypeError} If an async parser is passed at runtime.  Use
* {@link runWith} or {@link runWithAsync} for async parsers.
* @throws {TypeError} If two or more contexts share the same
* {@link SourceContext.id}.
* @throws {TypeError} If any context omits `phase` or declares an invalid
* phase value.
* @throws {TypeError} If any context returns a Promise or if a context's
* `[Symbol.asyncDispose]` returns a Promise.
* @throws {SuppressedError} If the runner throws and a context's disposal
* also throws.  The original error is available via `.suppressed` and the
* disposal error via `.error`.
* @since 0.10.0
*/
function runWithSync(parser, programName, contexts, options) {
	if (parser.mode !== "sync") throw new TypeError("Cannot use an async parser with runWithSync(). Use runWith() or runWithAsync() instead.");
	const args = options?.args ?? [];
	if (contexts.length === 0) return runParser(parser, programName, args, options);
	let result;
	let primaryError;
	let hasPrimaryError = false;
	try {
		result = runWithSyncBody(parser, programName, contexts, args, options);
	} catch (error) {
		hasPrimaryError = true;
		primaryError = error;
	}
	try {
		disposeContextsSync(contexts);
	} catch (disposeError) {
		if (hasPrimaryError) throw new SuppressedErrorCtor(disposeError, primaryError, "An error was suppressed during context disposal.");
		throw disposeError;
	}
	if (hasPrimaryError) throw primaryError;
	return result;
}
/**
* Runs any parser asynchronously with multiple source contexts.
*
* This function accepts parsers of any mode (sync or async) and always
* returns a Promise. Use this when working with async contexts or parsers.
*
* @template TParser The parser type.
* @template THelp Return type when help is shown.
* @template TError Return type when an error occurs.
* @param parser The parser to execute.
* @param programName Name of the program for help/error output.
* @param contexts Source contexts to use (priority: earlier overrides later).
* @param options Run options including args, help, version, etc.
* @returns Promise that resolves to the parsed result.
* @throws {TypeError} If two or more contexts share the same
* {@link SourceContext.id}.
* @throws {SuppressedError} If the runner throws and a context's disposal
* also throws.  The original error is available via `.suppressed` and the
* disposal error via `.error`.
* @since 0.10.0
*/
function runWithAsync(parser, programName, contexts, options) {
	return runWith(parser, programName, contexts, options);
}
/**
* Creates a new parser with annotations injected into its initial state.
*
* @param parser The original parser.
* @param annotations Annotations to inject.
* @returns A new parser with annotations in its initial state.
*/
function injectAnnotationsIntoParser(parser, annotations) {
	const newInitialState = injectAnnotations(parser.initialState, annotations);
	const descriptors = { ...Object.getOwnPropertyDescriptors(parser) };
	const initialState = descriptors.initialState;
	descriptors.initialState = initialState == null ? {
		value: newInitialState,
		writable: true,
		enumerable: true,
		configurable: true
	} : "get" in initialState || "set" in initialState ? {
		value: newInitialState,
		writable: true,
		enumerable: initialState.enumerable ?? true,
		configurable: initialState.configurable ?? true
	} : {
		value: newInitialState,
		writable: initialState.writable ?? true,
		enumerable: initialState.enumerable ?? true,
		configurable: initialState.configurable ?? true
	};
	return Object.create(Object.getPrototypeOf(parser), descriptors);
}

//#endregion
export { RunParserError, runParser, runParserAsync, runParserSync, runWith, runWithAsync, runWithSync };
