// @optique/core/dist/index.cjs
const require_internal_annotations = require('./internal/annotations.cjs');
const require_message = require('./message.cjs');
const require_usage = require('./usage.cjs');
const require_doc = require('./doc.cjs');
const require_internal_dependency = require('./internal/dependency.cjs');
const require_internal_parser = require('./internal/parser.cjs');
const require_completion = require('./completion.cjs');
const require_constructs = require('./constructs.cjs');
const require_modifiers = require('./modifiers.cjs');
const require_nonempty = require('./nonempty.cjs');
const require_valueparser = require('./valueparser.cjs');
const require_primitives = require('./primitives.cjs');
const require_facade = require('./facade.cjs');

exports.DuplicateOptionError = require_constructs.DuplicateOptionError;
exports.RunParserError = require_facade.RunParserError;
exports.WithDefaultError = require_modifiers.WithDefaultError;
exports.argument = require_primitives.argument;
exports.bash = require_completion.bash;
exports.checkBooleanOption = require_valueparser.checkBooleanOption;
exports.checkEnumOption = require_valueparser.checkEnumOption;
exports.choice = require_valueparser.choice;
exports.cidr = require_valueparser.cidr;
exports.cloneDocEntry = require_doc.cloneDocEntry;
exports.cloneUsage = require_usage.cloneUsage;
exports.cloneUsageTerm = require_usage.cloneUsageTerm;
exports.color = require_valueparser.color;
exports.command = require_primitives.command;
exports.commandLine = require_message.commandLine;
exports.concat = require_constructs.concat;
exports.conditional = require_constructs.conditional;
exports.constant = require_primitives.constant;
exports.createParserContext = require_internal_parser.createParserContext;
exports.deduplicateDocEntries = require_doc.deduplicateDocEntries;
exports.deduplicateDocFragments = require_doc.deduplicateDocFragments;
exports.dependency = require_internal_dependency.dependency;
exports.deriveFrom = require_internal_dependency.deriveFrom;
exports.deriveFromAsync = require_internal_dependency.deriveFromAsync;
exports.deriveFromSync = require_internal_dependency.deriveFromSync;
exports.domain = require_valueparser.domain;
exports.email = require_valueparser.email;
exports.ensureNonEmptyString = require_nonempty.ensureNonEmptyString;
exports.envVar = require_message.envVar;
exports.extractArgumentMetavars = require_usage.extractArgumentMetavars;
exports.extractCommandNames = require_usage.extractCommandNames;
exports.extractLiteralValues = require_usage.extractLiteralValues;
exports.extractOptionNames = require_usage.extractOptionNames;
exports.fail = require_primitives.fail;
exports.fileSize = require_valueparser.fileSize;
exports.firstOf = require_valueparser.firstOf;
exports.fish = require_completion.fish;
exports.flag = require_primitives.flag;
exports.float = require_valueparser.float;
exports.formatDocPage = require_doc.formatDocPage;
exports.formatMessage = require_message.formatMessage;
exports.formatUsage = require_usage.formatUsage;
exports.formatUsageTerm = require_usage.formatUsageTerm;
exports.getAnnotations = require_internal_annotations.getAnnotations;
exports.getDocPage = require_internal_parser.getDocPage;
exports.getDocPageAsync = require_internal_parser.getDocPageAsync;
exports.getDocPageSync = require_internal_parser.getDocPageSync;
exports.group = require_constructs.group;
exports.hostname = require_valueparser.hostname;
exports.integer = require_valueparser.integer;
exports.ip = require_valueparser.ip;
exports.ipv4 = require_valueparser.ipv4;
exports.ipv6 = require_valueparser.ipv6;
exports.isDependencySource = require_internal_dependency.isDependencySource;
exports.isDerivedValueParser = require_internal_dependency.isDerivedValueParser;
exports.isDocEntryHidden = require_doc.isDocEntryHidden;
exports.isDocHidden = require_usage.isDocHidden;
exports.isNonEmptyString = require_nonempty.isNonEmptyString;
exports.isSuggestionHidden = require_usage.isSuggestionHidden;
exports.isUsageHidden = require_usage.isUsageHidden;
exports.isValueParser = require_valueparser.isValueParser;
exports.json = require_valueparser.json;
exports.keyValue = require_valueparser.keyValue;
exports.lineBreak = require_message.lineBreak;
exports.link = require_message.link;
exports.locale = require_valueparser.locale;
exports.longestMatch = require_constructs.longestMatch;
exports.macAddress = require_valueparser.macAddress;
exports.map = require_modifiers.map;
exports.merge = require_constructs.merge;
exports.mergeHidden = require_usage.mergeHidden;
exports.message = require_message.message;
exports.metavar = require_message.metavar;
exports.multiple = require_modifiers.multiple;
exports.negatableFlag = require_primitives.negatableFlag;
exports.nonEmpty = require_modifiers.nonEmpty;
exports.normalizeUsage = require_usage.normalizeUsage;
exports.nu = require_completion.nu;
exports.object = require_constructs.object;
exports.option = require_primitives.option;
exports.optionName = require_message.optionName;
exports.optionNames = require_message.optionNames;
exports.optional = require_modifiers.optional;
exports.or = require_constructs.or;
exports.parse = require_internal_parser.parse;
exports.parseAsync = require_internal_parser.parseAsync;
exports.parseSync = require_internal_parser.parseSync;
exports.passThrough = require_primitives.passThrough;
exports.port = require_valueparser.port;
exports.portRange = require_valueparser.portRange;
exports.pwsh = require_completion.pwsh;
exports.runParser = require_facade.runParser;
exports.runParserAsync = require_facade.runParserAsync;
exports.runParserSync = require_facade.runParserSync;
exports.runWith = require_facade.runWith;
exports.runWithAsync = require_facade.runWithAsync;
exports.runWithSync = require_facade.runWithSync;
exports.semVer = require_valueparser.semVer;
exports.seq = require_constructs.seq;
exports.socketAddress = require_valueparser.socketAddress;
exports.string = require_valueparser.string;
exports.suggest = require_internal_parser.suggest;
exports.suggestAsync = require_internal_parser.suggestAsync;
exports.suggestSync = require_internal_parser.suggestSync;
exports.text = require_message.text;
exports.tuple = require_constructs.tuple;
exports.url = require_valueparser.url;
exports.uuid = require_valueparser.uuid;
exports.value = require_message.value;
exports.valueSet = require_message.valueSet;
exports.values = require_message.values;
exports.withDefault = require_modifiers.withDefault;
exports.zsh = require_completion.zsh;
