// @optique/core/dist/input-trace.js
//#region src/input-trace.ts
const symbolKeys = /* @__PURE__ */ new WeakMap();
let symbolCounter = 0;
/** Length-prefix a raw segment so that no delimiter escaping is needed. */
function lengthPrefix(s) {
	return `${s.length}:${s}`;
}
function serializeKey(key) {
	if (typeof key === "string") return lengthPrefix(`s${key}`);
	if (typeof key === "number") return lengthPrefix(`n${key}`);
	const registeredKey = Symbol.keyFor(key);
	if (registeredKey !== void 0) return lengthPrefix(`r${registeredKey}`);
	let id = symbolKeys.get(key);
	if (id === void 0) {
		id = `y${symbolCounter++}`;
		symbolKeys.set(key, id);
	}
	return lengthPrefix(id);
}
function serializePath(path) {
	if (path.length === 0) return "";
	return path.map(serializeKey).join("");
}
var InputTraceImpl = class InputTraceImpl {
	#entries;
	constructor(entries) {
		this.#entries = entries ?? /* @__PURE__ */ new Map();
	}
	get(path) {
		return this.#entries.get(serializePath(path));
	}
	set(path, entry) {
		const copy = new Map(this.#entries);
		copy.set(serializePath(path), entry);
		return new InputTraceImpl(copy);
	}
	delete(path) {
		const copy = new Map(this.#entries);
		copy.delete(serializePath(path));
		return new InputTraceImpl(copy);
	}
};
/**
* Creates a new empty {@link InputTrace}.
*
* @returns An empty immutable input trace.
* @internal
* @since 1.0.0
*/
function createInputTrace() {
	return new InputTraceImpl();
}

//#endregion
export { createInputTrace };
