// @optique/core/dist/modifiers.cjs
const require_internal_annotations = require('./internal/annotations.cjs');
const require_message = require('./message.cjs');
const require_mode_dispatch = require('./internal/mode-dispatch.cjs');
const require_internal_parser = require('./internal/parser.cjs');
const require_annotation_state = require('./annotation-state.cjs');
const require_execution_context = require('./execution-context.cjs');
const require_phase2_seed = require('./phase2-seed.cjs');
const require_dependency_metadata = require('./dependency-metadata.cjs');

//#region src/modifiers.ts
function isTerminalMultipleItemState(state) {
	const unwrapped = unwrapMultipleItemState(state).value;
	return unwrapped != null && typeof unwrapped === "object" && Object.hasOwn(unwrapped, "success") && typeof unwrapped.success === "boolean";
}
function isMatchedCommandParserState(parser, state) {
	if (Reflect.get(parser, Symbol.for("@optique/core/commandParser")) !== true) return false;
	const unwrapped = unwrapMultipleItemState(state).value;
	return Array.isArray(unwrapped) && unwrapped.length === 2 && unwrapped[0] === "matched" && typeof unwrapped[1] === "string";
}
function isUnstartedMultipleItemState(state, originalState) {
	const unwrappedState = unwrapMultipleItemState(state);
	if (unwrappedState.value == null) return true;
	if (originalState === void 0) return false;
	const unwrappedOriginalState = unwrapMultipleItemState(originalState);
	return unwrappedState.value === unwrappedOriginalState.value && (unwrappedState.viaBoundCliWrapper || unwrappedOriginalState.viaBoundCliWrapper || unwrappedOriginalState.value != null && typeof unwrappedOriginalState.value === "object");
}
function isBoundCliWrapperState(state) {
	return state != null && typeof state === "object" && Object.hasOwn(state, "hasCliValue") && typeof state.hasCliValue === "boolean" && (Object.hasOwn(state, "cliState") || state.hasCliValue === false);
}
function unwrapMultipleItemState(state) {
	let unwrapped = state;
	let viaBoundCliWrapper = false;
	while (true) {
		if (require_internal_annotations.isInjectedAnnotationWrapper(unwrapped)) {
			unwrapped = require_internal_annotations.unwrapInjectedAnnotationWrapper(unwrapped);
			continue;
		}
		if (Array.isArray(unwrapped) && unwrapped.length === 1) {
			unwrapped = unwrapped[0];
			continue;
		}
		if (isBoundCliWrapperState(unwrapped)) {
			viaBoundCliWrapper = true;
			unwrapped = unwrapped.cliState;
			continue;
		}
		return {
			value: unwrapped,
			viaBoundCliWrapper
		};
	}
}
function isPromiseLike(value) {
	return value != null && (typeof value === "object" || typeof value === "function") && "then" in value && typeof value.then === "function";
}
function normalizeOptionalLikeCompleteResult(result, shouldNormalize) {
	return result.success && shouldNormalize ? {
		...result,
		value: require_annotation_state.normalizeNestedDelegatedAnnotationState(result.value)
	} : result;
}
function shouldRetryOptionalLikeCompatibilityError(error, state) {
	return error instanceof TypeError && state != null && typeof state === "object" && require_internal_annotations.isInjectedAnnotationWrapper(state);
}
function completeOptionalLikeSync(parser, state, exec) {
	const hasCarrier = require_annotation_state.hasDelegatedAnnotationCarrier(state);
	const shouldRetryFalseResult = hasCarrier && require_annotation_state.isAnnotationWrappedInitialState(state);
	const run = (candidate, shouldNormalize) => normalizeOptionalLikeCompleteResult(parser.complete(candidate, exec), shouldNormalize);
	try {
		const result = run(state, hasCarrier);
		if (!result.success && shouldRetryFalseResult) return run(require_annotation_state.normalizeDelegatedAnnotationState(state), false);
		return result;
	} catch (error) {
		if (!shouldRetryOptionalLikeCompatibilityError(error, state)) throw error;
		return run(require_annotation_state.normalizeDelegatedAnnotationState(state), false);
	}
}
async function completeOptionalLikeAsync(parser, state, exec) {
	const hasCarrier = require_annotation_state.hasDelegatedAnnotationCarrier(state);
	const shouldRetryFalseResult = hasCarrier && require_annotation_state.isAnnotationWrappedInitialState(state);
	const run = async (candidate, shouldNormalize) => normalizeOptionalLikeCompleteResult(await parser.complete(candidate, exec), shouldNormalize);
	try {
		const result = await run(state, hasCarrier);
		if (!result.success && shouldRetryFalseResult) return await run(require_annotation_state.normalizeDelegatedAnnotationState(state), false);
		return result;
	} catch (error) {
		if (!shouldRetryOptionalLikeCompatibilityError(error, state)) throw error;
		return await run(require_annotation_state.normalizeDelegatedAnnotationState(state), false);
	}
}
function normalizeOptionalLikePhase2Seed(seed, shouldNormalize) {
	return seed == null ? null : shouldNormalize ? {
		...seed,
		value: require_annotation_state.normalizeNestedDelegatedAnnotationState(seed.value)
	} : seed;
}
function extractOptionalLikePhase2Seed(parser, state, exec) {
	if (!Array.isArray(state) && !(state != null && typeof state === "object")) return require_mode_dispatch.wrapForMode(parser.mode, null);
	const innerState = normalizeOptionalLikeInnerState(state, parser.initialState, parser);
	const hasCarrier = require_annotation_state.hasDelegatedAnnotationCarrier(innerState);
	const shouldRetryFalseResult = hasCarrier && require_annotation_state.isAnnotationWrappedInitialState(innerState);
	return require_mode_dispatch.dispatchByMode(parser.mode, () => {
		try {
			const result = parser.complete(innerState, exec);
			if (result.success) return normalizeOptionalLikePhase2Seed(require_phase2_seed.phase2SeedFromValueResult(result), hasCarrier);
			if (shouldRetryFalseResult) {
				const fallbackState = require_annotation_state.normalizeDelegatedAnnotationState(innerState);
				const fallbackResult = parser.complete(fallbackState, exec);
				if (fallbackResult.success) return normalizeOptionalLikePhase2Seed(require_phase2_seed.phase2SeedFromValueResult(fallbackResult), false);
			}
			const seed = require_phase2_seed.extractPhase2Seed(parser, innerState, exec);
			if (seed == null && hasCarrier) {
				const fallbackState = require_annotation_state.normalizeDelegatedAnnotationState(innerState);
				return normalizeOptionalLikePhase2Seed(require_phase2_seed.extractPhase2Seed(parser, fallbackState, exec), false);
			}
			return normalizeOptionalLikePhase2Seed(seed, hasCarrier);
		} catch (error) {
			if (!shouldRetryOptionalLikeCompatibilityError(error, innerState)) throw error;
			const fallbackState = require_annotation_state.normalizeDelegatedAnnotationState(innerState);
			const result = parser.complete(fallbackState, exec);
			if (result.success) return normalizeOptionalLikePhase2Seed(require_phase2_seed.phase2SeedFromValueResult(result), false);
			return normalizeOptionalLikePhase2Seed(require_phase2_seed.extractPhase2Seed(parser, fallbackState, exec), false);
		}
	}, async () => {
		try {
			const result = await parser.complete(innerState, exec);
			if (result.success) return normalizeOptionalLikePhase2Seed(require_phase2_seed.phase2SeedFromValueResult(result), hasCarrier);
			if (shouldRetryFalseResult) {
				const fallbackState = require_annotation_state.normalizeDelegatedAnnotationState(innerState);
				const fallbackResult = await parser.complete(fallbackState, exec);
				if (fallbackResult.success) return normalizeOptionalLikePhase2Seed(require_phase2_seed.phase2SeedFromValueResult(fallbackResult), false);
			}
			const seed = await require_phase2_seed.extractPhase2Seed(parser, innerState, exec);
			if (seed == null && hasCarrier) {
				const fallbackState = require_annotation_state.normalizeDelegatedAnnotationState(innerState);
				return normalizeOptionalLikePhase2Seed(await require_phase2_seed.extractPhase2Seed(parser, fallbackState, exec), false);
			}
			return normalizeOptionalLikePhase2Seed(seed, hasCarrier);
		} catch (error) {
			if (!shouldRetryOptionalLikeCompatibilityError(error, innerState)) throw error;
			const fallbackState = require_annotation_state.normalizeDelegatedAnnotationState(innerState);
			const result = await parser.complete(fallbackState, exec);
			if (result.success) return normalizeOptionalLikePhase2Seed(require_phase2_seed.phase2SeedFromValueResult(result), false);
			return normalizeOptionalLikePhase2Seed(await require_phase2_seed.extractPhase2Seed(parser, fallbackState, exec), false);
		}
	});
}
/**
* Computes the inner state to pass through to the wrapped parser inside
* {@link optional}/{@link withDefault}.  When the outer state is an
* array, the inner state is `state[0]`.  Otherwise, including the
* common case where `optional()` sits at top level and the outer state
* is either `undefined` or an annotation wrapper from `parseOptionalLike`/
* `parse({ annotations })`, we use the wrapped parser's
* `initialState`, propagating annotations from the outer state so that
* source-binding wrappers under `optional()`/`withDefault()` (e.g.,
* `bindEnv()`/`bindConfig()`) can resolve their fallbacks.
*
* @internal
*/
function deriveOptionalInnerParseState(outerState, parser) {
	if (Array.isArray(outerState)) {
		const innerState = outerState[0];
		if (require_internal_annotations.getAnnotations(outerState) != null && innerState != null && typeof innerState === "object") return require_annotation_state.getDelegatedAnnotationState(outerState, innerState);
		return innerState;
	}
	const initial = parser.initialState;
	if (initial != null && typeof initial !== "object") return initial;
	return initial != null && typeof initial === "object" ? require_annotation_state.getDelegatedAnnotationState(outerState, initial) : require_internal_annotations.inheritAnnotations(outerState, initial);
}
/**
* Internal helper for optional-style parsing logic shared by optional()
* and withDefault(). Handles the common pattern of:
* - Unwrapping optional state to inner parser state
* - Detecting if the inner parser actually matched (state changed or no consumption)
* - Returning success with undefined state when the inner parser fails without consuming
* @internal
*/
function parseOptionalStyleSync(context, parser) {
	const innerState = deriveOptionalInnerParseState(context.state, parser);
	const result = parser.parse({
		...context,
		state: innerState
	});
	return processOptionalStyleResult(result, innerState, context);
}
/**
* Internal async helper for optional-style parsing logic.
* @internal
*/
async function parseOptionalStyleAsync(context, parser) {
	const innerState = deriveOptionalInnerParseState(context.state, parser);
	const result = await parser.parse({
		...context,
		state: innerState
	});
	return processOptionalStyleResult(result, innerState, context);
}
/**
* Internal helper to process optional-style parse results.
* @internal
*/
function hasOptionalLikeParseStateChanged(previousState, nextState) {
	return previousState !== nextState && require_annotation_state.normalizeDelegatedAnnotationState(previousState) !== require_annotation_state.normalizeDelegatedAnnotationState(nextState);
}
function processOptionalStyleResult(result, innerState, context) {
	if (result.success) {
		if (hasOptionalLikeParseStateChanged(innerState, result.next.state) || result.consumed.length === 0) return {
			success: true,
			...result.provisional ? { provisional: true } : {},
			next: {
				...result.next,
				state: [result.next.state]
			},
			consumed: result.consumed
		};
		return {
			success: true,
			...result.provisional ? { provisional: true } : {},
			next: {
				...result.next,
				state: context.state
			},
			consumed: result.consumed
		};
	}
	if (result.consumed === 0) return {
		success: true,
		next: context,
		consumed: []
	};
	return result;
}
/**
* Creates a `shouldDeferCompletion` adapter that unwraps the outer state
* shape (`[TState] | undefined`) used by {@link optional} and
* {@link withDefault} before delegating to the inner parser's hook.
*
* When state is an array, the adapter unwraps `state[0]` and propagates
* annotations from the outer array.  Non-array objects (e.g. PromptBindState
* from `prompt()`) are passed through directly.  `undefined` returns `false`
* without calling the inner hook.
*
* @internal
*/
function adaptShouldDeferCompletion(innerCheck, parser) {
	return (state, exec) => {
		if (Array.isArray(state) || state != null && typeof state === "object") {
			const innerState = normalizeOptionalLikeInnerState(state, parser.initialState, parser);
			const hasCarrier = require_annotation_state.hasDelegatedAnnotationCarrier(innerState);
			const shouldRetryFalseResult = hasCarrier && require_annotation_state.isAnnotationWrappedInitialState(innerState);
			try {
				const result = innerCheck(innerState, exec);
				if (!result && shouldRetryFalseResult) return innerCheck(require_annotation_state.normalizeDelegatedAnnotationState(innerState), exec);
				return result;
			} catch (error) {
				if (!shouldRetryOptionalLikeCompatibilityError(error, innerState)) throw error;
				return innerCheck(require_annotation_state.normalizeDelegatedAnnotationState(innerState), exec);
			}
		}
		return false;
	};
}
function isAnnotationOnlyObjectState(state) {
	if (state == null || typeof state !== "object" || Array.isArray(state) || require_internal_annotations.isInjectedAnnotationWrapper(state)) return false;
	const keys = Reflect.ownKeys(state);
	return require_internal_annotations.getAnnotations(state) != null && keys.length === 1 && keys[0] === require_internal_annotations.annotationKey;
}
function normalizeOptionalLikeInnerState(state, initialState, parser) {
	if (Array.isArray(state)) return require_annotation_state.getDelegatedAnnotationState(state, state[0]);
	if (isAnnotationOnlyObjectState(state)) {
		if (parser != null && (parser.dependencyMetadata?.source != null || typeof parser.shouldDeferCompletion === "function")) return require_annotation_state.getDelegatedAnnotationState(state, initialState);
		return initialState;
	}
	if (state != null && typeof state === "object") return state;
	return initialState;
}
function normalizeOptionalLikeSuggestState(state, initialState, parser) {
	if (Array.isArray(state)) {
		const innerState = state[0];
		return innerState != null && typeof innerState === "object" ? require_annotation_state.getDelegatedAnnotationState(state, innerState) : innerState;
	}
	return normalizeOptionalLikeInnerState(state, initialState, parser);
}
/**
* Creates a parser that makes another parser optional, allowing it to succeed
* without consuming input if the wrapped parser fails to match.
* If the wrapped parser succeeds, this returns its value.
* If the wrapped parser fails, this returns `undefined` without consuming input.
* @template M The execution mode of the parser.
* @template TValue The type of the value returned by the wrapped parser.
* @template TState The type of the state used by the wrapped parser.
* @param parser The {@link Parser} to make optional.
* @returns A {@link Parser} that produces either the result of the wrapped parser
*          or `undefined` if the wrapped parser fails to match.
*/
function optional(parser) {
	const syncParser = parser;
	function* suggestSync(context, prefix) {
		const innerState = normalizeOptionalLikeSuggestState(context.state, syncParser.initialState, parser);
		yield* syncParser.suggest({
			...context,
			state: innerState
		}, prefix);
	}
	async function* suggestAsync(context, prefix) {
		const innerState = normalizeOptionalLikeSuggestState(context.state, syncParser.initialState, parser);
		const suggestions = parser.suggest({
			...context,
			state: innerState
		}, prefix);
		for await (const s of suggestions) yield s;
	}
	const optionalParser = {
		mode: parser.mode,
		$valueType: [],
		$stateType: [],
		...parser[require_internal_parser.unmatchedNonCliDependencySourceStateMarker] === true ? { [require_internal_parser.unmatchedNonCliDependencySourceStateMarker]: true } : {},
		placeholder: void 0,
		priority: parser.priority,
		usage: [{
			type: "optional",
			terms: parser.usage
		}],
		leadingNames: parser.leadingNames,
		acceptingAnyToken: false,
		initialState: void 0,
		canSkip(state, exec) {
			if (!Array.isArray(state)) return true;
			const innerState = normalizeOptionalLikeInnerState(state, parser.initialState, parser);
			return parser.canSkip?.(innerState, exec) === true;
		},
		...typeof parser.shouldDeferCompletion === "function" ? { shouldDeferCompletion: adaptShouldDeferCompletion(parser.shouldDeferCompletion.bind(parser), parser) } : {},
		getSuggestRuntimeNodes(state, path) {
			if (optionalParser.dependencyMetadata?.source != null) return [{
				path,
				parser: optionalParser,
				state
			}];
			const innerState = normalizeOptionalLikeSuggestState(state, parser.initialState, parser);
			return parser.getSuggestRuntimeNodes?.(innerState, path) ?? (parser.dependencyMetadata?.source != null ? [{
				path,
				parser,
				state: innerState
			}] : []);
		},
		[require_phase2_seed.extractPhase2SeedKey](state, exec) {
			return extractOptionalLikePhase2Seed(parser, state, exec);
		},
		parse(context) {
			return require_mode_dispatch.dispatchByMode(parser.mode, () => parseOptionalStyleSync(context, syncParser), () => parseOptionalStyleAsync(context, parser));
		},
		complete(state, exec) {
			if (!Array.isArray(state)) {
				const delegateToInner = (resolvedInnerState) => {
					const innerResult = require_mode_dispatch.dispatchByMode(parser.mode, () => completeOptionalLikeSync(syncParser, resolvedInnerState, exec), async () => await completeOptionalLikeAsync(parser, resolvedInnerState, exec));
					return require_mode_dispatch.mapModeValue(parser.mode, innerResult, (result) => result.success ? result : {
						success: true,
						value: void 0
					});
				};
				if (typeof parser.shouldDeferCompletion === "function" && state != null && typeof state === "object") {
					const innerState = normalizeOptionalLikeInnerState(state, parser.initialState, parser);
					return delegateToInner(innerState);
				}
				const sourceMetadata = parser.dependencyMetadata?.source;
				if (sourceMetadata?.preservesSourceValue !== false && sourceMetadata?.getMissingSourceValue != null) {
					const delegatedState = normalizeOptionalLikeInnerState(state, parser.initialState, parser);
					return require_mode_dispatch.dispatchByMode(parser.mode, () => completeOptionalLikeSync(syncParser, delegatedState, exec), async () => await completeOptionalLikeAsync(parser, delegatedState, exec));
				}
				if (parser[require_internal_parser.unmatchedNonCliDependencySourceStateMarker] === true && state != null && typeof state === "object") {
					const innerState = normalizeOptionalLikeInnerState(state, parser.initialState, parser);
					return delegateToInner(innerState);
				}
				return {
					success: true,
					value: void 0
				};
			}
			const innerElement = normalizeOptionalLikeInnerState(state, parser.initialState, parser);
			return require_mode_dispatch.dispatchByMode(parser.mode, () => completeOptionalLikeSync(syncParser, innerElement, exec), async () => await completeOptionalLikeAsync(parser, innerElement, exec));
		},
		suggest(context, prefix) {
			return require_mode_dispatch.dispatchIterableByMode(parser.mode, () => suggestSync(context, prefix), () => suggestAsync(context, prefix));
		},
		getDocFragments(state, defaultValue) {
			const innerState = state.kind === "unavailable" ? { kind: "unavailable" } : state.state === void 0 ? { kind: "unavailable" } : {
				kind: "available",
				state: state.state[0]
			};
			return syncParser.getDocFragments(innerState, defaultValue);
		}
	};
	if (typeof parser.normalizeValue === "function") {
		const innerNormalize = parser.normalizeValue.bind(parser);
		Object.defineProperty(optionalParser, "normalizeValue", {
			value(v) {
				if (v == null) return v;
				return innerNormalize(v);
			},
			configurable: true,
			enumerable: false
		});
	}
	if (typeof parser.validateValue === "function") {
		const innerValidate = parser.validateValue.bind(parser);
		Object.defineProperty(optionalParser, "validateValue", {
			value(v) {
				if (v === void 0) return require_mode_dispatch.wrapForMode(parser.mode, {
					success: true,
					value: v
				});
				return innerValidate(v);
			},
			configurable: true,
			enumerable: false
		});
	}
	if (parser.dependencyMetadata != null) {
		const composed = require_dependency_metadata.composeDependencyMetadata(parser.dependencyMetadata, "optional");
		if (composed != null) optionalParser.dependencyMetadata = composed;
	}
	require_internal_parser.defineInheritedAnnotationParser(optionalParser);
	require_internal_parser.defineSourceBindingOnlyAnnotationCompletionParser(optionalParser);
	return optionalParser;
}
/**
* Error type for structured error messages in {@link withDefault} default value callbacks.
* Unlike regular errors that only support string messages, this error type accepts
* a {@link Message} object that supports rich formatting, colors, and structured content.
*
* @example
* ```typescript
* withDefault(option("--url", url()), () => {
*   if (!process.env.INSTANCE_URL) {
*     throw new WithDefaultError(
*       message`Environment variable ${envVar("INSTANCE_URL")} is not set.`
*     );
*   }
*   return new URL(process.env.INSTANCE_URL);
* })
* ```
*
* @since 0.5.0
*/
var WithDefaultError = class extends Error {
	/**
	* The structured message associated with this error.
	*/
	errorMessage;
	/**
	* Creates a new WithDefaultError with a structured message.
	* @param message The structured {@link Message} describing the error.
	*/
	constructor(message$1) {
		super(require_message.formatMessage(message$1));
		this.errorMessage = message$1;
		this.name = "WithDefaultError";
	}
};
function withDefault(parser, defaultValue, options) {
	const syncParser = parser;
	const getDocDefaultValue = (upperDefaultValue) => {
		if (upperDefaultValue != null) return upperDefaultValue;
		if (options?.message) return void 0;
		if (typeof defaultValue !== "function") return defaultValue;
		try {
			return defaultValue();
		} catch {
			return void 0;
		}
	};
	function* suggestSync(context, prefix) {
		const innerState = normalizeOptionalLikeSuggestState(context.state, syncParser.initialState, parser);
		yield* syncParser.suggest({
			...context,
			state: innerState
		}, prefix);
	}
	async function* suggestAsync(context, prefix) {
		const innerState = normalizeOptionalLikeSuggestState(context.state, syncParser.initialState, parser);
		const suggestions = parser.suggest({
			...context,
			state: innerState
		}, prefix);
		for await (const s of suggestions) yield s;
	}
	const withDefaultParser = {
		mode: parser.mode,
		$valueType: [],
		$stateType: [],
		...parser[require_internal_parser.unmatchedNonCliDependencySourceStateMarker] === true ? { [require_internal_parser.unmatchedNonCliDependencySourceStateMarker]: true } : {},
		priority: parser.priority,
		usage: [{
			type: "optional",
			terms: parser.usage
		}],
		leadingNames: parser.leadingNames,
		acceptingAnyToken: false,
		initialState: void 0,
		canSkip(state, exec) {
			if (!Array.isArray(state)) return true;
			const innerState = normalizeOptionalLikeInnerState(state, parser.initialState, parser);
			return parser.canSkip?.(innerState, exec) === true;
		},
		...typeof parser.shouldDeferCompletion === "function" ? { shouldDeferCompletion: adaptShouldDeferCompletion(parser.shouldDeferCompletion.bind(parser), parser) } : {},
		getSuggestRuntimeNodes(state, path) {
			if (withDefaultParser.dependencyMetadata?.source != null) return [{
				path,
				parser: withDefaultParser,
				state
			}];
			const innerState = normalizeOptionalLikeSuggestState(state, parser.initialState, parser);
			return parser.getSuggestRuntimeNodes?.(innerState, path) ?? (parser.dependencyMetadata?.source != null ? [{
				path,
				parser,
				state: innerState
			}] : []);
		},
		[require_phase2_seed.extractPhase2SeedKey](state, exec) {
			return extractOptionalLikePhase2Seed(parser, state, exec);
		},
		parse(context) {
			return require_mode_dispatch.dispatchByMode(parser.mode, () => parseOptionalStyleSync(context, syncParser), () => parseOptionalStyleAsync(context, parser));
		},
		complete(state, exec) {
			function evaluateDefault() {
				const raw = typeof defaultValue === "function" ? defaultValue() : defaultValue;
				if (typeof parser.normalizeValue === "function") try {
					return parser.normalizeValue(raw);
				} catch {}
				return raw;
			}
			if (!Array.isArray(state)) {
				if (typeof parser.shouldDeferCompletion === "function" && state != null && typeof state === "object") {
					const innerState = normalizeOptionalLikeInnerState(state, parser.initialState, parser);
					const innerResult = require_mode_dispatch.dispatchByMode(parser.mode, () => completeOptionalLikeSync(syncParser, innerState, exec), async () => await completeOptionalLikeAsync(parser, innerState, exec));
					return innerResult;
				}
				if (parser[require_internal_parser.unmatchedNonCliDependencySourceStateMarker] === true && state != null && typeof state === "object") {
					const innerState = normalizeOptionalLikeInnerState(state, parser.initialState, parser);
					const innerResult = require_mode_dispatch.dispatchByMode(parser.mode, () => completeOptionalLikeSync(syncParser, innerState, exec), async () => await completeOptionalLikeAsync(parser, innerState, exec));
					const handleInnerResult = (result) => {
						if (result.success && result.value !== void 0) return result;
						try {
							return {
								success: true,
								value: evaluateDefault()
							};
						} catch (error) {
							return {
								success: false,
								error: error instanceof WithDefaultError ? error.errorMessage : require_message.message`${require_message.text(String(error))}`
							};
						}
					};
					return require_mode_dispatch.mapModeValue(parser.mode, innerResult, handleInnerResult);
				}
				try {
					const value = evaluateDefault();
					return {
						success: true,
						value
					};
				} catch (error) {
					return {
						success: false,
						error: error instanceof WithDefaultError ? error.errorMessage : require_message.message`${require_message.text(String(error))}`
					};
				}
			}
			const innerElement = normalizeOptionalLikeInnerState(state, parser.initialState, parser);
			return require_mode_dispatch.dispatchByMode(parser.mode, () => completeOptionalLikeSync(syncParser, innerElement, exec), async () => await completeOptionalLikeAsync(parser, innerElement, exec));
		},
		suggest(context, prefix) {
			return require_mode_dispatch.dispatchIterableByMode(parser.mode, () => suggestSync(context, prefix), () => suggestAsync(context, prefix));
		},
		getDocFragments(state, upperDefaultValue) {
			const innerState = state.kind === "unavailable" ? { kind: "unavailable" } : state.state === void 0 ? { kind: "unavailable" } : {
				kind: "available",
				state: state.state[0]
			};
			let docDefault = getDocDefaultValue(upperDefaultValue);
			if (docDefault != null && typeof parser.normalizeValue === "function") try {
				docDefault = parser.normalizeValue(docDefault);
			} catch {}
			const fragments = syncParser.getDocFragments(innerState, docDefault);
			if (options?.message) {
				const modifiedFragments = fragments.fragments.map((fragment) => {
					if (fragment.type === "entry") return {
						...fragment,
						default: options.message
					};
					return fragment;
				});
				return {
					...fragments,
					fragments: modifiedFragments
				};
			}
			return fragments;
		}
	};
	if ("placeholder" in parser) Object.defineProperty(withDefaultParser, "placeholder", {
		get() {
			return parser.placeholder;
		},
		configurable: true,
		enumerable: false
	});
	if (typeof parser.normalizeValue === "function") {
		const innerNormalize = parser.normalizeValue.bind(parser);
		Object.defineProperty(withDefaultParser, "normalizeValue", {
			value(v) {
				try {
					return innerNormalize(v);
				} catch {
					return v;
				}
			},
			configurable: true,
			enumerable: false
		});
	}
	if (typeof parser.validateValue === "function") {
		const innerValidate = parser.validateValue.bind(parser);
		Object.defineProperty(withDefaultParser, "validateValue", {
			value(v) {
				return innerValidate(v);
			},
			configurable: true,
			enumerable: false
		});
	}
	if (parser.dependencyMetadata != null) {
		const composed = require_dependency_metadata.composeDependencyMetadata(parser.dependencyMetadata, "withDefault", { defaultValue: () => {
			let v;
			try {
				v = typeof defaultValue === "function" ? defaultValue() : defaultValue;
			} catch (e) {
				const error = e instanceof WithDefaultError ? e.errorMessage : require_message.message`${e instanceof Error ? e.message : String(e)}`;
				return {
					success: false,
					error
				};
			}
			if (typeof parser.normalizeValue === "function") try {
				v = parser.normalizeValue(v);
			} catch {}
			return {
				success: true,
				value: v
			};
		} });
		if (composed != null) withDefaultParser.dependencyMetadata = composed;
	}
	require_internal_parser.defineInheritedAnnotationParser(withDefaultParser);
	require_internal_parser.defineSourceBindingOnlyAnnotationCompletionParser(withDefaultParser);
	return withDefaultParser;
}
/**
* Creates a parser that transforms the result value of another parser using
* a mapping function. This enables value transformation while preserving
* the original parser's parsing logic and state management.
*
* The `map()` function is useful for:
* - Converting parsed values to different types
* - Applying transformations like string formatting or boolean inversion
* - Computing derived values from parsed input
* - Creating reusable transformations that can be applied to any parser
*
* @template M The execution mode of the parser.
* @template T The type of the value produced by the original parser.
* @template U The type of the value produced by the mapping function.
* @template TState The type of the state used by the original parser.
* @param parser The {@link Parser} whose result will be transformed.
* @param transform A function that transforms the parsed value from type T to type U.
* @returns A {@link Parser} that produces the transformed value of type U
*          while preserving the original parser's state type and parsing behavior.
* @throws Any exception thrown by `transform` when completing a non-deferred
*   value.  Errors from deferred placeholder transforms are caught and the
*   mapped result falls back to `undefined` with `deferred: true`.
*
* ### Deferred prompt interaction
*
* During two-phase parsing, `map()` propagates the `deferred` flag from
* inner results but intentionally drops per-field `deferredKeys`.  The
* inner key set describes the *input* shape, but `transform` produces an
* arbitrary *output* shape where keys may be renamed, dropped, or reused
* with different semantics.  For `object()` results that are *not*
* wrapped in `map()`, per-field deferred stripping works normally.
*
* Because the `deferred` flag is propagated conservatively, mapped scalar
* results are treated as missing (`undefined`) during phase-two context
* collection—even when `transform` only used non-deferred fields.
* For example, `map(object({ apiKey: prompt(...), mode: option(...) }),
* v => v.mode)` makes phase-two contexts see `undefined` instead of the
* real `mode` value.  This is the intentional trade-off: the alternative
* (not propagating `deferred`) would leak placeholder values into context
* resolution when `transform` *does* use deferred fields.  The final
* parse always produces the correct result regardless.
*
* If the transform throws on a deferred placeholder value, the mapped
* result falls back to `undefined` with `deferred: true`, so the first
* pass does not abort.
*
* ### Transform purity
*
* The `transform` function must not mutate its input.  Object and array
* values may be shared placeholder references during deferred prompt
* resolution, and in-place mutations would corrupt the placeholder for
* subsequent parses.  Always return a new value:
*
* ```typescript
* // ✅ Correct — creates a new object
* map(parser, v => ({ ...v, host: "override" }))
*
* // ❌ Wrong — mutates the input in place
* map(parser, v => { v.host = "override"; return v; })
* ```
*
* @example
* ```typescript
* // Transform boolean flag to its inverse
* const parser = object({
*   disallow: map(option("--allow"), b => !b)
* });
*
* // Transform string to uppercase
* const upperParser = map(argument(string()), s => s.toUpperCase());
*
* // Transform number to formatted string
* const prefixedParser = map(option("-n", integer()), n => `value: ${n}`);
* ```
*/
function map(parser, transform) {
	const complete = (state, exec) => {
		return require_mode_dispatch.mapModeValue(parser.mode, parser.complete(state, exec), (result) => {
			if (!result.success) return result;
			if (result.deferred) try {
				return {
					success: true,
					value: transform(result.value),
					deferred: true
				};
			} catch {
				return {
					success: true,
					value: void 0,
					deferred: true
				};
			}
			return {
				success: true,
				value: transform(result.value)
			};
		});
	};
	const mappedParser = {
		...parser,
		$valueType: [],
		complete,
		[require_phase2_seed.extractPhase2SeedKey](state, exec) {
			return require_mode_dispatch.mapModeValue(parser.mode, require_phase2_seed.completeOrExtractPhase2Seed(parser, state, exec), (seed) => {
				if (seed == null) return null;
				if (seed.deferred) try {
					return {
						value: transform(seed.value),
						deferred: true
					};
				} catch {
					return {
						value: void 0,
						deferred: true
					};
				}
				return { value: transform(seed.value) };
			});
		},
		getSuggestRuntimeNodes(state, path) {
			if (mappedParser.dependencyMetadata?.source != null) return [{
				path,
				parser: mappedParser,
				state
			}];
			return parser.getSuggestRuntimeNodes?.(state, path) ?? (parser.dependencyMetadata?.source != null ? [{
				path,
				parser,
				state
			}] : []);
		},
		getDocFragments(state, _defaultValue) {
			return parser.getDocFragments(state, void 0);
		}
	};
	delete mappedParser.normalizeValue;
	delete mappedParser.validateValue;
	if ("placeholder" in parser) Object.defineProperty(mappedParser, "placeholder", {
		get() {
			try {
				return transform(parser.placeholder);
			} catch {
				return void 0;
			}
		},
		configurable: true,
		enumerable: false
	});
	if (parser.dependencyMetadata != null) {
		let composed = require_dependency_metadata.composeDependencyMetadata(parser.dependencyMetadata, "map");
		if (composed?.derived != null) {
			const innerReplay = composed.derived.replayParse;
			const applyMappedReplay = (r) => {
				if (!r.success) return r;
				if (r.deferred) try {
					return {
						success: true,
						value: transform(r.value),
						deferred: true
					};
				} catch {
					return {
						success: true,
						value: void 0,
						deferred: true
					};
				}
				return {
					success: true,
					value: transform(r.value)
				};
			};
			composed = {
				...composed,
				derived: {
					...composed.derived,
					replayParse: (rawInput, depValues) => {
						const result = innerReplay(rawInput, depValues);
						return isPromiseLike(result) ? result.then(applyMappedReplay) : applyMappedReplay(result);
					}
				}
			};
		}
		if (composed != null) mappedParser.dependencyMetadata = composed;
	}
	return mappedParser;
}
/**
* Creates a parser that allows multiple occurrences of a given parser.
* This parser can be used to parse multiple values of the same type,
* such as multiple command-line arguments or options.
* @template M The execution mode of the parser.
* @template TValue The type of the value that the parser produces.
* @template TState The type of the state used by the parser.
* @param parser The {@link Parser} to apply multiple times.
* @param options Optional configuration for the parser,
*                allowing you to specify the minimum and maximum number of
*                occurrences allowed.
* @returns A {@link Parser} that produces an array of values
*          of type {@link TValue} and an array of states
*          of type {@link TState}.
*/
function multiple(parser, options = {}) {
	const syncParser = parser;
	const { min = 0, max = Infinity } = options;
	const unwrapInjectedWrapper = require_internal_annotations.unwrapInjectedAnnotationWrapper;
	const completeSyncWithUnwrappedFallback = (state, exec) => {
		try {
			const result = syncParser.complete(state, exec);
			if (!result.success && require_internal_annotations.isInjectedAnnotationWrapper(state)) return syncParser.complete(unwrapInjectedWrapper(state), exec);
			return result;
		} catch (error) {
			if (!require_internal_annotations.isInjectedAnnotationWrapper(state)) throw error;
			return syncParser.complete(unwrapInjectedWrapper(state), exec);
		}
	};
	const parseSyncWithUnwrappedFallback = (context) => {
		try {
			const result = syncParser.parse(context);
			if (result.success || result.consumed !== 0 || !require_internal_annotations.isInjectedAnnotationWrapper(context.state)) return result;
			return syncParser.parse({
				...context,
				state: unwrapInjectedWrapper(context.state)
			});
		} catch (error) {
			if (!require_internal_annotations.isInjectedAnnotationWrapper(context.state)) throw error;
			return syncParser.parse({
				...context,
				state: unwrapInjectedWrapper(context.state)
			});
		}
	};
	const completeAsyncWithUnwrappedFallback = async (state, exec) => {
		try {
			const result = await parser.complete(state, exec);
			if (!result.success && require_internal_annotations.isInjectedAnnotationWrapper(state)) return await parser.complete(unwrapInjectedWrapper(state), exec);
			return result;
		} catch (error) {
			if (!require_internal_annotations.isInjectedAnnotationWrapper(state)) throw error;
			return await parser.complete(unwrapInjectedWrapper(state), exec);
		}
	};
	const parseAsyncWithUnwrappedFallback = async (context) => {
		try {
			const result = await parser.parse(context);
			if (result.success || result.consumed !== 0 || !require_internal_annotations.isInjectedAnnotationWrapper(context.state)) return result;
			return await parser.parse({
				...context,
				state: unwrapInjectedWrapper(context.state)
			});
		} catch (error) {
			if (!require_internal_annotations.isInjectedAnnotationWrapper(context.state)) throw error;
			return await parser.parse({
				...context,
				state: unwrapInjectedWrapper(context.state)
			});
		}
	};
	const extractPhase2SeedSyncWithUnwrappedFallback = (state, exec) => {
		try {
			const seed = require_phase2_seed.completeOrExtractPhase2Seed(syncParser, state, exec);
			if (seed == null && require_internal_annotations.isInjectedAnnotationWrapper(state)) return require_phase2_seed.completeOrExtractPhase2Seed(syncParser, unwrapInjectedWrapper(state), exec);
			return seed;
		} catch (error) {
			if (!require_internal_annotations.isInjectedAnnotationWrapper(state)) throw error;
			return require_phase2_seed.completeOrExtractPhase2Seed(syncParser, unwrapInjectedWrapper(state), exec);
		}
	};
	const extractPhase2SeedAsyncWithUnwrappedFallback = async (state, exec) => {
		try {
			const seed = await require_phase2_seed.completeOrExtractPhase2Seed(parser, state, exec);
			if (seed == null && require_internal_annotations.isInjectedAnnotationWrapper(state)) return await require_phase2_seed.completeOrExtractPhase2Seed(parser, unwrapInjectedWrapper(state), exec);
			return seed;
		} catch (error) {
			if (!require_internal_annotations.isInjectedAnnotationWrapper(state)) throw error;
			return await require_phase2_seed.completeOrExtractPhase2Seed(parser, unwrapInjectedWrapper(state), exec);
		}
	};
	const getInnerSuggestRuntimeNodes = (state, path) => parser.getSuggestRuntimeNodes?.(state, path) ?? (parser.dependencyMetadata?.source != null ? [{
		path,
		parser,
		state
	}] : []);
	const canExtendMultipleItem = (state, itemIndex, exec) => state != null && !isTerminalMultipleItemState(state) && (isMatchedCommandParserState(parser, state) || parser.canSkip?.(state, require_execution_context.withChildExecPath(exec, itemIndex)) !== true);
	const parseSync = (context) => {
		const currentItemState = context.state.at(-1);
		const currentItemIndex = context.state.length - 1;
		const canExtendCurrent = canExtendMultipleItem(currentItemState, currentItemIndex, context.exec);
		const canOpenFreshItem = context.state.length < max;
		if (!canExtendCurrent && !canOpenFreshItem) return {
			success: true,
			next: context,
			consumed: []
		};
		let added = !canExtendCurrent;
		let itemIndex = canExtendCurrent ? context.state.length - 1 : context.state.length;
		const currentItemStateWithAnnotations = canExtendCurrent ? currentItemState : require_internal_annotations.inheritAnnotations(context.state, syncParser.initialState);
		let result = parseSyncWithUnwrappedFallback(require_execution_context.withChildContext(context, itemIndex, currentItemStateWithAnnotations));
		if (!result.success) {
			if (result.consumed !== 0) return result;
			if (!added && canOpenFreshItem) {
				const nextInitialState = require_internal_annotations.inheritAnnotations(context.state, syncParser.initialState);
				itemIndex = context.state.length;
				result = parseSyncWithUnwrappedFallback(require_execution_context.withChildContext(context, itemIndex, nextInitialState));
				if (!result.success) {
					if (min === 0 && context.buffer.length === 0 && result.consumed === 0) return {
						success: true,
						next: context,
						consumed: []
					};
					return result;
				}
				added = true;
			} else if (min === 0 && context.buffer.length === 0) return {
				success: true,
				next: context,
				consumed: []
			};
			else return result;
		}
		const mergedExec = require_execution_context.mergeChildExec(context.exec, result.next.exec);
		if (added && result.consumed.length === 0 && result.next.optionsTerminated === context.optionsTerminated && isUnstartedMultipleItemState(result.next.state, currentItemStateWithAnnotations)) return {
			success: true,
			next: {
				...result.next,
				state: context.state,
				...mergedExec != null ? {
					trace: mergedExec.trace,
					exec: mergedExec,
					dependencyRegistry: mergedExec.dependencyRegistry
				} : {}
			},
			consumed: result.consumed
		};
		const itemAnnotationSource = added ? context.state : currentItemStateWithAnnotations;
		if (result.next.state === currentItemStateWithAnnotations && result.consumed.length > 0 && (!added || result.next.optionsTerminated !== context.optionsTerminated)) return {
			success: true,
			...result.provisional ? { provisional: true } : {},
			next: {
				...result.next,
				state: context.state,
				...mergedExec != null ? {
					trace: mergedExec.trace,
					exec: mergedExec,
					dependencyRegistry: mergedExec.dependencyRegistry
				} : {}
			},
			consumed: result.consumed
		};
		const nextItemState = require_internal_annotations.inheritAnnotations(itemAnnotationSource, result.next.state);
		return {
			success: true,
			...result.provisional ? { provisional: true } : {},
			next: {
				...result.next,
				state: require_internal_annotations.annotateFreshArray(context.state, [...added ? context.state : context.state.slice(0, -1), nextItemState]),
				...mergedExec != null ? {
					trace: mergedExec.trace,
					exec: mergedExec,
					dependencyRegistry: mergedExec.dependencyRegistry
				} : {}
			},
			consumed: result.consumed
		};
	};
	const parseAsync = async (context) => {
		const currentItemState = context.state.at(-1);
		const currentItemIndex = context.state.length - 1;
		const canExtendCurrent = canExtendMultipleItem(currentItemState, currentItemIndex, context.exec);
		const canOpenFreshItem = context.state.length < max;
		if (!canExtendCurrent && !canOpenFreshItem) return {
			success: true,
			next: context,
			consumed: []
		};
		let added = !canExtendCurrent;
		let itemIndex = canExtendCurrent ? context.state.length - 1 : context.state.length;
		const currentItemStateWithAnnotations = canExtendCurrent ? currentItemState : require_internal_annotations.inheritAnnotations(context.state, parser.initialState);
		let result = await parseAsyncWithUnwrappedFallback(require_execution_context.withChildContext(context, itemIndex, currentItemStateWithAnnotations));
		if (!result.success) {
			if (result.consumed !== 0) return result;
			if (!added && canOpenFreshItem) {
				const nextInitialState = require_internal_annotations.inheritAnnotations(context.state, parser.initialState);
				itemIndex = context.state.length;
				result = await parseAsyncWithUnwrappedFallback(require_execution_context.withChildContext(context, itemIndex, nextInitialState));
				if (!result.success) {
					if (min === 0 && context.buffer.length === 0 && result.consumed === 0) return {
						success: true,
						next: context,
						consumed: []
					};
					return result;
				}
				added = true;
			} else if (min === 0 && context.buffer.length === 0) return {
				success: true,
				next: context,
				consumed: []
			};
			else return result;
		}
		const mergedExec = require_execution_context.mergeChildExec(context.exec, result.next.exec);
		if (added && result.consumed.length === 0 && result.next.optionsTerminated === context.optionsTerminated && isUnstartedMultipleItemState(result.next.state, currentItemStateWithAnnotations)) return {
			success: true,
			next: {
				...result.next,
				state: context.state,
				...mergedExec != null ? {
					trace: mergedExec.trace,
					exec: mergedExec,
					dependencyRegistry: mergedExec.dependencyRegistry
				} : {}
			},
			consumed: result.consumed
		};
		const itemAnnotationSource = added ? context.state : currentItemStateWithAnnotations;
		if (result.next.state === currentItemStateWithAnnotations && result.consumed.length > 0 && (!added || result.next.optionsTerminated !== context.optionsTerminated)) return {
			success: true,
			...result.provisional ? { provisional: true } : {},
			next: {
				...result.next,
				state: context.state,
				...mergedExec != null ? {
					trace: mergedExec.trace,
					exec: mergedExec,
					dependencyRegistry: mergedExec.dependencyRegistry
				} : {}
			},
			consumed: result.consumed
		};
		const nextItemState = require_internal_annotations.inheritAnnotations(itemAnnotationSource, result.next.state);
		return {
			success: true,
			...result.provisional ? { provisional: true } : {},
			next: {
				...result.next,
				state: require_internal_annotations.annotateFreshArray(context.state, [...added ? context.state : context.state.slice(0, -1), nextItemState]),
				...mergedExec != null ? {
					trace: mergedExec.trace,
					exec: mergedExec,
					dependencyRegistry: mergedExec.dependencyRegistry
				} : {}
			},
			consumed: result.consumed
		};
	};
	const resultParser = {
		mode: parser.mode,
		$valueType: [],
		$stateType: [],
		priority: parser.priority,
		usage: [{
			type: "multiple",
			terms: parser.usage,
			min
		}],
		leadingNames: parser.leadingNames,
		acceptingAnyToken: min > 0 && (parser.acceptingAnyToken ?? false),
		initialState: [],
		canSkip(state, exec) {
			if (state.length < min) return false;
			const currentItemState = state.at(-1);
			return !canExtendMultipleItem(currentItemState, state.length - 1, exec);
		},
		getSuggestRuntimeNodes(state, path) {
			const innerNodes = state.flatMap((item, i) => [...getInnerSuggestRuntimeNodes(item, [...path, i])]);
			return resultParser.dependencyMetadata?.source != null ? [{
				path,
				parser: resultParser,
				state
			}, ...innerNodes] : innerNodes;
		},
		parse(context) {
			return require_mode_dispatch.dispatchByMode(parser.mode, () => parseSync(context), () => parseAsync(context));
		},
		complete(state, exec) {
			return require_mode_dispatch.dispatchByMode(parser.mode, () => {
				const result = [];
				const deferredIndices = /* @__PURE__ */ new Map();
				let hasDeferred = false;
				for (let i = 0; i < state.length; i++) {
					const valueResult = completeSyncWithUnwrappedFallback(state[i], require_execution_context.withChildExecPath(exec, i));
					if (valueResult.success) {
						const unwrappedValue = unwrapInjectedWrapper(valueResult.value);
						result.push(unwrappedValue);
						if (valueResult.deferred) if (valueResult.deferredKeys) deferredIndices.set(i, valueResult.deferredKeys);
						else if (unwrappedValue == null || typeof unwrappedValue !== "object") deferredIndices.set(i, null);
						else hasDeferred = true;
					} else return {
						success: false,
						error: valueResult.error
					};
				}
				return validateMultipleResult(result, deferredIndices, hasDeferred);
			}, async () => {
				const values = [];
				const deferredIndices = /* @__PURE__ */ new Map();
				let hasDeferred = false;
				for (let i = 0; i < state.length; i++) {
					const valueResult = await completeAsyncWithUnwrappedFallback(state[i], require_execution_context.withChildExecPath(exec, i));
					if (valueResult.success) {
						const unwrappedValue = unwrapInjectedWrapper(valueResult.value);
						values.push(unwrappedValue);
						if (valueResult.deferred) if (valueResult.deferredKeys) deferredIndices.set(i, valueResult.deferredKeys);
						else if (unwrappedValue == null || typeof unwrappedValue !== "object") deferredIndices.set(i, null);
						else hasDeferred = true;
					} else return {
						success: false,
						error: valueResult.error
					};
				}
				return validateMultipleResult(values, deferredIndices, hasDeferred);
			});
		},
		[require_phase2_seed.extractPhase2SeedKey](state, exec) {
			return require_mode_dispatch.dispatchByMode(parser.mode, () => {
				const values = [];
				const deferredIndices = /* @__PURE__ */ new Map();
				let hasDeferred = false;
				let hasAnySeed = false;
				for (let i = 0; i < state.length; i++) {
					const seed = extractPhase2SeedSyncWithUnwrappedFallback(state[i], require_execution_context.withChildExecPath(exec, i));
					if (seed == null) continue;
					hasAnySeed = true;
					values[i] = seed.value;
					if (seed.deferred) if (seed.deferredKeys) deferredIndices.set(i, seed.deferredKeys);
					else if (seed.value == null || typeof seed.value !== "object") deferredIndices.set(i, null);
					else hasDeferred = true;
				}
				if (!hasAnySeed) return null;
				return {
					value: values,
					...deferredIndices.size > 0 || hasDeferred ? {
						deferred: true,
						...deferredIndices.size > 0 ? { deferredKeys: deferredIndices } : {}
					} : {}
				};
			}, async () => {
				const values = [];
				const deferredIndices = /* @__PURE__ */ new Map();
				let hasDeferred = false;
				let hasAnySeed = false;
				for (let i = 0; i < state.length; i++) {
					const seed = await extractPhase2SeedAsyncWithUnwrappedFallback(state[i], require_execution_context.withChildExecPath(exec, i));
					if (seed == null) continue;
					hasAnySeed = true;
					values[i] = seed.value;
					if (seed.deferred) if (seed.deferredKeys) deferredIndices.set(i, seed.deferredKeys);
					else if (seed.value == null || typeof seed.value !== "object") deferredIndices.set(i, null);
					else hasDeferred = true;
				}
				if (!hasAnySeed) return null;
				return {
					value: values,
					...deferredIndices.size > 0 || hasDeferred ? {
						deferred: true,
						...deferredIndices.size > 0 ? { deferredKeys: deferredIndices } : {}
					} : {}
				};
			});
		},
		suggest(context, prefix) {
			const currentItemState = context.state.at(-1);
			const currentItemIndex = context.state.length - 1;
			const canExtendCurrent = canExtendMultipleItem(currentItemState, currentItemIndex, context.exec);
			const canOpenNew = context.state.length < max;
			if (!canExtendCurrent && !canOpenNew) return require_mode_dispatch.dispatchIterableByMode(parser.mode, function* () {}, async function* () {});
			const itemIndex = canExtendCurrent ? context.state.length - 1 : context.state.length;
			const suggestInitialState = canExtendCurrent ? currentItemState : require_internal_annotations.inheritAnnotations(context.state, parser.initialState);
			const suggestFallbackState = unwrapInjectedWrapper(suggestInitialState);
			const hasSuggestFallbackState = suggestFallbackState !== suggestInitialState;
			const collectSelectedValuesSync = () => {
				const selectedValues = /* @__PURE__ */ new Set();
				for (let i = 0; i < context.state.length; i++) {
					const childContext = require_execution_context.withChildContext(context, i, context.state[i]);
					const completed = completeSyncWithUnwrappedFallback(childContext.state, childContext.exec);
					if (completed.success) selectedValues.add(String(unwrapInjectedWrapper(completed.value)));
				}
				return selectedValues;
			};
			const collectSelectedValuesAsync = async () => {
				const selectedValues = /* @__PURE__ */ new Set();
				for (let i = 0; i < context.state.length; i++) {
					const childContext = require_execution_context.withChildContext(context, i, context.state[i]);
					const completed = await completeAsyncWithUnwrappedFallback(childContext.state, childContext.exec);
					if (completed.success) selectedValues.add(String(unwrapInjectedWrapper(completed.value)));
				}
				return selectedValues;
			};
			const shouldInclude = (selectedValues, suggestion) => {
				if (suggestion.kind === "literal") return !selectedValues.has(suggestion.text);
				return true;
			};
			const suggestionKey = (suggestion) => {
				const description = suggestion.description == null ? "" : require_message.formatMessage(suggestion.description);
				if (suggestion.kind === "literal") return JSON.stringify([
					"literal",
					suggestion.text,
					description
				]);
				return JSON.stringify([
					"file",
					suggestion.type,
					suggestion.pattern ?? "",
					suggestion.includeHidden === true,
					suggestion.extensions == null ? "" : suggestion.extensions.toSorted().join("\0"),
					description
				]);
			};
			return require_mode_dispatch.dispatchIterableByMode(parser.mode, function* () {
				const selectedValues = collectSelectedValuesSync();
				const emitted = /* @__PURE__ */ new Set();
				const yieldUnique = function* (suggestions) {
					for (const s of suggestions) {
						const key = suggestionKey(s);
						if (shouldInclude(selectedValues, s) && !emitted.has(key)) {
							emitted.add(key);
							yield s;
						}
					}
				};
				let shouldTryFallback = false;
				try {
					yield* yieldUnique(syncParser.suggest(require_execution_context.withChildContext(context, itemIndex, suggestInitialState), prefix));
				} catch (error) {
					if (!hasSuggestFallbackState) throw error;
					shouldTryFallback = true;
				}
				if (shouldTryFallback) yield* yieldUnique(syncParser.suggest(require_execution_context.withChildContext(context, itemIndex, suggestFallbackState), prefix));
			}, async function* () {
				const selectedValues = await collectSelectedValuesAsync();
				const emitted = /* @__PURE__ */ new Set();
				const yieldUnique = async function* (suggestions) {
					for await (const s of suggestions) {
						const key = suggestionKey(s);
						if (shouldInclude(selectedValues, s) && !emitted.has(key)) {
							emitted.add(key);
							yield s;
						}
					}
				};
				let shouldTryFallback = false;
				try {
					yield* yieldUnique(parser.suggest(require_execution_context.withChildContext(context, itemIndex, suggestInitialState), prefix));
				} catch (error) {
					if (!hasSuggestFallbackState) throw error;
					shouldTryFallback = true;
				}
				if (shouldTryFallback) yield* yieldUnique(parser.suggest(require_execution_context.withChildContext(context, itemIndex, suggestFallbackState), prefix));
			});
		},
		getDocFragments(state, defaultValue) {
			const latestState = state.kind === "available" && state.state.length > 0 ? state.state.at(-1) : void 0;
			const latestInnerState = latestState != null && require_internal_annotations.isInjectedAnnotationWrapper(latestState) ? unwrapInjectedWrapper(latestState) : latestState;
			const innerState = state.kind === "unavailable" ? { kind: "unavailable" } : latestInnerState !== void 0 ? {
				kind: "available",
				state: latestInnerState
			} : { kind: "unavailable" };
			return syncParser.getDocFragments(innerState, defaultValue != null && defaultValue.length > 0 ? defaultValue[0] : void 0);
		}
	};
	function validateMultipleResult(result, deferredIndices, hasDeferred = false) {
		if (result.length < min) {
			const customMessage = options.errors?.tooFew;
			return {
				success: false,
				error: customMessage ? typeof customMessage === "function" ? customMessage(min, result.length) : customMessage : require_message.message`Expected at least ${require_message.text(min.toLocaleString("en"))} values, but got only ${require_message.text(result.length.toLocaleString("en"))}.`
			};
		} else if (result.length > max) {
			const customMessage = options.errors?.tooMany;
			return {
				success: false,
				error: customMessage ? typeof customMessage === "function" ? customMessage(max, result.length) : customMessage : require_message.message`Expected at most ${require_message.text(max.toLocaleString("en"))} values, but got ${require_message.text(result.length.toLocaleString("en"))}.`
			};
		}
		const isDeferred = deferredIndices.size > 0 || hasDeferred;
		return {
			success: true,
			value: result,
			...isDeferred ? {
				deferred: true,
				...deferredIndices.size > 0 ? { deferredKeys: deferredIndices } : {}
			} : {}
		};
	}
	Object.defineProperty(resultParser, "placeholder", {
		get() {
			try {
				if (min > 0 && "placeholder" in parser) return Array.from({ length: min }, () => parser.placeholder);
			} catch {}
			return [];
		},
		configurable: true,
		enumerable: false
	});
	if (typeof parser.normalizeValue === "function") {
		const innerNormalize = parser.normalizeValue.bind(parser);
		Object.defineProperty(resultParser, "normalizeValue", {
			value(values) {
				if (!Array.isArray(values)) return values;
				let changed = false;
				const result = values.map((v) => {
					try {
						const n = innerNormalize(v);
						if (n !== v) changed = true;
						return n;
					} catch {
						return v;
					}
				});
				return changed ? result : values;
			},
			configurable: true,
			enumerable: false
		});
	}
	{
		const innerValidate = typeof parser.validateValue === "function" ? parser.validateValue.bind(parser) : void 0;
		const validateArity = (values) => {
			if (values.length < min) {
				const customMessage = options.errors?.tooFew;
				return {
					success: false,
					error: customMessage ? typeof customMessage === "function" ? customMessage(min, values.length) : customMessage : require_message.message`Expected at least ${require_message.text(min.toLocaleString("en"))} values, but got only ${require_message.text(values.length.toLocaleString("en"))}.`
				};
			}
			if (values.length > max) {
				const customMessage = options.errors?.tooMany;
				return {
					success: false,
					error: customMessage ? typeof customMessage === "function" ? customMessage(max, values.length) : customMessage : require_message.message`Expected at most ${require_message.text(max.toLocaleString("en"))} values, but got ${require_message.text(values.length.toLocaleString("en"))}.`
				};
			}
			return {
				success: true,
				value: values
			};
		};
		Object.defineProperty(resultParser, "validateValue", {
			value(values) {
				if (!Array.isArray(values)) {
					const actualType = values === null ? "null" : typeof values;
					return require_mode_dispatch.wrapForMode(parser.mode, {
						success: false,
						error: require_message.message`Expected an array of values, but received ${actualType}.`
					});
				}
				const arity = validateArity(values);
				if (!arity.success) return require_mode_dispatch.wrapForMode(parser.mode, arity);
				if (innerValidate == null) return require_mode_dispatch.wrapForMode(parser.mode, arity);
				return require_mode_dispatch.dispatchByMode(parser.mode, () => {
					let changed = false;
					const normalized = [];
					for (const v of values) {
						const r = innerValidate(v);
						if (!r.success) return r;
						normalized.push(r.value);
						if (r.value !== v) changed = true;
					}
					return {
						success: true,
						value: changed ? normalized : values
					};
				}, async () => {
					let changed = false;
					const normalized = [];
					for (const v of values) {
						const r = await innerValidate(v);
						if (!r.success) return r;
						normalized.push(r.value);
						if (r.value !== v) changed = true;
					}
					return {
						success: true,
						value: changed ? normalized : values
					};
				});
			},
			configurable: true,
			enumerable: false
		});
	}
	if (parser.dependencyMetadata?.source != null) {
		const innerSource = parser.dependencyMetadata.source;
		Object.defineProperty(resultParser, "dependencyMetadata", {
			value: {
				...parser.dependencyMetadata,
				source: {
					...innerSource,
					preservesSourceValue: false,
					extractSourceValue: (state) => {
						if (!Array.isArray(state)) return innerSource.extractSourceValue(state);
						const scan = (index) => {
							for (let i = index; i >= 0; i--) {
								const result = innerSource.extractSourceValue(state[i]);
								if (result == null) continue;
								if (isPromiseLike(result)) return Promise.resolve(result).then((resolved) => resolved ?? scan(i - 1));
								return result;
							}
							return void 0;
						};
						return scan(state.length - 1);
					}
				}
			},
			configurable: true,
			enumerable: false
		});
	}
	return resultParser;
}
/**
* Creates a parser that requires the wrapped parser to consume at least one
* input token to succeed. If the wrapped parser succeeds without consuming
* any tokens, this parser fails with an error.
*
* This modifier is useful with `longestMatch()` for implementing conditional
* default values. When used together, `nonEmpty()` prevents a parser with
* only default values from matching when no input is provided, allowing
* another branch (like a help command) to match instead.
*
* @template M The execution mode of the parser.
* @template T The type of value produced by the wrapped parser.
* @template TState The type of state used by the wrapped parser.
* @param parser The {@link Parser} that must consume input to succeed.
* @returns A {@link Parser} that fails if the wrapped parser consumes no input.
*
* @example
* ```typescript
* // Without nonEmpty(): activeParser always wins (consumes 0 tokens)
* // With nonEmpty(): helpParser wins when no options are provided
* const activeParser = nonEmpty(object({
*   cwd: withDefault(option("--cwd", string()), "./default"),
*   key: optional(option("--key", string())),
* }));
*
* const helpParser = object({
*   mode: constant("help"),
* });
*
* const parser = longestMatch(activeParser, helpParser);
*
* // cli           → helpParser matches (activeParser fails with nonEmpty)
* // cli --key foo → activeParser matches (consumes tokens)
* ```
*
* @since 0.10.0
*/
function nonEmpty(parser) {
	const syncParser = parser;
	const initialState = parser.initialState;
	const processNonEmptyResult = (result) => {
		if (!result.success) return result;
		if (result.consumed.length === 0) return {
			success: false,
			consumed: 0,
			error: require_message.message`Parser must consume at least one token.`
		};
		return result;
	};
	const parseSync = (context) => {
		const result = syncParser.parse(context);
		return processNonEmptyResult(result);
	};
	const parseAsync = async (context) => {
		const result = await parser.parse(context);
		return processNonEmptyResult(result);
	};
	const nonEmptyParser = {
		mode: parser.mode,
		$valueType: parser.$valueType,
		$stateType: parser.$stateType,
		priority: parser.priority,
		usage: parser.usage,
		leadingNames: parser.leadingNames,
		acceptingAnyToken: parser.acceptingAnyToken,
		initialState,
		...typeof parser.canSkip === "function" ? { canSkip(state, exec) {
			const unwrappedState = require_internal_annotations.unwrapInjectedAnnotationWrapper(state);
			if (unwrappedState === initialState) return false;
			return parser.canSkip?.(unwrappedState, exec) === true;
		} } : {},
		...typeof parser.shouldDeferCompletion === "function" ? { shouldDeferCompletion: parser.shouldDeferCompletion.bind(parser) } : {},
		getSuggestRuntimeNodes(state, path) {
			return parser.getSuggestRuntimeNodes?.(state, path) ?? (parser.dependencyMetadata?.source != null ? [{
				path,
				parser,
				state
			}] : []);
		},
		parse(context) {
			return require_mode_dispatch.dispatchByMode(parser.mode, () => parseSync(context), () => parseAsync(context));
		},
		complete(state, exec) {
			return parser.complete(state, exec);
		},
		suggest(context, prefix) {
			return parser.suggest(context, prefix);
		},
		getDocFragments(state, defaultValue) {
			return syncParser.getDocFragments(state, defaultValue);
		}
	};
	if ("placeholder" in parser) Object.defineProperty(nonEmptyParser, "placeholder", {
		get() {
			return parser.placeholder;
		},
		configurable: true,
		enumerable: false
	});
	Object.defineProperty(nonEmptyParser, require_phase2_seed.extractPhase2SeedKey, {
		value(state, exec) {
			return require_phase2_seed.extractPhase2Seed(parser, state, exec);
		},
		configurable: true,
		enumerable: false
	});
	if (typeof parser.normalizeValue === "function") Object.defineProperty(nonEmptyParser, "normalizeValue", {
		value: parser.normalizeValue.bind(parser),
		configurable: true,
		enumerable: false
	});
	if (typeof parser.validateValue === "function") Object.defineProperty(nonEmptyParser, "validateValue", {
		value: parser.validateValue.bind(parser),
		configurable: true,
		enumerable: false
	});
	return nonEmptyParser;
}

//#endregion
exports.WithDefaultError = WithDefaultError;
exports.map = map;
exports.multiple = multiple;
exports.nonEmpty = nonEmpty;
exports.optional = optional;
exports.withDefault = withDefault;
