// @optique/core/dist/primitives.cjs
const require_internal_annotations = require('./internal/annotations.cjs');
const require_message = require('./message.cjs');
const require_validate = require('./validate.cjs');
const require_usage = require('./usage.cjs');
const require_mode_dispatch = require('./internal/mode-dispatch.cjs');
const require_internal_dependency = require('./internal/dependency.cjs');
const require_dependency_runtime = require('./dependency-runtime.cjs');
const require_annotation_state = require('./annotation-state.cjs');
const require_command_alias = require('./internal/command-alias.cjs');
const require_execution_context = require('./execution-context.cjs');
const require_phase2_seed = require('./phase2-seed.cjs');
const require_suggestion = require('./suggestion.cjs');
const require_usage_internals = require('./usage-internals.cjs');
const require_dependency_metadata = require('./dependency-metadata.cjs');
const require_valueparser = require('./valueparser.cjs');

//#region src/primitives.ts
/**
* A shared empty set used as the `leadingNames` value for parsers that
* do not match any specific name at the first buffer position.
*/
const EMPTY_LEADING_NAMES = /* @__PURE__ */ new Set();
function hasParsedOptionValue(state, valueParser) {
	if (valueParser != null) return state != null && typeof state === "object" && "success" in state && typeof state.success === "boolean";
	return state != null && "success" in state && state.success && state.value === true;
}
function isTerminalValueState(state) {
	return state != null && typeof state === "object" && "success" in state && typeof state.success === "boolean";
}
/**
* Helper function to create the stored state for an option or argument value.
*
* Primitive parsers now keep only the plain parse result as their local state.
* Any dependency-aware replay information is recorded separately in the
* execution trace.
* @internal
*/
function createOptionParseState(parseResult) {
	return parseResult;
}
function buildTraceEntry(kind, rawInput, consumed, valueParser, parseResult, optionNames$1) {
	const entry = {
		kind,
		rawInput,
		consumed,
		preliminaryResult: parseResult,
		...optionNames$1 != null ? { optionNames: optionNames$1 } : {},
		metavar: valueParser.metavar
	};
	if (require_internal_dependency.isDerivedValueParser(valueParser)) {
		const defaults = require_internal_dependency.getSnapshottedDefaultDependencyValues(parseResult);
		if (defaults != null) return {
			...entry,
			defaultDependencyValues: defaults
		};
	}
	return entry;
}
function recordTrace(context, entry) {
	if (context.exec?.trace == null) return context;
	const trace = context.exec.trace.set(context.exec.path, entry);
	return {
		...context,
		trace,
		exec: {
			...context.exec,
			trace
		}
	};
}
function resolveDerivedCompletionSync(derivedMetadata, state, exec) {
	if (derivedMetadata?.derived == null || exec?.dependencyRuntime == null) return state;
	const traceEntry = exec.trace?.get(exec.path);
	if (traceEntry?.rawInput == null) return traceEntry?.preliminaryResult ?? state;
	if (traceEntry.preliminaryResult != null) {
		const resolution = exec.dependencyRuntime.resolveDependencies({
			dependencyIds: derivedMetadata.derived.dependencyIds,
			defaultValues: traceEntry.defaultDependencyValues
		});
		if (resolution.kind === "resolved" && resolution.usedDefaults.length > 0 && resolution.usedDefaults.every((used) => used)) return traceEntry.preliminaryResult;
	}
	const replayed = require_dependency_runtime.replayDerivedParser({
		path: exec.path,
		parser: { dependencyMetadata: { derived: derivedMetadata.derived } },
		state,
		defaultDependencyValues: traceEntry.defaultDependencyValues
	}, traceEntry.rawInput, exec.dependencyRuntime);
	return replayed ?? traceEntry.preliminaryResult ?? state;
}
async function resolveDerivedCompletionAsync(derivedMetadata, state, exec) {
	if (derivedMetadata?.derived == null || exec?.dependencyRuntime == null) return state;
	const traceEntry = exec.trace?.get(exec.path);
	if (traceEntry?.rawInput == null) return traceEntry?.preliminaryResult ?? state;
	if (traceEntry.preliminaryResult != null) {
		const resolution = exec.dependencyRuntime.resolveDependencies({
			dependencyIds: derivedMetadata.derived.dependencyIds,
			defaultValues: traceEntry.defaultDependencyValues
		});
		if (resolution.kind === "resolved" && resolution.usedDefaults.length > 0 && resolution.usedDefaults.every((used) => used)) return traceEntry.preliminaryResult;
	}
	const replayed = await require_dependency_runtime.replayDerivedParserAsync({
		path: exec.path,
		parser: { dependencyMetadata: { derived: derivedMetadata.derived } },
		state,
		defaultDependencyValues: traceEntry.defaultDependencyValues
	}, traceEntry.rawInput, exec.dependencyRuntime);
	return replayed ?? traceEntry.preliminaryResult ?? state;
}
/**
* Creates a parser that always succeeds without consuming any input and
* produces a constant value of the type {@link T}.
* @template T The type of the constant value produced by the parser.
*/
function constant(value) {
	const result = {
		$valueType: [],
		$stateType: [],
		mode: "sync",
		priority: 0,
		usage: [],
		leadingNames: EMPTY_LEADING_NAMES,
		acceptingAnyToken: false,
		initialState: value,
		canSkip() {
			return true;
		},
		parse(context) {
			return {
				success: true,
				next: context,
				consumed: []
			};
		},
		complete(state, _exec) {
			return {
				success: true,
				value: state
			};
		},
		suggest(_context, _prefix) {
			return [];
		},
		getDocFragments(_state, _defaultValue) {
			return { fragments: [] };
		}
	};
	Object.defineProperty(result, "placeholder", {
		value,
		configurable: true,
		enumerable: false,
		writable: false
	});
	return result;
}
/**
* Creates a parser that always fails without consuming any input.
*
* This is the counterpart to {@link constant}: while `constant(value)` always
* succeeds, `fail<T>()` always fails.  At the type level it is declared to
* produce a value of type `T`, so it composes seamlessly with other parsers
* that expect `Parser<"sync", T, …>`.
*
* The primary use case is as the inner parser for
* `bindConfig(fail<T>(), { … })` when a value should *only* come from a
* config file and has no corresponding CLI flag or argument.  Because `fail()`
* always fails, `bindConfig` will always fall back to the config file (or the
* supplied default).
*
* @template T The type of value this parser is declared to produce.
* @returns A {@link Parser} that always fails at parse time and always fails at
*          complete time.
* @since 1.0.0
*/
function fail() {
	return {
		$valueType: [],
		$stateType: [],
		mode: "sync",
		priority: 0,
		usage: [],
		leadingNames: EMPTY_LEADING_NAMES,
		acceptingAnyToken: false,
		initialState: void 0,
		canSkip() {
			return false;
		},
		parse(_context) {
			return {
				success: false,
				consumed: 0,
				error: require_message.message`No value provided.`
			};
		},
		complete(_state, _exec) {
			return {
				success: false,
				error: require_message.message`No value provided.`
			};
		},
		suggest(_context, _prefix) {
			return [];
		},
		getDocFragments(_state, _defaultValue) {
			return { fragments: [] };
		}
	};
}
/**
* Internal helper to get suggestions from a value parser, using dependency values
* if the parser is a derived parser and dependency values are available.
* @internal
*/
function* getSuggestionsWithDependency(valueParser, prefix, dependencyRegistry, exec) {
	if (!valueParser.suggest) return;
	if (require_internal_dependency.isDerivedValueParser(valueParser) && require_internal_dependency.suggestWithDependency in valueParser) {
		const derived = valueParser;
		const suggestWithDep = derived[require_internal_dependency.suggestWithDependency];
		if (suggestWithDep && dependencyRegistry) {
			const dependencyRuntime = exec?.dependencyRuntime;
			const depIds = require_internal_dependency.getDependencyIds(derived);
			const defaultsFn = require_internal_dependency.getDefaultValuesFunction(derived);
			const defaults = defaultsFn?.();
			const dependencyValues = [];
			let hasAnyValue = false;
			for (let i = 0; i < depIds.length; i++) {
				const depId = depIds[i];
				if (dependencyRegistry.has(depId)) {
					dependencyValues.push(dependencyRegistry.get(depId));
					hasAnyValue = true;
				} else if (dependencyRuntime?.isSourceFailed(depId)) return;
				else if (defaults && i < defaults.length) dependencyValues.push(defaults[i]);
				else {
					yield* valueParser.suggest(prefix);
					return;
				}
			}
			if (hasAnyValue) {
				const depValue = depIds.length === 1 ? dependencyValues[0] : dependencyValues;
				yield* suggestWithDep(prefix, depValue);
				return;
			}
		}
	}
	yield* valueParser.suggest(prefix);
}
/**
* Internal sync helper for option suggest functionality.
* @internal
*/
function* suggestOptionSync(optionNames$1, valueParser, hidden, context, prefix) {
	if (hidden) return;
	const equalsIndex = prefix.indexOf("=");
	if (equalsIndex >= 0) {
		const optionPart = prefix.slice(0, equalsIndex);
		const valuePart = prefix.slice(equalsIndex + 1);
		if (optionNames$1.includes(optionPart)) {
			if (valueParser && valueParser.suggest) {
				const valueSuggestions = getSuggestionsWithDependency(valueParser, valuePart, context.dependencyRegistry, context.exec);
				for (const suggestion of valueSuggestions) if (suggestion.kind === "literal") yield {
					kind: "literal",
					text: `${optionPart}=${suggestion.text}`,
					description: suggestion.description
				};
				else yield {
					kind: "literal",
					text: `${optionPart}=${suggestion.pattern || ""}`,
					description: suggestion.description
				};
			}
		}
	} else {
		const expectingValue = context.buffer.length > 0 && optionNames$1.includes(context.buffer[context.buffer.length - 1]);
		if (!expectingValue && (prefix.startsWith("--") || prefix.startsWith("-") || prefix.startsWith("/") || prefix.startsWith("+"))) {
			for (const optionName$1 of optionNames$1) if (optionName$1.startsWith(prefix)) {
				if (prefix === "-" && optionName$1.length !== 2) continue;
				yield {
					kind: "literal",
					text: optionName$1
				};
			}
		}
		if (valueParser && valueParser.suggest) {
			let shouldSuggestValues = false;
			if (context.buffer.length > 0) {
				const lastToken = context.buffer[context.buffer.length - 1];
				if (optionNames$1.includes(lastToken)) shouldSuggestValues = true;
			} else if (require_annotation_state.isAnnotationWrappedInitialState(context.state) && context.buffer.length === 0 && (context.exec?.path?.length ?? 0) === 0 && !(prefix.startsWith("--") || prefix.startsWith("-") || prefix.startsWith("/") || prefix.startsWith("+"))) shouldSuggestValues = true;
			if (shouldSuggestValues) yield* getSuggestionsWithDependency(valueParser, prefix, context.dependencyRegistry, context.exec);
		}
	}
}
/**
* Internal async helper to get suggestions from a value parser, using dependency values
* if the parser is a derived parser and dependency values are available.
* @internal
*/
async function* getSuggestionsWithDependencyAsync(valueParser, prefix, dependencyRegistry, exec) {
	if (!valueParser.suggest) return;
	if (require_internal_dependency.isDerivedValueParser(valueParser) && require_internal_dependency.suggestWithDependency in valueParser) {
		const derived = valueParser;
		const suggestWithDep = derived[require_internal_dependency.suggestWithDependency];
		if (suggestWithDep && dependencyRegistry) {
			const dependencyRuntime = exec?.dependencyRuntime;
			const depIds = require_internal_dependency.getDependencyIds(derived);
			const defaultsFn = require_internal_dependency.getDefaultValuesFunction(derived);
			const defaults = defaultsFn?.();
			const dependencyValues = [];
			let hasAnyValue = false;
			for (let i = 0; i < depIds.length; i++) {
				const depId = depIds[i];
				if (dependencyRegistry.has(depId)) {
					dependencyValues.push(dependencyRegistry.get(depId));
					hasAnyValue = true;
				} else if (dependencyRuntime?.isSourceFailed(depId)) return;
				else if (defaults && i < defaults.length) dependencyValues.push(defaults[i]);
				else {
					for await (const suggestion of valueParser.suggest(prefix)) yield suggestion;
					return;
				}
			}
			if (hasAnyValue) {
				const depValue = depIds.length === 1 ? dependencyValues[0] : dependencyValues;
				for await (const suggestion of suggestWithDep(prefix, depValue)) yield suggestion;
				return;
			}
		}
	}
	for await (const suggestion of valueParser.suggest(prefix)) yield suggestion;
}
/**
* Internal async helper for option suggest functionality.
* @internal
*/
async function* suggestOptionAsync(optionNames$1, valueParser, hidden, context, prefix) {
	if (hidden) return;
	const equalsIndex = prefix.indexOf("=");
	if (equalsIndex >= 0) {
		const optionPart = prefix.slice(0, equalsIndex);
		const valuePart = prefix.slice(equalsIndex + 1);
		if (optionNames$1.includes(optionPart)) {
			if (valueParser && valueParser.suggest) {
				const valueSuggestions = getSuggestionsWithDependencyAsync(valueParser, valuePart, context.dependencyRegistry, context.exec);
				for await (const suggestion of valueSuggestions) if (suggestion.kind === "literal") yield {
					kind: "literal",
					text: `${optionPart}=${suggestion.text}`,
					description: suggestion.description
				};
				else yield {
					kind: "literal",
					text: `${optionPart}=${suggestion.pattern || ""}`,
					description: suggestion.description
				};
			}
		}
	} else {
		const expectingValue = context.buffer.length > 0 && optionNames$1.includes(context.buffer[context.buffer.length - 1]);
		if (!expectingValue && (prefix.startsWith("--") || prefix.startsWith("-") || prefix.startsWith("/") || prefix.startsWith("+"))) {
			for (const optionName$1 of optionNames$1) if (optionName$1.startsWith(prefix)) {
				if (prefix === "-" && optionName$1.length !== 2) continue;
				yield {
					kind: "literal",
					text: optionName$1
				};
			}
		}
		if (valueParser && valueParser.suggest) {
			let shouldSuggestValues = false;
			if (context.buffer.length > 0) {
				const lastToken = context.buffer[context.buffer.length - 1];
				if (optionNames$1.includes(lastToken)) shouldSuggestValues = true;
			} else if (require_annotation_state.isAnnotationWrappedInitialState(context.state) && context.buffer.length === 0 && (context.exec?.path?.length ?? 0) === 0 && !(prefix.startsWith("--") || prefix.startsWith("-") || prefix.startsWith("/") || prefix.startsWith("+"))) shouldSuggestValues = true;
			if (shouldSuggestValues) for await (const suggestion of getSuggestionsWithDependencyAsync(valueParser, prefix, context.dependencyRegistry, context.exec)) yield suggestion;
		}
	}
}
/**
* Internal sync helper for argument suggest functionality.
* @internal
*/
function* suggestArgumentSync(valueParser, hidden, prefix, dependencyRegistry, exec) {
	if (hidden) return;
	if (valueParser.suggest) yield* getSuggestionsWithDependency(valueParser, prefix, dependencyRegistry, exec);
}
/**
* Internal async helper for argument suggest functionality.
* @internal
*/
async function* suggestArgumentAsync(valueParser, hidden, prefix, dependencyRegistry, exec) {
	if (hidden) return;
	if (valueParser.suggest) yield* getSuggestionsWithDependencyAsync(valueParser, prefix, dependencyRegistry, exec);
}
function option(...args) {
	const lastArg = args.at(-1);
	const secondLastArg = args.at(-2);
	let valueParser;
	let optionNames$1;
	let options = {};
	if (require_valueparser.isValueParser(lastArg)) {
		valueParser = lastArg;
		optionNames$1 = args.slice(0, -1);
	} else if (typeof lastArg === "object" && lastArg != null) {
		options = lastArg;
		if (require_valueparser.isValueParser(secondLastArg)) {
			valueParser = secondLastArg;
			optionNames$1 = args.slice(0, -2);
		} else {
			valueParser = void 0;
			optionNames$1 = args.slice(0, -1);
		}
	} else {
		optionNames$1 = args;
		valueParser = void 0;
	}
	require_validate.validateOptionNames(optionNames$1, "Option");
	const mode = valueParser?.mode ?? "sync";
	const isAsync = mode === "async";
	const syncValueParser = valueParser;
	const dependencyMetadata = valueParser != null ? require_dependency_metadata.extractDependencyMetadata(valueParser) : void 0;
	const formatInvalidValueError = (error) => options.errors?.invalidValue ? typeof options.errors.invalidValue === "function" ? options.errors.invalidValue(error) : options.errors.invalidValue : require_message.message`${require_message.optionNames(optionNames$1)}: ${error}`;
	const result = {
		mode,
		$valueType: [],
		$stateType: [],
		priority: 10,
		usage: [valueParser == null ? {
			type: "optional",
			terms: [{
				type: "option",
				names: optionNames$1,
				...options.hidden != null && { hidden: options.hidden }
			}]
		} : {
			type: "option",
			names: optionNames$1,
			metavar: valueParser.metavar,
			...options.hidden != null && { hidden: options.hidden }
		}],
		leadingNames: new Set(optionNames$1),
		acceptingAnyToken: false,
		initialState: valueParser == null ? {
			success: true,
			value: false
		} : void 0,
		canSkip(state) {
			return valueParser == null || isTerminalValueState(state);
		},
		parse(context) {
			if (context.optionsTerminated) return {
				success: false,
				consumed: 0,
				error: options.errors?.optionsTerminated ?? require_message.message`No more options can be parsed.`
			};
			else if (context.buffer.length < 1) return {
				success: false,
				consumed: 0,
				error: options.errors?.endOfInput ?? require_message.message`Expected an option, but got end of input.`
			};
			if (context.buffer[0] === "--") return {
				success: true,
				next: {
					...context,
					buffer: context.buffer.slice(1),
					state: context.state,
					optionsTerminated: true
				},
				consumed: context.buffer.slice(0, 1)
			};
			if (optionNames$1.includes(context.buffer[0])) {
				if (hasParsedOptionValue(context.state, valueParser)) return {
					success: false,
					consumed: 1,
					error: options.errors?.duplicate ? typeof options.errors.duplicate === "function" ? options.errors.duplicate(context.buffer[0]) : options.errors.duplicate : require_message.message`${require_message.optionName(context.buffer[0])} cannot be used multiple times.`
				};
				if (valueParser == null) return {
					success: true,
					next: {
						...context,
						state: {
							success: true,
							value: true
						},
						buffer: context.buffer.slice(1)
					},
					consumed: context.buffer.slice(0, 1)
				};
				if (context.buffer.length < 2) return {
					success: false,
					consumed: 1,
					error: options.errors?.endOfInput ?? require_message.message`${require_message.optionName(context.buffer[0])} requires ${require_message.metavar(valueParser.metavar)}.`
				};
				const rawInput = context.buffer[1];
				return require_mode_dispatch.dispatchByMode(mode, () => {
					const parseResult = syncValueParser.parse(rawInput);
					const next = recordTrace(context, buildTraceEntry("option-value", rawInput, context.buffer.slice(0, 2), syncValueParser, parseResult, optionNames$1));
					return {
						success: true,
						next: {
							...next,
							state: createOptionParseState(parseResult),
							buffer: context.buffer.slice(2)
						},
						consumed: context.buffer.slice(0, 2)
					};
				}, async () => {
					const parseResult = await valueParser.parse(rawInput);
					const next = recordTrace(context, buildTraceEntry("option-value", rawInput, context.buffer.slice(0, 2), valueParser, parseResult, optionNames$1));
					return {
						success: true,
						next: {
							...next,
							state: createOptionParseState(parseResult),
							buffer: context.buffer.slice(2)
						},
						consumed: context.buffer.slice(0, 2)
					};
				});
			}
			const prefixes = optionNames$1.filter((name) => name.startsWith("--") || name.startsWith("/") || name.startsWith("-") && name.length > 2).map((name) => name.startsWith("/") ? `${name}:` : `${name}=`);
			for (const prefix of prefixes) {
				if (!context.buffer[0].startsWith(prefix)) continue;
				if (hasParsedOptionValue(context.state, valueParser)) {
					const optionName$1 = prefix.slice(0, -1);
					return {
						success: false,
						consumed: 1,
						error: options.errors?.duplicate ? typeof options.errors.duplicate === "function" ? options.errors.duplicate(prefix) : options.errors.duplicate : require_message.message`${require_message.optionName(optionName$1)} cannot be used multiple times.`
					};
				}
				const rawInput = context.buffer[0].slice(prefix.length);
				if (valueParser == null) return {
					success: false,
					consumed: 1,
					error: options.errors?.unexpectedValue ? typeof options.errors.unexpectedValue === "function" ? options.errors.unexpectedValue(rawInput) : options.errors.unexpectedValue : require_message.message`Option ${require_message.optionName(prefix)} is a Boolean flag, but got a value: ${rawInput}.`
				};
				return require_mode_dispatch.dispatchByMode(mode, () => {
					const parseResult = syncValueParser.parse(rawInput);
					const next = recordTrace(context, buildTraceEntry("option-value", rawInput, context.buffer.slice(0, 1), syncValueParser, parseResult, optionNames$1));
					return {
						success: true,
						next: {
							...next,
							state: createOptionParseState(parseResult),
							buffer: context.buffer.slice(1)
						},
						consumed: context.buffer.slice(0, 1)
					};
				}, async () => {
					const parseResult = await valueParser.parse(rawInput);
					const next = recordTrace(context, buildTraceEntry("option-value", rawInput, context.buffer.slice(0, 1), valueParser, parseResult, optionNames$1));
					return {
						success: true,
						next: {
							...next,
							state: createOptionParseState(parseResult),
							buffer: context.buffer.slice(1)
						},
						consumed: context.buffer.slice(0, 1)
					};
				});
			}
			if (valueParser == null) {
				const shortOptions = optionNames$1.filter((name) => name.match(/^-[^-]$/));
				for (const shortOption of shortOptions) {
					if (!context.buffer[0].startsWith(shortOption)) continue;
					if (hasParsedOptionValue(context.state, valueParser)) return {
						success: false,
						consumed: 1,
						error: options.errors?.duplicate ? typeof options.errors.duplicate === "function" ? options.errors.duplicate(shortOption) : options.errors.duplicate : require_message.message`${require_message.optionName(shortOption)} cannot be used multiple times.`
					};
					return {
						success: true,
						next: {
							...context,
							state: {
								success: true,
								value: true
							},
							buffer: [`-${context.buffer[0].slice(2)}`, ...context.buffer.slice(1)]
						},
						consumed: [context.buffer[0].slice(0, 2)]
					};
				}
			}
			const invalidOption = context.buffer[0];
			if (options.errors?.noMatch) {
				const candidates = /* @__PURE__ */ new Set();
				for (const name of require_usage.extractOptionNames(context.usage)) candidates.add(name);
				const suggestions = require_suggestion.findSimilar(invalidOption, candidates, require_suggestion.DEFAULT_FIND_SIMILAR_OPTIONS);
				const errorMessage = typeof options.errors.noMatch === "function" ? options.errors.noMatch(invalidOption, suggestions) : options.errors.noMatch;
				return {
					success: false,
					consumed: 0,
					error: errorMessage
				};
			}
			const baseError = require_message.message`No matched option for ${require_message.optionName(invalidOption)}.`;
			return {
				success: false,
				consumed: 0,
				error: require_suggestion.createErrorWithSuggestions(baseError, invalidOption, context.usage, "option")
			};
		},
		complete(state, exec) {
			const missing = valueParser == null ? {
				success: true,
				value: false
			} : {
				success: false,
				error: options.errors?.missing ? typeof options.errors.missing === "function" ? options.errors.missing(optionNames$1) : options.errors.missing : require_message.message`Missing option ${require_message.optionNames(optionNames$1)}.`
			};
			const completeSync = () => {
				if (state == null) return missing;
				const resolvedState = valueParser != null && dependencyMetadata?.derived != null ? resolveDerivedCompletionSync(dependencyMetadata, state, exec) : state;
				return resolvedState.success ? resolvedState : {
					success: false,
					error: formatInvalidValueError(resolvedState.error)
				};
			};
			const completeAsync = async () => {
				if (state == null) return missing;
				if (valueParser == null || dependencyMetadata?.derived == null) return completeSync();
				const resolved = await resolveDerivedCompletionAsync(dependencyMetadata, state, exec);
				return resolved.success ? resolved : {
					success: false,
					error: formatInvalidValueError(resolved.error)
				};
			};
			return require_mode_dispatch.dispatchByMode(mode, completeSync, completeAsync);
		},
		suggest(context, prefix) {
			if (isAsync) return suggestOptionAsync(optionNames$1, valueParser, require_usage.isSuggestionHidden(options.hidden), context, prefix);
			return suggestOptionSync(optionNames$1, valueParser, require_usage.isSuggestionHidden(options.hidden), context, prefix);
		},
		getDocFragments(_state, defaultValue) {
			if (require_usage.isDocHidden(options.hidden)) return {
				fragments: [],
				description: options.description
			};
			const choicesMessage = valueParser?.choices != null && valueParser.choices.length > 0 ? require_message.valueSet(valueParser.choices.map((c) => valueParser.format(c)), {
				fallback: "",
				type: "unit"
			}) : void 0;
			const fragments = [{
				type: "entry",
				term: {
					type: "option",
					names: optionNames$1,
					metavar: valueParser?.metavar
				},
				description: options.description,
				default: defaultValue != null && valueParser != null ? require_message.message`${valueParser.format(defaultValue)}` : void 0,
				choices: choicesMessage
			}];
			return {
				fragments,
				description: options.description
			};
		},
		[Symbol.for("Deno.customInspect")]() {
			return `option(${optionNames$1.map((o) => JSON.stringify(o)).join(", ")})`;
		}
	};
	if (valueParser != null && typeof valueParser.normalize === "function") {
		const normalize = valueParser.normalize.bind(valueParser);
		Object.defineProperty(result, "normalizeValue", {
			value(v) {
				try {
					return normalize(v);
				} catch {
					return v;
				}
			},
			configurable: true,
			enumerable: false
		});
	}
	if (valueParser == null) Object.defineProperty(result, "validateValue", {
		value(v) {
			if (typeof v !== "boolean") {
				const actualType = v === null ? "null" : typeof v;
				return require_mode_dispatch.wrapForMode(mode, {
					success: false,
					error: formatInvalidValueError(require_message.message`Expected a boolean value, but received ${actualType}.`)
				});
			}
			return require_mode_dispatch.wrapForMode(mode, {
				success: true,
				value: v
			});
		},
		configurable: true,
		enumerable: false
	});
	else if (!require_internal_dependency.isDerivedValueParser(valueParser)) {
		const vp = valueParser;
		const wrapParseResult = (parsed) => parsed.success ? parsed : {
			success: false,
			error: formatInvalidValueError(parsed.error)
		};
		Object.defineProperty(result, "validateValue", {
			value(v) {
				if (typeof vp.validate === "function") return require_mode_dispatch.wrapForMode(mode, wrapParseResult(vp.validate(v)));
				let stringified;
				try {
					stringified = vp.format(v);
				} catch {
					return require_mode_dispatch.wrapForMode(mode, {
						success: true,
						value: v
					});
				}
				if (typeof stringified !== "string") return require_mode_dispatch.wrapForMode(mode, {
					success: true,
					value: v
				});
				return require_mode_dispatch.dispatchByMode(mode, () => wrapParseResult(vp.parse(stringified)), async () => wrapParseResult(await vp.parse(stringified)));
			},
			configurable: true,
			enumerable: false
		});
	}
	if (valueParser != null) Object.defineProperty(result, "placeholder", {
		get() {
			try {
				return valueParser.placeholder;
			} catch {
				return void 0;
			}
		},
		configurable: true,
		enumerable: false
	});
	else Object.defineProperty(result, "placeholder", {
		value: false,
		configurable: true,
		enumerable: false,
		writable: false
	});
	if (dependencyMetadata != null) Object.defineProperty(result, "dependencyMetadata", {
		value: dependencyMetadata,
		configurable: true,
		enumerable: false
	});
	return result;
}
/**
* Creates a parser for command-line flags that must be explicitly provided.
* Unlike {@link option}, this parser fails if the flag is not present, making
* it suitable for required boolean flags that don't have a meaningful default.
*
* The key difference from {@link option} is:
* - {@link option} without a value parser: Returns `false` when not present
* - {@link flag}: Fails parsing when not present, only produces `true`
*
* This is useful for dependent options where the presence of a flag changes
* the shape of the result type.
*
* @param args The {@link OptionName}s to parse, followed by an optional
*             {@link FlagOptions} object that allows you to specify
*             a description or other metadata.
* @returns A {@link Parser} that produces `true` when the flag is present
*          and fails when it is not present.
*
* @example
* ```typescript
* // Basic flag usage
* const parser = flag("-f", "--force");
* // Succeeds with true: parse(parser, ["-f"])
* // Fails: parse(parser, [])
*
* // With description
* const verboseFlag = flag("-v", "--verbose", {
*   description: "Enable verbose output"
* });
* ```
*/
function flag(...args) {
	const lastArg = args.at(-1);
	let optionNames$1;
	let options = {};
	if (typeof lastArg === "object" && lastArg != null && !Array.isArray(lastArg)) {
		options = lastArg;
		optionNames$1 = args.slice(0, -1);
	} else optionNames$1 = args;
	require_validate.validateOptionNames(optionNames$1, "Flag");
	const result = {
		$valueType: [],
		$stateType: [],
		mode: "sync",
		priority: 10,
		usage: [{
			type: "option",
			names: optionNames$1,
			...options.hidden != null && { hidden: options.hidden }
		}],
		leadingNames: new Set(optionNames$1),
		acceptingAnyToken: false,
		initialState: void 0,
		canSkip(state) {
			return isTerminalValueState(state);
		},
		parse(context) {
			if (context.optionsTerminated) return {
				success: false,
				consumed: 0,
				error: options.errors?.optionsTerminated ?? require_message.message`No more options can be parsed.`
			};
			else if (context.buffer.length < 1) return {
				success: false,
				consumed: 0,
				error: options.errors?.endOfInput ?? require_message.message`Expected an option, but got end of input.`
			};
			if (context.buffer[0] === "--") return {
				success: true,
				next: {
					...context,
					buffer: context.buffer.slice(1),
					state: context.state,
					optionsTerminated: true
				},
				consumed: context.buffer.slice(0, 1)
			};
			if (optionNames$1.includes(context.buffer[0])) {
				if (context.state?.success) return {
					success: false,
					consumed: 1,
					error: options.errors?.duplicate ? typeof options.errors.duplicate === "function" ? options.errors.duplicate(context.buffer[0]) : options.errors.duplicate : require_message.message`${require_message.optionName(context.buffer[0])} cannot be used multiple times.`
				};
				return {
					success: true,
					next: {
						...context,
						state: {
							success: true,
							value: true
						},
						buffer: context.buffer.slice(1)
					},
					consumed: context.buffer.slice(0, 1)
				};
			}
			const prefixes = optionNames$1.filter((name) => name.startsWith("--") || name.startsWith("/") || name.startsWith("-") && name.length > 2).map((name) => name.startsWith("/") ? `${name}:` : `${name}=`);
			for (const prefix of prefixes) if (context.buffer[0].startsWith(prefix)) {
				const value = context.buffer[0].slice(prefix.length);
				return {
					success: false,
					consumed: 1,
					error: require_message.message`Flag ${require_message.optionName(prefix.slice(0, -1))} does not accept a value, but got: ${value}.`
				};
			}
			const shortOptions = optionNames$1.filter((name) => name.match(/^-[^-]$/));
			for (const shortOption of shortOptions) {
				if (!context.buffer[0].startsWith(shortOption)) continue;
				if (context.state?.success) return {
					success: false,
					consumed: 1,
					error: options.errors?.duplicate ? typeof options.errors.duplicate === "function" ? options.errors.duplicate(shortOption) : options.errors.duplicate : require_message.message`${require_message.optionName(shortOption)} cannot be used multiple times.`
				};
				return {
					success: true,
					next: {
						...context,
						state: {
							success: true,
							value: true
						},
						buffer: [`-${context.buffer[0].slice(2)}`, ...context.buffer.slice(1)]
					},
					consumed: [context.buffer[0].slice(0, 2)]
				};
			}
			const invalidOption = context.buffer[0];
			if (options.errors?.noMatch) {
				const candidates = /* @__PURE__ */ new Set();
				for (const name of require_usage.extractOptionNames(context.usage)) candidates.add(name);
				const suggestions = require_suggestion.findSimilar(invalidOption, candidates, require_suggestion.DEFAULT_FIND_SIMILAR_OPTIONS);
				const errorMessage = typeof options.errors.noMatch === "function" ? options.errors.noMatch(invalidOption, suggestions) : options.errors.noMatch;
				return {
					success: false,
					consumed: 0,
					error: errorMessage
				};
			}
			const baseError = require_message.message`No matched option for ${require_message.optionName(invalidOption)}.`;
			return {
				success: false,
				consumed: 0,
				error: require_suggestion.createErrorWithSuggestions(baseError, invalidOption, context.usage, "option")
			};
		},
		complete(state, _exec) {
			if (state == null) return {
				success: false,
				error: options.errors?.missing ? typeof options.errors.missing === "function" ? options.errors.missing(optionNames$1) : options.errors.missing : require_message.message`Required flag ${require_message.optionNames(optionNames$1)} is missing.`
			};
			if (state.success) return {
				success: true,
				value: true
			};
			return {
				success: false,
				error: require_message.message`${require_message.optionNames(optionNames$1)}: ${state.error}`
			};
		},
		suggest(_context, prefix) {
			if (require_usage.isSuggestionHidden(options.hidden)) return [];
			const suggestions = [];
			if (prefix.startsWith("--") || prefix.startsWith("-") || prefix.startsWith("/") || prefix.startsWith("+")) {
				for (const optionName$1 of optionNames$1) if (optionName$1.startsWith(prefix)) {
					if (prefix === "-" && optionName$1.length !== 2) continue;
					suggestions.push({
						kind: "literal",
						text: optionName$1
					});
				}
			}
			return suggestions;
		},
		getDocFragments(_state, _defaultValue) {
			if (require_usage.isDocHidden(options.hidden)) return {
				fragments: [],
				description: options.description
			};
			const fragments = [{
				type: "entry",
				term: {
					type: "option",
					names: optionNames$1
				},
				description: options.description
			}];
			return {
				fragments,
				description: options.description
			};
		},
		[Symbol.for("Deno.customInspect")]() {
			return `flag(${optionNames$1.map((o) => JSON.stringify(o)).join(", ")})`;
		}
	};
	Object.defineProperty(result, "placeholder", {
		value: true,
		configurable: true,
		enumerable: false,
		writable: false
	});
	return result;
}
function normalizeNegatableFlagNameList(names) {
	return typeof names === "string" ? [names] : names;
}
function validateNoDuplicateOptionNames(names, label) {
	const seen = /* @__PURE__ */ new Set();
	for (const name of names) {
		if (seen.has(name)) throw new TypeError(`${label} has a duplicate name: "${name}".`);
		seen.add(name);
	}
}
function validateNoNegatableFlagNameCollision(positiveNames, negativeNames) {
	const positiveSet = new Set(positiveNames);
	for (const name of negativeNames) if (positiveSet.has(name)) throw new TypeError(`Negatable flag name is both positive and negative: "${name}".`);
}
function formatNegatableFlagDuplicateError(options, token) {
	return options.errors?.duplicate ? typeof options.errors.duplicate === "function" ? options.errors.duplicate(token) : options.errors.duplicate : require_message.message`${require_message.optionName(token)} cannot be used multiple times.`;
}
function formatNegatableFlagConflictError(options, previousToken, token) {
	return options.errors?.conflict ? typeof options.errors.conflict === "function" ? options.errors.conflict(previousToken, token) : options.errors.conflict : require_message.message`${require_message.optionName(previousToken)} and ${require_message.optionName(token)} cannot be used together.`;
}
function formatNegatableFlagUnexpectedValueError(options, optionName$1, value) {
	return options.errors?.unexpectedValue ? typeof options.errors.unexpectedValue === "function" ? options.errors.unexpectedValue(optionName$1, value) : options.errors.unexpectedValue : require_message.message`Flag ${require_message.optionName(optionName$1)} does not accept a value, but got: ${value}.`;
}
function parseMatchedNegatableFlag(context, token, value, consumed, consumedOnFailure, buffer, options) {
	if (context.state != null) return {
		success: false,
		consumed: consumedOnFailure,
		error: context.state.value === value ? formatNegatableFlagDuplicateError(options, token) : formatNegatableFlagConflictError(options, context.state.token, token)
	};
	return {
		success: true,
		next: {
			...context,
			state: {
				value,
				token
			},
			buffer
		},
		consumed
	};
}
/**
* Creates a parser for a pair of command-line flags that explicitly enable or
* disable a Boolean value.
*
* The positive names produce `true`; the negative names produce `false`.
* Unlike {@link option}, this parser fails when neither side is present,
* matching {@link flag} semantics. Wrap it in {@link optional} for a
* tri-state override or {@link withDefault} for a concrete fallback.
*
* @param names The positive and negative option names to parse.
* @param options Optional metadata and error customization.
* @returns A {@link Parser} that produces `true` for positive names and
*          `false` for negative names.
* @throws {TypeError} If any option name is invalid, duplicated within one
*         side, or shared by the positive and negative sides.
* @since 1.1.0
*/
function negatableFlag(names, options = {}) {
	const positiveNames = normalizeNegatableFlagNameList(names.positive);
	const negativeNames = normalizeNegatableFlagNameList(names.negative);
	require_validate.validateOptionNames(positiveNames, "Positive flag");
	require_validate.validateOptionNames(negativeNames, "Negative flag");
	validateNoDuplicateOptionNames(positiveNames, "Positive flag");
	validateNoDuplicateOptionNames(negativeNames, "Negative flag");
	validateNoNegatableFlagNameCollision(positiveNames, negativeNames);
	const optionNames$1 = [...positiveNames, ...negativeNames];
	const valueByName = /* @__PURE__ */ new Map();
	for (const name of positiveNames) valueByName.set(name, true);
	for (const name of negativeNames) valueByName.set(name, false);
	const result = {
		$valueType: [],
		$stateType: [],
		mode: "sync",
		priority: 10,
		usage: [{
			type: "option",
			names: optionNames$1,
			...options.hidden != null && { hidden: options.hidden }
		}],
		leadingNames: new Set(optionNames$1),
		acceptingAnyToken: false,
		initialState: void 0,
		canSkip(state) {
			return state != null;
		},
		parse(context) {
			if (context.optionsTerminated) return {
				success: false,
				consumed: 0,
				error: options.errors?.optionsTerminated ?? require_message.message`No more options can be parsed.`
			};
			else if (context.buffer.length < 1) return {
				success: false,
				consumed: 0,
				error: options.errors?.endOfInput ?? require_message.message`Expected an option, but got end of input.`
			};
			if (context.buffer[0] === "--") return {
				success: true,
				next: {
					...context,
					buffer: context.buffer.slice(1),
					state: context.state,
					optionsTerminated: true
				},
				consumed: context.buffer.slice(0, 1)
			};
			const directValue = valueByName.get(context.buffer[0]);
			if (directValue != null) return parseMatchedNegatableFlag(context, context.buffer[0], directValue, context.buffer.slice(0, 1), 1, context.buffer.slice(1), options);
			for (const name of optionNames$1) {
				if (!name.startsWith("--") && !name.startsWith("/") && !name.startsWith("+") && !(name.startsWith("-") && name.length > 2)) continue;
				const prefix = name.startsWith("/") ? `${name}:` : `${name}=`;
				if (context.buffer[0].startsWith(prefix)) {
					const value = context.buffer[0].slice(prefix.length);
					return {
						success: false,
						consumed: 1,
						error: formatNegatableFlagUnexpectedValueError(options, prefix.slice(0, -1), value)
					};
				}
			}
			for (const shortOption of optionNames$1) {
				if (!shortOption.match(/^-[^-]$/)) continue;
				if (!context.buffer[0].startsWith(shortOption)) continue;
				return parseMatchedNegatableFlag(context, shortOption, valueByName.get(shortOption), [context.buffer[0].slice(0, 2)], 1, [`-${context.buffer[0].slice(2)}`, ...context.buffer.slice(1)], options);
			}
			const invalidOption = context.buffer[0];
			if (options.errors?.noMatch) {
				const candidates = /* @__PURE__ */ new Set();
				for (const name of require_usage.extractOptionNames(context.usage)) candidates.add(name);
				const suggestions = require_suggestion.findSimilar(invalidOption, candidates, require_suggestion.DEFAULT_FIND_SIMILAR_OPTIONS);
				return {
					success: false,
					consumed: 0,
					error: typeof options.errors.noMatch === "function" ? options.errors.noMatch(invalidOption, suggestions) : options.errors.noMatch
				};
			}
			const baseError = require_message.message`No matched option for ${require_message.optionName(invalidOption)}.`;
			return {
				success: false,
				consumed: 0,
				error: require_suggestion.createErrorWithSuggestions(baseError, invalidOption, context.usage, "option")
			};
		},
		complete(state, _exec) {
			if (state == null) return {
				success: false,
				error: options.errors?.missing ? typeof options.errors.missing === "function" ? options.errors.missing(positiveNames, negativeNames) : options.errors.missing : require_message.message`Required flag ${require_message.optionNames(optionNames$1)} is missing.`
			};
			return {
				success: true,
				value: state.value
			};
		},
		suggest(_context, prefix) {
			if (require_usage.isSuggestionHidden(options.hidden)) return [];
			const suggestions = [];
			if (prefix.startsWith("--") || prefix.startsWith("-") || prefix.startsWith("/") || prefix.startsWith("+")) {
				for (const optionName$1 of optionNames$1) if (optionName$1.startsWith(prefix)) {
					if (prefix === "-" && optionName$1.length !== 2) continue;
					suggestions.push({
						kind: "literal",
						text: optionName$1
					});
				}
			}
			return suggestions;
		},
		getDocFragments(_state, _defaultValue) {
			if (require_usage.isDocHidden(options.hidden)) return {
				fragments: [],
				description: options.description
			};
			const fragments = [{
				type: "entry",
				term: {
					type: "option",
					names: optionNames$1
				},
				description: options.description
			}];
			return {
				fragments,
				description: options.description
			};
		},
		[Symbol.for("Deno.customInspect")]() {
			const args = [`positive: ${JSON.stringify(positiveNames)}`, `negative: ${JSON.stringify(negativeNames)}`];
			return `negatableFlag({ ${args.join(", ")} })`;
		}
	};
	Object.defineProperty(result, "validateValue", {
		value(v) {
			if (typeof v !== "boolean") {
				const actualType = v === null ? "null" : typeof v;
				return {
					success: false,
					error: require_message.message`${require_message.optionNames(optionNames$1)}: Expected a boolean value, but received ${actualType}.`
				};
			}
			return {
				success: true,
				value: v
			};
		},
		configurable: true,
		enumerable: false,
		writable: false
	});
	Object.defineProperty(result, "placeholder", {
		value: false,
		configurable: true,
		enumerable: false,
		writable: false
	});
	return result;
}
/**
* Creates a parser that expects a single argument value.
* This parser is typically used for positional arguments
* that are not options or flags.
* @template M The execution mode of the parser.
* @template T The type of the value produced by the parser.
* @param valueParser The {@link ValueParser} that defines how to parse
*                    the argument value.
* @param options Optional configuration for the argument parser,
*                allowing you to specify a description or other metadata.
* @returns A {@link Parser} that expects a single argument value and produces
*          the parsed value of type {@link T}.
*/
function argument(valueParser, options = {}) {
	const isAsync = valueParser.mode === "async";
	const syncValueParser = valueParser;
	const dependencyMetadata = require_dependency_metadata.extractDependencyMetadata(valueParser);
	const formatInvalidValueError = (error) => options.errors?.invalidValue ? typeof options.errors.invalidValue === "function" ? options.errors.invalidValue(error) : options.errors.invalidValue : require_message.message`${require_message.metavar(valueParser.metavar)}: ${error}`;
	const optionPattern = /^--?[a-z0-9-]+$/i;
	const term = {
		type: "argument",
		metavar: valueParser.metavar,
		...options.hidden != null && { hidden: options.hidden }
	};
	const result = {
		mode: valueParser.mode,
		$valueType: [],
		$stateType: [],
		priority: 5,
		usage: [term],
		leadingNames: EMPTY_LEADING_NAMES,
		acceptingAnyToken: true,
		initialState: void 0,
		canSkip(state) {
			return isTerminalValueState(state);
		},
		parse(context) {
			const localState = require_annotation_state.normalizeInjectedAnnotationState(context.state);
			if (context.buffer.length < 1) return {
				success: false,
				consumed: 0,
				error: options.errors?.endOfInput ?? require_message.message`Expected an argument, but got end of input.`
			};
			let i = 0;
			let optionsTerminated = context.optionsTerminated;
			if (!optionsTerminated) {
				if (context.buffer[i] === "--") {
					optionsTerminated = true;
					i++;
				} else if (context.buffer[i].match(optionPattern)) return {
					success: false,
					consumed: i,
					error: require_message.message`Expected an argument, but got an option: ${require_message.optionName(context.buffer[i])}.`
				};
			}
			if (context.buffer.length < i + 1) return {
				success: false,
				consumed: i,
				error: require_message.message`Expected an argument, but got end of input.`
			};
			if (localState != null) return {
				success: false,
				consumed: i,
				error: options.errors?.multiple ? typeof options.errors.multiple === "function" ? options.errors.multiple(valueParser.metavar) : options.errors.multiple : require_message.message`The argument ${require_message.metavar(valueParser.metavar)} cannot be used multiple times.`
			};
			const rawInput = context.buffer[i];
			return require_mode_dispatch.dispatchByMode(valueParser.mode, () => {
				const parseResult = syncValueParser.parse(rawInput);
				const next = recordTrace(context, buildTraceEntry("argument-value", rawInput, context.buffer.slice(0, i + 1), syncValueParser, parseResult));
				return {
					success: true,
					next: {
						...next,
						buffer: context.buffer.slice(i + 1),
						state: createOptionParseState(parseResult),
						optionsTerminated
					},
					consumed: context.buffer.slice(0, i + 1)
				};
			}, async () => {
				const parseResult = await valueParser.parse(rawInput);
				const next = recordTrace(context, buildTraceEntry("argument-value", rawInput, context.buffer.slice(0, i + 1), valueParser, parseResult));
				return {
					success: true,
					next: {
						...next,
						buffer: context.buffer.slice(i + 1),
						state: createOptionParseState(parseResult),
						optionsTerminated
					},
					consumed: context.buffer.slice(0, i + 1)
				};
			});
		},
		complete(state, exec) {
			const missing = {
				success: false,
				error: options.errors?.endOfInput ?? require_message.message`Expected a ${require_message.metavar(valueParser.metavar)}, but too few arguments.`
			};
			const completeSync = () => {
				if (state == null) return missing;
				const resolvedState = dependencyMetadata?.derived != null ? resolveDerivedCompletionSync(dependencyMetadata, state, exec) : state;
				return resolvedState.success ? resolvedState : {
					success: false,
					error: formatInvalidValueError(resolvedState.error)
				};
			};
			const completeAsync = async () => {
				if (state == null) return missing;
				if (dependencyMetadata?.derived == null) return completeSync();
				const resolved = await resolveDerivedCompletionAsync(dependencyMetadata, state, exec);
				return resolved.success ? resolved : {
					success: false,
					error: formatInvalidValueError(resolved.error)
				};
			};
			return require_mode_dispatch.dispatchByMode(valueParser.mode, completeSync, completeAsync);
		},
		suggest(context, prefix) {
			if (require_annotation_state.normalizeInjectedAnnotationState(context.state) != null) return require_mode_dispatch.dispatchIterableByMode(valueParser.mode, function* () {}, async function* () {});
			if (isAsync) return suggestArgumentAsync(valueParser, require_usage.isSuggestionHidden(options.hidden), prefix, context.dependencyRegistry, context.exec);
			return suggestArgumentSync(valueParser, require_usage.isSuggestionHidden(options.hidden), prefix, context.dependencyRegistry, context.exec);
		},
		getDocFragments(_state, defaultValue) {
			if (require_usage.isDocHidden(options.hidden)) return {
				fragments: [],
				description: options.description
			};
			const choicesMessage = valueParser.choices != null && valueParser.choices.length > 0 ? require_message.valueSet(valueParser.choices.map((c) => valueParser.format(c)), {
				fallback: "",
				type: "unit"
			}) : void 0;
			const fragments = [{
				type: "entry",
				term,
				description: options.description,
				default: defaultValue == null ? void 0 : require_message.message`${valueParser.format(defaultValue)}`,
				choices: choicesMessage
			}];
			return {
				fragments,
				description: options.description
			};
		},
		[Symbol.for("Deno.customInspect")]() {
			return `argument()`;
		}
	};
	if (typeof valueParser.normalize === "function") {
		const normalize = valueParser.normalize.bind(valueParser);
		Object.defineProperty(result, "normalizeValue", {
			value(v) {
				try {
					return normalize(v);
				} catch {
					return v;
				}
			},
			configurable: true,
			enumerable: false
		});
	}
	if (!require_internal_dependency.isDerivedValueParser(valueParser)) {
		const vp = valueParser;
		const vpMode = valueParser.mode;
		const wrapParseResult = (parsed) => parsed.success ? parsed : {
			success: false,
			error: formatInvalidValueError(parsed.error)
		};
		Object.defineProperty(result, "validateValue", {
			value(v) {
				if (typeof vp.validate === "function") return require_mode_dispatch.wrapForMode(vpMode, wrapParseResult(vp.validate(v)));
				let stringified;
				try {
					stringified = vp.format(v);
				} catch {
					return require_mode_dispatch.wrapForMode(vpMode, {
						success: true,
						value: v
					});
				}
				if (typeof stringified !== "string") return require_mode_dispatch.wrapForMode(vpMode, {
					success: true,
					value: v
				});
				return require_mode_dispatch.dispatchByMode(vpMode, () => wrapParseResult(syncValueParser.parse(stringified)), async () => wrapParseResult(await vp.parse(stringified)));
			},
			configurable: true,
			enumerable: false
		});
	}
	Object.defineProperty(result, "placeholder", {
		get() {
			try {
				return valueParser.placeholder;
			} catch {
				return void 0;
			}
		},
		configurable: true,
		enumerable: false
	});
	if (dependencyMetadata != null) Object.defineProperty(result, "dependencyMetadata", {
		value: dependencyMetadata,
		configurable: true,
		enumerable: false
	});
	return result;
}
function normalizeCommandState(state) {
	return require_annotation_state.normalizeInjectedAnnotationState(state);
}
function getCommandParseChildState(commandState, childState, parser) {
	return require_annotation_state.getWrappedChildParseState(commandState, childState, parser);
}
function getCommandChildState(commandState, childState, parser) {
	return require_annotation_state.getWrappedChildState(commandState, childState, parser);
}
function createCommandState(sourceState, state) {
	return require_internal_annotations.annotateFreshArray(sourceState, state);
}
function appendCommandPath(exec, name) {
	if (exec == null) return void 0;
	return {
		...exec,
		commandPath: [...exec.commandPath ?? [], name]
	};
}
function getCommandNames(name, options) {
	return [
		name,
		...getVisibleCommandAliases(options),
		...getHiddenCommandAliases(options)
	];
}
function getVisibleCommandAliases(options) {
	const aliases = options.aliases;
	if (aliases == null) return [];
	if (!isNonEmptyStringArray(aliases)) throw new TypeError("Command aliases must be a non-empty array of strings.");
	return aliases;
}
function getHiddenCommandAliases(options) {
	const hiddenAliases = options[require_command_alias.hiddenCommandAliasesKey];
	if (hiddenAliases == null) return [];
	if (!isNonEmptyStringArray(hiddenAliases)) throw new TypeError("Hidden command aliases must be a non-empty array of strings.");
	return hiddenAliases;
}
function isNonEmptyStringArray(value) {
	if (!Array.isArray(value) || value.length === 0) return false;
	for (let i = 0; i < value.length; i++) if (typeof value[i] !== "string") return false;
	return true;
}
function validateUniqueCommandNames(names) {
	const seen = /* @__PURE__ */ new Set();
	for (const name of names) {
		if (seen.has(name)) throw new TypeError(`Command has a duplicate name: "${name}".`);
		seen.add(name);
	}
}
function* suggestCommandSync(context, prefix, name, aliases, parser, options) {
	if (require_usage.isSuggestionHidden(options.hidden)) return;
	const state = normalizeCommandState(context.state);
	if (state === void 0) {
		for (const commandName of [name, ...aliases]) if (commandName.startsWith(prefix)) yield {
			kind: "literal",
			text: commandName,
			...options.description && { description: options.description }
		};
	} else if (state[0] === "matched") yield* parser.suggest(require_execution_context.withChildContext(context, name, getCommandChildState(context.state, parser.initialState, parser), parser.usage), prefix);
	else if (state[0] === "parsing") yield* parser.suggest(require_execution_context.withChildContext(context, name, getCommandChildState(context.state, state[1], parser), parser.usage), prefix);
}
async function* suggestCommandAsync(context, prefix, name, aliases, parser, options) {
	if (require_usage.isSuggestionHidden(options.hidden)) return;
	const state = normalizeCommandState(context.state);
	if (state === void 0) {
		for (const commandName of [name, ...aliases]) if (commandName.startsWith(prefix)) yield {
			kind: "literal",
			text: commandName,
			...options.description && { description: options.description }
		};
	} else if (state[0] === "matched") {
		const suggestions = parser.suggest(require_execution_context.withChildContext(context, name, getCommandChildState(context.state, parser.initialState, parser), parser.usage), prefix);
		for await (const s of suggestions) yield s;
	} else if (state[0] === "parsing") {
		const suggestions = parser.suggest(require_execution_context.withChildContext(context, name, getCommandChildState(context.state, state[1], parser), parser.usage), prefix);
		for await (const s of suggestions) yield s;
	}
}
/**
* Creates a parser that matches a specific subcommand name and then applies
* an inner parser to the remaining arguments.
* This is useful for building CLI tools with subcommands like git, npm, etc.
* @template M The execution mode of the parser.
* @template T The type of the value returned by the inner parser.
* @template TState The type of the state used by the inner parser.
* @param name The subcommand name to match (e.g., `"show"`, `"edit"`).
* @param parser The {@link Parser} to apply after the command is matched.
* @param options Optional configuration for the command parser, such as
*                a description for documentation.
* @returns A {@link Parser} that matches the command name and delegates
*          to the inner parser for the remaining arguments.
* @throws {TypeError} If `name` is empty, whitespace-only, contains
*         embedded whitespace, or contains control characters.
*/
function command(name, parser, options = {}) {
	const commandNames = getCommandNames(name, options);
	const aliases = getVisibleCommandAliases(options);
	const hiddenAliases = getHiddenCommandAliases(options);
	require_validate.validateCommandNames(commandNames, "Command");
	validateUniqueCommandNames(commandNames);
	const isAsync = parser.mode === "async";
	const syncInnerParser = parser;
	const asyncInnerParser = parser;
	const result = {
		[Symbol.for("@optique/core/commandParser")]: true,
		mode: parser.mode,
		$valueType: [],
		$stateType: [],
		priority: 15,
		usage: [{
			type: "command",
			name,
			...aliases.length > 0 && { aliases },
			...hiddenAliases.length > 0 && { hiddenAliases },
			...options.usageLine != null && { usageLine: options.usageLine },
			...options.hidden != null && { hidden: options.hidden }
		}, ...parser.usage],
		leadingNames: new Set(commandNames),
		acceptingAnyToken: false,
		initialState: void 0,
		canSkip(state, exec) {
			const normalizedState = normalizeCommandState(state);
			if (normalizedState === void 0) return false;
			if (normalizedState[0] === "matched") return parser.canSkip?.(getCommandChildState(state, parser.initialState, parser), require_execution_context.withChildExecPath(exec, name)) === true;
			return parser.canSkip?.(getCommandChildState(state, normalizedState[1], parser), require_execution_context.withChildExecPath(exec, name)) === true;
		},
		getSuggestRuntimeNodes(state, path) {
			const normalizedState = normalizeCommandState(state);
			if (normalizedState === void 0) return [];
			const childState = normalizedState[0] === "matched" ? getCommandChildState(state, parser.initialState, parser) : getCommandChildState(state, normalizedState[1], parser);
			const childPath = [...path, name];
			return parser.getSuggestRuntimeNodes?.(childState, childPath) ?? (parser.dependencyMetadata?.source != null ? [{
				path: childPath,
				parser,
				state: childState
			}] : []);
		},
		parse(context) {
			const state = normalizeCommandState(context.state);
			if (state === void 0) {
				if (context.buffer.length < 1 || !commandNames.includes(context.buffer[0])) {
					const actual = context.buffer.length > 0 ? context.buffer[0] : null;
					const leadingCmds = require_usage_internals.extractLeadingCommandNames(context.usage);
					const rawSuggestions = actual ? require_suggestion.findSimilar(actual, leadingCmds, require_suggestion.DEFAULT_FIND_SIMILAR_OPTIONS) : [];
					const suggestions = require_suggestion.expandCommandAliasSuggestions(context.usage, rawSuggestions);
					if (options.errors?.notMatched) {
						const errorMessage = options.errors.notMatched;
						return {
							success: false,
							consumed: 0,
							error: typeof errorMessage === "function" ? errorMessage(name, actual, suggestions) : errorMessage
						};
					}
					if (actual == null) return {
						success: false,
						consumed: 0,
						error: require_message.message`Expected command ${require_message.optionName(name)}, but got end of input.`
					};
					const baseError = require_message.message`Expected command ${require_message.optionName(name)}, but got ${actual}.`;
					const suggestionMsg = require_suggestion.createSuggestionMessage(suggestions);
					return {
						success: false,
						consumed: 0,
						error: suggestionMsg.length > 0 ? [
							...baseError,
							require_message.text("\n\n"),
							...suggestionMsg
						] : baseError
					};
				}
				return {
					success: true,
					next: {
						...context,
						buffer: context.buffer.slice(1),
						state: createCommandState(context.state, ["matched", name]),
						...context.exec != null ? { exec: appendCommandPath(context.exec, name) } : {}
					},
					consumed: context.buffer.slice(0, 1)
				};
			} else if (state[0] === "matched" || state[0] === "parsing") {
				const innerState = state[0] === "matched" ? getCommandParseChildState(context.state, parser.initialState, parser) : getCommandParseChildState(context.state, state[1], parser);
				const wrapState = (parseResult) => {
					if (parseResult.success) {
						const mergedExec = require_execution_context.mergeChildExec(context.exec, parseResult.next.exec);
						return {
							success: true,
							next: {
								...parseResult.next,
								state: createCommandState(context.state, ["parsing", parseResult.next.state]),
								...mergedExec != null ? {
									exec: mergedExec,
									dependencyRegistry: mergedExec.dependencyRegistry
								} : {}
							},
							consumed: parseResult.consumed
						};
					}
					return parseResult;
				};
				return require_mode_dispatch.dispatchByMode(parser.mode, () => wrapState(syncInnerParser.parse(require_execution_context.withChildContext(context, name, innerState, parser.usage))), async () => wrapState(await parser.parse(require_execution_context.withChildContext(context, name, innerState, parser.usage))));
			}
			return {
				success: false,
				consumed: 0,
				error: options.errors?.invalidState ?? require_message.message`Invalid command state.`
			};
		},
		complete(state, exec) {
			const normalizedState = normalizeCommandState(state);
			if (typeof normalizedState === "undefined") return {
				success: false,
				error: options.errors?.notFound ?? require_message.message`Command ${require_message.optionName(name)} was not matched.`
			};
			else if (normalizedState[0] === "matched") {
				const childExec = require_execution_context.withChildExecPath(exec, name);
				const childContext = {
					buffer: [],
					optionsTerminated: false,
					usage: parser.usage,
					state: getCommandParseChildState(state, parser.initialState, parser),
					...childExec != null ? {
						exec: childExec,
						trace: childExec.trace,
						dependencyRegistry: childExec.dependencyRegistry
					} : {}
				};
				return require_mode_dispatch.dispatchByMode(parser.mode, () => {
					const parseResult = syncInnerParser.parse(childContext);
					const nextExec = parseResult.success ? require_execution_context.mergeChildExec(childExec, parseResult.next.exec) : childExec;
					return syncInnerParser.complete(parseResult.success ? getCommandChildState(state, parseResult.next.state, parser) : getCommandChildState(state, syncInnerParser.initialState, parser), nextExec);
				}, async () => {
					const parseResult = await asyncInnerParser.parse(childContext);
					const nextExec = parseResult.success ? require_execution_context.mergeChildExec(childExec, parseResult.next.exec) : childExec;
					return asyncInnerParser.complete(parseResult.success ? getCommandChildState(state, parseResult.next.state, parser) : getCommandChildState(state, parser.initialState, parser), nextExec);
				});
			} else if (normalizedState[0] === "parsing") {
				const childExec = require_execution_context.withChildExecPath(exec, name);
				return require_mode_dispatch.dispatchByMode(parser.mode, () => syncInnerParser.complete(getCommandChildState(state, normalizedState[1], parser), childExec), async () => await asyncInnerParser.complete(getCommandChildState(state, normalizedState[1], parser), childExec));
			}
			return {
				success: false,
				error: options.errors?.invalidState ?? require_message.message`Invalid command state during completion.`
			};
		},
		[require_phase2_seed.extractPhase2SeedKey](state, exec) {
			const normalizedState = normalizeCommandState(state);
			if (typeof normalizedState === "undefined") return require_mode_dispatch.wrapForMode(parser.mode, null);
			if (normalizedState[0] === "matched") {
				const childExec = require_execution_context.withChildExecPath(exec, name);
				const childContext = {
					buffer: [],
					optionsTerminated: false,
					usage: parser.usage,
					state: getCommandParseChildState(state, parser.initialState, parser),
					...childExec != null ? {
						exec: childExec,
						trace: childExec.trace,
						dependencyRegistry: childExec.dependencyRegistry
					} : {}
				};
				return require_mode_dispatch.dispatchByMode(parser.mode, () => {
					const parseResult = syncInnerParser.parse(childContext);
					const nextExec = parseResult.success ? require_execution_context.mergeChildExec(childExec, parseResult.next.exec) : childExec;
					return require_phase2_seed.completeOrExtractPhase2Seed(syncInnerParser, parseResult.success ? getCommandChildState(state, parseResult.next.state, parser) : getCommandChildState(state, syncInnerParser.initialState, parser), nextExec);
				}, async () => {
					const parseResult = await asyncInnerParser.parse(childContext);
					const nextExec = parseResult.success ? require_execution_context.mergeChildExec(childExec, parseResult.next.exec) : childExec;
					return await require_phase2_seed.completeOrExtractPhase2Seed(asyncInnerParser, parseResult.success ? getCommandChildState(state, parseResult.next.state, parser) : getCommandChildState(state, parser.initialState, parser), nextExec);
				});
			}
			if (normalizedState[0] === "parsing") return require_phase2_seed.completeOrExtractPhase2Seed(parser, getCommandChildState(state, normalizedState[1], parser), require_execution_context.withChildExecPath(exec, name));
			return require_mode_dispatch.wrapForMode(parser.mode, null);
		},
		suggest(context, prefix) {
			if (isAsync) return suggestCommandAsync(context, prefix, name, aliases, parser, options);
			return suggestCommandSync(context, prefix, name, aliases, parser, options);
		},
		getDocFragments(state, defaultValue) {
			const commandState = state.kind === "available" ? normalizeCommandState(state.state) : void 0;
			if (state.kind === "unavailable" || typeof commandState === "undefined") {
				if (require_usage.isDocHidden(options.hidden)) return {
					fragments: [],
					description: options.description
				};
				return {
					description: options.description,
					fragments: [{
						type: "entry",
						term: {
							type: "command",
							name
						},
						description: options.brief ?? options.description
					}]
				};
			}
			const innerState = commandState[0] === "parsing" ? {
				kind: "available",
				state: getCommandChildState(state.state, commandState[1], parser)
			} : {
				kind: "available",
				state: getCommandChildState(state.state, parser.initialState, parser)
			};
			const innerFragments = parser.getDocFragments(innerState, defaultValue);
			return {
				...innerFragments,
				brief: innerFragments.brief ?? options.brief,
				description: innerFragments.description ?? options.description,
				footer: innerFragments.footer ?? options.footer
			};
		},
		[Symbol.for("Deno.customInspect")]() {
			return `command(${JSON.stringify(name)})`;
		}
	};
	if (typeof parser.normalizeValue === "function") Object.defineProperty(result, "normalizeValue", {
		value: parser.normalizeValue.bind(parser),
		configurable: true,
		enumerable: false
	});
	if (typeof parser.validateValue === "function") Object.defineProperty(result, "validateValue", {
		value: parser.validateValue.bind(parser),
		configurable: true,
		enumerable: false
	});
	return result;
}
/**
* Creates a parser that collects unrecognized options and passes them through.
* This is useful for building wrapper CLI tools that need to forward unknown
* options to an underlying tool.
*
* **Important**: This parser intentionally weakens Optique's strict parsing
* philosophy where "all input must be recognized." The benefit is enabling
* legitimate wrapper/proxy tool patterns, but the trade-off is that typos
* in pass-through options won't be caught.
*
* @param options Configuration for how to capture options.
* @returns A {@link Parser} that captures unrecognized options as an array
*          of strings.
*
* @example
* ```typescript
* // Default format: only captures --opt=val
* const parser = object({
*   debug: option("--debug"),
*   extra: passThrough(),
* });
*
* // mycli --debug --foo=bar --baz=qux
* // → { debug: true, extra: ["--foo=bar", "--baz=qux"] }
*
* // nextToken format: captures --opt val pairs
* const parser = object({
*   debug: option("--debug"),
*   extra: passThrough({ format: "nextToken" }),
* });
*
* // mycli --debug --foo bar
* // → { debug: true, extra: ["--foo", "bar"] }
*
* // greedy format: captures all remaining tokens
* const parser = command("exec", object({
*   container: argument(string()),
*   args: passThrough({ format: "greedy" }),
* }));
*
* // myproxy exec mycontainer --verbose -it bash
* // → { container: "mycontainer", args: ["--verbose", "-it", "bash"] }
* ```
*
* @since 0.8.0
*/
function passThrough(options = {}) {
	const format = options.format ?? "equalsOnly";
	const optionPattern = /^-[a-z0-9-]|^--[a-z0-9-]+/i;
	const equalsOptionPattern = /^--[a-z0-9-]+=/i;
	return {
		$valueType: [],
		$stateType: [],
		mode: "sync",
		priority: -10,
		usage: [{
			type: "passthrough",
			...options.hidden != null && { hidden: options.hidden }
		}],
		leadingNames: EMPTY_LEADING_NAMES,
		acceptingAnyToken: false,
		initialState: [],
		canSkip() {
			return true;
		},
		parse(context) {
			if (context.buffer.length < 1) return {
				success: false,
				consumed: 0,
				error: require_message.message`No input to pass through.`
			};
			const token = context.buffer[0];
			if (format === "greedy") {
				const captured = [...context.buffer];
				return {
					success: true,
					next: {
						...context,
						buffer: [],
						state: require_internal_annotations.annotateFreshArray(context.state, [...context.state, ...captured])
					},
					consumed: captured
				};
			}
			if (context.optionsTerminated) return {
				success: false,
				consumed: 0,
				error: require_message.message`Options terminated; cannot capture pass-through options.`
			};
			if (format === "equalsOnly") {
				if (!equalsOptionPattern.test(token)) return {
					success: false,
					consumed: 0,
					error: require_message.message`Expected --option=value format, but got: ${token}.`
				};
				return {
					success: true,
					next: {
						...context,
						buffer: context.buffer.slice(1),
						state: require_internal_annotations.annotateFreshArray(context.state, [...context.state, token])
					},
					consumed: [token]
				};
			}
			if (format === "nextToken") {
				if (!optionPattern.test(token)) return {
					success: false,
					consumed: 0,
					error: require_message.message`Expected option, but got: ${token}.`
				};
				if (token.includes("=")) return {
					success: true,
					next: {
						...context,
						buffer: context.buffer.slice(1),
						state: require_internal_annotations.annotateFreshArray(context.state, [...context.state, token])
					},
					consumed: [token]
				};
				const nextToken = context.buffer[1];
				if (nextToken !== void 0 && !optionPattern.test(nextToken)) return {
					success: true,
					next: {
						...context,
						buffer: context.buffer.slice(2),
						state: require_internal_annotations.annotateFreshArray(context.state, [
							...context.state,
							token,
							nextToken
						])
					},
					consumed: [token, nextToken]
				};
				return {
					success: true,
					next: {
						...context,
						buffer: context.buffer.slice(1),
						state: require_internal_annotations.annotateFreshArray(context.state, [...context.state, token])
					},
					consumed: [token]
				};
			}
			return {
				success: false,
				consumed: 0,
				error: require_message.message`Unknown passThrough format: ${format}.`
			};
		},
		complete(state, _exec) {
			if (require_internal_annotations.getAnnotations(state) == null) return {
				success: true,
				value: state
			};
			const copied = [...state];
			return {
				success: true,
				value: copied
			};
		},
		suggest(_context, _prefix) {
			return [];
		},
		getDocFragments(_state, _defaultValue) {
			if (require_usage.isDocHidden(options.hidden)) return {
				fragments: [],
				description: options.description
			};
			return {
				fragments: [{
					type: "entry",
					term: { type: "passthrough" },
					description: options.description
				}],
				description: options.description
			};
		},
		[Symbol.for("Deno.customInspect")]() {
			return `passThrough(${format})`;
		}
	};
}

//#endregion
exports.argument = argument;
exports.command = command;
exports.constant = constant;
exports.fail = fail;
exports.flag = flag;
exports.negatableFlag = negatableFlag;
exports.option = option;
exports.passThrough = passThrough;
