// @optique/core/dist/usage.js
import { getDisplayWidth } from "./displaywidth.js";
import { validateProgramName } from "./validate.js";

//#region src/usage.ts
/**
* Returns whether the term should be hidden from usage output.
*/
function isUsageHidden(hidden) {
	return hidden === true || hidden === "usage" || hidden === "help";
}
/**
* Returns whether the term should be hidden from documentation output.
*/
function isDocHidden(hidden) {
	return hidden === true || hidden === "doc" || hidden === "help";
}
/**
* Returns whether the term should be hidden from suggestion/error candidates.
*/
function isSuggestionHidden(hidden) {
	return hidden === true;
}
/**
* Merges two hidden visibility settings by taking the union of restrictions.
*/
function mergeHidden(a, b) {
	if (a == null) return b;
	if (b == null) return a;
	if (a === true || b === true) return true;
	if (a === false) return b;
	if (b === false) return a;
	if (a === b) return a;
	if (a === "help" || b === "help") return "help";
	if ((a === "usage" || a === "doc") && (b === "usage" || b === "doc")) return "help";
	return a;
}
/**
* Extracts all option names from a usage description.
*
* This function recursively traverses a {@link Usage} tree and collects all
* option names defined within it, including those nested inside optional,
* multiple, and exclusive terms.
*
* @param usage The usage description to extract option names from.
* @param includeHidden Whether to include fully hidden options (`hidden: true`)
*   in the result.  Defaults to `false`.
* @returns A set containing all option names found in the usage description.
*
* @example
* ```typescript
* const usage: Usage = [
*   { type: "option", names: ["--verbose", "-v"] },
*   { type: "option", names: ["--quiet", "-q"] },
* ];
* const names = extractOptionNames(usage);
* // names = Set(["--verbose", "-v", "--quiet", "-q"])
* ```
*/
function extractOptionNames(usage, includeHidden) {
	const names = /* @__PURE__ */ new Set();
	function traverseUsage(terms) {
		if (!terms || !Array.isArray(terms)) return;
		for (const term of terms) if (term.type === "option") {
			if (!includeHidden && isSuggestionHidden(term.hidden)) continue;
			for (const name of term.names) names.add(name);
		} else if (term.type === "optional" || term.type === "multiple" || term.type === "sequence") traverseUsage(term.terms);
		else if (term.type === "exclusive") for (const exclusiveUsage of term.terms) traverseUsage(exclusiveUsage);
	}
	traverseUsage(usage);
	return names;
}
/**
* Extracts all command names from a Usage array.
*
* This function recursively traverses the usage structure and collects
* all command names, similar to {@link extractOptionNames}.
*
* @param usage The usage structure to extract command names from.
* @param includeHidden Whether to include fully hidden commands
*   (`hidden: true`) in the result.  Defaults to `false`.
* @returns A set of all command names found in the usage structure.
*
* @example
* ```typescript
* const usage: Usage = [
*   { type: "command", name: "build" },
*   { type: "command", name: "test" },
* ];
* const names = extractCommandNames(usage);
* // names = Set(["build", "test"])
* ```
* @since 0.7.0
*/
function extractCommandNames(usage, includeHidden) {
	const names = /* @__PURE__ */ new Set();
	function traverseUsage(terms) {
		if (!terms || !Array.isArray(terms)) return;
		for (const term of terms) if (term.type === "command") {
			if (!includeHidden && isSuggestionHidden(term.hidden)) continue;
			names.add(term.name);
			for (const alias of term.aliases ?? []) names.add(alias);
			if (includeHidden) for (const alias of term.hiddenAliases ?? []) names.add(alias);
		} else if (term.type === "optional" || term.type === "multiple" || term.type === "sequence") traverseUsage(term.terms);
		else if (term.type === "exclusive") for (const exclusiveUsage of term.terms) traverseUsage(exclusiveUsage);
	}
	traverseUsage(usage);
	return names;
}
/**
* Extracts all literal values from a usage description.
*
* This function recursively traverses the usage tree and collects all
* `literal` term values.  Literal values represent fixed strings that
* the user must type (e.g., conditional discriminator values like
* `"server"` in `conditional(option("--mode", string()), { server: ... })`).
*
* @param usage The usage description to extract literal values from.
* @returns A set of all literal values found in the usage description.
* @since 1.0.0
*/
function extractLiteralValues(usage) {
	const values = /* @__PURE__ */ new Set();
	function traverseUsage(terms) {
		if (!terms || !Array.isArray(terms)) return;
		for (const term of terms) if (term.type === "literal") values.add(term.value);
		else if (term.type === "optional" || term.type === "multiple" || term.type === "sequence") traverseUsage(term.terms);
		else if (term.type === "exclusive") for (const branch of term.terms) traverseUsage(branch);
	}
	traverseUsage(usage);
	return values;
}
/**
* Extracts all argument metavars from a Usage array.
*
* This function recursively traverses the usage structure and collects
* all argument metavariable names, similar to {@link extractOptionNames}
* and {@link extractCommandNames}.
*
* @param usage The usage structure to extract argument metavars from.
* @returns A Set of all argument metavars found in the usage structure.
*
* @example
* ```typescript
* const usage: Usage = [
*   { type: "argument", metavar: "FILE" },
*   { type: "argument", metavar: "OUTPUT" },
* ];
* const metavars = extractArgumentMetavars(usage);
* // metavars = Set(["FILE", "OUTPUT"])
* ```
* @since 0.9.0
*/
function extractArgumentMetavars(usage) {
	const metavars = /* @__PURE__ */ new Set();
	function traverseUsage(terms) {
		if (!terms || !Array.isArray(terms)) return;
		for (const term of terms) if (term.type === "argument") {
			if (isSuggestionHidden(term.hidden)) continue;
			metavars.add(term.metavar);
		} else if (term.type === "optional" || term.type === "multiple" || term.type === "sequence") traverseUsage(term.terms);
		else if (term.type === "exclusive") for (const exclusiveUsage of term.terms) traverseUsage(exclusiveUsage);
	}
	traverseUsage(usage);
	return metavars;
}
/**
* Formats a usage description into a human-readable string representation
* suitable for command-line help text.
*
* This function converts a structured {@link Usage} description into a
* formatted string that follows common CLI conventions. It supports various
* formatting options including colors and compact option display.
* @param programName The name of the program or command for which the usage
*                    description is being formatted. This is typically the
*                    name of the executable or script that the user will run.
* @param usage The usage description to format, consisting of an array
*              of usage terms representing the command-line structure.
* @param options Optional formatting options to customize the output.
*                See {@link UsageFormatOptions} for available options.
* @returns A formatted string representation of the usage description.
* @throws {TypeError} If `programName` is not a string, is empty,
*         whitespace-only, or contains control characters.
*/
function formatUsage(programName, usage, options = {}) {
	validateProgramName(programName);
	usage = normalizeUsage(filterUsageForDisplay(usage));
	if (options.expandCommands) {
		const lastTerm = usage.at(-1);
		if (usage.length > 0 && usage.slice(0, -1).every((t) => t.type === "command") && lastTerm.type === "exclusive" && lastTerm.terms.every((t) => t.length > 0 && (t[0].type === "command" || t[0].type === "option" || t[0].type === "argument" || t[0].type === "optional" && t[0].terms.length === 1 && (t[0].terms[0].type === "command" || t[0].terms[0].type === "option" || t[0].terms[0].type === "argument")))) {
			const lines = [];
			for (let command of lastTerm.terms) {
				const firstTerm = command[0];
				if (firstTerm?.type === "command" && isUsageHidden(firstTerm.hidden)) continue;
				if (usage.length > 1) command = [...usage.slice(0, -1), ...command];
				lines.push(formatUsage(programName, command, options));
			}
			if (lines.length > 0) return lines.join("\n");
		}
	}
	let output = options.colors ? `\x1b[1m${programName}\x1b[0m` : programName;
	let lineWidth = getDisplayWidth(programName);
	let first = true;
	for (const { text, width } of formatUsageTerms(usage, options)) {
		if (first) {
			first = false;
			if (options.maxWidth != null && lineWidth + 1 + width > options.maxWidth) {
				output += "\n";
				lineWidth = 0;
			} else {
				output += " ";
				lineWidth += 1;
			}
		} else if (options.maxWidth != null && lineWidth > 0 && lineWidth + width > options.maxWidth) {
			if (output.endsWith(" ")) output = output.slice(0, -1);
			output += "\n";
			lineWidth = 0;
			if (text === " ") continue;
		}
		output += text;
		lineWidth += width;
	}
	return output;
}
/**
* Normalizes a usage description by flattening nested exclusive terms,
* sorting terms for better readability, and ensuring consistent structure
* throughout the usage tree.
*
* This function performs three main operations:
*
* 1. *Stripping*: Removes degenerate terms that would render as empty or
*    malformed output, such as options with no names, commands with empty
*    names, arguments with empty metavars, and container terms (`optional`,
*    `multiple`, `exclusive`) whose top-level terms array is empty after
*    recursive normalization.  Exclusive branches representing valid
*    zero-token alternatives (e.g., `conditional()` default branches or
*    `optional(constant(...))`) and empty-value literals are preserved.
*    Only branches that become empty because all their content was
*    malformed are removed.
*
* 2. *Flattening*: Recursively processes all usage terms and merges any
*    nested exclusive terms into their parent exclusive term to avoid
*    redundant nesting. For example, an exclusive term containing another
*    exclusive term will have its nested terms flattened into the parent.
*    Similarly, nested optional terms are collapsed:
*    `optional(optional(X))` becomes `optional(X)` when the outer optional
*    contains only a single inner optional term.
*
* 3. *Sorting*: Reorders terms to improve readability by placing:
*    - Commands (subcommands) first
*    - Options and other terms in the middle
*    - Positional arguments last (including optional/multiple wrappers around
*      arguments)
*
* The sorting logic also recognizes when optional or multiple terms contain
* positional arguments and treats them as arguments for sorting purposes.
*
* @param usage The usage description to normalize.
* @returns A normalized usage description with degenerate terms removed,
*          nested exclusive and optional terms flattened, and remaining
*          terms sorted for optimal readability.
*/
function normalizeUsage(usage) {
	const terms = usage.map(normalizeUsageTerm).filter(isNonDegenerateTerm);
	terms.sort((a, b) => {
		const aCmd = a.type === "command";
		const bCmd = b.type === "command";
		const aArg = a.type === "argument" || (a.type === "optional" || a.type === "multiple") && a.terms.at(-1)?.type === "argument";
		const bArg = b.type === "argument" || (b.type === "optional" || b.type === "multiple") && b.terms.at(-1)?.type === "argument";
		return aCmd === bCmd ? aArg === bArg ? 0 : aArg ? 1 : -1 : aCmd ? -1 : 1;
	});
	return terms;
}
function normalizeUsageTerm(term) {
	if (term.type === "optional") {
		const normalized = normalizeUsage(term.terms);
		if (normalized.length === 1 && normalized[0].type === "optional") return normalized[0];
		return {
			type: "optional",
			terms: normalized
		};
	} else if (term.type === "multiple") return {
		type: "multiple",
		terms: normalizeUsage(term.terms),
		min: term.min
	};
	else if (term.type === "sequence") return {
		type: "sequence",
		terms: term.terms.map(normalizeUsageTerm).filter(isNonDegenerateTerm)
	};
	else if (term.type === "exclusive") {
		const terms = [];
		for (const usage of term.terms) {
			const normalized = normalizeUsage(usage);
			if (normalized.length >= 1 && normalized[0].type === "exclusive") {
				const rest = normalized.slice(1);
				for (const subUsage of normalized[0].terms) terms.push([...subUsage, ...rest]);
			} else if (normalized.length > 0 || !containsMalformedLeaf(usage)) terms.push(normalized);
		}
		return {
			type: "exclusive",
			terms
		};
	} else {
		if (term.type === "option") return {
			...term,
			names: [...term.names]
		};
		else if (term.type === "command") {
			if (term.usageLine == null || typeof term.usageLine === "function") return { ...term };
			return {
				...term,
				usageLine: cloneUsage(term.usageLine)
			};
		}
		return { ...term };
	}
}
function isNonDegenerateTerm(term) {
	if (term.type === "option") return term.names.length > 0;
	if (term.type === "command") return term.name !== "";
	if (term.type === "argument") return term.metavar.length > 0;
	if (term.type === "optional" || term.type === "multiple" || term.type === "exclusive" || term.type === "sequence") return term.terms.length > 0;
	return true;
}
function containsMalformedLeaf(usage) {
	for (const term of usage) {
		if (term.type === "option" && term.names.length === 0) return true;
		if (term.type === "command" && term.name === "") return true;
		if (term.type === "argument" && term.metavar.length === 0) return true;
		if (term.type === "optional" || term.type === "multiple" || term.type === "sequence") {
			if (containsMalformedLeaf(term.terms)) return true;
		}
		if (term.type === "exclusive") {
			for (const branch of term.terms) if (containsMalformedLeaf(branch)) return true;
		}
	}
	return false;
}
/**
* Creates a deep clone of a single {@link UsageTerm}.  Recursive term
* variants (`optional`, `multiple`, `exclusive`) are cloned recursively.
* For `command` terms, a function-valued `usageLine` is preserved by
* reference (functions are stateless callbacks), while an array-valued
* `usageLine` is deep-cloned.
*
* @param term The usage term to clone.
* @returns A structurally equal but referentially distinct copy.
* @since 1.0.0
*/
function cloneUsageTerm(term) {
	switch (term.type) {
		case "argument": return { ...term };
		case "option": return {
			...term,
			names: [...term.names]
		};
		case "command": {
			if (term.usageLine == null || typeof term.usageLine === "function") return {
				...term,
				...term.aliases != null ? { aliases: [...term.aliases] } : {},
				...term.hiddenAliases != null ? { hiddenAliases: [...term.hiddenAliases] } : {}
			};
			return {
				...term,
				...term.aliases != null ? { aliases: [...term.aliases] } : {},
				...term.hiddenAliases != null ? { hiddenAliases: [...term.hiddenAliases] } : {},
				usageLine: term.usageLine.map(cloneUsageTerm)
			};
		}
		case "optional": return {
			type: "optional",
			terms: term.terms.map(cloneUsageTerm)
		};
		case "multiple": return {
			type: "multiple",
			terms: term.terms.map(cloneUsageTerm),
			min: term.min
		};
		case "exclusive": return {
			type: "exclusive",
			terms: term.terms.map((u) => u.map(cloneUsageTerm))
		};
		case "sequence": return {
			type: "sequence",
			terms: term.terms.map(cloneUsageTerm)
		};
		case "literal":
		case "passthrough":
		case "ellipsis": return { ...term };
	}
}
/**
* Creates a deep clone of a {@link Usage} array and all of its terms.
*
* @param usage The usage array to clone.
* @returns A mutable array of deeply cloned usage terms.
* @since 1.0.0
*/
function cloneUsage(usage) {
	return usage.map(cloneUsageTerm);
}
function filterUsageForDisplay(usage, isHidden = isUsageHidden) {
	const terms = [];
	for (const term of usage) {
		if ((term.type === "argument" || term.type === "option" || term.type === "command" || term.type === "passthrough") && isHidden(term.hidden)) continue;
		if (term.type === "option" && term.names.length === 0) continue;
		if (term.type === "command" && term.name === "") continue;
		if (term.type === "argument" && term.metavar.length === 0) continue;
		if (term.type === "literal" && term.value === "") continue;
		if (term.type === "optional") {
			const filtered = filterUsageForDisplay(term.terms, isHidden);
			if (filtered.length > 0) terms.push({
				type: "optional",
				terms: filtered
			});
			continue;
		}
		if (term.type === "multiple") {
			const filtered = filterUsageForDisplay(term.terms, isHidden);
			if (filtered.length > 0) terms.push({
				type: "multiple",
				terms: filtered,
				min: term.min
			});
			continue;
		}
		if (term.type === "exclusive") {
			const filteredBranches = term.terms.map((branch) => {
				const first = branch[0];
				if (first?.type === "command" && isHidden(first.hidden)) return [];
				return filterUsageForDisplay(branch, isHidden);
			}).filter((branch) => branch.length > 0);
			if (filteredBranches.length > 0) terms.push({
				type: "exclusive",
				terms: filteredBranches
			});
			continue;
		}
		if (term.type === "sequence") {
			const filtered = filterUsageForDisplay(term.terms, isHidden);
			if (filtered.length > 0) terms.push({
				type: "sequence",
				terms: filtered
			});
			continue;
		}
		terms.push(term);
	}
	return terms;
}
function* formatUsageTerms(terms, options) {
	let i = 0;
	for (const t of terms) {
		if (i > 0) yield {
			text: " ",
			width: 1
		};
		yield* formatUsageTermInternal(t, options);
		i++;
	}
}
/**
* Formats a single {@link UsageTerm} into a string representation
* suitable for command-line help text.
* @param term The usage term to format, which can be an argument,
*             option, command, optional term, exclusive term, or multiple term.
* @param options Optional formatting options to customize the output.
*                See {@link UsageTermFormatOptions} for available options.
* @returns A formatted string representation of the usage term.
*/
function formatUsageTerm(term, options = {}) {
	const hiddenCheck = options.context === "doc" ? isDocHidden : isUsageHidden;
	const visibleTerms = filterUsageForDisplay([term], hiddenCheck);
	if (visibleTerms.length < 1) return "";
	let lineWidth = 0;
	let output = "";
	for (const { text, width } of formatUsageTermInternal(visibleTerms[0], options)) {
		if (options.maxWidth != null && lineWidth > 0 && lineWidth + width > options.maxWidth) {
			if (output.endsWith(" ")) output = output.slice(0, -1);
			output += "\n";
			lineWidth = 0;
			if (text === " ") continue;
		}
		output += text;
		lineWidth += width;
	}
	return output;
}
function* formatUsageTermInternal(term, options) {
	const optionsSeparator = options.optionsSeparator ?? "/";
	if (term.type === "argument") yield {
		text: options?.colors ? `\x1b[4m${term.metavar}\x1b[0m` : term.metavar,
		width: getDisplayWidth(term.metavar)
	};
	else if (term.type === "option") if (options?.onlyShortestOptions) {
		const shortestName = term.names.reduce((a, b) => getDisplayWidth(a) <= getDisplayWidth(b) ? a : b);
		yield {
			text: options?.colors ? `\x1b[3m${shortestName}\x1b[0m` : shortestName,
			width: getDisplayWidth(shortestName)
		};
	} else {
		let i = 0;
		for (const optionName of term.names) {
			if (i > 0) yield {
				text: options?.colors ? `\x1b[2m${optionsSeparator}\x1b[0m` : optionsSeparator,
				width: getDisplayWidth(optionsSeparator)
			};
			yield {
				text: options?.colors ? `\x1b[3m${optionName}\x1b[0m` : optionName,
				width: getDisplayWidth(optionName)
			};
			i++;
		}
		if (term.metavar != null) {
			yield {
				text: " ",
				width: 1
			};
			yield {
				text: options?.colors ? `\x1b[4m\x1b[2m${term.metavar}\x1b[0m` : term.metavar,
				width: getDisplayWidth(term.metavar)
			};
		}
	}
	else if (term.type === "command") yield {
		text: options?.colors ? `\x1b[1m${term.name}\x1b[0m` : term.name,
		width: getDisplayWidth(term.name)
	};
	else if (term.type === "optional") {
		yield {
			text: options?.colors ? `\x1b[2m[\x1b[0m` : "[",
			width: 1
		};
		yield* formatUsageTerms(term.terms, options);
		yield {
			text: options?.colors ? `\x1b[2m]\x1b[0m` : "]",
			width: 1
		};
	} else if (term.type === "exclusive") {
		yield {
			text: options?.colors ? `\x1b[2m(\x1b[0m` : "(",
			width: 1
		};
		let i = 0;
		for (const termGroup of term.terms) {
			if (i > 0) {
				yield {
					text: " ",
					width: 1
				};
				yield {
					text: "|",
					width: 1
				};
				yield {
					text: " ",
					width: 1
				};
			}
			yield* formatUsageTerms(termGroup, options);
			i++;
		}
		yield {
			text: options?.colors ? `\x1b[2m)\x1b[0m` : ")",
			width: 1
		};
	} else if (term.type === "sequence") yield* formatUsageTerms(term.terms, options);
	else if (term.type === "multiple") {
		if (term.min < 1) yield {
			text: options?.colors ? `\x1b[2m[\x1b[0m` : "[",
			width: 1
		};
		for (let i = 0; i < Math.max(1, term.min); i++) {
			if (i > 0) yield {
				text: " ",
				width: 1
			};
			yield* formatUsageTerms(term.terms, options);
		}
		yield {
			text: options?.colors ? `\x1b[2m...\x1b[0m` : "...",
			width: 3
		};
		if (term.min < 1) yield {
			text: options?.colors ? `\x1b[2m]\x1b[0m` : "]",
			width: 1
		};
	} else if (term.type === "literal") yield {
		text: term.value,
		width: getDisplayWidth(term.value)
	};
	else if (term.type === "passthrough") {
		const text = "[...]";
		yield {
			text: options?.colors ? `\x1b[2m${text}\x1b[0m` : text,
			width: 5
		};
	} else if (term.type === "ellipsis") {
		const text = "...";
		yield {
			text: options?.colors ? `\x1b[2m${text}\x1b[0m` : text,
			width: 3
		};
	} else throw new TypeError(`Unknown usage term type: ${term["type"]}.`);
}

//#endregion
export { cloneUsage, cloneUsageTerm, extractArgumentMetavars, extractCommandNames, extractLiteralValues, extractOptionNames, formatUsage, formatUsageTerm, isDocHidden, isSuggestionHidden, isUsageHidden, mergeHidden, normalizeUsage };
