// @optique/core/dist/valueparser.cjs
const require_message = require('./message.cjs');
const require_internal_dependency = require('./internal/dependency.cjs');
const require_suggestion = require('./suggestion.cjs');
const require_nonempty = require('./nonempty.cjs');

//#region src/valueparser.ts
/**
* A predicate function that checks if an object is a {@link ValueParser}.
* @param object The object to check.
* @return `true` if the object is a {@link ValueParser}, `false` otherwise.
* @throws {TypeError} If the object looks like a value parser (has `mode`,
*   `metavar`, `parse`, and `format`) but is missing the required
*   `placeholder` property.
*/
function isValueParser(object) {
	if (typeof object !== "object" || object == null || !("mode" in object) || object.mode !== "sync" && object.mode !== "async") return false;
	const hasMetavar = "metavar" in object && typeof object.metavar === "string";
	const hasParse = "parse" in object && typeof object.parse === "function";
	const hasFormat = "format" in object && typeof object.format === "function";
	const hasPlaceholder = "placeholder" in object;
	if (hasMetavar && hasParse && hasFormat && !hasPlaceholder) throw new TypeError("Value parser is missing the required placeholder property. All value parsers must define a placeholder value.");
	return hasMetavar && hasParse && hasFormat && hasPlaceholder;
}
/**
* Implementation of the choice parser for both string and number types.
*/
function choice(choices, options = {}) {
	if (choices.length < 1) throw new TypeError("Expected at least one choice, but got an empty array.");
	if (choices.some((c) => c === "")) throw new TypeError("Empty strings are not allowed as choices.");
	for (const c of choices) if (typeof c !== "string" && typeof c !== "number") throw new TypeError(`Expected every choice to be a string or number, but got ${typeof c}.`);
	const isNumber = typeof choices[0] === "number";
	for (const c of choices) if (isNumber ? typeof c !== "number" : typeof c !== "string") throw new TypeError(`Expected every choice to be the same type, but got both ${isNumber ? "number" : "string"} and ${typeof c}.`);
	if (isNumber && choices.some((v) => Number.isNaN(v))) throw new TypeError("NaN is not allowed in number choices.");
	const metavar$1 = options.metavar ?? "TYPE";
	require_nonempty.ensureNonEmptyString(metavar$1);
	if (isNumber) {
		const numberChoices = (() => {
			const seen = /* @__PURE__ */ new Set();
			let hasPositiveZero = false;
			let hasNegativeZero = false;
			const result = [];
			for (const v of choices) {
				if (Object.is(v, -0)) {
					if (hasNegativeZero) continue;
					hasNegativeZero = true;
				} else if (Object.is(v, 0)) {
					if (hasPositiveZero) continue;
					hasPositiveZero = true;
				} else {
					if (seen.has(v)) continue;
					seen.add(v);
				}
				result.push(v);
			}
			return result;
		})();
		const numberInvalidChoice = options.errors?.invalidChoice;
		const numberStrings = numberChoices.map((v) => Object.is(v, -0) ? "-0" : String(v));
		const frozenNumberChoices = Object.freeze(numberChoices);
		return {
			mode: "sync",
			metavar: metavar$1,
			placeholder: choices[0],
			choices: frozenNumberChoices,
			parse(input) {
				const index = numberStrings.indexOf(input);
				if (index >= 0) return {
					success: true,
					value: numberChoices[index]
				};
				if (/^[+-]?(\d+\.?\d*|\.\d+)$/.test(input)) {
					const parsed = Number(input);
					if (Number.isFinite(parsed)) {
						const canonical = Object.is(parsed, -0) ? "-0" : String(parsed);
						const normalizedInput = normalizeDecimal(input);
						const normalizedCanonical = normalizeDecimal(expandScientific(canonical));
						if (normalizedInput === normalizedCanonical) {
							const fallbackIndex = numberChoices.findIndex((v) => Object.is(v, parsed));
							if (fallbackIndex >= 0) return {
								success: true,
								value: numberChoices[fallbackIndex]
							};
						}
						if (parsed === 0 && normalizedInput.replace(/^-/, "") === "0" && !numberChoices.some((v) => Object.is(v, -0))) {
							const zeroIndex = numberChoices.indexOf(0);
							if (zeroIndex >= 0) return {
								success: true,
								value: numberChoices[zeroIndex]
							};
						}
					}
				}
				if (/^[+-]?(\d+\.?\d*|\.\d+)[eE][+-]?\d+$/.test(input)) {
					const parsed = Number(input);
					if (Number.isFinite(parsed)) {
						const canonical = Object.is(parsed, -0) ? "-0" : String(parsed);
						if (/[eE]/.test(canonical)) {
							const normalizedInput = normalizeDecimal(expandScientific(input));
							const normalizedCanonical = normalizeDecimal(expandScientific(canonical));
							if (normalizedInput === normalizedCanonical) {
								const fallbackIndex = numberChoices.findIndex((v) => Object.is(v, parsed));
								if (fallbackIndex >= 0) return {
									success: true,
									value: numberChoices[fallbackIndex]
								};
							}
						}
					}
				}
				return {
					success: false,
					error: formatNumberChoiceError(input, numberChoices, numberChoices, numberInvalidChoice)
				};
			},
			format(value) {
				return Object.is(value, -0) ? "-0" : String(value);
			},
			suggest(prefix) {
				return numberStrings.filter((valueStr, i) => !Number.isNaN(numberChoices[i]) && valueStr.startsWith(prefix)).map((valueStr) => ({
					kind: "literal",
					text: valueStr
				}));
			}
		};
	}
	const stringChoices = Object.freeze([...new Set(choices)]);
	const stringOptions = options;
	checkBooleanOption(stringOptions, "caseInsensitive");
	const caseInsensitive = stringOptions.caseInsensitive ?? false;
	const normalizedValues = caseInsensitive ? stringChoices.map((v) => v.toLowerCase()) : stringChoices;
	if (caseInsensitive) {
		const seen = /* @__PURE__ */ new Map();
		for (let i = 0; i < stringChoices.length; i++) {
			const nv = normalizedValues[i];
			const original = stringChoices[i];
			const prev = seen.get(nv);
			if (prev !== void 0 && prev !== original) throw new TypeError(`Ambiguous choices for case-insensitive matching: ${JSON.stringify(prev)} and ${JSON.stringify(original)} both normalize to ${JSON.stringify(nv)}.`);
			seen.set(nv, original);
		}
	}
	const stringInvalidChoice = stringOptions.errors?.invalidChoice;
	return {
		mode: "sync",
		metavar: metavar$1,
		placeholder: choices[0],
		choices: stringChoices,
		parse(input) {
			const normalizedInput = caseInsensitive ? input.toLowerCase() : input;
			const index = normalizedValues.indexOf(normalizedInput);
			if (index < 0) return {
				success: false,
				error: formatStringChoiceError(input, stringChoices, stringInvalidChoice)
			};
			return {
				success: true,
				value: stringChoices[index]
			};
		},
		format(value) {
			return String(value);
		},
		suggest(prefix) {
			const normalizedPrefix = caseInsensitive ? prefix.toLowerCase() : prefix;
			return stringChoices.filter((value) => {
				const normalizedValue = caseInsensitive ? value.toLowerCase() : value;
				return normalizedValue.startsWith(normalizedPrefix);
			}).map((value) => ({
				kind: "literal",
				text: value
			}));
		}
	};
}
/**
* Validates that an option value, if present, is a boolean.
* Throws a {@link TypeError} if the value is defined but not a boolean.
*
* @template T The type of the options object.
* @param options The options object to check.
* @param key The key of the option to validate.
* @throws {TypeError} If the option value is defined but not a boolean.
* @since 1.0.0
*/
function checkBooleanOption(options, key) {
	const value = options?.[key];
	if (value !== void 0 && typeof value !== "boolean") throw new TypeError(`Expected ${String(key)} to be a boolean, but got ${typeof value}: ${String(value)}.`);
}
/**
* Validates that an option value, if present, is one of the allowed values.
* Throws a {@link TypeError} if the value is defined but not in the allowed
* list.
*
* @template T The type of the options object.
* @param options The options object to check.
* @param key The key of the option to validate.
* @param allowed The list of allowed values.
* @throws {TypeError} If the option value is defined but not in the allowed
*   list.
* @since 1.0.0
*/
function checkEnumOption(options, key, allowed) {
	const value = options?.[key];
	if (value !== void 0 && (typeof value !== "string" || !allowed.includes(value))) {
		const rendered = typeof value === "string" ? JSON.stringify(value) : typeof value === "symbol" ? value.toString() : String(value);
		throw new TypeError(`Expected ${String(key)} to be one of ${allowed.map((v) => JSON.stringify(v)).join(", ")}, but got ${typeof value}: ${rendered}.`);
	}
}
/**
* Expands a numeric string in scientific notation (e.g., `"1e+21"`,
* `"1.5e-3"`, `".1e-6"`) into plain decimal form for normalization.
* Used for both canonical `String(number)` output and user input.
* Returns the input unchanged if it does not contain scientific notation.
*/
function expandScientific(s) {
	const match = /^([+-]?)(\d+\.?\d*|\.\d+)[eE]([+-]?\d+)$/.exec(s);
	if (!match) return s;
	const [, rawSign, mantissa, expStr] = match;
	const sign = rawSign === "-" ? "-" : "";
	const exp = parseInt(expStr, 10);
	const dotPos = mantissa.indexOf(".");
	const digits = mantissa.replace(".", "");
	const intLen = dotPos >= 0 ? dotPos : digits.length;
	const newIntLen = intLen + exp;
	let result;
	if (newIntLen >= digits.length) result = digits + "0".repeat(newIntLen - digits.length);
	else if (newIntLen <= 0) result = "0." + "0".repeat(-newIntLen) + digits;
	else result = digits.slice(0, newIntLen) + "." + digits.slice(newIntLen);
	return sign + result;
}
/**
* Normalizes a plain decimal string by stripping leading zeros from the
* integer part and trailing zeros from the fractional part, so that two
* strings representing the same mathematical value compare as equal.
*/
function normalizeDecimal(s) {
	let sign = "";
	let str = s;
	if (str.startsWith("-") || str.startsWith("+")) {
		if (str[0] === "-") sign = "-";
		str = str.slice(1);
	}
	const dot = str.indexOf(".");
	let int;
	let frac;
	if (dot >= 0) {
		int = str.slice(0, dot);
		frac = str.slice(dot + 1);
	} else {
		int = str;
		frac = "";
	}
	int = int.replace(/^0+/, "") || "0";
	frac = frac.replace(/0+$/, "");
	return sign + (frac ? int + "." + frac : int);
}
/**
* Formats error message for string choice parser.
*/
function formatStringChoiceError(input, choices, invalidChoice) {
	if (invalidChoice) return typeof invalidChoice === "function" ? invalidChoice(input, choices) : invalidChoice;
	return formatDefaultChoiceError(input, choices);
}
/**
* Formats error message for number choice parser.
*/
function formatNumberChoiceError(input, validChoices, allChoices, invalidChoice) {
	if (invalidChoice) return typeof invalidChoice === "function" ? invalidChoice(input, validChoices) : invalidChoice;
	return formatDefaultChoiceError(input, allChoices);
}
/**
* Formats default error message for choice parser.
*/
function formatDefaultChoiceError(input, choices) {
	const choiceStrings = choices.filter((c) => typeof c === "string" || !Number.isNaN(c)).map((c) => Object.is(c, -0) ? "-0" : String(c));
	if (choiceStrings.length === 0 && choices.length > 0) return require_message.message`No valid choices are configured, but got ${input}.`;
	return require_message.message`Expected one of ${require_message.valueSet(choiceStrings, {
		fallback: "",
		locale: "en-US"
	})}, but got ${input}.`;
}
/**
* Creates a {@link ValueParser} for strings.
*
* This parser validates that the input is a string and optionally checks
* if it matches a specified regular expression pattern.
*
* **Security note**: When using the `pattern` option with user-defined or
* complex patterns, be aware of potential Regular Expression Denial of Service
* (ReDoS) attacks. See {@link StringOptions.pattern} for more details.
*
* @param options Configuration options for the string parser.
* @returns A {@link ValueParser} that parses strings according to the
*          specified options.
* @throws {TypeError} If `options.pattern` is provided but is not a
*         `RegExp` instance.
*/
function string(options = {}) {
	if (options.pattern != null && !(options.pattern instanceof RegExp)) throw new TypeError(`Expected pattern to be a RegExp, but got: ${Object.prototype.toString.call(options.pattern)}`);
	const metavar$1 = options.metavar ?? "STRING";
	require_nonempty.ensureNonEmptyString(metavar$1);
	const patternSource = options.pattern?.source ?? null;
	const patternFlags = options.pattern?.flags ?? null;
	const patternMismatch = options.errors?.patternMismatch;
	return {
		mode: "sync",
		metavar: metavar$1,
		placeholder: options.placeholder ?? "",
		parse(input) {
			if (patternSource != null && patternFlags != null) {
				const pattern = new RegExp(patternSource, patternFlags);
				if (pattern.test(input)) return {
					success: true,
					value: input
				};
				return {
					success: false,
					error: patternMismatch ? typeof patternMismatch === "function" ? patternMismatch(input, pattern) : patternMismatch : require_message.message`Expected a string matching pattern ${require_message.text(patternSource)}, but got ${input}.`
				};
			}
			return {
				success: true,
				value: input
			};
		},
		format(value) {
			return value;
		}
	};
}
function keyValue(options = {}) {
	const separator = options.separator ?? "=";
	if (typeof separator !== "string") throw new TypeError(`Expected separator to be a string, but got ${typeof separator}: ${String(separator)}.`);
	require_nonempty.ensureNonEmptyString(separator);
	checkBooleanOption(options, "allowEmptyKey");
	checkBooleanOption(options, "allowEmptyValue");
	checkEnumOption(options, "split", ["first", "last"]);
	const split = options.split ?? "first";
	const allowEmptyKey = options.allowEmptyKey ?? false;
	const allowEmptyValue = options.allowEmptyValue ?? true;
	const metavar$1 = options.metavar ?? `KEY${separator}VALUE`;
	require_nonempty.ensureNonEmptyString(metavar$1);
	const rawKeyParser = options.key ?? string({ placeholder: "KEY" });
	const rawValueParser = options.value ?? string({ placeholder: allowEmptyValue ? "" : "VALUE" });
	checkKeyValueChildParser("key", rawKeyParser);
	checkKeyValueChildParser("value", rawValueParser);
	const keyParser = rawKeyParser;
	const valueParser = rawValueParser;
	const hasNormalize = typeof keyParser.normalize === "function" || typeof valueParser.normalize === "function";
	const hasSuggest = typeof keyParser.suggest === "function" || typeof valueParser.suggest === "function";
	function findSeparator(input) {
		return split === "last" ? input.lastIndexOf(separator) : input.indexOf(separator);
	}
	function emptyKeyError(input) {
		const custom = options.errors?.emptyKey;
		if (custom != null) return typeof custom === "function" ? custom(input) : custom;
		return require_message.message`Expected a non-empty key in ${input}.`;
	}
	function emptyValueError(input) {
		const custom = options.errors?.emptyValue;
		if (custom != null) return typeof custom === "function" ? custom(input) : custom;
		return require_message.message`Expected a non-empty value in ${input}.`;
	}
	function missingSeparatorError(input) {
		const custom = options.errors?.missingSeparator;
		if (custom != null) return typeof custom === "function" ? custom(input, separator) : custom;
		return require_message.message`Expected ${separator} in ${input}.`;
	}
	function invalidKeyError(error) {
		const custom = options.errors?.invalidKey;
		if (custom != null) return typeof custom === "function" ? custom(error) : custom;
		return [require_message.text("Invalid key: "), ...require_message.cloneMessage(error)];
	}
	function invalidValueError(error) {
		const custom = options.errors?.invalidValue;
		if (custom != null) return typeof custom === "function" ? custom(error) : custom;
		return [require_message.text("Invalid value: "), ...require_message.cloneMessage(error)];
	}
	function validateParts(input, key, value) {
		if (!allowEmptyKey && key === "") return {
			success: false,
			error: emptyKeyError(input)
		};
		if (!allowEmptyValue && value === "") return {
			success: false,
			error: emptyValueError(input)
		};
		const keyResult = keyParser.parse(key);
		if (!keyResult.success) return {
			success: false,
			error: invalidKeyError(keyResult.error)
		};
		const valueResult = valueParser.parse(value);
		if (!valueResult.success) return {
			success: false,
			error: invalidValueError(valueResult.error)
		};
		return validateResultParts(input, keyResult, valueResult);
	}
	function validateKeyForSuggestion(key) {
		if (!allowEmptyKey && key === "") return false;
		if (!partsRoundTrip(key, "")) return false;
		const keyResult = keyParser.parse(key);
		if (!keyResult.success) return false;
		if (!allowEmptyKey && keyResult.value === "") return false;
		const formattedKeyResult = formatValueParserValue(keyParser, keyResult.value);
		if (!formattedKeyResult.success) return false;
		const formattedKey = formattedKeyResult.value;
		return formattedKey == null || partsRoundTrip(formattedKey, "");
	}
	function validateKeyPatternForSuggestion(key) {
		if (!allowEmptyKey && key === "") return false;
		return partsRoundTrip(key, "");
	}
	function fallbackInput(key, value) {
		return `${typeof key === "string" ? key : ""}${separator}${typeof value === "string" ? value : ""}`;
	}
	function partsRoundTrip(key, value) {
		const input = `${key}${separator}${value}`;
		const index = findSeparator(input);
		if (index < 0) return false;
		return input.slice(0, index) === key && input.slice(index + separator.length) === value;
	}
	function validateResultParts(input, keyResult, valueResult) {
		if (!allowEmptyKey && keyResult.value === "") return {
			success: false,
			error: emptyKeyError(input)
		};
		if (!allowEmptyValue && valueResult.value === "") return {
			success: false,
			error: emptyValueError(input)
		};
		const formattedKeyResult = formatValueParserValue(keyParser, keyResult.value);
		if (!formattedKeyResult.success) return {
			success: false,
			error: invalidKeyError(formattedKeyResult.error)
		};
		const formattedValueResult = formatValueParserValue(valueParser, valueResult.value);
		if (!formattedValueResult.success) return {
			success: false,
			error: invalidValueError(formattedValueResult.error)
		};
		const formattedKey = formattedKeyResult.value;
		const formattedValue = formattedValueResult.value;
		if (!allowEmptyKey && formattedKey === "") return {
			success: false,
			error: emptyKeyError(input)
		};
		if (!allowEmptyValue && formattedValue === "") return {
			success: false,
			error: emptyValueError(input)
		};
		if (split === "first" && formattedKey != null && formattedKey.includes(separator)) return {
			success: false,
			error: invalidKeyError(require_message.message`Expected a key without ${separator}, but got ${formattedKey}.`)
		};
		if (split === "last" && formattedValue != null && formattedValue.includes(separator)) return {
			success: false,
			error: invalidValueError(require_message.message`Expected a value without ${separator}, but got ${formattedValue}.`)
		};
		if (formattedKey != null && formattedValue != null) {
			if (!partsRoundTrip(formattedKey, formattedValue)) return split === "first" ? {
				success: false,
				error: invalidKeyError(require_message.message`Expected a key that round-trips with ${separator}, but got ${formattedKey}.`)
			} : {
				success: false,
				error: invalidValueError(require_message.message`Expected a value that round-trips with ${separator}, but got ${formattedValue}.`)
			};
		}
		return makeKeyValueSuccess(keyResult, valueResult);
	}
	function validateTuple(value) {
		if (!allowEmptyKey && value[0] === "") return {
			success: false,
			error: emptyKeyError(fallbackInput(value[0], value[1]))
		};
		if (!allowEmptyValue && value[1] === "") return {
			success: false,
			error: emptyValueError(fallbackInput(value[0], value[1]))
		};
		const keyResult = validateValueParserValue(keyParser, value[0]);
		if (!keyResult.success) return {
			success: false,
			error: invalidKeyError(keyResult.error)
		};
		const valueResult = validateValueParserValue(valueParser, value[1]);
		if (!valueResult.success) return {
			success: false,
			error: invalidValueError(valueResult.error)
		};
		return validateResultParts(fallbackInput(value[0], value[1]), keyResult, valueResult);
	}
	return {
		mode: "sync",
		metavar: metavar$1,
		placeholder: [keyParser.placeholder, valueParser.placeholder],
		parse(input) {
			const index = findSeparator(input);
			if (index < 0) return {
				success: false,
				error: missingSeparatorError(input)
			};
			const key = input.slice(0, index);
			const value = input.slice(index + separator.length);
			return validateParts(input, key, value);
		},
		format(value) {
			return `${keyParser.format(value[0])}${separator}${valueParser.format(value[1])}`;
		},
		validate(value) {
			if (!isKeyValueTuple(value)) return {
				success: false,
				error: require_message.message`Expected a key-value tuple.`
			};
			return validateTuple(value);
		},
		...hasNormalize ? { normalize(value) {
			if (!isKeyValueTuple(value)) return value;
			const normalized = [typeof keyParser.normalize === "function" ? keyParser.normalize(value[0]) : value[0], typeof valueParser.normalize === "function" ? valueParser.normalize(value[1]) : value[1]];
			return validateTuple(normalized).success ? normalized : value;
		} } : {},
		...hasSuggest ? { *suggest(prefix) {
			const index = findSeparator(prefix);
			const suggestions = [];
			if (index < 0 || split === "last") {
				if (typeof keyParser.suggest === "function") for (const suggestion of keyParser.suggest(prefix)) {
					const key = suggestionTextValue(suggestion);
					if (suggestion.kind === "literal" ? !validateKeyForSuggestion(key) : !validateKeyPatternForSuggestion(key)) continue;
					suggestions.push(prefixKeyValueSuggestion(suggestion, "", separator));
				}
			}
			if (index < 0) {
				yield* require_suggestion.deduplicateSuggestions(suggestions);
				return;
			}
			if (typeof valueParser.suggest === "function") {
				const key = prefix.slice(0, index);
				if (!validateKeyForSuggestion(key)) {
					yield* require_suggestion.deduplicateSuggestions(suggestions);
					return;
				}
				const valuePrefix = prefix.slice(index + separator.length);
				for (const suggestion of valueParser.suggest(valuePrefix)) {
					const value = suggestionTextValue(suggestion);
					if (!partsRoundTrip(key, value)) continue;
					if (suggestion.kind === "literal" && !validateParts(`${key}${separator}${value}`, key, value).success) continue;
					suggestions.push(prefixKeyValueSuggestion(suggestion, `${key}${separator}`));
				}
			}
			yield* require_suggestion.deduplicateSuggestions(suggestions);
		} } : {}
	};
}
function checkKeyValueChildParser(role, parser) {
	if (!isValueParser(parser)) throw new TypeError(`The ${role} option for keyValue() must be a value parser.`);
	if (parser.mode !== "sync") throw new TypeError(`keyValue() only supports sync ${role} parsers, but an async one was given.`);
	if (require_internal_dependency.isDerivedValueParser(parser)) throw new TypeError(`keyValue() does not support dependency-derived ${role} parsers (created via deriveFrom() or dependency().derive()); pass the derived parser directly to option() or argument() instead.`);
}
function validateValueParserValue(parser, value) {
	if (typeof parser.validate === "function") return parser.validate(value);
	let formatted;
	try {
		formatted = parser.format(value);
	} catch {
		return {
			success: true,
			value
		};
	}
	if (typeof formatted !== "string") return {
		success: false,
		error: stringFormatError()
	};
	return parser.parse(formatted);
}
function formatValueParserValue(parser, value) {
	let formatted;
	try {
		formatted = parser.format(value);
	} catch {
		return {
			success: true,
			value: void 0
		};
	}
	if (typeof formatted !== "string") return {
		success: false,
		error: stringFormatError()
	};
	return {
		success: true,
		value: formatted
	};
}
function stringFormatError() {
	return require_message.message`Expected a value formatted as a string.`;
}
function suggestionTextValue(suggestion) {
	return suggestion.kind === "literal" ? suggestion.text : suggestion.pattern ?? "";
}
function prefixKeyValueSuggestion(suggestion, prefix, suffix = "") {
	const textValue = suggestionTextValue(suggestion);
	const text$1 = `${prefix}${textValue}${suffix}`;
	if (suggestion.kind === "file" && suffix === "") return {
		...suggestion,
		pattern: text$1
	};
	return suggestion.description == null ? {
		kind: "literal",
		text: text$1
	} : {
		kind: "literal",
		text: text$1,
		description: suggestion.description
	};
}
function isKeyValueTuple(value) {
	return Array.isArray(value) && value.length === 2 && 0 in value && 1 in value;
}
function makeKeyValueSuccess(keyResult, valueResult) {
	const deferredKeys = /* @__PURE__ */ new Map();
	let hasDeferred = false;
	collectKeyValueDeferredKeys(deferredKeys, 0, keyResult, () => {
		hasDeferred = true;
	});
	collectKeyValueDeferredKeys(deferredKeys, 1, valueResult, () => {
		hasDeferred = true;
	});
	return {
		success: true,
		value: [keyResult.value, valueResult.value],
		...hasDeferred ? {
			deferred: true,
			...deferredKeys.size > 0 ? { deferredKeys } : {}
		} : {}
	};
}
function collectKeyValueDeferredKeys(deferredKeys, index, result, markDeferred) {
	if (result.deferredKeys != null) {
		markDeferred();
		deferredKeys.set(index, result.deferredKeys);
		return;
	}
	if (result.deferred !== true) return;
	markDeferred();
	deferredKeys.set(index, null);
}
/**
* Creates a ValueParser for parsing integer values from strings.
*
* This function provides two modes of operation:
*
* - Regular mode: Returns JavaScript numbers
*   (safe up to `Number.MAX_SAFE_INTEGER`)
* - `bigint` mode: Returns `bigint` values for arbitrarily large integers
*
* The parser validates that the input is a valid integer and optionally
* enforces minimum and maximum value constraints.
*
* @example
* ```typescript
* // Create a parser for regular integers
* const portParser = integer({ min: 1, max: 0xffff });
*
* // Create a parser for BigInt values
* const bigIntParser = integer({ type: "bigint", min: 0n });
*
* // Use the parser
* const result = portParser.parse("8080");
* if (result.success) {
*   console.log(`Port: ${result.value}`);
* } else {
*   console.error(result.error);
* }
* ```
*
* @param options Configuration options specifying the type and constraints.
* @returns A {@link ValueParser} that converts string input to the specified
*          integer type.
* @throws {TypeError} If `options.type` is provided but is neither `"number"`
*   nor `"bigint"`.
* @throws {RangeError} If the configured min/max range for number mode contains
*   no safe integers.
*/
function integer(options) {
	if (options?.type !== void 0 && options.type !== "number" && options.type !== "bigint") throw new TypeError(`Expected type to be "number" or "bigint", but got: ${String(options.type)}.`);
	if (options?.type !== "bigint") {
		if (options?.min != null && !Number.isFinite(options.min)) throw new RangeError(`Expected min to be a finite number, but got: ${options.min}`);
		if (options?.max != null && !Number.isFinite(options.max)) throw new RangeError(`Expected max to be a finite number, but got: ${options.max}`);
	}
	if (options?.min != null && options?.max != null && options.min > options.max) throw new RangeError(`Expected min to be less than or equal to max, but got min: ${options.min} and max: ${options.max}.`);
	if (options?.type === "bigint") {
		const metavar$2 = options.metavar ?? "INTEGER";
		require_nonempty.ensureNonEmptyString(metavar$2);
		return {
			mode: "sync",
			metavar: metavar$2,
			placeholder: options?.placeholder ?? (options?.min != null && options.min > 0n ? options.min : options?.max != null && options.max < 0n ? options.max : 0n),
			parse(input) {
				if (!input.match(/^-?\d+$/)) return {
					success: false,
					error: options.errors?.invalidInteger ? typeof options.errors.invalidInteger === "function" ? options.errors.invalidInteger(input) : options.errors.invalidInteger : require_message.message`Expected a valid integer, but got ${input}.`
				};
				const value = BigInt(input);
				if (options.min != null && value < options.min) return {
					success: false,
					error: options.errors?.belowMinimum ? typeof options.errors.belowMinimum === "function" ? options.errors.belowMinimum(value, options.min) : options.errors.belowMinimum : require_message.message`Expected a value greater than or equal to ${require_message.text(options.min.toLocaleString("en"))}, but got ${input}.`
				};
				else if (options.max != null && value > options.max) return {
					success: false,
					error: options.errors?.aboveMaximum ? typeof options.errors.aboveMaximum === "function" ? options.errors.aboveMaximum(value, options.max) : options.errors.aboveMaximum : require_message.message`Expected a value less than or equal to ${require_message.text(options.max.toLocaleString("en"))}, but got ${input}.`
				};
				return {
					success: true,
					value
				};
			},
			format(value) {
				return value.toString();
			}
		};
	}
	const metavar$1 = options?.metavar ?? "INTEGER";
	require_nonempty.ensureNonEmptyString(metavar$1);
	const maxSafe = BigInt(Number.MAX_SAFE_INTEGER);
	const minSafe = BigInt(Number.MIN_SAFE_INTEGER);
	const safeMin = Math.max(options?.min ?? Number.MIN_SAFE_INTEGER, Number.MIN_SAFE_INTEGER);
	const safeMax = Math.min(options?.max ?? Number.MAX_SAFE_INTEGER, Number.MAX_SAFE_INTEGER);
	const firstAllowed = Math.ceil(safeMin);
	const lastAllowed = Math.floor(safeMax);
	if (firstAllowed > lastAllowed) throw new RangeError("The configured integer range contains no safe integers. Use type: \"bigint\" instead.");
	const unsafeIntegerError = options?.errors?.unsafeInteger;
	function makeUnsafeIntegerError(input) {
		return {
			success: false,
			error: unsafeIntegerError ? typeof unsafeIntegerError === "function" ? unsafeIntegerError(input) : unsafeIntegerError : require_message.message`Expected a safe integer between ${require_message.text(Number.MIN_SAFE_INTEGER.toLocaleString("en"))} and ${require_message.text(Number.MAX_SAFE_INTEGER.toLocaleString("en"))}, but got ${input}. Use type: "bigint" for large values.`
		};
	}
	return {
		mode: "sync",
		metavar: metavar$1,
		placeholder: options?.placeholder ?? (firstAllowed > 0 ? firstAllowed : lastAllowed < 0 ? lastAllowed : 0),
		parse(input) {
			if (!input.match(/^-?\d+$/)) return {
				success: false,
				error: options?.errors?.invalidInteger ? typeof options.errors.invalidInteger === "function" ? options.errors.invalidInteger(input) : options.errors.invalidInteger : require_message.message`Expected a valid integer, but got ${input}.`
			};
			let n;
			try {
				n = BigInt(input);
			} catch {
				return makeUnsafeIntegerError(input);
			}
			if (n > maxSafe || n < minSafe) return makeUnsafeIntegerError(input);
			const value = Number(input);
			if (options?.min != null && value < options.min) return {
				success: false,
				error: options.errors?.belowMinimum ? typeof options.errors.belowMinimum === "function" ? options.errors.belowMinimum(value, options.min) : options.errors.belowMinimum : require_message.message`Expected a value greater than or equal to ${require_message.text(options.min.toLocaleString("en"))}, but got ${input}.`
			};
			else if (options?.max != null && value > options.max) return {
				success: false,
				error: options.errors?.aboveMaximum ? typeof options.errors.aboveMaximum === "function" ? options.errors.aboveMaximum(value, options.max) : options.errors.aboveMaximum : require_message.message`Expected a value less than or equal to ${require_message.text(options.max.toLocaleString("en"))}, but got ${input}.`
			};
			return {
				success: true,
				value
			};
		},
		format(value) {
			return value.toString();
		}
	};
}
/**
* Creates a {@link ValueParser} for floating-point numbers.
*
* This parser validates that the input is a valid floating-point number
* and optionally enforces minimum and maximum value constraints.
* @param options Configuration options for the float parser.
* @returns A {@link ValueParser} that parses strings into floating-point
*          numbers.
*/
function float(options = {}) {
	if (options.min != null && !Number.isFinite(options.min)) throw new RangeError(`Expected min to be a finite number, but got: ${options.min}`);
	if (options.max != null && !Number.isFinite(options.max)) throw new RangeError(`Expected max to be a finite number, but got: ${options.max}`);
	if (options.min != null && options.max != null && options.min > options.max) throw new RangeError(`Expected min to be less than or equal to max, but got min: ${options.min} and max: ${options.max}.`);
	const floatRegex = /^[+-]?(?:(?:\d+\.?\d*)|(?:\d*\.\d+))(?:[eE][+-]?\d+)?$/;
	const metavar$1 = options.metavar ?? "NUMBER";
	require_nonempty.ensureNonEmptyString(metavar$1);
	return {
		mode: "sync",
		metavar: metavar$1,
		placeholder: options?.placeholder ?? (options?.min != null && options.min > 0 ? options.min : options?.max != null && options.max < 0 ? options.max : 0),
		parse(input) {
			const invalidNumber = (i) => ({
				success: false,
				error: options.errors?.invalidNumber ? typeof options.errors.invalidNumber === "function" ? options.errors.invalidNumber(i) : options.errors.invalidNumber : require_message.message`Expected a valid number, but got ${i}.`
			});
			let value;
			const lowerInput = input.toLowerCase();
			if (lowerInput === "nan" && options.allowNaN) value = NaN;
			else if ((lowerInput === "infinity" || lowerInput === "+infinity") && options.allowInfinity) value = Infinity;
			else if (lowerInput === "-infinity" && options.allowInfinity) value = -Infinity;
			else if (floatRegex.test(input)) {
				value = Number(input);
				if (!Number.isFinite(value) && !options.allowInfinity) return invalidNumber(input);
				if (Number.isNaN(value)) return invalidNumber(input);
			} else return invalidNumber(input);
			if (options.min != null && value < options.min) return {
				success: false,
				error: options.errors?.belowMinimum ? typeof options.errors.belowMinimum === "function" ? options.errors.belowMinimum(value, options.min) : options.errors.belowMinimum : require_message.message`Expected a value greater than or equal to ${require_message.text(options.min.toLocaleString("en"))}, but got ${input}.`
			};
			else if (options.max != null && value > options.max) return {
				success: false,
				error: options.errors?.aboveMaximum ? typeof options.errors.aboveMaximum === "function" ? options.errors.aboveMaximum(value, options.max) : options.errors.aboveMaximum : require_message.message`Expected a value less than or equal to ${require_message.text(options.max.toLocaleString("en"))}, but got ${input}.`
			};
			return {
				success: true,
				value
			};
		},
		format(value) {
			return value.toString();
		}
	};
}
const SI_MULTIPLIERS = {
	b: 1,
	kb: 1e3,
	mb: 1e6,
	gb: 1e9,
	tb: 1e12,
	pb: 1e15,
	eb: 1e18
};
const IEC_MULTIPLIERS = {
	b: 1,
	kb: 1024,
	mb: 1024 ** 2,
	gb: 1024 ** 3,
	tb: 1024 ** 4,
	pb: 1024 ** 5,
	eb: 1024 ** 6,
	kib: 1024,
	mib: 1024 ** 2,
	gib: 1024 ** 3,
	tib: 1024 ** 4,
	pib: 1024 ** 5,
	eib: 1024 ** 6
};
const IEC_ONLY_MULTIPLIERS = {
	kib: 1024,
	mib: 1024 ** 2,
	gib: 1024 ** 3,
	tib: 1024 ** 4,
	pib: 1024 ** 5,
	eib: 1024 ** 6
};
const FILE_SIZE_REGEX = /^([+-]?(?:\d+\.?\d*|\d*\.\d+))\s*([a-zA-Z]*)$/;
/**
* Formats `bytes` using the most readable unit from the given ordered list.
* Falls back to `"${bytes}B"`.
*
* A unit is chosen only when the formatted value round-trips exactly: the
* rounded quotient must lie in [1, 1000) and `rounded * size === bytes` must
* hold in float64 arithmetic.  This guarantees `parse(format(x)) === x`.
*/
function formatWithUnits(bytes, units) {
	if (bytes === 0) return "0B";
	const absBytes = Math.abs(bytes);
	for (const [unit, size] of units) {
		if (absBytes < size) continue;
		const v = bytes / size;
		const rounded = Math.round(v * 100) / 100;
		const absRounded = Math.abs(rounded);
		if (absRounded >= 1 && absRounded < 1e3 && rounded * size === bytes) return `${rounded}${unit}`;
	}
	return `${bytes}B`;
}
const FORMAT_UNITS_DEFAULT = [
	["EiB", 1024 ** 6],
	["PiB", 1024 ** 5],
	["TiB", 1024 ** 4],
	["GiB", 1024 ** 3],
	["MiB", 1024 ** 2],
	["KiB", 1024],
	["EB", 1e18],
	["PB", 1e15],
	["TB", 1e12],
	["GB", 1e9],
	["MB", 1e6],
	["KB", 1e3]
];
const FORMAT_UNITS_SI_AS_BINARY = [
	["EB", 1024 ** 6],
	["PB", 1024 ** 5],
	["TB", 1024 ** 4],
	["GB", 1024 ** 3],
	["MB", 1024 ** 2],
	["KB", 1024]
];
const BIGINT_SI_MULTIPLIERS = {
	b: 1n,
	kb: 1000n,
	mb: 1000000n,
	gb: 1000000000n,
	tb: 1000000000000n,
	pb: 1000000000000000n,
	eb: 1000000000000000000n
};
const BIGINT_IEC_MULTIPLIERS = {
	b: 1n,
	kb: 1024n,
	mb: 1024n ** 2n,
	gb: 1024n ** 3n,
	tb: 1024n ** 4n,
	pb: 1024n ** 5n,
	eb: 1024n ** 6n,
	kib: 1024n,
	mib: 1024n ** 2n,
	gib: 1024n ** 3n,
	tib: 1024n ** 4n,
	pib: 1024n ** 5n,
	eib: 1024n ** 6n
};
const BIGINT_IEC_ONLY_MULTIPLIERS = {
	kib: 1024n,
	mib: 1024n ** 2n,
	gib: 1024n ** 3n,
	tib: 1024n ** 4n,
	pib: 1024n ** 5n,
	eib: 1024n ** 6n
};
const BIGINT_FORMAT_UNITS_DEFAULT = [
	["EiB", 1024n ** 6n],
	["PiB", 1024n ** 5n],
	["TiB", 1024n ** 4n],
	["GiB", 1024n ** 3n],
	["MiB", 1024n ** 2n],
	["KiB", 1024n],
	["EB", 1000000000000000000n],
	["PB", 1000000000000000n],
	["TB", 1000000000000n],
	["GB", 1000000000n],
	["MB", 1000000n],
	["KB", 1000n]
];
const BIGINT_FORMAT_UNITS_SI_AS_BINARY = [
	["EB", 1024n ** 6n],
	["PB", 1024n ** 5n],
	["TB", 1024n ** 4n],
	["GB", 1024n ** 3n],
	["MB", 1024n ** 2n],
	["KB", 1024n]
];
/**
* Formats a bigint byte count using the most readable unit.  Falls back to
* `"${value}B"`.  Tries exact integers, then 1 and 2 decimal places.
*/
function formatBigIntBytes(value, units) {
	if (value === 0n) return "0B";
	const absValue = value < 0n ? -value : value;
	for (const [unit, size] of units) {
		if (absValue < size) continue;
		if (value % size === 0n) {
			const v = value / size;
			const absV = v < 0n ? -v : v;
			if (absV >= 1n && absV < 1000n) return `${v}${unit}`;
		}
		const v100 = value * 100n;
		if (v100 % size === 0n) {
			const q = v100 / size;
			const absQ = q < 0n ? -q : q;
			if (absQ >= 100n && absQ < 100000n) {
				const intPart = q / 100n;
				const decPart = absQ % 100n;
				if (decPart % 10n === 0n) return `${intPart}.${decPart / 10n}${unit}`;
				return `${intPart}.${String(decPart).padStart(2, "0")}${unit}`;
			}
		}
	}
	return `${value}B`;
}
/**
* Core computation: parses `numStr` as a decimal rational, multiplies by
* `mBig`, and returns the exact integer result as a `bigint`, or `null` when
* the result would be fractional.  The safe-integer range check is NOT applied
* here; callers add it as needed.
*/
function parseExactBytesRaw(numStr, mBig) {
	const negative = numStr.startsWith("-");
	const absStr = numStr.startsWith("+") || numStr.startsWith("-") ? numStr.slice(1) : numStr;
	const dotIdx = absStr.indexOf(".");
	const intPart = dotIdx < 0 ? absStr : absStr.slice(0, dotIdx);
	const fracPart = dotIdx < 0 ? "" : absStr.slice(dotIdx + 1);
	const numeratorStr = ((intPart || "0") + fracPart).replace(/^0+/, "") || "0";
	const numerator = BigInt(numeratorStr) * (negative ? -1n : 1n);
	const denominator = 10n ** BigInt(fracPart.length);
	const bytesNumerator = numerator * mBig;
	if (bytesNumerator % denominator !== 0n) return null;
	return bytesNumerator / denominator;
}
/**
* Parses `numStr` as a decimal rational and multiplies by `multiplier`,
* returning the exact integer result as a safe `number`, or `null` when the
* result would be fractional or outside `Number.MAX_SAFE_INTEGER`.
*
* All arithmetic is performed with `bigint` to avoid float64 precision loss
* (e.g. `"1.0000000000000001"` must not silently round to `1`).
*/
function parseExactBytes(numStr, multiplier) {
	const bytes = parseExactBytesRaw(numStr, BigInt(multiplier));
	if (bytes == null) return null;
	const maxSafe = BigInt(Number.MAX_SAFE_INTEGER);
	if (bytes < -maxSafe || bytes > maxSafe) return null;
	return Number(bytes);
}
const FILE_SIZE_UNITS = [
	"B",
	"KB",
	"MB",
	"GB",
	"TB",
	"PB",
	"EB",
	"KiB",
	"MiB",
	"GiB",
	"TiB",
	"PiB",
	"EiB"
];
/**
* Creates a {@link ValueParser} for human-readable file/data size strings such
* as `"10MB"`, `"1.5GiB"`, or `"512B"`.  The parsed value is a `number` or
* `bigint` representing the equivalent byte count.
*
* Supported units:
*
* | Unit | Bytes (default) |
* |------|----------------|
* | B    | 1              |
* | KB   | 1 000          |
* | MB   | 1 000 000      |
* | GB   | 1 000 000 000  |
* | KiB  | 1 024          |
* | MiB  | 1 048 576      |
* | GiB  | 1 073 741 824  |
* | …    | …              |
*
* Unit suffixes are matched case-insensitively, so `"1kb"`, `"1KB"`, and
* `"1Kb"` are all equivalent.
*
* @param options Configuration options for the file size parser.
* @returns A {@link ValueParser} that parses file size strings into byte
*          counts.
* @throws {TypeError} If `type` is neither `"number"` nor `"bigint"`, if
*   `metavar` is an empty string, if `allowNegative` or `siAsBinary` is not
*   a boolean, or if `defaultUnit` is not a valid {@link FileSizeUnit}.
* @since 1.1.0
*/
function fileSize(options = {}) {
	if (options.type !== void 0 && options.type !== "number" && options.type !== "bigint") throw new TypeError(`Expected type to be "number" or "bigint", but got: ${String(options.type)}.`);
	const metavar$1 = options.metavar ?? "SIZE";
	require_nonempty.ensureNonEmptyString(metavar$1);
	checkBooleanOption(options, "allowNegative");
	checkBooleanOption(options, "siAsBinary");
	checkEnumOption(options, "defaultUnit", FILE_SIZE_UNITS);
	const siAsBinary = options.siAsBinary ?? false;
	function invalidFormatError(input) {
		return {
			success: false,
			error: options.errors?.invalidFormat ? typeof options.errors.invalidFormat === "function" ? options.errors.invalidFormat(input) : options.errors.invalidFormat : require_message.message`Expected a file size like ${"10MB"} or ${"1.5GiB"}, but got ${input}.`
		};
	}
	if (options.type === "bigint") {
		const bigintMultiplierMap = siAsBinary ? BIGINT_IEC_MULTIPLIERS : {
			...BIGINT_SI_MULTIPLIERS,
			...BIGINT_IEC_ONLY_MULTIPLIERS
		};
		const bigintFormatUnits = siAsBinary ? BIGINT_FORMAT_UNITS_SI_AS_BINARY : BIGINT_FORMAT_UNITS_DEFAULT;
		const numberFormatUnits = siAsBinary ? FORMAT_UNITS_SI_AS_BINARY : FORMAT_UNITS_DEFAULT;
		return {
			mode: "sync",
			metavar: metavar$1,
			placeholder: options.placeholder ?? 0n,
			parse(input) {
				const match = FILE_SIZE_REGEX.exec(input.trim());
				if (match == null) return invalidFormatError(input);
				const numStr = match[1];
				const unitStr = match[2].toLowerCase();
				let mBig;
				if (unitStr === "") {
					if (options.defaultUnit == null) return invalidFormatError(input);
					mBig = bigintMultiplierMap[options.defaultUnit.toLowerCase()];
				} else {
					const m = bigintMultiplierMap[unitStr];
					if (m == null) return invalidFormatError(input);
					mBig = m;
				}
				const bytes = parseExactBytesRaw(numStr, mBig);
				if (bytes == null) return invalidFormatError(input);
				if (!(options.allowNegative ?? false) && bytes < 0n) return {
					success: false,
					error: options.errors?.negativeNotAllowed ? typeof options.errors.negativeNotAllowed === "function" ? options.errors.negativeNotAllowed(bytes) : options.errors.negativeNotAllowed : require_message.message`Expected a non-negative file size, but got ${input}.`
				};
				return {
					success: true,
					value: bytes
				};
			},
			format(value) {
				const maxSafe = BigInt(Number.MAX_SAFE_INTEGER);
				if (value >= -maxSafe && value <= maxSafe) return formatWithUnits(Number(value), numberFormatUnits);
				return formatBigIntBytes(value, bigintFormatUnits);
			}
		};
	}
	const multiplierMap = siAsBinary ? IEC_MULTIPLIERS : {
		...SI_MULTIPLIERS,
		...IEC_ONLY_MULTIPLIERS
	};
	const formatUnits = siAsBinary ? FORMAT_UNITS_SI_AS_BINARY : FORMAT_UNITS_DEFAULT;
	return {
		mode: "sync",
		metavar: metavar$1,
		placeholder: options.placeholder ?? 0,
		parse(input) {
			const match = FILE_SIZE_REGEX.exec(input.trim());
			if (match == null) return invalidFormatError(input);
			const numStr = match[1];
			const unitStr = match[2].toLowerCase();
			let multiplier;
			if (unitStr === "") {
				if (options.defaultUnit == null) return invalidFormatError(input);
				multiplier = multiplierMap[options.defaultUnit.toLowerCase()];
			} else {
				const m = multiplierMap[unitStr];
				if (m == null) return invalidFormatError(input);
				multiplier = m;
			}
			const bytes = parseExactBytes(numStr, multiplier);
			if (bytes == null) return invalidFormatError(input);
			if (!(options.allowNegative ?? false) && bytes < 0) return {
				success: false,
				error: options.errors?.negativeNotAllowed ? typeof options.errors.negativeNotAllowed === "function" ? options.errors.negativeNotAllowed(bytes) : options.errors.negativeNotAllowed : require_message.message`Expected a non-negative file size, but got ${input}.`
			};
			return {
				success: true,
				value: bytes
			};
		},
		format(value) {
			return formatWithUnits(value, formatUnits);
		}
	};
}
const CSS_NAMED_COLORS = {
	aliceblue: {
		r: 240,
		g: 248,
		b: 255,
		a: 1
	},
	antiquewhite: {
		r: 250,
		g: 235,
		b: 215,
		a: 1
	},
	aqua: {
		r: 0,
		g: 255,
		b: 255,
		a: 1
	},
	aquamarine: {
		r: 127,
		g: 255,
		b: 212,
		a: 1
	},
	azure: {
		r: 240,
		g: 255,
		b: 255,
		a: 1
	},
	beige: {
		r: 245,
		g: 245,
		b: 220,
		a: 1
	},
	bisque: {
		r: 255,
		g: 228,
		b: 196,
		a: 1
	},
	black: {
		r: 0,
		g: 0,
		b: 0,
		a: 1
	},
	blanchedalmond: {
		r: 255,
		g: 235,
		b: 205,
		a: 1
	},
	blue: {
		r: 0,
		g: 0,
		b: 255,
		a: 1
	},
	blueviolet: {
		r: 138,
		g: 43,
		b: 226,
		a: 1
	},
	brown: {
		r: 165,
		g: 42,
		b: 42,
		a: 1
	},
	burlywood: {
		r: 222,
		g: 184,
		b: 135,
		a: 1
	},
	cadetblue: {
		r: 95,
		g: 158,
		b: 160,
		a: 1
	},
	chartreuse: {
		r: 127,
		g: 255,
		b: 0,
		a: 1
	},
	chocolate: {
		r: 210,
		g: 105,
		b: 30,
		a: 1
	},
	coral: {
		r: 255,
		g: 127,
		b: 80,
		a: 1
	},
	cornflowerblue: {
		r: 100,
		g: 149,
		b: 237,
		a: 1
	},
	cornsilk: {
		r: 255,
		g: 248,
		b: 220,
		a: 1
	},
	crimson: {
		r: 220,
		g: 20,
		b: 60,
		a: 1
	},
	cyan: {
		r: 0,
		g: 255,
		b: 255,
		a: 1
	},
	darkblue: {
		r: 0,
		g: 0,
		b: 139,
		a: 1
	},
	darkcyan: {
		r: 0,
		g: 139,
		b: 139,
		a: 1
	},
	darkgoldenrod: {
		r: 184,
		g: 134,
		b: 11,
		a: 1
	},
	darkgray: {
		r: 169,
		g: 169,
		b: 169,
		a: 1
	},
	darkgreen: {
		r: 0,
		g: 100,
		b: 0,
		a: 1
	},
	darkgrey: {
		r: 169,
		g: 169,
		b: 169,
		a: 1
	},
	darkkhaki: {
		r: 189,
		g: 183,
		b: 107,
		a: 1
	},
	darkmagenta: {
		r: 139,
		g: 0,
		b: 139,
		a: 1
	},
	darkolivegreen: {
		r: 85,
		g: 107,
		b: 47,
		a: 1
	},
	darkorange: {
		r: 255,
		g: 140,
		b: 0,
		a: 1
	},
	darkorchid: {
		r: 153,
		g: 50,
		b: 204,
		a: 1
	},
	darkred: {
		r: 139,
		g: 0,
		b: 0,
		a: 1
	},
	darksalmon: {
		r: 233,
		g: 150,
		b: 122,
		a: 1
	},
	darkseagreen: {
		r: 143,
		g: 188,
		b: 143,
		a: 1
	},
	darkslateblue: {
		r: 72,
		g: 61,
		b: 139,
		a: 1
	},
	darkslategray: {
		r: 47,
		g: 79,
		b: 79,
		a: 1
	},
	darkslategrey: {
		r: 47,
		g: 79,
		b: 79,
		a: 1
	},
	darkturquoise: {
		r: 0,
		g: 206,
		b: 209,
		a: 1
	},
	darkviolet: {
		r: 148,
		g: 0,
		b: 211,
		a: 1
	},
	deeppink: {
		r: 255,
		g: 20,
		b: 147,
		a: 1
	},
	deepskyblue: {
		r: 0,
		g: 191,
		b: 255,
		a: 1
	},
	dimgray: {
		r: 105,
		g: 105,
		b: 105,
		a: 1
	},
	dimgrey: {
		r: 105,
		g: 105,
		b: 105,
		a: 1
	},
	dodgerblue: {
		r: 30,
		g: 144,
		b: 255,
		a: 1
	},
	firebrick: {
		r: 178,
		g: 34,
		b: 34,
		a: 1
	},
	floralwhite: {
		r: 255,
		g: 250,
		b: 240,
		a: 1
	},
	forestgreen: {
		r: 34,
		g: 139,
		b: 34,
		a: 1
	},
	fuchsia: {
		r: 255,
		g: 0,
		b: 255,
		a: 1
	},
	gainsboro: {
		r: 220,
		g: 220,
		b: 220,
		a: 1
	},
	ghostwhite: {
		r: 248,
		g: 248,
		b: 255,
		a: 1
	},
	gold: {
		r: 255,
		g: 215,
		b: 0,
		a: 1
	},
	goldenrod: {
		r: 218,
		g: 165,
		b: 32,
		a: 1
	},
	gray: {
		r: 128,
		g: 128,
		b: 128,
		a: 1
	},
	green: {
		r: 0,
		g: 128,
		b: 0,
		a: 1
	},
	greenyellow: {
		r: 173,
		g: 255,
		b: 47,
		a: 1
	},
	grey: {
		r: 128,
		g: 128,
		b: 128,
		a: 1
	},
	honeydew: {
		r: 240,
		g: 255,
		b: 240,
		a: 1
	},
	hotpink: {
		r: 255,
		g: 105,
		b: 180,
		a: 1
	},
	indianred: {
		r: 205,
		g: 92,
		b: 92,
		a: 1
	},
	indigo: {
		r: 75,
		g: 0,
		b: 130,
		a: 1
	},
	ivory: {
		r: 255,
		g: 255,
		b: 240,
		a: 1
	},
	khaki: {
		r: 240,
		g: 230,
		b: 140,
		a: 1
	},
	lavender: {
		r: 230,
		g: 230,
		b: 250,
		a: 1
	},
	lavenderblush: {
		r: 255,
		g: 240,
		b: 245,
		a: 1
	},
	lawngreen: {
		r: 124,
		g: 252,
		b: 0,
		a: 1
	},
	lemonchiffon: {
		r: 255,
		g: 250,
		b: 205,
		a: 1
	},
	lightblue: {
		r: 173,
		g: 216,
		b: 230,
		a: 1
	},
	lightcoral: {
		r: 240,
		g: 128,
		b: 128,
		a: 1
	},
	lightcyan: {
		r: 224,
		g: 255,
		b: 255,
		a: 1
	},
	lightgoldenrodyellow: {
		r: 250,
		g: 250,
		b: 210,
		a: 1
	},
	lightgray: {
		r: 211,
		g: 211,
		b: 211,
		a: 1
	},
	lightgreen: {
		r: 144,
		g: 238,
		b: 144,
		a: 1
	},
	lightgrey: {
		r: 211,
		g: 211,
		b: 211,
		a: 1
	},
	lightpink: {
		r: 255,
		g: 182,
		b: 193,
		a: 1
	},
	lightsalmon: {
		r: 255,
		g: 160,
		b: 122,
		a: 1
	},
	lightseagreen: {
		r: 32,
		g: 178,
		b: 170,
		a: 1
	},
	lightskyblue: {
		r: 135,
		g: 206,
		b: 250,
		a: 1
	},
	lightslategray: {
		r: 119,
		g: 136,
		b: 153,
		a: 1
	},
	lightslategrey: {
		r: 119,
		g: 136,
		b: 153,
		a: 1
	},
	lightsteelblue: {
		r: 176,
		g: 196,
		b: 222,
		a: 1
	},
	lightyellow: {
		r: 255,
		g: 255,
		b: 224,
		a: 1
	},
	lime: {
		r: 0,
		g: 255,
		b: 0,
		a: 1
	},
	limegreen: {
		r: 50,
		g: 205,
		b: 50,
		a: 1
	},
	linen: {
		r: 250,
		g: 240,
		b: 230,
		a: 1
	},
	magenta: {
		r: 255,
		g: 0,
		b: 255,
		a: 1
	},
	maroon: {
		r: 128,
		g: 0,
		b: 0,
		a: 1
	},
	mediumaquamarine: {
		r: 102,
		g: 205,
		b: 170,
		a: 1
	},
	mediumblue: {
		r: 0,
		g: 0,
		b: 205,
		a: 1
	},
	mediumorchid: {
		r: 186,
		g: 85,
		b: 211,
		a: 1
	},
	mediumpurple: {
		r: 147,
		g: 112,
		b: 219,
		a: 1
	},
	mediumseagreen: {
		r: 60,
		g: 179,
		b: 113,
		a: 1
	},
	mediumslateblue: {
		r: 123,
		g: 104,
		b: 238,
		a: 1
	},
	mediumspringgreen: {
		r: 0,
		g: 250,
		b: 154,
		a: 1
	},
	mediumturquoise: {
		r: 72,
		g: 209,
		b: 204,
		a: 1
	},
	mediumvioletred: {
		r: 199,
		g: 21,
		b: 133,
		a: 1
	},
	midnightblue: {
		r: 25,
		g: 25,
		b: 112,
		a: 1
	},
	mintcream: {
		r: 245,
		g: 255,
		b: 250,
		a: 1
	},
	mistyrose: {
		r: 255,
		g: 228,
		b: 225,
		a: 1
	},
	moccasin: {
		r: 255,
		g: 228,
		b: 181,
		a: 1
	},
	navajowhite: {
		r: 255,
		g: 222,
		b: 173,
		a: 1
	},
	navy: {
		r: 0,
		g: 0,
		b: 128,
		a: 1
	},
	oldlace: {
		r: 253,
		g: 245,
		b: 230,
		a: 1
	},
	olive: {
		r: 128,
		g: 128,
		b: 0,
		a: 1
	},
	olivedrab: {
		r: 107,
		g: 142,
		b: 35,
		a: 1
	},
	orange: {
		r: 255,
		g: 165,
		b: 0,
		a: 1
	},
	orangered: {
		r: 255,
		g: 69,
		b: 0,
		a: 1
	},
	orchid: {
		r: 218,
		g: 112,
		b: 214,
		a: 1
	},
	palegoldenrod: {
		r: 238,
		g: 232,
		b: 170,
		a: 1
	},
	palegreen: {
		r: 152,
		g: 251,
		b: 152,
		a: 1
	},
	paleturquoise: {
		r: 175,
		g: 238,
		b: 238,
		a: 1
	},
	palevioletred: {
		r: 219,
		g: 112,
		b: 147,
		a: 1
	},
	papayawhip: {
		r: 255,
		g: 239,
		b: 213,
		a: 1
	},
	peachpuff: {
		r: 255,
		g: 218,
		b: 185,
		a: 1
	},
	peru: {
		r: 205,
		g: 133,
		b: 63,
		a: 1
	},
	pink: {
		r: 255,
		g: 192,
		b: 203,
		a: 1
	},
	plum: {
		r: 221,
		g: 160,
		b: 221,
		a: 1
	},
	powderblue: {
		r: 176,
		g: 224,
		b: 230,
		a: 1
	},
	purple: {
		r: 128,
		g: 0,
		b: 128,
		a: 1
	},
	rebeccapurple: {
		r: 102,
		g: 51,
		b: 153,
		a: 1
	},
	red: {
		r: 255,
		g: 0,
		b: 0,
		a: 1
	},
	rosybrown: {
		r: 188,
		g: 143,
		b: 143,
		a: 1
	},
	royalblue: {
		r: 65,
		g: 105,
		b: 225,
		a: 1
	},
	saddlebrown: {
		r: 139,
		g: 69,
		b: 19,
		a: 1
	},
	salmon: {
		r: 250,
		g: 128,
		b: 114,
		a: 1
	},
	sandybrown: {
		r: 244,
		g: 164,
		b: 96,
		a: 1
	},
	seagreen: {
		r: 46,
		g: 139,
		b: 87,
		a: 1
	},
	seashell: {
		r: 255,
		g: 245,
		b: 238,
		a: 1
	},
	sienna: {
		r: 160,
		g: 82,
		b: 45,
		a: 1
	},
	silver: {
		r: 192,
		g: 192,
		b: 192,
		a: 1
	},
	skyblue: {
		r: 135,
		g: 206,
		b: 235,
		a: 1
	},
	slateblue: {
		r: 106,
		g: 90,
		b: 205,
		a: 1
	},
	slategray: {
		r: 112,
		g: 128,
		b: 144,
		a: 1
	},
	slategrey: {
		r: 112,
		g: 128,
		b: 144,
		a: 1
	},
	snow: {
		r: 255,
		g: 250,
		b: 250,
		a: 1
	},
	springgreen: {
		r: 0,
		g: 255,
		b: 127,
		a: 1
	},
	steelblue: {
		r: 70,
		g: 130,
		b: 180,
		a: 1
	},
	tan: {
		r: 210,
		g: 180,
		b: 140,
		a: 1
	},
	teal: {
		r: 0,
		g: 128,
		b: 128,
		a: 1
	},
	thistle: {
		r: 216,
		g: 191,
		b: 216,
		a: 1
	},
	tomato: {
		r: 255,
		g: 99,
		b: 71,
		a: 1
	},
	transparent: {
		r: 0,
		g: 0,
		b: 0,
		a: 0
	},
	turquoise: {
		r: 64,
		g: 224,
		b: 208,
		a: 1
	},
	violet: {
		r: 238,
		g: 130,
		b: 238,
		a: 1
	},
	wheat: {
		r: 245,
		g: 222,
		b: 179,
		a: 1
	},
	white: {
		r: 255,
		g: 255,
		b: 255,
		a: 1
	},
	whitesmoke: {
		r: 245,
		g: 245,
		b: 245,
		a: 1
	},
	yellow: {
		r: 255,
		g: 255,
		b: 0,
		a: 1
	},
	yellowgreen: {
		r: 154,
		g: 205,
		b: 50,
		a: 1
	}
};
const CSS_NAMED_COLOR_KEYS = Object.keys(CSS_NAMED_COLORS);
function hslToRgb(h, s, l) {
	h = (h % 360 + 360) % 360;
	const a = s * Math.min(l, 1 - l);
	const f = (n) => {
		const k = (n + h / 30) % 12;
		return l - a * Math.max(-1, Math.min(k - 3, 9 - k, 1));
	};
	return [
		Math.round(f(0) * 255),
		Math.round(f(8) * 255),
		Math.round(f(4) * 255)
	];
}
function rgbToHsl(r, g, b) {
	const rn = r / 255, gn = g / 255, bn = b / 255;
	const max = Math.max(rn, gn, bn);
	const min = Math.min(rn, gn, bn);
	const l = (max + min) / 2;
	if (max === min) return [
		0,
		0,
		l * 100
	];
	const d = max - min;
	const s = l > .5 ? d / (2 - max - min) : d / (max + min);
	let h;
	if (max === rn) h = ((gn - bn) / d + (gn < bn ? 6 : 0)) / 6;
	else if (max === gn) h = ((bn - rn) / d + 2) / 6;
	else h = ((rn - gn) / d + 4) / 6;
	return [
		h * 360,
		s * 100,
		l * 100
	];
}
const COLOR_HEX_SHORT_REGEX = /^#([0-9a-fA-F])([0-9a-fA-F])([0-9a-fA-F])([0-9a-fA-F])?$/;
const COLOR_HEX_LONG_REGEX = /^#([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})?$/;
const COLOR_NUM_PATTERN = "(?:\\d+(?:\\.\\d*)?|\\d*\\.\\d+)";
const COLOR_HUE_PATTERN = `(?:-?${COLOR_NUM_PATTERN})`;
const COLOR_RGB_REGEX = new RegExp(`^rgba?\\(\\s*(\\d{1,3})\\s*,\\s*(\\d{1,3})\\s*,\\s*(\\d{1,3})\\s*(?:,\\s*(${COLOR_NUM_PATTERN}))?\\s*\\)$`, "i");
const COLOR_HSL_REGEX = new RegExp(`^hsla?\\(\\s*(${COLOR_HUE_PATTERN})\\s*,\\s*(${COLOR_NUM_PATTERN})%\\s*,\\s*(${COLOR_NUM_PATTERN})%\\s*(?:,\\s*(${COLOR_NUM_PATTERN}))?\\s*\\)$`, "i");
const VALID_COLOR_FORMATS = [
	"hex",
	"rgb",
	"hsl",
	"named"
];
/**
* Creates a value parser that accepts CSS color strings and returns a
* structured {@link Color} object with normalized RGBA components.
*
* Supported input notations by default:
* - Hex: `#rgb`, `#rrggbb`, `#rgba`, `#rrggbbaa`
* - RGB: `rgb(r, g, b)`, `rgba(r, g, b, a)`
* - HSL: `hsl(h, s%, l%)`, `hsla(h, s%, l%, a)`
* - Named: all 148 CSS Level 4 named colors (e.g., `red`, `rebeccapurple`)
*
* The `format()` method always outputs canonical lowercase hex
* (`#rrggbb` when fully opaque, `#rrggbbaa` otherwise).
*
* @param options Configuration options for the parser.
* @returns A sync value parser producing {@link Color} objects.
* @throws {TypeError} If {@link ColorOptions.metavar} is an empty string, or
*   if {@link ColorOptions.formats} contains an invalid format name.
* @throws {RangeError} If {@link ValueParser.format} is called with a
*   {@link Color} whose `r`, `g`, or `b` channel is not an integer in
*   0–255, or whose `a` channel is not a finite number in 0–1.
* @since 1.1.0
*/
function color(options = {}) {
	const metavar$1 = options.metavar ?? "COLOR";
	require_nonempty.ensureNonEmptyString(metavar$1);
	if (options.formats !== void 0) {
		for (const fmt of options.formats) if (!VALID_COLOR_FORMATS.includes(fmt)) throw new TypeError(`Expected formats to contain only ${VALID_COLOR_FORMATS.map((v) => JSON.stringify(v)).join(", ")}, but got: ${JSON.stringify(fmt)}.`);
	}
	const allowedFormats = options.formats ?? VALID_COLOR_FORMATS;
	const allowHex = allowedFormats.includes("hex");
	const allowRgb = allowedFormats.includes("rgb");
	const allowHsl = allowedFormats.includes("hsl");
	const allowNamed = allowedFormats.includes("named");
	const defaultPlaceholder = {
		r: 0,
		g: 0,
		b: 0,
		a: 1
	};
	const formatExamples = [
		...allowHex ? ["#ff0000"] : [],
		...allowRgb ? ["rgb(255, 0, 0)"] : [],
		...allowHsl ? ["hsl(0, 100%, 50%)"] : [],
		...allowNamed ? ["red"] : []
	];
	function invalidFormatError(input) {
		return {
			success: false,
			error: options.errors?.invalidFormat ? typeof options.errors.invalidFormat === "function" ? options.errors.invalidFormat(input) : options.errors.invalidFormat : require_message.message`Expected a CSS color like ${require_message.valueSet(formatExamples, {
				fallback: "a valid color",
				type: "disjunction"
			})}, but got ${input}.`
		};
	}
	return {
		mode: "sync",
		metavar: metavar$1,
		placeholder: options.placeholder ?? defaultPlaceholder,
		parse(input) {
			const trimmed = input.trim();
			if (allowHex) {
				let m = COLOR_HEX_LONG_REGEX.exec(trimmed);
				if (m != null) {
					const r = parseInt(m[1], 16);
					const g = parseInt(m[2], 16);
					const b = parseInt(m[3], 16);
					const a = m[4] !== void 0 ? parseInt(m[4], 16) / 255 : 1;
					return {
						success: true,
						value: {
							r,
							g,
							b,
							a
						}
					};
				}
				m = COLOR_HEX_SHORT_REGEX.exec(trimmed);
				if (m != null) {
					const r = parseInt(m[1] + m[1], 16);
					const g = parseInt(m[2] + m[2], 16);
					const b = parseInt(m[3] + m[3], 16);
					const a = m[4] !== void 0 ? parseInt(m[4] + m[4], 16) / 255 : 1;
					return {
						success: true,
						value: {
							r,
							g,
							b,
							a
						}
					};
				}
			}
			if (allowRgb) {
				const m = COLOR_RGB_REGEX.exec(trimmed);
				if (m != null) {
					const r = parseInt(m[1], 10);
					const g = parseInt(m[2], 10);
					const b = parseInt(m[3], 10);
					const aRaw = m[4] !== void 0 ? parseFloat(m[4]) : 1;
					if (r > 255 || g > 255 || b > 255 || !Number.isFinite(aRaw) || aRaw < 0 || aRaw > 1) return invalidFormatError(input);
					const a = Math.round(aRaw * 255) / 255;
					return {
						success: true,
						value: {
							r,
							g,
							b,
							a
						}
					};
				}
			}
			if (allowHsl) {
				const m = COLOR_HSL_REGEX.exec(trimmed);
				if (m != null) {
					const h = parseFloat(m[1]);
					const s = parseFloat(m[2]);
					const l = parseFloat(m[3]);
					const aRaw = m[4] !== void 0 ? parseFloat(m[4]) : 1;
					if (!Number.isFinite(h) || !Number.isFinite(s) || s < 0 || s > 100 || !Number.isFinite(l) || l < 0 || l > 100 || !Number.isFinite(aRaw) || aRaw < 0 || aRaw > 1) return invalidFormatError(input);
					const [r, g, b] = hslToRgb(h, s / 100, l / 100);
					const a = Math.round(aRaw * 255) / 255;
					return {
						success: true,
						value: {
							r,
							g,
							b,
							a
						}
					};
				}
			}
			if (allowNamed) {
				const key = trimmed.toLowerCase();
				if (Object.hasOwn(CSS_NAMED_COLORS, key)) return {
					success: true,
					value: { ...CSS_NAMED_COLORS[key] }
				};
			}
			return invalidFormatError(input);
		},
		format(value) {
			const { r, g, b, a } = value;
			if (!Number.isInteger(r) || r < 0 || r > 255 || !Number.isInteger(g) || g < 0 || g > 255 || !Number.isInteger(b) || b < 0 || b > 255 || !Number.isFinite(a) || a < 0 || a > 1) throw new RangeError(`Color components out of range: r=${r}, g=${g}, b=${b}, a=${a}.`);
			const aStr = parseFloat(a.toFixed(4));
			if (allowHex) {
				const rh = r.toString(16).padStart(2, "0");
				const gh = g.toString(16).padStart(2, "0");
				const bh = b.toString(16).padStart(2, "0");
				if (a === 1) return `#${rh}${gh}${bh}`;
				const ah = Math.round(a * 255).toString(16).padStart(2, "0");
				return `#${rh}${gh}${bh}${ah}`;
			}
			if (allowRgb) return a === 1 ? `rgb(${r}, ${g}, ${b})` : `rgba(${r}, ${g}, ${b}, ${aStr})`;
			if (allowHsl) {
				const [h, s, l] = rgbToHsl(r, g, b);
				const hStr = parseFloat(h.toFixed(4));
				const sStr = parseFloat(s.toFixed(4));
				const lStr = parseFloat(l.toFixed(4));
				return a === 1 ? `hsl(${hStr}, ${sStr}%, ${lStr}%)` : `hsla(${hStr}, ${sStr}%, ${lStr}%, ${aStr})`;
			}
			for (const [name, c] of Object.entries(CSS_NAMED_COLORS)) if (c.r === r && c.g === g && c.b === b && c.a === a) return name;
			throw new RangeError(`No CSS named color matches { r: ${r}, g: ${g}, b: ${b}, a: ${a} }.`);
		},
		normalize(value) {
			if (!Number.isInteger(value.r) || value.r < 0 || value.r > 255 || !Number.isInteger(value.g) || value.g < 0 || value.g > 255 || !Number.isInteger(value.b) || value.b < 0 || value.b > 255 || !Number.isFinite(value.a) || value.a < 0 || value.a > 1) return value;
			const a = Math.round(value.a * 255) / 255;
			return a === value.a ? value : {
				...value,
				a
			};
		},
		*suggest(prefix) {
			if (allowNamed) {
				const lowerPrefix = prefix.toLowerCase();
				for (const name of CSS_NAMED_COLOR_KEYS) if (name.startsWith(lowerPrefix)) yield {
					kind: "literal",
					text: name
				};
			}
		}
	};
}
/**
* The set of URL schemes that are considered "special" by the WHATWG URL
* Standard.  These schemes always use the `://` authority syntax.
* Non-special schemes use only `:` (e.g., `mailto:`, `urn:`).
*/
const SPECIAL_URL_SCHEMES = new Set([
	"ftp",
	"file",
	"http",
	"https",
	"ws",
	"wss"
]);
/**
* Creates a {@link ValueParser} for URL values.
*
* This parser validates that the input is a well-formed URL and optionally
* restricts the allowed protocols. The parsed result is a JavaScript `URL`
* object.
* @param options Configuration options for the URL parser.
* @returns A {@link ValueParser} that converts string input to `URL` objects.
* @throws {TypeError} If any `allowedProtocols` entry is not a valid protocol
*   string ending with a colon (e.g., `"https:"`).
*/
function url(options = {}) {
	const originalProtocolsList = [];
	const normalizedProtocolsList = [];
	if (options.allowedProtocols != null) {
		const seen = /* @__PURE__ */ new Set();
		for (const protocol of options.allowedProtocols) {
			if (typeof protocol !== "string" || !/^[a-z][a-z0-9+\-.]*:$/i.test(protocol)) {
				const rendered = typeof protocol === "string" ? JSON.stringify(protocol) : String(protocol);
				throw new TypeError(`Each allowed protocol must be a valid protocol ending with a colon (e.g., "https:"), got: ${rendered}.`);
			}
			const normalized = protocol.toLowerCase();
			if (seen.has(normalized)) continue;
			seen.add(normalized);
			originalProtocolsList.push(protocol);
			normalizedProtocolsList.push(normalized);
		}
		if (originalProtocolsList.length === 0) throw new TypeError("allowedProtocols must not be empty.");
	}
	const originalProtocols = options.allowedProtocols != null ? Object.freeze(originalProtocolsList) : void 0;
	const allowedProtocols = options.allowedProtocols != null ? Object.freeze(normalizedProtocolsList) : void 0;
	const metavar$1 = options.metavar ?? "URL";
	require_nonempty.ensureNonEmptyString(metavar$1);
	const invalidUrl = options.errors?.invalidUrl;
	const disallowedProtocol = options.errors?.disallowedProtocol;
	return {
		mode: "sync",
		metavar: metavar$1,
		get placeholder() {
			return new URL(`${allowedProtocols?.[0] ?? "http:"}//0.invalid`);
		},
		parse(input) {
			if (!URL.canParse(input)) return {
				success: false,
				error: invalidUrl ? typeof invalidUrl === "function" ? invalidUrl(input) : invalidUrl : require_message.message`Invalid URL: ${input}.`
			};
			const url$1 = new URL(input);
			if (allowedProtocols != null && !allowedProtocols.includes(url$1.protocol)) return {
				success: false,
				error: disallowedProtocol ? typeof disallowedProtocol === "function" ? disallowedProtocol(url$1.protocol, originalProtocols) : disallowedProtocol : [
					{
						type: "text",
						text: "URL protocol "
					},
					{
						type: "value",
						value: url$1.protocol
					},
					{
						type: "text",
						text: " is not allowed. Allowed protocols: "
					},
					...require_message.valueSet(originalProtocols, {
						fallback: "",
						locale: "en-US"
					}),
					{
						type: "text",
						text: "."
					}
				]
			};
			return {
				success: true,
				value: url$1
			};
		},
		format(value) {
			return value.href;
		},
		*suggest(prefix) {
			if (allowedProtocols && prefix.length > 0 && !prefix.includes(":")) for (const protocol of allowedProtocols) {
				const cleanProtocol = protocol.replace(/:+$/, "");
				if (cleanProtocol.startsWith(prefix.toLowerCase())) {
					const suffix = SPECIAL_URL_SCHEMES.has(cleanProtocol) ? "://" : ":";
					yield {
						kind: "literal",
						text: `${cleanProtocol}${suffix}`
					};
				}
			}
		}
	};
}
/**
* Creates a {@link ValueParser} for locale values.
*
* This parser validates that the input is a well-formed locale identifier
* according to the Unicode Locale Identifier standard (BCP 47).
* The parsed result is a JavaScript `Intl.Locale` object.
* @param options Configuration options for the locale parser.
* @returns A {@link ValueParser} that converts string input to `Intl.Locale`
*          objects.
*/
function locale(options = {}) {
	const metavar$1 = options.metavar ?? "LOCALE";
	require_nonempty.ensureNonEmptyString(metavar$1);
	return {
		mode: "sync",
		metavar: metavar$1,
		placeholder: new Intl.Locale("und"),
		parse(input) {
			let locale$1;
			try {
				locale$1 = new Intl.Locale(input);
			} catch (e) {
				if (e instanceof RangeError) return {
					success: false,
					error: options.errors?.invalidLocale ? typeof options.errors.invalidLocale === "function" ? options.errors.invalidLocale(input) : options.errors.invalidLocale : require_message.message`Invalid locale: ${input}.`
				};
				throw e;
			}
			return {
				success: true,
				value: locale$1
			};
		},
		format(value) {
			return value.toString();
		},
		*suggest(prefix) {
			const commonLocales = [
				"en",
				"en-US",
				"en-GB",
				"en-CA",
				"en-AU",
				"en-NZ",
				"en-IE",
				"en-ZA",
				"en-IN",
				"es",
				"es-ES",
				"es-MX",
				"es-AR",
				"es-CL",
				"es-CO",
				"es-PE",
				"es-VE",
				"es-EC",
				"es-GT",
				"es-CU",
				"es-BO",
				"es-DO",
				"es-HN",
				"es-PY",
				"es-SV",
				"es-NI",
				"es-CR",
				"es-PA",
				"es-UY",
				"es-PR",
				"fr",
				"fr-FR",
				"fr-CA",
				"fr-BE",
				"fr-CH",
				"fr-LU",
				"fr-MC",
				"de",
				"de-DE",
				"de-AT",
				"de-CH",
				"de-BE",
				"de-LU",
				"de-LI",
				"it",
				"it-IT",
				"it-CH",
				"it-SM",
				"it-VA",
				"pt",
				"pt-BR",
				"pt-PT",
				"pt-AO",
				"pt-MZ",
				"pt-CV",
				"pt-GW",
				"pt-ST",
				"pt-TL",
				"ru",
				"ru-RU",
				"ru-BY",
				"ru-KZ",
				"ru-KG",
				"ru-MD",
				"uk",
				"uk-UA",
				"be",
				"be-BY",
				"bg",
				"bg-BG",
				"cs",
				"cs-CZ",
				"sk",
				"sk-SK",
				"sl",
				"sl-SI",
				"hr",
				"hr-HR",
				"sr",
				"sr-RS",
				"mk",
				"mk-MK",
				"ja",
				"ja-JP",
				"ko",
				"ko-KR",
				"zh",
				"zh-CN",
				"zh-TW",
				"zh-HK",
				"zh-SG",
				"zh-MO",
				"ar",
				"ar-SA",
				"ar-AE",
				"ar-BH",
				"ar-DZ",
				"ar-EG",
				"ar-IQ",
				"ar-JO",
				"ar-KW",
				"ar-LB",
				"ar-LY",
				"ar-MA",
				"ar-OM",
				"ar-QA",
				"ar-SY",
				"ar-TN",
				"ar-YE",
				"hi",
				"hi-IN",
				"bn",
				"bn-BD",
				"bn-IN",
				"ur",
				"ur-PK",
				"ta",
				"ta-IN",
				"te",
				"te-IN",
				"mr",
				"mr-IN",
				"gu",
				"gu-IN",
				"kn",
				"kn-IN",
				"ml",
				"ml-IN",
				"pa",
				"pa-IN",
				"tr",
				"tr-TR",
				"tr-CY",
				"pl",
				"pl-PL",
				"nl",
				"nl-NL",
				"nl-BE",
				"nl-SR",
				"sv",
				"sv-SE",
				"sv-FI",
				"da",
				"da-DK",
				"no",
				"no-NO",
				"nb",
				"nb-NO",
				"nn",
				"nn-NO",
				"fi",
				"fi-FI",
				"is",
				"is-IS",
				"el",
				"el-GR",
				"el-CY",
				"hu",
				"hu-HU",
				"ro",
				"ro-RO",
				"et",
				"et-EE",
				"lv",
				"lv-LV",
				"lt",
				"lt-LT",
				"mt",
				"mt-MT",
				"ga",
				"ga-IE",
				"cy",
				"cy-GB",
				"eu",
				"eu-ES",
				"ca",
				"ca-ES",
				"ca-AD",
				"th",
				"th-TH",
				"vi",
				"vi-VN",
				"id",
				"id-ID",
				"ms",
				"ms-MY",
				"ms-BN",
				"ms-SG",
				"tl",
				"tl-PH",
				"km",
				"km-KH",
				"my",
				"my-MM",
				"lo",
				"lo-LA",
				"si",
				"si-LK",
				"ne",
				"ne-NP",
				"sw",
				"sw-TZ",
				"sw-KE",
				"am",
				"am-ET",
				"ha",
				"ha-NG",
				"yo",
				"yo-NG",
				"ig",
				"ig-NG",
				"zu",
				"zu-ZA",
				"xh",
				"xh-ZA",
				"af",
				"af-ZA",
				"he",
				"he-IL",
				"fa",
				"fa-IR",
				"fa-AF"
			];
			for (const locale$1 of commonLocales) if (locale$1.toLowerCase().startsWith(prefix.toLowerCase())) yield {
				kind: "literal",
				text: locale$1
			};
		}
	};
}
/**
* Creates a {@link ValueParser} for UUID values.
*
* This parser validates that the input is a well-formed UUID string in the
* standard format: `xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx` where each `x`
* is a hexadecimal digit.
*
* By default, the parser enforces strict [RFC 9562] validation: it requires
* a standardized version digit (1 through 8) and the RFC 9562 variant bits
* (`10xx`).  The nil and max UUIDs are accepted as special standard values.
* Set `strict: false` to disable the default RFC 9562 version/variant
* checks.  An explicit {@link UuidOptions.allowedVersions} list still
* constrains the version nibble even in lenient mode.
*
* [RFC 9562]: https://www.rfc-editor.org/rfc/rfc9562
* @param options Configuration options for the UUID parser.
* @returns A {@link ValueParser} that converts string input to {@link Uuid}
*          strings.
* @throws {TypeError} If any element of
*   {@link UuidOptions.allowedVersions} is not an integer.
* @throws {RangeError} If any element of
*   {@link UuidOptions.allowedVersions} is outside the range 1 to 8.
*/
function uuid(options = {}) {
	const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
	const metavar$1 = options.metavar ?? "UUID";
	require_nonempty.ensureNonEmptyString(metavar$1);
	checkBooleanOption(options, "strict");
	const strict = options.strict !== false;
	const allowedVersions = options.allowedVersions != null ? (() => {
		const unique = /* @__PURE__ */ new Set();
		for (const v of options.allowedVersions) {
			if (!Number.isInteger(v)) throw new TypeError(`Expected every element of allowedVersions to be an integer, but got value "${typeof v === "symbol" ? v.toString() : String(v)}" of type "${Array.isArray(v) ? "array" : v === null ? "null" : typeof v}".`);
			if (v < 1 || v > 8) throw new RangeError(`Expected every element of allowedVersions to be between 1 and 8, but got: ${v}.`);
			unique.add(v);
		}
		return Object.freeze([...unique]);
	})() : null;
	const invalidUuid = options.errors?.invalidUuid;
	const disallowedVersion = options.errors?.disallowedVersion;
	const invalidVariant = options.errors?.invalidVariant;
	return {
		mode: "sync",
		metavar: metavar$1,
		placeholder: "00000000-0000-0000-0000-000000000000",
		parse(input) {
			if (!uuidRegex.test(input)) return {
				success: false,
				error: invalidUuid ? typeof invalidUuid === "function" ? invalidUuid(input) : invalidUuid : require_message.message`Expected a valid UUID in format ${"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"}, but got ${input}.`
			};
			const lower = input.toLowerCase();
			if (lower === "00000000-0000-0000-0000-000000000000" || lower === "ffffffff-ffff-ffff-ffff-ffffffffffff") return {
				success: true,
				value: input
			};
			const versionChar = input.charAt(14);
			const version = parseInt(versionChar, 16);
			if (allowedVersions != null && allowedVersions.length > 0) {
				if (!allowedVersions.includes(version)) return {
					success: false,
					error: disallowedVersion ? typeof disallowedVersion === "function" ? disallowedVersion(version, allowedVersions) : disallowedVersion : (() => {
						let expectedVersions = require_message.message``;
						let i = 0;
						for (const v of allowedVersions) {
							expectedVersions = i < 1 ? require_message.message`${expectedVersions}${v.toLocaleString("en")}` : i + 1 >= allowedVersions.length ? require_message.message`${expectedVersions}, or ${v.toLocaleString("en")}` : require_message.message`${expectedVersions}, ${v.toLocaleString("en")}`;
							i++;
						}
						return require_message.message`Expected UUID version ${expectedVersions}, but got version ${version.toLocaleString("en")}.`;
					})()
				};
			} else if (strict && (version < 1 || version > 8)) return {
				success: false,
				error: disallowedVersion ? typeof disallowedVersion === "function" ? disallowedVersion(version, [
					1,
					2,
					3,
					4,
					5,
					6,
					7,
					8
				]) : disallowedVersion : require_message.message`Expected UUID version 1 through 8, but got version ${version.toLocaleString("en")}.`
			};
			if (strict) {
				const variantChar = input.charAt(19).toLowerCase();
				if (variantChar !== "8" && variantChar !== "9" && variantChar !== "a" && variantChar !== "b") return {
					success: false,
					error: invalidVariant ? typeof invalidVariant === "function" ? invalidVariant(input) : invalidVariant : require_message.message`Expected RFC 9562 variant (8, 9, a, or b at position 20), but got ${variantChar} in ${input}.`
				};
			}
			return {
				success: true,
				value: input
			};
		},
		format(value) {
			return value;
		}
	};
}
/**
* Creates a ValueParser for TCP/UDP port numbers.
*
* This parser validates that the input is a valid port number (1-65535 by default)
* and optionally enforces range constraints and well-known port restrictions.
*
* Port numbers are validated according to the following rules:
* - Must be a valid integer (no decimals, no scientific notation)
* - Must be within the range `[min, max]` (default `[1, 65535]`)
* - If `disallowWellKnown` is `true`, ports 1-1023 are rejected
*
* The parser provides two modes of operation:
* - Regular mode: Returns JavaScript numbers (safe for all port values)
* - `bigint` mode: Returns `bigint` values for consistency with other numeric types
*
* @example
* ```typescript
* // Basic port parser (1-65535)
* option("--port", port())
*
* // Custom range (non-privileged ports only)
* option("--port", port({ min: 1024, max: 65535 }))
*
* // Disallow well-known ports (reject 1-1023)
* option("--port", port({ disallowWellKnown: true }))
*
* // Development ports only
* option("--dev-port", port({ min: 3000, max: 9000 }))
*
* // Using bigint type
* option("--port", port({ type: "bigint" }))
* ```
*
* @param options Configuration options specifying the type and constraints.
* @returns A {@link ValueParser} that converts string input to port numbers.
* @throws {TypeError} If `options.type` is provided but is neither `"number"`
*   nor `"bigint"`.
* @since 0.10.0
*/
function port(options) {
	checkBooleanOption(options, "disallowWellKnown");
	if (options?.type !== void 0 && options.type !== "number" && options.type !== "bigint") throw new TypeError(`Expected type to be "number" or "bigint", but got: ${String(options.type)}.`);
	if (options?.type === "bigint") {
		const metavar$2 = options.metavar ?? "PORT";
		require_nonempty.ensureNonEmptyString(metavar$2);
		const min$1 = options.min ?? 1n;
		const max$1 = options.max ?? 65535n;
		if (min$1 > max$1) throw new RangeError(`Expected min to be less than or equal to max, but got min: ${min$1} and max: ${max$1}.`);
		if (options.disallowWellKnown && min$1 < 1024n && max$1 < 1024n) throw new RangeError(`disallowWellKnown is incompatible with the configured port range: all ports ${min$1}..${max$1} are well-known.`);
		return {
			mode: "sync",
			metavar: metavar$2,
			placeholder: options.placeholder ?? (options.disallowWellKnown && min$1 < 1024n ? 1024n > min$1 ? 1024n : min$1 : min$1),
			parse(input) {
				if (!input.match(/^-?\d+$/)) return {
					success: false,
					error: options.errors?.invalidPort ? typeof options.errors.invalidPort === "function" ? options.errors.invalidPort(input) : options.errors.invalidPort : require_message.message`Expected a valid port number, but got ${input}.`
				};
				const value = BigInt(input);
				if (value < min$1) return {
					success: false,
					error: options.errors?.belowMinimum ? typeof options.errors.belowMinimum === "function" ? options.errors.belowMinimum(value, min$1) : options.errors.belowMinimum : require_message.message`Expected a port number greater than or equal to ${require_message.text(min$1.toLocaleString("en"))}, but got ${input}.`
				};
				if (value > max$1) return {
					success: false,
					error: options.errors?.aboveMaximum ? typeof options.errors.aboveMaximum === "function" ? options.errors.aboveMaximum(value, max$1) : options.errors.aboveMaximum : require_message.message`Expected a port number less than or equal to ${require_message.text(max$1.toLocaleString("en"))}, but got ${input}.`
				};
				if (options.disallowWellKnown && value >= 1n && value <= 1023n) return {
					success: false,
					error: options.errors?.wellKnownNotAllowed ? typeof options.errors.wellKnownNotAllowed === "function" ? options.errors.wellKnownNotAllowed(value) : options.errors.wellKnownNotAllowed : require_message.message`Port ${value.toLocaleString("en")} is a well-known port (1-1023) and may require elevated privileges.`
				};
				return {
					success: true,
					value
				};
			},
			format(value) {
				return value.toString();
			}
		};
	}
	if (options?.min != null && !Number.isFinite(options.min)) throw new RangeError(`Expected min to be a finite number, but got: ${options.min}`);
	if (options?.max != null && !Number.isFinite(options.max)) throw new RangeError(`Expected max to be a finite number, but got: ${options.max}`);
	const metavar$1 = options?.metavar ?? "PORT";
	require_nonempty.ensureNonEmptyString(metavar$1);
	const min = options?.min ?? 1;
	const max = options?.max ?? 65535;
	if (min > max) throw new RangeError(`Expected min to be less than or equal to max, but got min: ${min} and max: ${max}.`);
	if (options?.disallowWellKnown && min < 1024 && max < 1024) throw new RangeError(`disallowWellKnown is incompatible with the configured port range: all ports ${min}..${max} are well-known.`);
	return {
		mode: "sync",
		metavar: metavar$1,
		placeholder: options?.placeholder ?? (options?.disallowWellKnown && min < 1024 ? Math.max(1024, min) : min),
		parse(input) {
			if (!input.match(/^-?\d+$/)) return {
				success: false,
				error: options?.errors?.invalidPort ? typeof options.errors.invalidPort === "function" ? options.errors.invalidPort(input) : options.errors.invalidPort : require_message.message`Expected a valid port number, but got ${input}.`
			};
			const value = Number.parseInt(input);
			if (value < min) return {
				success: false,
				error: options?.errors?.belowMinimum ? typeof options.errors.belowMinimum === "function" ? options.errors.belowMinimum(value, min) : options.errors.belowMinimum : require_message.message`Expected a port number greater than or equal to ${require_message.text(min.toLocaleString("en"))}, but got ${input}.`
			};
			if (value > max) return {
				success: false,
				error: options?.errors?.aboveMaximum ? typeof options.errors.aboveMaximum === "function" ? options.errors.aboveMaximum(value, max) : options.errors.aboveMaximum : require_message.message`Expected a port number less than or equal to ${require_message.text(max.toLocaleString("en"))}, but got ${input}.`
			};
			if (options?.disallowWellKnown && value >= 1 && value <= 1023) return {
				success: false,
				error: options.errors?.wellKnownNotAllowed ? typeof options.errors.wellKnownNotAllowed === "function" ? options.errors.wellKnownNotAllowed(value) : options.errors.wellKnownNotAllowed : require_message.message`Port ${value.toLocaleString("en")} is a well-known port (1-1023) and may require elevated privileges.`
			};
			return {
				success: true,
				value
			};
		},
		format(value) {
			return value.toString();
		}
	};
}
function isPrivateIp(octets) {
	if (octets[0] === 10) return true;
	if (octets[0] === 172 && octets[1] >= 16 && octets[1] <= 31) return true;
	if (octets[0] === 192 && octets[1] === 168) return true;
	return false;
}
function isLoopbackIp(octets) {
	return octets[0] === 127;
}
function isLinkLocalIp(octets) {
	return octets[0] === 169 && octets[1] === 254;
}
function isMulticastIp(octets) {
	return octets[0] >= 224 && octets[0] <= 239;
}
function isBroadcastIp(octets) {
	return octets.every((o) => o === 255);
}
function isZeroIp(octets) {
	return octets.every((o) => o === 0);
}
/**
* Creates a value parser for IPv4 addresses.
*
* This parser validates IPv4 addresses in dotted-decimal notation (e.g.,
* "192.168.1.1") and provides options to filter specific IP address types
* such as private, loopback, link-local, multicast, broadcast, and zero
* addresses.
*
* @param options The parser options.
* @returns A value parser for IPv4 addresses.
* @throws {TypeError} If the metavar is an empty string.
* @since 0.10.0
* @example
* ```typescript
* import { ipv4 } from "@optique/core/valueparser";
*
* // Basic IPv4 parser (allows all types)
* const address = ipv4();
*
* // Public IPs only (no private/loopback)
* const publicIp = ipv4({
*   allowPrivate: false,
*   allowLoopback: false
* });
*
* // Server binding (allow 0.0.0.0 and private IPs)
* const bindAddress = ipv4({
*   allowZero: true,
*   allowPrivate: true
* });
* ```
*/
function ipv4(options) {
	const metavar$1 = options?.metavar ?? "IPV4";
	require_nonempty.ensureNonEmptyString(metavar$1);
	const allowPrivate = options?.allowPrivate ?? true;
	const allowLoopback = options?.allowLoopback ?? true;
	const allowLinkLocal = options?.allowLinkLocal ?? true;
	const allowMulticast = options?.allowMulticast ?? true;
	const allowBroadcast = options?.allowBroadcast ?? true;
	const allowZero = options?.allowZero ?? true;
	return {
		mode: "sync",
		metavar: metavar$1,
		placeholder: allowZero ? "0.0.0.0" : allowLoopback ? "127.0.0.1" : "192.0.2.1",
		parse(input) {
			const octets = parseIpv4Octets(input);
			if (octets === null) {
				const errorMsg = options?.errors?.invalidIpv4;
				const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? require_message.message`Expected a valid IPv4 address, but got ${input}.`;
				return {
					success: false,
					error: msg
				};
			}
			const ipAddress = octets.join(".");
			if (!allowPrivate && isPrivateIp(octets)) {
				const errorMsg = options?.errors?.privateNotAllowed;
				const msg = typeof errorMsg === "function" ? errorMsg(ipAddress) : errorMsg ?? require_message.message`${ipAddress} is a private IP address.`;
				return {
					success: false,
					error: msg
				};
			}
			if (!allowLoopback && isLoopbackIp(octets)) {
				const errorMsg = options?.errors?.loopbackNotAllowed;
				const msg = typeof errorMsg === "function" ? errorMsg(ipAddress) : errorMsg ?? require_message.message`${ipAddress} is a loopback address.`;
				return {
					success: false,
					error: msg
				};
			}
			if (!allowLinkLocal && isLinkLocalIp(octets)) {
				const errorMsg = options?.errors?.linkLocalNotAllowed;
				const msg = typeof errorMsg === "function" ? errorMsg(ipAddress) : errorMsg ?? require_message.message`${ipAddress} is a link-local address.`;
				return {
					success: false,
					error: msg
				};
			}
			if (!allowMulticast && isMulticastIp(octets)) {
				const errorMsg = options?.errors?.multicastNotAllowed;
				const msg = typeof errorMsg === "function" ? errorMsg(ipAddress) : errorMsg ?? require_message.message`${ipAddress} is a multicast address.`;
				return {
					success: false,
					error: msg
				};
			}
			if (!allowBroadcast && isBroadcastIp(octets)) {
				const errorMsg = options?.errors?.broadcastNotAllowed;
				const msg = typeof errorMsg === "function" ? errorMsg(ipAddress) : errorMsg ?? require_message.message`${ipAddress} is the broadcast address.`;
				return {
					success: false,
					error: msg
				};
			}
			if (!allowZero && isZeroIp(octets)) {
				const errorMsg = options?.errors?.zeroNotAllowed;
				const msg = typeof errorMsg === "function" ? errorMsg(ipAddress) : errorMsg ?? require_message.message`${ipAddress} is the zero address.`;
				return {
					success: false,
					error: msg
				};
			}
			return {
				success: true,
				value: ipAddress
			};
		},
		format(value) {
			return value;
		}
	};
}
/**
* Creates a value parser for DNS hostnames.
*
* Validates hostnames according to RFC 1123:
* - Labels separated by dots
* - Each label: 1-63 characters
* - Labels can contain alphanumeric characters and hyphens
* - Labels cannot start or end with a hyphen
* - Total length ≤ 253 characters (default)
* - Dotted all-numeric strings (e.g., `192.168.0.1`) are rejected as they
*   resemble IPv4 addresses rather than DNS hostnames
*
* @param options - Options for hostname validation.
* @returns A value parser for hostnames.
* @throws {TypeError} If `allowWildcard`, `allowUnderscore`, or
*   `allowLocalhost` is not a boolean.
* @throws {RangeError} If `maxLength` is not a positive integer.
* @since 0.10.0
*
* @example
* ```typescript
* import { hostname } from "@optique/core/valueparser";
*
* // Basic hostname parser
* const host = hostname();
*
* // Allow wildcards for certificate validation
* const domain = hostname({ allowWildcard: true });
*
* // Reject localhost
* const remoteHost = hostname({ allowLocalhost: false });
* ```
*/
function hostname(options) {
	const metavar$1 = options?.metavar ?? "HOST";
	require_nonempty.ensureNonEmptyString(metavar$1);
	checkBooleanOption(options, "allowWildcard");
	checkBooleanOption(options, "allowUnderscore");
	checkBooleanOption(options, "allowLocalhost");
	const allowWildcard = options?.allowWildcard ?? false;
	const allowUnderscore = options?.allowUnderscore ?? false;
	const allowLocalhost = options?.allowLocalhost ?? true;
	const maxLength = options?.maxLength ?? 253;
	if (!Number.isInteger(maxLength) || maxLength < 1) throw new RangeError("maxLength must be an integer greater than or equal to 1.");
	return {
		mode: "sync",
		metavar: metavar$1,
		placeholder: options?.placeholder ?? (allowLocalhost ? maxLength >= 9 ? "localhost" : "a.bc" : maxLength >= 11 ? "example.com" : "a.bc"),
		parse(input) {
			if (input.length > maxLength) {
				const errorMsg = options?.errors?.tooLong;
				const msg = typeof errorMsg === "function" ? errorMsg(input, maxLength) : errorMsg ?? require_message.message`Hostname ${input} is too long (maximum ${require_message.text(maxLength.toString())} characters).`;
				return {
					success: false,
					error: msg
				};
			}
			if (!allowLocalhost && input.toLowerCase() === "localhost") {
				const errorMsg = options?.errors?.localhostNotAllowed;
				const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? require_message.message`Hostname 'localhost' is not allowed.`;
				return {
					success: false,
					error: msg
				};
			}
			if (input.startsWith("*.")) {
				if (!allowWildcard) {
					const errorMsg = options?.errors?.wildcardNotAllowed;
					const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? require_message.message`Wildcard hostname ${input} is not allowed.`;
					return {
						success: false,
						error: msg
					};
				}
				const rest = input.slice(2);
				if (!rest || rest.includes("*")) {
					const errorMsg = options?.errors?.invalidHostname;
					const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? require_message.message`Expected a valid hostname, but got ${input}.`;
					return {
						success: false,
						error: msg
					};
				}
				if (!allowLocalhost && rest.toLowerCase() === "localhost") {
					const errorMsg = options?.errors?.localhostNotAllowed;
					const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? require_message.message`Hostname 'localhost' is not allowed.`;
					return {
						success: false,
						error: msg
					};
				}
			}
			if (!allowUnderscore && input.includes("_")) {
				const errorMsg = options?.errors?.underscoreNotAllowed;
				const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? require_message.message`Hostname ${input} contains underscore, which is not allowed.`;
				return {
					success: false,
					error: msg
				};
			}
			if (input.length === 0) {
				const errorMsg = options?.errors?.invalidHostname;
				const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? require_message.message`Expected a valid hostname, but got ${input}.`;
				return {
					success: false,
					error: msg
				};
			}
			const labels = input.split(".");
			for (const label of labels) {
				if (label.length === 0) {
					const errorMsg = options?.errors?.invalidHostname;
					const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? require_message.message`Expected a valid hostname, but got ${input}.`;
					return {
						success: false,
						error: msg
					};
				}
				if (label.length > 63) {
					const errorMsg = options?.errors?.invalidHostname;
					const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? require_message.message`Expected a valid hostname, but got ${input}.`;
					return {
						success: false,
						error: msg
					};
				}
				if (label === "*" && allowWildcard && input.startsWith("*.") && label === labels[0]) continue;
				if (label.startsWith("-") || label.endsWith("-")) {
					const errorMsg = options?.errors?.invalidHostname;
					const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? require_message.message`Expected a valid hostname, but got ${input}.`;
					return {
						success: false,
						error: msg
					};
				}
				const allowedPattern = allowUnderscore ? /^[a-zA-Z0-9_-]+$/ : /^[a-zA-Z0-9-]+$/;
				if (!allowedPattern.test(label)) {
					const errorMsg = options?.errors?.invalidHostname;
					const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? require_message.message`Expected a valid hostname, but got ${input}.`;
					return {
						success: false,
						error: msg
					};
				}
			}
			if (labels.length >= 2 && labels.every((l) => /^[0-9]+$/.test(l))) {
				const errorMsg = options?.errors?.invalidHostname;
				const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? require_message.message`Expected a valid hostname, but got ${input}.`;
				return {
					success: false,
					error: msg
				};
			}
			return {
				success: true,
				value: input
			};
		},
		format(value) {
			return value;
		}
	};
}
function email(options) {
	const metavar$1 = options?.metavar ?? "EMAIL";
	require_nonempty.ensureNonEmptyString(metavar$1);
	const allowMultiple = options?.allowMultiple ?? false;
	if (options?.placeholder != null) {
		if (allowMultiple && !Array.isArray(options.placeholder)) throw new TypeError("email() placeholder must be an array when allowMultiple is true.");
		if (!allowMultiple && typeof options.placeholder !== "string") throw new TypeError("email() placeholder must be a string when allowMultiple is false.");
	}
	const allowDisplayName = options?.allowDisplayName ?? false;
	const lowercase = options?.lowercase ?? false;
	const allowedDomains = options?.allowedDomains != null ? Object.freeze([...options.allowedDomains]) : void 0;
	if (allowedDomains != null) {
		if (allowedDomains.length === 0) throw new TypeError("allowedDomains must not be empty.");
		for (let i = 0; i < allowedDomains.length; i++) {
			const entry = allowedDomains[i];
			if (typeof entry !== "string") throw new TypeError(`allowedDomains[${i}] must be a string, got ${typeof entry}.`);
			if (entry !== entry.trim()) throw new TypeError(`allowedDomains[${i}] must not have leading or trailing whitespace: ${JSON.stringify(entry)}`);
			if (entry.startsWith("@")) throw new TypeError(`allowedDomains[${i}] must not start with "@": ${JSON.stringify(entry)}`);
			if (entry === "" || !entry.includes(".")) throw new TypeError(`allowedDomains[${i}] is not a valid domain: ${JSON.stringify(entry)}`);
			if (entry.startsWith(".") || entry.endsWith(".") || entry.startsWith("-") || entry.endsWith("-")) throw new TypeError(`allowedDomains[${i}] is not a valid domain: ${JSON.stringify(entry)}`);
			const labels = entry.split(".");
			for (const label of labels) if (label.length === 0 || label.length > 63 || label.startsWith("-") || label.endsWith("-") || !/^[a-zA-Z0-9-]+$/.test(label)) throw new TypeError(`allowedDomains[${i}] is not a valid domain: ${JSON.stringify(entry)}`);
			if (labels.length === 4 && labels.every((label) => /^[0-9]+$/.test(label))) throw new TypeError(`allowedDomains[${i}] is not a valid domain: ${JSON.stringify(entry)}`);
		}
	}
	const invalidEmail = options?.errors?.invalidEmail;
	const domainNotAllowed = options?.errors?.domainNotAllowed;
	const atextRegex = /^[a-zA-Z0-9._+-]+$/;
	const encoder = new TextEncoder();
	function validateEmail(input) {
		const trimmed = input.trim();
		let emailAddr = trimmed;
		if (allowDisplayName) {
			const displayNameMatch = trimmed.match(/^((?:"(?:[^"\\]|\\.)*"|[^<>"])+)\s*<([^<>]+)>$/);
			if (displayNameMatch && /\S/.test(displayNameMatch[1].replace(/"((?:[^"\\]|\\.)*)"/g, (_match, inner) => inner))) emailAddr = displayNameMatch[2].trim();
		}
		let atIndex = -1;
		if (emailAddr.startsWith("\"")) {
			const closingQuoteIndex = emailAddr.indexOf("\"", 1);
			if (closingQuoteIndex === -1) return null;
			atIndex = emailAddr.indexOf("@", closingQuoteIndex);
		} else atIndex = emailAddr.indexOf("@");
		if (atIndex === -1) return null;
		const lastAtIndex = emailAddr.lastIndexOf("@");
		if (atIndex !== lastAtIndex) return null;
		const localPart = emailAddr.substring(0, atIndex);
		const domain$1 = emailAddr.substring(atIndex + 1);
		if (!localPart || localPart.length === 0) return null;
		let isValidLocal = false;
		if (localPart.startsWith("\"") && localPart.endsWith("\"")) isValidLocal = localPart.length >= 2;
		else {
			const localParts = localPart.split(".");
			if (localPart.startsWith(".") || localPart.endsWith(".")) return null;
			isValidLocal = localParts.length > 0 && localParts.every((part) => part.length > 0 && atextRegex.test(part));
		}
		if (!isValidLocal) return null;
		if (encoder.encode(localPart).length > 64) return null;
		if (!domain$1 || domain$1.length === 0) return null;
		if (!domain$1.includes(".")) return null;
		if (domain$1.startsWith(".") || domain$1.endsWith(".") || domain$1.startsWith("-") || domain$1.endsWith("-")) return null;
		const domainLabels = domain$1.split(".");
		for (const label of domainLabels) {
			if (label.length === 0 || label.length > 63) return null;
			if (label.startsWith("-") || label.endsWith("-")) return null;
			if (!/^[a-zA-Z0-9-]+$/.test(label)) return null;
		}
		if (domainLabels.length === 4 && domainLabels.every((label) => /^[0-9]+$/.test(label))) return null;
		if (encoder.encode(emailAddr).length > 254) return null;
		const resultEmail = emailAddr;
		if (!lowercase) return resultEmail;
		const lastAt = resultEmail.lastIndexOf("@");
		return resultEmail.slice(0, lastAt) + resultEmail.slice(lastAt).toLowerCase();
	}
	/**
	* Splits an input string on commas, respecting quoted segments and
	* angle-bracket display-name syntax per RFC 5322.
	*/
	function splitEmails(input) {
		const result = [];
		let current = "";
		let inQuotes = false;
		let inAngleBrackets = false;
		let escaped = false;
		for (const char of input) {
			if (escaped) escaped = false;
			else if (char === "\\" && inQuotes) escaped = true;
			else if (char === "\"" && !inAngleBrackets) {
				if (inQuotes) inQuotes = false;
				else if (current.trim() === "") inQuotes = true;
			} else if (char === "<" && !inQuotes) inAngleBrackets = true;
			else if (char === ">" && !inQuotes) inAngleBrackets = false;
			else if (char === "," && !inQuotes && !inAngleBrackets) {
				result.push(current);
				current = "";
				continue;
			}
			current += char;
		}
		result.push(current);
		return result;
	}
	return {
		mode: "sync",
		metavar: metavar$1,
		placeholder: options?.placeholder ?? (options?.allowMultiple ? [`user@${options?.allowedDomains?.[0] ?? "example.com"}`] : `user@${options?.allowedDomains?.[0] ?? "example.com"}`),
		parse(input) {
			if (allowMultiple) {
				const emails = splitEmails(input).map((e) => e.trim());
				const validatedEmails = [];
				for (const email$1 of emails) {
					const validated = validateEmail(email$1);
					if (validated === null) {
						const errorMsg = invalidEmail;
						const msg = typeof errorMsg === "function" ? errorMsg(email$1) : errorMsg ?? require_message.message`Expected a valid email address, but got ${email$1}.`;
						return {
							success: false,
							error: msg
						};
					}
					if (allowedDomains != null) {
						const atIndex = validated.indexOf("@");
						const domain$1 = validated.substring(atIndex + 1).toLowerCase();
						const isAllowed = allowedDomains.some((allowed) => domain$1 === allowed.toLowerCase());
						if (!isAllowed) {
							const errorMsg = domainNotAllowed;
							if (typeof errorMsg === "function") return {
								success: false,
								error: errorMsg(validated, allowedDomains)
							};
							const msg = errorMsg ?? [
								{
									type: "text",
									text: "Email domain "
								},
								{
									type: "value",
									value: domain$1
								},
								{
									type: "text",
									text: " is not allowed. Allowed domains: "
								},
								...require_message.valueSet(allowedDomains, {
									fallback: "",
									locale: "en-US"
								}),
								{
									type: "text",
									text: "."
								}
							];
							return {
								success: false,
								error: msg
							};
						}
					}
					validatedEmails.push(validated);
				}
				return {
					success: true,
					value: validatedEmails
				};
			} else {
				const validated = validateEmail(input);
				if (validated === null) {
					const errorMsg = invalidEmail;
					const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? require_message.message`Expected a valid email address, but got ${input}.`;
					return {
						success: false,
						error: msg
					};
				}
				if (allowedDomains != null) {
					const atIndex = validated.indexOf("@");
					const domain$1 = validated.substring(atIndex + 1).toLowerCase();
					const isAllowed = allowedDomains.some((allowed) => domain$1 === allowed.toLowerCase());
					if (!isAllowed) {
						const errorMsg = domainNotAllowed;
						if (typeof errorMsg === "function") return {
							success: false,
							error: errorMsg(validated, allowedDomains)
						};
						const msg = errorMsg ?? [
							{
								type: "text",
								text: "Email domain "
							},
							{
								type: "value",
								value: domain$1
							},
							{
								type: "text",
								text: " is not allowed. Allowed domains: "
							},
							...require_message.valueSet(allowedDomains, {
								fallback: "",
								locale: "en-US"
							}),
							{
								type: "text",
								text: "."
							}
						];
						return {
							success: false,
							error: msg
						};
					}
				}
				return {
					success: true,
					value: validated
				};
			}
		},
		format(value) {
			if (Array.isArray(value)) return value.join(", ");
			return value;
		}
	};
}
/**
* Creates a value parser for socket addresses in "host:port" format.
*
* Validates socket addresses with support for:
* - Hostnames, IPv4 addresses, and IPv6 addresses
* - Configurable host:port separator
* - Optional default port
* - Host type filtering (hostname only, IP only, or both)
* - Port range validation
*
* @param options - Options for socket address validation.
* @returns A value parser for socket addresses.
* @throws {TypeError} If `separator` is an empty string.
* @throws {TypeError} If `separator` contains digit characters, since digits
*   in the separator would cause ambiguous splitting of port input.
* @throws {TypeError} If `host.version` is provided but is not `4`, `6`, or
*   `"both"`.
* @since 0.10.0
*
* @example
* ```typescript
* import { socketAddress } from "@optique/core/valueparser";
*
* // Basic socket address parser
* const endpoint = socketAddress({ requirePort: true });
*
* // With default port
* const server = socketAddress({ defaultPort: 80 });
*
* // IP addresses only
* const bind = socketAddress({
*   defaultPort: 8080,
*   host: { type: "ip" }
* });
* ```
*/
function socketAddress(options) {
	const separator = options?.separator ?? ":";
	if (separator === "") throw new TypeError("Expected separator to not be empty.");
	if (/\p{Nd}/u.test(separator)) throw new TypeError(`Expected separator to not contain digits, but got: ${JSON.stringify(separator)}.`);
	const formatExample = `host${JSON.stringify(separator).slice(1, -1)}port`;
	const metavar$1 = options?.metavar ?? `HOST${separator}PORT`;
	require_nonempty.ensureNonEmptyString(metavar$1);
	const defaultPort = options?.defaultPort;
	const requirePort = options?.requirePort ?? false;
	const hostType = options?.host?.type ?? "both";
	const rawHostVersion = options?.host?.version;
	if (rawHostVersion !== void 0 && rawHostVersion !== 4 && rawHostVersion !== 6 && rawHostVersion !== "both") throw new TypeError(`Expected host.version to be 4, 6, or "both", but got ${Array.isArray(rawHostVersion) ? "array" : typeof rawHostVersion}: ${String(rawHostVersion)}.`);
	const hasLegacyIpOptions = options?.host?.ip !== void 0;
	const hasNewIpOptions = rawHostVersion !== void 0 || options?.host?.ipv4 !== void 0 || options?.host?.ipv6 !== void 0;
	const hostVersion = rawHostVersion ?? (hasLegacyIpOptions && !hasNewIpOptions ? 4 : "both");
	const hostnameParser = hostname({
		...options?.host?.hostname,
		metavar: "HOST"
	});
	const ipv4Options = {
		...options?.host?.ip,
		...options?.host?.ipv4
	};
	const ipv4Parser = ipv4({
		...ipv4Options,
		metavar: "HOST"
	});
	const ipv6Parser = ipv6({
		...options?.host?.ipv6,
		metavar: "HOST"
	});
	const disambiguationParser = hostType === "ip" ? hostname({
		metavar: "HOST",
		allowWildcard: true,
		allowUnderscore: true,
		maxLength: Math.max(253, options?.host?.hostname?.maxLength ?? 0)
	}) : hostnameParser;
	const separatorIsHostChar = /^[a-zA-Z0-9._-]+$/.test(separator);
	const portParser = port({
		...options?.port,
		metavar: "PORT",
		type: "number"
	});
	function looksLikeIpv4(input) {
		return /^\d+\.\d+\.\d+\.\d+$/.test(input);
	}
	function looksLikeIpv6(input) {
		const colonCount = input.match(/:/g)?.length ?? 0;
		return colonCount >= 2 && (input.includes("::") || /^[0-9a-fA-F:.]+$/.test(input));
	}
	function looksLikeBareIpv6(input) {
		if (!looksLikeIpv6(input)) return false;
		if (parseAndNormalizeIpv6(input) !== null) return true;
		return input.includes("::") || !input.includes(".");
	}
	function looksLikeAltIpv4Literal(input) {
		if (/^0[xX][0-9a-fA-F]+$/.test(input)) {
			const n = parseInt(input.slice(2), 16);
			return n <= 4294967295;
		}
		if (/^0[0-7]+$/.test(input)) {
			const n = parseInt(input.slice(1), 8);
			return n <= 4294967295;
		}
		const parts = input.split(".");
		if (parts.length >= 2 && parts.length <= 4) {
			const numericOrHex = /^(?:[0-9]+|0[xX][0-9a-fA-F]+)$/;
			if (parts.every((p) => numericOrHex.test(p)) && parts.some((p) => /^0[xX]/i.test(p) || p.length > 1 && p[0] === "0")) {
				const values = [];
				for (const p of parts) if (/^0[xX]/i.test(p)) values.push(parseInt(p.slice(2), 16));
				else if (p.length > 1 && p[0] === "0") {
					if (/[89]/.test(p)) return false;
					values.push(parseInt(p, 8));
				} else values.push(Number(p));
				const lastMax = 256 ** (5 - parts.length);
				return values.slice(0, -1).every((v) => v <= 255) && values[values.length - 1] < lastMax;
			}
		}
		return false;
	}
	function parseIpHost(hostInput) {
		if (hostVersion === 4) return ipv4Parser.parse(hostInput);
		if (hostVersion === 6) return ipv6Parser.parse(hostInput);
		if (looksLikeIpv6(hostInput)) {
			const result = ipv6Parser.parse(hostInput);
			if (result.success) {
				const mappedOctets = extractIpv4FromMapped(result.value);
				if (mappedOctets !== null) {
					const restrictionError = checkIpv4MappedRestrictions(mappedOctets, result.value, ipv4Options, void 0);
					if (restrictionError !== null) return restrictionError;
				}
			}
			return result;
		}
		return ipv4Parser.parse(hostInput);
	}
	function parseHost(hostInput) {
		if (hostType === "hostname") {
			if (looksLikeAltIpv4Literal(hostInput)) return {
				success: false,
				error: require_message.message`${hostInput} appears to be a non-standard IPv4 address notation.`
			};
			if (looksLikeIpv4(hostInput)) return {
				success: false,
				error: require_message.message`Expected a valid hostname, but got ${hostInput}.`
			};
			return hostnameParser.parse(hostInput);
		} else if (hostType === "ip") return parseIpHost(hostInput);
		else {
			if (looksLikeAltIpv4Literal(hostInput)) return {
				success: false,
				error: require_message.message`${hostInput} appears to be a non-standard IPv4 address notation.`
			};
			if (looksLikeIpv4(hostInput)) return parseIpHost(hostInput);
			if (looksLikeIpv6(hostInput)) return parseIpHost(hostInput);
			return hostnameParser.parse(hostInput);
		}
	}
	function makeInvalidFormatError(input) {
		const errorMsg = options?.errors?.invalidFormat;
		const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? require_message.message`Expected a socket address in format ${require_message.text(formatExample)}, but got ${input}.`;
		return {
			success: false,
			error: msg
		};
	}
	function makeMissingPortError(input) {
		const errorMsg = options?.errors?.missingPort;
		const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? require_message.message`Port number is required but was not specified.`;
		return {
			success: false,
			error: msg
		};
	}
	function parseBracketedHost(input, trimmed, canOmitPort) {
		if (!trimmed.startsWith("[")) return void 0;
		const closingBracket = trimmed.indexOf("]");
		if (closingBracket === -1) return makeInvalidFormatError(input);
		const hostInput = trimmed.slice(1, closingBracket);
		if (!looksLikeBareIpv6(hostInput)) return makeInvalidFormatError(input);
		const rest = trimmed.slice(closingBracket + 1).trimStart();
		const hostResult = parseHost(hostInput);
		if (rest === "") {
			if (!hostResult.success) {
				if (options?.errors?.invalidFormat) return makeInvalidFormatError(input);
				return {
					success: false,
					error: hostResult.error
				};
			}
			if (canOmitPort) return {
				success: true,
				value: {
					host: hostResult.value,
					port: defaultPort
				}
			};
			return makeMissingPortError(input);
		}
		if (!rest.startsWith(separator)) return makeInvalidFormatError(input);
		const portPart = rest.slice(separator.length).trim();
		if (portPart === "") {
			if (!hostResult.success) {
				if (options?.errors?.invalidFormat) return makeInvalidFormatError(input);
				return {
					success: false,
					error: hostResult.error
				};
			}
			return makeMissingPortError(input);
		}
		const portResult = portParser.parse(portPart);
		if (!hostResult.success) {
			if (options?.errors?.invalidFormat) return makeInvalidFormatError(input);
			return {
				success: false,
				error: hostResult.error
			};
		}
		if (portResult.success) return {
			success: true,
			value: {
				host: hostResult.value,
				port: portResult.value
			}
		};
		if (options?.errors?.invalidFormat) return makeInvalidFormatError(input);
		if (/^[0-9]+$/.test(portPart)) return {
			success: false,
			error: portResult.error
		};
		return makeInvalidFormatError(input);
	}
	function isSocketAddressValue(value) {
		return typeof value === "object" && value !== null && "host" in value && typeof value.host === "string" && "port" in value && typeof value.port === "number";
	}
	function formatSocketAddressValue(value) {
		const ipv6Host = parseAndNormalizeIpv6(value.host);
		if (separator === ":" && ipv6Host !== null) return `[${ipv6Host}]${separator}${value.port}`;
		return `${value.host}${separator}${value.port}`;
	}
	const parser = {
		mode: "sync",
		metavar: metavar$1,
		get placeholder() {
			return {
				host: hostType === "ip" ? hostVersion === 6 ? ipv6Parser.placeholder : ipv4Parser.placeholder : hostnameParser.placeholder,
				port: defaultPort ?? portParser.placeholder
			};
		},
		parse(input) {
			const trimmed = input.trim();
			const searchInput = input.trimStart();
			const canOmitPort = defaultPort !== void 0 && !requirePort;
			if (separator === ":") {
				const bracketedResult = parseBracketedHost(input, trimmed, canOmitPort);
				if (bracketedResult !== void 0) return bracketedResult;
				if (looksLikeBareIpv6(trimmed)) {
					const hostResult = parseHost(trimmed);
					if (hostResult.success) {
						if (canOmitPort) return {
							success: true,
							value: {
								host: hostResult.value,
								port: defaultPort
							}
						};
						return makeMissingPortError(input);
					}
					if (options?.errors?.invalidFormat) return makeInvalidFormatError(input);
					return {
						success: false,
						error: hostResult.error
					};
				}
			}
			let firstHostError;
			let validHostInvalidPortError;
			let trailingSepHost;
			let trailingSepHostError;
			let anySeparatorFound = false;
			let trailingSepInTrimmedRegion = false;
			let searchFrom = searchInput.length;
			while (searchFrom > 0) {
				const separatorIndex = searchInput.lastIndexOf(separator, searchFrom - 1);
				if (separatorIndex === -1) break;
				anySeparatorFound = true;
				const hostPart = searchInput.substring(0, separatorIndex).trim();
				const portPart = searchInput.substring(separatorIndex + separator.length).trim();
				if (portPart === "") {
					if (trailingSepHost === void 0 && trailingSepHostError === void 0) {
						if (separatorIndex + separator.length > trimmed.length) trailingSepInTrimmedRegion = true;
						const hostResult = parseHost(hostPart);
						if (hostResult.success) trailingSepHost = hostResult.value;
						else trailingSepHostError = {
							hostPart,
							error: hostResult.error
						};
					}
				} else {
					const portResult = portParser.parse(portPart);
					if (portResult.success) {
						const hostResult = parseHost(hostPart);
						if (hostResult.success) return {
							success: true,
							value: {
								host: hostResult.value,
								port: portResult.value
							}
						};
						if (firstHostError === void 0 || (looksLikeIpv4(hostPart) || looksLikeAltIpv4Literal(hostPart)) && !looksLikeIpv4(firstHostError.hostPart) && !looksLikeAltIpv4Literal(firstHostError.hostPart)) firstHostError = {
							hostPart,
							error: hostResult.error
						};
					} else if (validHostInvalidPortError === void 0 && hostPart !== "" && /^[0-9]+$/.test(portPart)) if (looksLikeIpv4(hostPart) || looksLikeAltIpv4Literal(hostPart)) {
						const hostResult = parseHost(hostPart);
						if (!hostResult.success) {
							if (firstHostError === void 0 || !looksLikeIpv4(firstHostError.hostPart) && !looksLikeAltIpv4Literal(firstHostError.hostPart)) firstHostError = {
								hostPart,
								error: hostResult.error
							};
						} else validHostInvalidPortError = portResult.error;
					} else {
						validHostInvalidPortError = portResult.error;
						const hostResult = parseHost(hostPart);
						if (!hostResult.success && firstHostError === void 0) firstHostError = {
							hostPart,
							error: hostResult.error
						};
					}
					else if ((firstHostError === void 0 || !looksLikeIpv4(firstHostError.hostPart) && !looksLikeAltIpv4Literal(firstHostError.hostPart)) && (looksLikeIpv4(hostPart) || looksLikeAltIpv4Literal(hostPart))) {
						const hostResult = parseHost(hostPart);
						if (!hostResult.success) firstHostError = {
							hostPart,
							error: hostResult.error
						};
					}
				}
				searchFrom = separatorIndex;
			}
			if (validHostInvalidPortError !== void 0) {
				const errorMsg$1 = options?.errors?.invalidFormat;
				if (errorMsg$1) {
					const msg = typeof errorMsg$1 === "function" ? errorMsg$1(input) : errorMsg$1;
					return {
						success: false,
						error: msg
					};
				}
				if (firstHostError !== void 0 && firstHostError.hostPart !== "") {
					const portSplitHostIsDegenerate = separatorIsHostChar ? firstHostError.hostPart.replaceAll(separator, "") === "" : firstHostError.hostPart.includes(separator);
					if (portSplitHostIsDegenerate) return {
						success: false,
						error: require_message.message`Expected a socket address in format ${require_message.text(formatExample)}, but got ${input}.`
					};
					if (!disambiguationParser.parse(trimmed).success) return {
						success: false,
						error: firstHostError.error
					};
				}
				if (disambiguationParser.parse(trimmed).success) return {
					success: false,
					error: require_message.message`Expected a socket address in format ${require_message.text(formatExample)}, but got ${input}.`
				};
				return {
					success: false,
					error: validHostInvalidPortError
				};
			}
			if (firstHostError !== void 0) {
				if (looksLikeIpv4(firstHostError.hostPart) || looksLikeAltIpv4Literal(firstHostError.hostPart)) {
					const errorMsg$1 = options?.errors?.invalidFormat;
					if (errorMsg$1) {
						const msg = typeof errorMsg$1 === "function" ? errorMsg$1(input) : errorMsg$1;
						return {
							success: false,
							error: msg
						};
					}
					return {
						success: false,
						error: firstHostError.error
					};
				}
			}
			let hostOnlyResult;
			if (!requirePort || trailingSepHost !== void 0 || trailingSepHostError !== void 0) {
				hostOnlyResult = parseHost(trimmed);
				if (canOmitPort && hostOnlyResult.success && !trailingSepInTrimmedRegion) return {
					success: true,
					value: {
						host: hostOnlyResult.value,
						port: defaultPort
					}
				};
			}
			if (trailingSepHost !== void 0 && hostOnlyResult !== void 0 && (!hostOnlyResult.success || trailingSepInTrimmedRegion)) {
				const errorMsg$1 = options?.errors?.missingPort;
				const msg = typeof errorMsg$1 === "function" ? errorMsg$1(input) : errorMsg$1 ?? require_message.message`Port number is required but was not specified.`;
				return {
					success: false,
					error: msg
				};
			}
			if (trailingSepHostError !== void 0 && hostOnlyResult !== void 0 && (!hostOnlyResult.success || trailingSepInTrimmedRegion)) {
				const errorMsg$1 = options?.errors?.invalidFormat;
				if (errorMsg$1) {
					const msg = typeof errorMsg$1 === "function" ? errorMsg$1(input) : errorMsg$1;
					return {
						success: false,
						error: msg
					};
				}
				const trailingHostIsDegenerate = separatorIsHostChar ? trailingSepHostError.hostPart.replaceAll(separator, "") === "" : trailingSepHostError.hostPart.includes(separator);
				if (trailingSepHostError.hostPart !== "" && !trailingHostIsDegenerate && !disambiguationParser.parse(trimmed).success) return {
					success: false,
					error: trailingSepHostError.error
				};
			}
			if (!canOmitPort && !requirePort && hostOnlyResult !== void 0 && hostOnlyResult.success) {
				const errorMsg$1 = options?.errors?.missingPort;
				const msg = typeof errorMsg$1 === "function" ? errorMsg$1(input) : errorMsg$1 ?? require_message.message`Port number is required but was not specified.`;
				return {
					success: false,
					error: msg
				};
			}
			if (!canOmitPort && !anySeparatorFound) {
				hostOnlyResult = hostOnlyResult ?? parseHost(trimmed);
				if (hostOnlyResult.success) {
					const errorMsg$1 = options?.errors?.missingPort;
					const msg = typeof errorMsg$1 === "function" ? errorMsg$1(input) : errorMsg$1 ?? require_message.message`Port number is required but was not specified.`;
					return {
						success: false,
						error: msg
					};
				}
			}
			if (firstHostError !== void 0) {
				const errorMsg$1 = options?.errors?.invalidFormat;
				if (errorMsg$1) {
					const msg = typeof errorMsg$1 === "function" ? errorMsg$1(input) : errorMsg$1;
					return {
						success: false,
						error: msg
					};
				}
				const hostPartIsDegenerate = separatorIsHostChar ? firstHostError.hostPart.replaceAll(separator, "") === "" : firstHostError.hostPart.includes(separator);
				if (firstHostError.hostPart !== "" && !hostPartIsDegenerate && !disambiguationParser.parse(trimmed).success) return {
					success: false,
					error: firstHostError.error
				};
			}
			const errorMsg = options?.errors?.invalidFormat;
			if (errorMsg) {
				const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg;
				return {
					success: false,
					error: msg
				};
			}
			if (hostOnlyResult !== void 0 && !hostOnlyResult.success) {
				if (looksLikeIpv4(trimmed) || looksLikeAltIpv4Literal(trimmed)) return {
					success: false,
					error: hostOnlyResult.error
				};
			}
			return {
				success: false,
				error: require_message.message`Expected a socket address in format ${require_message.text(formatExample)}, but got ${input}.`
			};
		},
		format(value) {
			return formatSocketAddressValue(value);
		},
		normalize(value) {
			if (!isSocketAddressValue(value)) return value;
			try {
				const result = parser.parse(formatSocketAddressValue(value));
				return result.success ? result.value : value;
			} catch {
				return value;
			}
		}
	};
	return parser;
}
function portRange(options) {
	checkBooleanOption(options, "disallowWellKnown");
	checkBooleanOption(options, "allowSingle");
	if (options?.type !== void 0 && options.type !== "number" && options.type !== "bigint") throw new TypeError(`Expected type to be "number" or "bigint", but got: ${String(options.type)}.`);
	const separator = options?.separator ?? "-";
	if (separator === "") throw new TypeError("Expected separator to not be empty.");
	if (/\p{Nd}/u.test(separator)) throw new TypeError(`Expected separator to not contain digits, but got: ${JSON.stringify(separator)}.`);
	const metavar$1 = options?.metavar ?? `PORT${separator}PORT`;
	require_nonempty.ensureNonEmptyString(metavar$1);
	const allowSingle = options?.allowSingle ?? false;
	const isBigInt = options?.type === "bigint";
	const portParser = isBigInt ? port({
		type: "bigint",
		min: options.min,
		max: options.max,
		disallowWellKnown: options.disallowWellKnown,
		errors: options.errors
	}) : port({
		type: "number",
		min: options?.min,
		max: options?.max,
		disallowWellKnown: options?.disallowWellKnown,
		errors: options?.errors
	});
	return {
		mode: "sync",
		metavar: metavar$1,
		get placeholder() {
			return isBigInt ? {
				start: portParser.placeholder,
				end: portParser.placeholder
			} : {
				start: portParser.placeholder,
				end: portParser.placeholder
			};
		},
		parse(input) {
			const trimmed = input.trim();
			const separatorIndex = trimmed.indexOf(separator);
			if (separatorIndex === -1) {
				if (!allowSingle) {
					const errorMsg = options?.errors?.invalidFormat;
					if (typeof errorMsg === "function") return {
						success: false,
						error: errorMsg(input)
					};
					const msg = errorMsg ?? [
						{
							type: "text",
							text: `Expected a port range in format start${separator}end, but got `
						},
						{
							type: "value",
							value: input
						},
						{
							type: "text",
							text: "."
						}
					];
					return {
						success: false,
						error: msg
					};
				}
				const portResult = portParser.parse(trimmed);
				if (!portResult.success) return portResult;
				const portValue = portResult.value;
				return {
					success: true,
					value: isBigInt ? {
						start: portValue,
						end: portValue
					} : {
						start: portValue,
						end: portValue
					}
				};
			}
			const startPart = trimmed.substring(0, separatorIndex);
			const endPart = trimmed.substring(separatorIndex + separator.length);
			const startResult = portParser.parse(startPart);
			if (!startResult.success) return startResult;
			const endResult = portParser.parse(endPart);
			if (!endResult.success) return endResult;
			const startValue = startResult.value;
			const endValue = endResult.value;
			if (isBigInt) {
				const start = startValue;
				const end = endValue;
				if (start > end) {
					const errorMsg = options.errors?.invalidRange;
					const msg = typeof errorMsg === "function" ? errorMsg(start, end) : errorMsg ?? require_message.message`Start port ${startPart} must be less than or equal to end port ${endPart}.`;
					return {
						success: false,
						error: msg
					};
				}
				return {
					success: true,
					value: {
						start,
						end
					}
				};
			} else {
				const start = startValue;
				const end = endValue;
				if (start > end) {
					const errorMsg = options?.errors?.invalidRange;
					const msg = typeof errorMsg === "function" ? errorMsg(start, end) : errorMsg ?? require_message.message`Start port ${startPart} must be less than or equal to end port ${endPart}.`;
					return {
						success: false,
						error: msg
					};
				}
				return {
					success: true,
					value: {
						start,
						end
					}
				};
			}
		},
		format(value) {
			return `${value.start}${separator}${value.end}`;
		}
	};
}
/**
* Creates a value parser for MAC (Media Access Control) addresses.
*
* Validates MAC-48 addresses (6 octets, 12 hex digits) in various formats:
* - Colon-separated: `00:1A:2B:3C:4D:5E` (1–2 hex digits per octet)
* - Hyphen-separated: `00-1A-2B-3C-4D-5E` (1–2 hex digits per octet)
* - Dot-separated (Cisco): `001A.2B3C.4D5E` (exactly 4 hex digits per group)
* - No separator: `001A2B3C4D5E` (exactly 12 hex digits)
*
* Colon-separated and hyphen-separated formats accept single-digit octets
* (e.g., `0:1:2:3:4:5`), which are automatically zero-padded to canonical
* two-digit form (e.g., `00:01:02:03:04:05`).
*
* Returns the MAC address as a formatted string according to `case` and
* `outputSeparator` options.
*
* @param options Configuration options for the MAC address parser.
* @returns A parser that validates MAC addresses and returns formatted strings.
* @since 0.10.0
*
* @example
* ```typescript
* import { macAddress } from "@optique/core/valueparser";
*
* // Accept any format
* const mac = macAddress();
*
* // Normalize to uppercase colon-separated
* const normalizedMac = macAddress({
*   outputSeparator: ":",
*   case: "upper"
* });
* ```
*/
function macAddress(options) {
	checkEnumOption(options, "separator", [
		":",
		"-",
		".",
		"none",
		"any"
	]);
	checkEnumOption(options, "outputSeparator", [
		":",
		"-",
		".",
		"none"
	]);
	checkEnumOption(options, "case", [
		"preserve",
		"upper",
		"lower"
	]);
	const separator = options?.separator ?? "any";
	const caseOption = options?.case ?? "preserve";
	const outputSeparator = options?.outputSeparator;
	const metavar$1 = options?.metavar ?? "MAC";
	const colonRegex = /^([0-9a-fA-F]{1,2}):([0-9a-fA-F]{1,2}):([0-9a-fA-F]{1,2}):([0-9a-fA-F]{1,2}):([0-9a-fA-F]{1,2}):([0-9a-fA-F]{1,2})$/;
	const hyphenRegex = /^([0-9a-fA-F]{1,2})-([0-9a-fA-F]{1,2})-([0-9a-fA-F]{1,2})-([0-9a-fA-F]{1,2})-([0-9a-fA-F]{1,2})-([0-9a-fA-F]{1,2})$/;
	const dotRegex = /^([0-9a-fA-F]{4})\.([0-9a-fA-F]{4})\.([0-9a-fA-F]{4})$/;
	const noneRegex = /^([0-9a-fA-F]{12})$/;
	function joinOctets(octets, sep) {
		let formatted = octets;
		if (caseOption === "upper") formatted = octets.map((o) => o.toUpperCase());
		else if (caseOption === "lower") formatted = octets.map((o) => o.toLowerCase());
		if (sep === ".") return [
			formatted[0] + formatted[1],
			formatted[2] + formatted[3],
			formatted[4] + formatted[5]
		].join(".");
		if (sep === "none") return formatted.join("");
		return formatted.join(sep);
	}
	let macParsing = false;
	function normalizeMacValue(value, fallback) {
		if (macParsing) return value;
		macParsing = true;
		try {
			const result = parserObj.parse(value);
			return result.success ? result.value : fallback;
		} catch {
			return fallback;
		} finally {
			macParsing = false;
		}
	}
	const parserObj = {
		mode: "sync",
		metavar: metavar$1,
		get placeholder() {
			const octets = [
				"00",
				"00",
				"00",
				"00",
				"00",
				"00"
			];
			const sep = outputSeparator ?? (separator === "any" ? ":" : separator);
			if (sep === ".") return `${octets[0]}${octets[1]}.${octets[2]}${octets[3]}.${octets[4]}${octets[5]}`;
			if (sep === "none") return octets.join("");
			return octets.join(sep);
		},
		parse(input) {
			let octets = [];
			let inputSeparator;
			if (separator === ":" || separator === "any") {
				const match = colonRegex.exec(input);
				if (match) {
					octets = match.slice(1, 7);
					inputSeparator = ":";
				}
			}
			if (octets.length === 0 && (separator === "-" || separator === "any")) {
				const match = hyphenRegex.exec(input);
				if (match) {
					octets = match.slice(1, 7);
					inputSeparator = "-";
				}
			}
			if (octets.length === 0 && (separator === "." || separator === "any")) {
				const match = dotRegex.exec(input);
				if (match) {
					const groups = match.slice(1, 4);
					octets = groups.flatMap((group) => [group.slice(0, 2), group.slice(2, 4)]);
					inputSeparator = ".";
				}
			}
			if (octets.length === 0 && (separator === "none" || separator === "any")) {
				const match = noneRegex.exec(input);
				if (match) {
					const hex = match[1];
					octets = [];
					for (let i = 0; i < 12; i += 2) octets.push(hex.slice(i, i + 2));
					inputSeparator = "none";
				}
			}
			if (octets.length === 0) {
				const errorMsg = options?.errors?.invalidMacAddress;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(input)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Expected a valid MAC address, but got "
					},
					{
						type: "value",
						value: input
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			octets = octets.map((o) => o.padStart(2, "0"));
			const finalSeparator = outputSeparator ?? inputSeparator ?? ":";
			return {
				success: true,
				value: joinOctets(octets, finalSeparator)
			};
		},
		format(value) {
			if (typeof value !== "string") return metavar$1;
			return normalizeMacValue(value, value);
		},
		normalize(value) {
			if (typeof value !== "string") return value;
			return normalizeMacValue(value, value);
		}
	};
	return parserObj;
}
/**
* Creates a value parser for domain names.
*
* Validates domain names according to RFC 1035 with configurable options for
* subdomain filtering, TLD restrictions, minimum label requirements, and case
* normalization.
*
* @param options Parser options for domain validation.
* @returns A parser that accepts valid domain names as strings.
* @throws {RangeError} If `maxLength` is not a positive integer.
* @throws {RangeError} If `minLabels` is not a positive integer.
* @throws {TypeError} If `allowSubdomains` or `lowercase` is not a boolean.
* @throws {TypeError} If any `allowedTlds` entry is not a string, is empty,
*   contains dots, has leading/trailing whitespace, or is not a valid DNS
*   label.
* @throws {TypeError} If `allowSubdomains` is `false` and `minLabels` is
*   greater than 2, since non-subdomain domains have exactly 2 labels.
*
* @example
* ``` typescript
* import { option } from "@optique/core/primitives";
* import { domain } from "@optique/core/valueparser";
*
* // Accept any valid domain
* option("--domain", domain())
*
* // Root domains only (no subdomains)
* option("--root", domain({ allowSubdomains: false }))
*
* // Restrict to specific TLDs
* option("--domain", domain({ allowedTlds: ["com", "org", "net"] }))
*
* // Normalize to lowercase
* option("--domain", domain({ lowercase: true }))
* ```
*
* @since 0.10.0
*/
function domain(options) {
	const metavar$1 = options?.metavar ?? "DOMAIN";
	checkBooleanOption(options, "allowSubdomains");
	checkBooleanOption(options, "lowercase");
	const allowSubdomains = options?.allowSubdomains ?? true;
	const allowedTlds = options?.allowedTlds != null ? Object.freeze([...options.allowedTlds]) : void 0;
	const labelRegex = /^[a-zA-Z0-9]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?$/;
	if (allowedTlds !== void 0) {
		if (allowedTlds.length === 0) throw new TypeError("allowedTlds must not be empty.");
		for (const [i, tld] of allowedTlds.entries()) {
			if (typeof tld !== "string") {
				const actualType = Array.isArray(tld) ? "array" : typeof tld;
				throw new TypeError(`allowedTlds[${i}] must be a string, but got ${actualType}.`);
			}
			if (tld.length === 0) throw new TypeError(`allowedTlds[${i}] must not be an empty string.`);
			if (tld.includes(".")) throw new TypeError(`allowedTlds[${i}] must not contain dots: ${JSON.stringify(tld)}.`);
			if (tld !== tld.trim()) throw new TypeError(`allowedTlds[${i}] must not have leading or trailing whitespace: ${JSON.stringify(tld)}.`);
			if (!labelRegex.test(tld)) throw new TypeError(`allowedTlds[${i}] is not a valid DNS label: ${JSON.stringify(tld)}.`);
		}
	}
	const allowedTldsLower = allowedTlds != null ? Object.freeze(allowedTlds.map((t) => t.toLowerCase())) : void 0;
	const minLabels = options?.minLabels ?? 2;
	const maxLength = options?.maxLength ?? 253;
	const lowercase = options?.lowercase ?? false;
	if (!Number.isInteger(maxLength) || maxLength < 1) throw new RangeError("maxLength must be an integer greater than or equal to 1.");
	if (!Number.isInteger(minLabels) || minLabels < 1) throw new RangeError("minLabels must be an integer greater than or equal to 1.");
	if (!allowSubdomains && minLabels > 2) throw new TypeError("allowSubdomains: false is incompatible with minLabels > 2, as non-subdomain domains have exactly 2 labels.");
	const invalidDomain = options?.errors?.invalidDomain;
	const tooLong = options?.errors?.tooLong;
	const tooFewLabels = options?.errors?.tooFewLabels;
	const subdomainsNotAllowed = options?.errors?.subdomainsNotAllowed;
	const tldNotAllowed = options?.errors?.tldNotAllowed;
	const domainParserObj = {
		mode: "sync",
		metavar: metavar$1,
		placeholder: options?.placeholder ?? `example.${allowedTldsLower?.[0] ?? "com"}`,
		parse(input) {
			if (input.length > maxLength) {
				const errorMsg = tooLong;
				const msg = typeof errorMsg === "function" ? errorMsg(input, maxLength) : errorMsg ?? require_message.message`Domain ${input} is too long (maximum ${require_message.text(maxLength.toString())} characters).`;
				return {
					success: false,
					error: msg
				};
			}
			if (input.length === 0 || input.startsWith(".") || input.endsWith(".")) {
				const errorMsg = invalidDomain;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(input)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Expected a valid domain name, but got "
					},
					{
						type: "value",
						value: input
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			if (input.includes("..")) {
				const errorMsg = invalidDomain;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(input)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Expected a valid domain name, but got "
					},
					{
						type: "value",
						value: input
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			const labels = input.split(".");
			for (const label of labels) if (!labelRegex.test(label)) {
				const errorMsg = invalidDomain;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(input)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Expected a valid domain name, but got "
					},
					{
						type: "value",
						value: input
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			if (labels.length >= 2 && labels.every((l) => /^[0-9]+$/.test(l))) {
				const errorMsg = invalidDomain;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(input)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Expected a valid domain name, but got "
					},
					{
						type: "value",
						value: input
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			if (labels.length < minLabels) {
				const errorMsg = tooFewLabels;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(input, minLabels)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Domain "
					},
					{
						type: "value",
						value: input
					},
					{
						type: "text",
						text: ` must have at least ${minLabels} labels.`
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			if (!allowSubdomains && labels.length > 2) {
				const errorMsg = subdomainsNotAllowed;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(input)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Subdomains are not allowed, but got "
					},
					{
						type: "value",
						value: input
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			if (allowedTlds !== void 0 && allowedTldsLower !== void 0) {
				const tld = labels[labels.length - 1];
				const tldLower = tld.toLowerCase();
				if (!allowedTldsLower.includes(tldLower)) {
					const errorMsg = tldNotAllowed;
					if (typeof errorMsg === "function") return {
						success: false,
						error: errorMsg(tld, allowedTlds)
					};
					const msg = errorMsg ?? [
						{
							type: "text",
							text: "Top-level domain "
						},
						{
							type: "value",
							value: tld
						},
						{
							type: "text",
							text: " is not allowed. Allowed TLDs: "
						},
						...require_message.valueSet(allowedTlds, {
							fallback: "",
							locale: "en-US"
						}),
						{
							type: "text",
							text: "."
						}
					];
					return {
						success: false,
						error: msg
					};
				}
			}
			const result = lowercase ? input.toLowerCase() : input;
			return {
				success: true,
				value: result
			};
		},
		format(value) {
			if (typeof value !== "string") return metavar$1;
			if (!lowercase) return value;
			return value.split(".").length >= minLabels ? value.toLowerCase() : value;
		}
	};
	if (lowercase) {
		const domParser = domainParserObj;
		let domParsing = false;
		Object.defineProperty(domainParserObj, "format", {
			value(v) {
				if (typeof v !== "string") return metavar$1;
				if (domParsing) return v;
				domParsing = true;
				try {
					const result = domParser.parse(v);
					return result.success ? result.value : v;
				} catch {
					return v;
				} finally {
					domParsing = false;
				}
			},
			configurable: true,
			enumerable: true
		});
		Object.defineProperty(domainParserObj, "normalize", {
			value(v) {
				if (typeof v !== "string") return v;
				if (domParsing) return v;
				domParsing = true;
				try {
					const result = domParser.parse(v);
					return result.success ? result.value : v;
				} catch {
					return v;
				} finally {
					domParsing = false;
				}
			},
			configurable: true,
			enumerable: true
		});
	}
	return domainParserObj;
}
/**
* Creates a value parser for IPv6 addresses.
*
* Validates and normalizes IPv6 addresses to canonical form (lowercase,
* compressed using `::` notation where appropriate).
*
* @param options Configuration options for IPv6 validation.
* @returns A value parser that validates IPv6 addresses.
*
* @example
* ```typescript
* // Basic IPv6 parser
* option("--ipv6", ipv6())
*
* // Global unicast only (no link-local, no unique local)
* option("--public-ipv6", ipv6({
*   allowLinkLocal: false,
*   allowUniqueLocal: false
* }))
* ```
*
* @since 0.10.0
*/
function ipv6(options) {
	const allowLoopback = options?.allowLoopback ?? true;
	const allowLinkLocal = options?.allowLinkLocal ?? true;
	const allowUniqueLocal = options?.allowUniqueLocal ?? true;
	const allowMulticast = options?.allowMulticast ?? true;
	const allowZero = options?.allowZero ?? true;
	const errors = options?.errors;
	const metavar$1 = options?.metavar ?? "IPV6";
	const ipv6ParserObj = {
		mode: "sync",
		metavar: metavar$1,
		placeholder: allowZero ? "::" : allowLoopback ? "::1" : "2001:db8::1",
		parse(input) {
			const normalized = parseAndNormalizeIpv6(input);
			if (normalized === null) {
				const errorMsg = errors?.invalidIpv6;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(input)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Expected a valid IPv6 address, but got "
					},
					{
						type: "value",
						value: input
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			if (!allowZero && normalized === "::") {
				const errorMsg = errors?.zeroNotAllowed;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(normalized)
				};
				const msg = errorMsg ?? [{
					type: "value",
					value: normalized
				}, {
					type: "text",
					text: " is the zero address."
				}];
				return {
					success: false,
					error: msg
				};
			}
			if (!allowLoopback && normalized === "::1") {
				const errorMsg = errors?.loopbackNotAllowed;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(normalized)
				};
				const msg = errorMsg ?? [{
					type: "value",
					value: normalized
				}, {
					type: "text",
					text: " is a loopback address."
				}];
				return {
					success: false,
					error: msg
				};
			}
			const groups = expandIpv6(normalized);
			if (groups === null) {
				const errorMsg = errors?.invalidIpv6;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(input)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Expected a valid IPv6 address, but got "
					},
					{
						type: "value",
						value: input
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			const firstGroup = parseInt(groups[0], 16);
			if (!allowLinkLocal && (firstGroup & 65472) === 65152) {
				const errorMsg = errors?.linkLocalNotAllowed;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(normalized)
				};
				const msg = errorMsg ?? [{
					type: "value",
					value: normalized
				}, {
					type: "text",
					text: " is a link-local address."
				}];
				return {
					success: false,
					error: msg
				};
			}
			if (!allowUniqueLocal && (firstGroup & 65024) === 64512) {
				const errorMsg = errors?.uniqueLocalNotAllowed;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(normalized)
				};
				const msg = errorMsg ?? [{
					type: "value",
					value: normalized
				}, {
					type: "text",
					text: " is a unique local address."
				}];
				return {
					success: false,
					error: msg
				};
			}
			if (!allowMulticast && (firstGroup & 65280) === 65280) {
				const errorMsg = errors?.multicastNotAllowed;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(normalized)
				};
				const msg = errorMsg ?? [{
					type: "value",
					value: normalized
				}, {
					type: "text",
					text: " is a multicast address."
				}];
				return {
					success: false,
					error: msg
				};
			}
			return {
				success: true,
				value: normalized
			};
		},
		format(value) {
			if (typeof value !== "string") return metavar$1;
			return parseAndNormalizeIpv6(value) ?? value;
		}
	};
	let ipv6Parsing = false;
	Object.defineProperty(ipv6ParserObj, "format", {
		value(v) {
			if (typeof v !== "string") return metavar$1;
			if (ipv6Parsing) return v;
			ipv6Parsing = true;
			try {
				const result = ipv6ParserObj.parse(v);
				return result.success ? result.value : v;
			} catch {
				return v;
			} finally {
				ipv6Parsing = false;
			}
		},
		configurable: true,
		enumerable: true
	});
	Object.defineProperty(ipv6ParserObj, "normalize", {
		value(v) {
			if (typeof v !== "string") return v;
			if (ipv6Parsing) return v;
			ipv6Parsing = true;
			try {
				const result = ipv6ParserObj.parse(v);
				return result.success ? result.value : v;
			} catch {
				return v;
			} finally {
				ipv6Parsing = false;
			}
		},
		configurable: true,
		enumerable: true
	});
	return ipv6ParserObj;
}
/**
* Parses a dotted-decimal IPv4 string into four validated octets.
* Returns null if the input is not a valid strict IPv4 address
* (exactly four decimal octets 0–255, no leading zeros, no whitespace,
* no non-decimal characters).
*/
function parseIpv4Octets(input) {
	const parts = input.split(".");
	if (parts.length !== 4) return null;
	const octets = [
		0,
		0,
		0,
		0
	];
	for (let i = 0; i < 4; i++) {
		const part = parts[i];
		if (part.length === 0) return null;
		if (part.trim() !== part) return null;
		if (part.length > 1 && part[0] === "0") return null;
		if (!/^[0-9]+$/.test(part)) return null;
		const octet = Number(part);
		if (!Number.isInteger(octet) || octet < 0 || octet > 255) return null;
		octets[i] = octet;
	}
	return octets;
}
/**
* Parses and normalizes an IPv6 address to canonical form.
* Returns null if the input is not a valid IPv6 address.
*/
function parseAndNormalizeIpv6(input) {
	if (input.length === 0) return null;
	const ipv4MappedMatch = input.match(/^(.+):(\d+\.\d+\.\d+\.\d+)$/);
	if (ipv4MappedMatch) {
		const ipv6Part = ipv4MappedMatch[1];
		const ipv4Part = ipv4MappedMatch[2];
		const octets = parseIpv4Octets(ipv4Part);
		if (octets === null) return null;
		const group1 = octets[0] << 8 | octets[1];
		const group2 = octets[2] << 8 | octets[3];
		const fullAddress = `${ipv6Part}:${group1.toString(16)}:${group2.toString(16)}`;
		return parseAndNormalizeIpv6(fullAddress);
	}
	const compressionCount = (input.match(/::/g) || []).length;
	if (compressionCount > 1) return null;
	let groups;
	if (input.includes("::")) {
		const parts = input.split("::");
		if (parts.length > 2) return null;
		const leftGroups = parts[0] ? parts[0].split(":") : [];
		const rightGroups = parts[1] ? parts[1].split(":") : [];
		const totalGroups = leftGroups.length + rightGroups.length;
		if (totalGroups >= 8) return null;
		const zeroCount = 8 - totalGroups;
		const zeros = Array(zeroCount).fill("0");
		groups = [
			...leftGroups,
			...zeros,
			...rightGroups
		];
	} else {
		groups = input.split(":");
		if (groups.length !== 8) return null;
	}
	for (let i = 0; i < groups.length; i++) {
		const group = groups[i];
		if (group.length === 0 || group.length > 4) return null;
		if (!/^[0-9a-fA-F]+$/.test(group)) return null;
		groups[i] = parseInt(group, 16).toString(16);
	}
	return compressIpv6(groups);
}
/**
* Expands a compressed IPv6 address to 8 groups of 4 hex digits.
* Returns null if the input is invalid.
*/
function expandIpv6(input) {
	if (input.includes("::")) {
		const parts = input.split("::");
		if (parts.length > 2) return null;
		const leftGroups = parts[0] ? parts[0].split(":").filter((g) => g) : [];
		const rightGroups = parts[1] ? parts[1].split(":").filter((g) => g) : [];
		const totalGroups = leftGroups.length + rightGroups.length;
		if (totalGroups >= 8) return null;
		const zeroCount = 8 - totalGroups;
		const zeros = Array(zeroCount).fill("0");
		const groups = [
			...leftGroups,
			...zeros,
			...rightGroups
		];
		return groups.map((g) => g.padStart(4, "0"));
	} else {
		const groups = input.split(":");
		if (groups.length !== 8) return null;
		return groups.map((g) => g.padStart(4, "0"));
	}
}
/**
* Compresses an IPv6 address by replacing the longest sequence of zeros with ::.
*/
function compressIpv6(groups) {
	let longestStart = -1;
	let longestLength = 0;
	let currentStart = -1;
	let currentLength = 0;
	for (let i = 0; i < groups.length; i++) if (groups[i] === "0") if (currentStart === -1) {
		currentStart = i;
		currentLength = 1;
	} else currentLength++;
	else {
		if (currentLength > longestLength) {
			longestStart = currentStart;
			longestLength = currentLength;
		}
		currentStart = -1;
		currentLength = 0;
	}
	if (currentLength > longestLength) {
		longestStart = currentStart;
		longestLength = currentLength;
	}
	if (longestLength < 2) return groups.join(":");
	const before = groups.slice(0, longestStart);
	const after = groups.slice(longestStart + longestLength);
	if (before.length === 0 && after.length === 0) return "::";
	else if (before.length === 0) return "::" + after.join(":");
	else if (after.length === 0) return before.join(":") + "::";
	else return before.join(":") + "::" + after.join(":");
}
/**
* Extracts IPv4 octets from an IPv4-mapped IPv6 address.
* Returns null if the address is not an IPv4-mapped address
* (i.e., not in the `::ffff:x.x.x.x` range).
*/
function extractIpv4FromMapped(normalizedIpv6) {
	const groups = expandIpv6(normalizedIpv6);
	if (groups === null) return null;
	for (let i = 0; i < 5; i++) if (parseInt(groups[i], 16) !== 0) return null;
	if (parseInt(groups[5], 16) !== 65535) return null;
	const high = parseInt(groups[6], 16);
	const low = parseInt(groups[7], 16);
	return [
		high >> 8 & 255,
		high & 255,
		low >> 8 & 255,
		low & 255
	];
}
/**
* Checks IPv4 restrictions against octets extracted from an IPv4-mapped
* IPv6 address.  The check uses the base address, consistent with how
* the `ipv4()` parser validates the address part of a regular IPv4 CIDR.
*
* Returns an error result if a restriction is violated, or null if all
* checks pass.
*/
function checkIpv4MappedRestrictions(octets, normalizedIpv6, ipv4Opts, errors) {
	const allowPrivate = ipv4Opts?.allowPrivate ?? true;
	const allowLoopback = ipv4Opts?.allowLoopback ?? true;
	const allowLinkLocal = ipv4Opts?.allowLinkLocal ?? true;
	const allowMulticast = ipv4Opts?.allowMulticast ?? true;
	const allowBroadcast = ipv4Opts?.allowBroadcast ?? true;
	const allowZero = ipv4Opts?.allowZero ?? true;
	if (!allowPrivate && isPrivateIp(octets)) {
		const errorMsg = errors?.privateNotAllowed;
		const msg = typeof errorMsg === "function" ? errorMsg(normalizedIpv6) : errorMsg ?? require_message.message`${normalizedIpv6} is a private IP address.`;
		return {
			success: false,
			error: msg
		};
	}
	if (!allowLoopback && isLoopbackIp(octets)) {
		const errorMsg = errors?.loopbackNotAllowed;
		const msg = typeof errorMsg === "function" ? errorMsg(normalizedIpv6) : errorMsg ?? require_message.message`${normalizedIpv6} is a loopback address.`;
		return {
			success: false,
			error: msg
		};
	}
	if (!allowLinkLocal && isLinkLocalIp(octets)) {
		const errorMsg = errors?.linkLocalNotAllowed;
		const msg = typeof errorMsg === "function" ? errorMsg(normalizedIpv6) : errorMsg ?? require_message.message`${normalizedIpv6} is a link-local address.`;
		return {
			success: false,
			error: msg
		};
	}
	if (!allowMulticast && isMulticastIp(octets)) {
		const errorMsg = errors?.multicastNotAllowed;
		const msg = typeof errorMsg === "function" ? errorMsg(normalizedIpv6) : errorMsg ?? require_message.message`${normalizedIpv6} is a multicast address.`;
		return {
			success: false,
			error: msg
		};
	}
	if (!allowBroadcast && isBroadcastIp(octets)) {
		const errorMsg = errors?.broadcastNotAllowed;
		const msg = typeof errorMsg === "function" ? errorMsg(normalizedIpv6) : errorMsg ?? require_message.message`${normalizedIpv6} is the broadcast address.`;
		return {
			success: false,
			error: msg
		};
	}
	if (!allowZero && isZeroIp(octets)) {
		const errorMsg = errors?.zeroNotAllowed;
		const msg = typeof errorMsg === "function" ? errorMsg(normalizedIpv6) : errorMsg ?? require_message.message`${normalizedIpv6} is the zero address.`;
		return {
			success: false,
			error: msg
		};
	}
	return null;
}
/**
* Creates a value parser that accepts both IPv4 and IPv6 addresses.
*
* By default, accepts both IPv4 and IPv6 addresses. Use the `version` option
* to restrict to a specific IP version.
*
* @param options Configuration options for IP validation.
* @returns A value parser that validates IP addresses.
*
* @example
* ```typescript
* // Accept both IPv4 and IPv6
* option("--ip", ip())
*
* // IPv4 only
* option("--ipv4", ip({ version: 4 }))
*
* // Public IPs only (both versions)
* option("--public-ip", ip({
*   ipv4: { allowPrivate: false, allowLoopback: false },
*   ipv6: { allowLinkLocal: false, allowUniqueLocal: false }
* }))
* ```
*
* @since 0.10.0
*/
function ip(options) {
	const version = options?.version ?? "both";
	const metavar$1 = options?.metavar ?? "IP";
	const errors = options?.errors;
	const ipv4Parser = version === 4 || version === "both" ? ipv4({
		...options?.ipv4,
		errors: {
			invalidIpv4: errors?.invalidIP,
			privateNotAllowed: errors?.privateNotAllowed,
			loopbackNotAllowed: errors?.loopbackNotAllowed,
			linkLocalNotAllowed: errors?.linkLocalNotAllowed,
			multicastNotAllowed: errors?.multicastNotAllowed,
			broadcastNotAllowed: errors?.broadcastNotAllowed,
			zeroNotAllowed: errors?.zeroNotAllowed
		}
	}) : null;
	const ipv6Parser = version === 6 || version === "both" ? ipv6({
		...options?.ipv6,
		errors: {
			invalidIpv6: errors?.invalidIP,
			loopbackNotAllowed: errors?.loopbackNotAllowed,
			linkLocalNotAllowed: errors?.linkLocalNotAllowed,
			uniqueLocalNotAllowed: errors?.uniqueLocalNotAllowed,
			multicastNotAllowed: errors?.multicastNotAllowed,
			zeroNotAllowed: errors?.zeroNotAllowed
		}
	}) : null;
	const mappedIpv4Opts = version === "both" ? {
		allowPrivate: options?.ipv4?.allowPrivate,
		allowLoopback: options?.ipv4?.allowLoopback,
		allowLinkLocal: options?.ipv4?.allowLinkLocal,
		allowMulticast: options?.ipv4?.allowMulticast,
		allowBroadcast: options?.ipv4?.allowBroadcast,
		allowZero: options?.ipv4?.allowZero
	} : void 0;
	const mappedIpv4Errors = version === "both" ? {
		privateNotAllowed: errors?.privateNotAllowed,
		loopbackNotAllowed: errors?.loopbackNotAllowed,
		linkLocalNotAllowed: errors?.linkLocalNotAllowed,
		multicastNotAllowed: errors?.multicastNotAllowed,
		broadcastNotAllowed: errors?.broadcastNotAllowed,
		zeroNotAllowed: errors?.zeroNotAllowed
	} : void 0;
	const ipParserObj = {
		mode: "sync",
		metavar: metavar$1,
		placeholder: version === 6 ? ipv6Parser.placeholder : ipv4Parser.placeholder,
		parse(input) {
			let ipv4Error = null;
			let ipv6Error = null;
			if (ipv4Parser !== null) {
				const result = ipv4Parser.parse(input);
				if (result.success) return result;
				ipv4Error = result;
				if (version === 4) return result;
			}
			if (ipv6Parser !== null) {
				const result = ipv6Parser.parse(input);
				if (result.success) {
					if (version === "both") {
						const mappedOctets = extractIpv4FromMapped(result.value);
						if (mappedOctets !== null) {
							const restrictionError = checkIpv4MappedRestrictions(mappedOctets, result.value, mappedIpv4Opts, mappedIpv4Errors);
							if (restrictionError !== null) return restrictionError;
						}
					}
					return result;
				}
				ipv6Error = result;
				if (version === 6) return result;
			}
			if (ipv4Error !== null && !ipv4Error.success) {
				const isGeneric = ipv4Error.error.some((term) => term.type === "text" && term.text.includes("Expected"));
				if (!isGeneric) return ipv4Error;
			}
			if (ipv6Error !== null && !ipv6Error.success) {
				const isGeneric = ipv6Error.error.some((term) => term.type === "text" && term.text.includes("Expected"));
				if (!isGeneric) return ipv6Error;
			}
			const errorMsg = errors?.invalidIP;
			if (typeof errorMsg === "function") return {
				success: false,
				error: errorMsg(input)
			};
			const msg = errorMsg ?? [
				{
					type: "text",
					text: "Expected a valid IP address, but got "
				},
				{
					type: "value",
					value: input
				},
				{
					type: "text",
					text: "."
				}
			];
			return {
				success: false,
				error: msg
			};
		},
		format(value) {
			if (typeof value !== "string") return metavar$1;
			return parseAndNormalizeIpv6(value) ?? value;
		}
	};
	let ipParsing = false;
	Object.defineProperty(ipParserObj, "format", {
		value(v) {
			if (typeof v !== "string") return metavar$1;
			if (ipParsing) return v;
			ipParsing = true;
			try {
				const result = ipParserObj.parse(v);
				return result.success ? result.value : v;
			} catch {
				return v;
			} finally {
				ipParsing = false;
			}
		},
		configurable: true,
		enumerable: true
	});
	Object.defineProperty(ipParserObj, "normalize", {
		value(v) {
			if (typeof v !== "string") return v;
			if (ipParsing) return v;
			ipParsing = true;
			try {
				const result = ipParserObj.parse(v);
				return result.success ? result.value : v;
			} catch {
				return v;
			} finally {
				ipParsing = false;
			}
		},
		configurable: true,
		enumerable: true
	});
	return ipParserObj;
}
/**
* Creates a value parser for CIDR notation (IP address with prefix length).
*
* Parses and validates CIDR notation like `192.168.0.0/24` or `2001:db8::/32`.
* Returns a structured object with the normalized IP address, prefix length,
* and IP version.
*
* @param options Configuration options for CIDR validation.
* @returns A value parser that validates CIDR notation.
*
* @example
* ```typescript
* // Accept both IPv4 and IPv6 CIDR
* option("--network", cidr())
*
* // IPv4 CIDR only with prefix constraints
* option("--subnet", cidr({
*   version: 4,
*   minPrefix: 16,
*   maxPrefix: 24
* }))
* ```
*
* @since 0.10.0
*/
function cidr(options) {
	if (options?.minPrefix != null && !Number.isFinite(options.minPrefix)) throw new RangeError(`Expected minPrefix to be a finite number, but got: ${options.minPrefix}`);
	if (options?.maxPrefix != null && !Number.isFinite(options.maxPrefix)) throw new RangeError(`Expected maxPrefix to be a finite number, but got: ${options.maxPrefix}`);
	if (options?.minPrefix != null && options?.maxPrefix != null && options.minPrefix > options.maxPrefix) throw new RangeError(`Expected minPrefix to be less than or equal to maxPrefix, but got minPrefix: ${options.minPrefix} and maxPrefix: ${options.maxPrefix}.`);
	const version = options?.version ?? "both";
	const maxPrefixForVersion = version === 4 ? 32 : version === 6 ? 128 : 128;
	if (options?.minPrefix != null && (options.minPrefix < 0 || options.minPrefix > maxPrefixForVersion)) throw new RangeError(`Expected minPrefix to be between 0 and ${maxPrefixForVersion} for IPv${version === "both" ? "4/6" : version}, but got minPrefix: ${options.minPrefix}.`);
	if (options?.maxPrefix != null && (options.maxPrefix < 0 || options.maxPrefix > maxPrefixForVersion)) throw new RangeError(`Expected maxPrefix to be between 0 and ${maxPrefixForVersion} for IPv${version === "both" ? "4/6" : version}, but got maxPrefix: ${options.maxPrefix}.`);
	const minPrefix = options?.minPrefix;
	const maxPrefix = options?.maxPrefix;
	const errors = options?.errors;
	const metavar$1 = options?.metavar ?? "CIDR";
	const genericIpSentinel = [];
	const ipv4Parser = version === 4 || version === "both" ? ipv4({
		...options?.ipv4,
		errors: {
			invalidIpv4: genericIpSentinel,
			privateNotAllowed: errors?.privateNotAllowed,
			loopbackNotAllowed: errors?.loopbackNotAllowed,
			linkLocalNotAllowed: errors?.linkLocalNotAllowed,
			multicastNotAllowed: errors?.multicastNotAllowed,
			broadcastNotAllowed: errors?.broadcastNotAllowed,
			zeroNotAllowed: errors?.zeroNotAllowed
		}
	}) : null;
	const ipv6Parser = version === 6 || version === "both" ? ipv6({
		...options?.ipv6,
		errors: {
			invalidIpv6: genericIpSentinel,
			loopbackNotAllowed: errors?.loopbackNotAllowed,
			linkLocalNotAllowed: errors?.linkLocalNotAllowed,
			multicastNotAllowed: errors?.multicastNotAllowed,
			zeroNotAllowed: errors?.zeroNotAllowed,
			uniqueLocalNotAllowed: errors?.uniqueLocalNotAllowed
		}
	}) : null;
	const mappedIpv4Opts = version === "both" ? {
		allowPrivate: options?.ipv4?.allowPrivate,
		allowLoopback: options?.ipv4?.allowLoopback,
		allowLinkLocal: options?.ipv4?.allowLinkLocal,
		allowMulticast: options?.ipv4?.allowMulticast,
		allowBroadcast: options?.ipv4?.allowBroadcast,
		allowZero: options?.ipv4?.allowZero
	} : void 0;
	const mappedIpv4Errors = version === "both" ? {
		privateNotAllowed: errors?.privateNotAllowed,
		loopbackNotAllowed: errors?.loopbackNotAllowed,
		linkLocalNotAllowed: errors?.linkLocalNotAllowed,
		multicastNotAllowed: errors?.multicastNotAllowed,
		broadcastNotAllowed: errors?.broadcastNotAllowed,
		zeroNotAllowed: errors?.zeroNotAllowed
	} : void 0;
	const cidrParserObj = {
		mode: "sync",
		metavar: metavar$1,
		get placeholder() {
			return version === 6 || version === "both" && (minPrefix ?? 0) > 32 ? {
				address: ipv6Parser.placeholder,
				prefix: minPrefix ?? 0,
				version: 6
			} : {
				address: ipv4Parser.placeholder,
				prefix: minPrefix ?? 0,
				version: 4
			};
		},
		parse(input) {
			const slashIndex = input.lastIndexOf("/");
			if (slashIndex === -1) {
				const errorMsg = errors?.invalidCidr;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(input)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Expected a valid CIDR notation, but got "
					},
					{
						type: "value",
						value: input
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			const ipPart = input.slice(0, slashIndex);
			const prefixPart = input.slice(slashIndex + 1);
			const prefix = parseInt(prefixPart, 10);
			if (!Number.isInteger(prefix) || prefixPart !== prefix.toString() || prefix < 0) {
				const errorMsg = errors?.invalidCidr;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(input)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Expected a valid CIDR notation, but got "
					},
					{
						type: "value",
						value: input
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			let ipVersion = null;
			let normalizedIp = null;
			let ipv4Error = null;
			let ipv6Error = null;
			if (ipv4Parser !== null) {
				const result = ipv4Parser.parse(ipPart);
				if (result.success) {
					ipVersion = 4;
					normalizedIp = result.value;
					if (prefix > 32) {
						const errorMsg = errors?.invalidPrefix;
						if (typeof errorMsg === "function") return {
							success: false,
							error: errorMsg(prefix, 4)
						};
						const msg = errorMsg ?? [
							{
								type: "text",
								text: "Expected a prefix length between 0 and "
							},
							{
								type: "text",
								text: "32"
							},
							{
								type: "text",
								text: " for IPv4, but got "
							},
							{
								type: "text",
								text: prefix.toString()
							},
							{
								type: "text",
								text: "."
							}
						];
						return {
							success: false,
							error: msg
						};
					}
				} else ipv4Error = result;
			}
			if (ipVersion === null && ipv6Parser !== null) {
				const result = ipv6Parser.parse(ipPart);
				if (result.success) {
					ipVersion = 6;
					normalizedIp = result.value;
					if (prefix > 128) {
						const errorMsg = errors?.invalidPrefix;
						if (typeof errorMsg === "function") return {
							success: false,
							error: errorMsg(prefix, 6)
						};
						const msg = errorMsg ?? [
							{
								type: "text",
								text: "Expected a prefix length between 0 and "
							},
							{
								type: "text",
								text: "128"
							},
							{
								type: "text",
								text: " for IPv6, but got "
							},
							{
								type: "text",
								text: prefix.toString()
							},
							{
								type: "text",
								text: "."
							}
						];
						return {
							success: false,
							error: msg
						};
					}
				} else ipv6Error = result;
			}
			if (ipVersion === null || normalizedIp === null) {
				const candidates = [[
					ipv4Error,
					4,
					32
				], [
					ipv6Error,
					6,
					128
				]];
				for (const [err, ver, maxPfx] of candidates) if (err !== null && !err.success && err.error !== genericIpSentinel) {
					if (prefix > maxPfx) {
						const errorMsg$1 = errors?.invalidPrefix;
						if (typeof errorMsg$1 === "function") return {
							success: false,
							error: errorMsg$1(prefix, ver)
						};
						const msg$1 = errorMsg$1 ?? [
							{
								type: "text",
								text: "Expected a prefix length between 0 and "
							},
							{
								type: "text",
								text: maxPfx.toString()
							},
							{
								type: "text",
								text: ` for IPv${ver}, but got `
							},
							{
								type: "text",
								text: prefix.toString()
							},
							{
								type: "text",
								text: "."
							}
						];
						return {
							success: false,
							error: msg$1
						};
					}
					if (minPrefix !== void 0 && prefix < minPrefix) {
						const errorMsg$1 = errors?.prefixBelowMinimum;
						if (typeof errorMsg$1 === "function") return {
							success: false,
							error: errorMsg$1(prefix, minPrefix)
						};
						const msg$1 = errorMsg$1 ?? [
							{
								type: "text",
								text: "Expected a prefix length greater than or equal to "
							},
							{
								type: "text",
								text: minPrefix.toString()
							},
							{
								type: "text",
								text: ", but got "
							},
							{
								type: "text",
								text: prefix.toString()
							},
							{
								type: "text",
								text: "."
							}
						];
						return {
							success: false,
							error: msg$1
						};
					}
					if (maxPrefix !== void 0 && prefix > maxPrefix) {
						const errorMsg$1 = errors?.prefixAboveMaximum;
						if (typeof errorMsg$1 === "function") return {
							success: false,
							error: errorMsg$1(prefix, maxPrefix)
						};
						const msg$1 = errorMsg$1 ?? [
							{
								type: "text",
								text: "Expected a prefix length less than or equal to "
							},
							{
								type: "text",
								text: maxPrefix.toString()
							},
							{
								type: "text",
								text: ", but got "
							},
							{
								type: "text",
								text: prefix.toString()
							},
							{
								type: "text",
								text: "."
							}
						];
						return {
							success: false,
							error: msg$1
						};
					}
					return err;
				}
				const errorMsg = errors?.invalidCidr;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(input)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Expected a valid CIDR notation, but got "
					},
					{
						type: "value",
						value: input
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			if (minPrefix !== void 0 && prefix < minPrefix) {
				const errorMsg = errors?.prefixBelowMinimum;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(prefix, minPrefix)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Expected a prefix length greater than or equal to "
					},
					{
						type: "text",
						text: minPrefix.toString()
					},
					{
						type: "text",
						text: ", but got "
					},
					{
						type: "text",
						text: prefix.toString()
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			if (maxPrefix !== void 0 && prefix > maxPrefix) {
				const errorMsg = errors?.prefixAboveMaximum;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(prefix, maxPrefix)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Expected a prefix length less than or equal to "
					},
					{
						type: "text",
						text: maxPrefix.toString()
					},
					{
						type: "text",
						text: ", but got "
					},
					{
						type: "text",
						text: prefix.toString()
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			if (version === "both" && ipVersion === 6 && normalizedIp !== null) {
				const mappedOctets = extractIpv4FromMapped(normalizedIp);
				if (mappedOctets !== null) {
					const restrictionError = checkIpv4MappedRestrictions(mappedOctets, normalizedIp, mappedIpv4Opts, mappedIpv4Errors);
					if (restrictionError !== null) return restrictionError;
				}
			}
			return {
				success: true,
				value: {
					address: normalizedIp,
					prefix,
					version: ipVersion
				}
			};
		},
		format: ((_) => metavar$1)
	};
	let cidrParsing = false;
	Object.defineProperty(cidrParserObj, "format", {
		value(value) {
			if (typeof value !== "object" || value == null || !("address" in value) || !("prefix" in value) || !("version" in value)) return metavar$1;
			if (cidrParsing) return `${value.address}/${value.prefix}`;
			cidrParsing = true;
			try {
				const raw = `${value.address}/${value.prefix}`;
				const result = cidrParserObj.parse(raw);
				return result.success && result.value.version === value.version ? `${result.value.address}/${result.value.prefix}` : raw;
			} catch {
				return `${value.address}/${value.prefix}`;
			} finally {
				cidrParsing = false;
			}
		},
		configurable: true,
		enumerable: true
	});
	Object.defineProperty(cidrParserObj, "normalize", {
		value(v) {
			if (typeof v !== "object" || v == null || !("address" in v) || !("prefix" in v) || !("version" in v)) return v;
			if (cidrParsing) return v;
			cidrParsing = true;
			const formatted = `${v.address}/${v.prefix}`;
			try {
				const result = cidrParserObj.parse(formatted);
				if (result.success && result.value.version === v.version) return result.value;
				return v;
			} catch {
				return v;
			} finally {
				cidrParsing = false;
			}
		},
		configurable: true,
		enumerable: true
	});
	return cidrParserObj;
}
const SEMVER_REGEX = /^(?<prefix>v)?(?<major>0|[1-9]\d*)\.(?<minor>0|[1-9]\d*)\.(?<patch>0|[1-9]\d*)(?:-(?<pre>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?<meta>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$/;
const SEMVER_SUGGESTIONS = Object.freeze([
	"0.0.0",
	"1.0.0",
	"1.0.0-alpha",
	"1.0.0-alpha.1",
	"1.0.0+build.1",
	"1.0.0-alpha.1+build.1"
]);
const SEMVER_SUGGESTIONS_WITH_PREFIX = Object.freeze([...SEMVER_SUGGESTIONS, ...SEMVER_SUGGESTIONS.map((s) => `v${s}`)]);
function semVer(options = {}) {
	const metavar$1 = options.metavar ?? "SEMVER";
	require_nonempty.ensureNonEmptyString(metavar$1);
	checkEnumOption(options, "type", ["string", "object"]);
	checkBooleanOption(options, "allowPrefix");
	const allowPrefix = options.allowPrefix ?? false;
	const objectMode = options.type === "object";
	const errorOption = options.errors?.invalidSemVer;
	const suggestions = allowPrefix ? SEMVER_SUGGESTIONS_WITH_PREFIX : SEMVER_SUGGESTIONS;
	function makeError(input) {
		return {
			success: false,
			error: errorOption ? typeof errorOption === "function" ? errorOption(input) : errorOption : require_message.message`Expected a valid Semantic Versioning 2.0.0 string (e.g. ${"1.0.0"}), but got ${input}.`
		};
	}
	if (objectMode) return {
		mode: "sync",
		metavar: metavar$1,
		placeholder: {
			major: 0,
			minor: 0,
			patch: 0
		},
		parse(input) {
			const m = SEMVER_REGEX.exec(input);
			if (m == null) return makeError(input);
			if (!allowPrefix && m.groups.prefix != null) return makeError(input);
			const major = parseInt(m.groups.major, 10);
			const minor = parseInt(m.groups.minor, 10);
			const patch = parseInt(m.groups.patch, 10);
			if (!Number.isSafeInteger(major) || !Number.isSafeInteger(minor) || !Number.isSafeInteger(patch)) return makeError(input);
			const result = {
				major,
				minor,
				patch,
				...m.groups.pre != null ? { preRelease: m.groups.pre } : {},
				...m.groups.meta != null ? { metadata: m.groups.meta } : {}
			};
			return {
				success: true,
				value: result
			};
		},
		format(value) {
			let s = `${value.major}.${value.minor}.${value.patch}`;
			if (value.preRelease != null) s += `-${value.preRelease}`;
			if (value.metadata != null) s += `+${value.metadata}`;
			return s;
		},
		*suggest(prefix) {
			for (const s of suggestions) if (s.startsWith(prefix)) yield {
				kind: "literal",
				text: s
			};
		}
	};
	return {
		mode: "sync",
		metavar: metavar$1,
		placeholder: "0.0.0",
		parse(input) {
			const m = SEMVER_REGEX.exec(input);
			if (m == null) return makeError(input);
			if (!allowPrefix && m.groups.prefix != null) return makeError(input);
			const canonical = `${m.groups.major}.${m.groups.minor}.${m.groups.patch}` + (m.groups.pre != null ? `-${m.groups.pre}` : "") + (m.groups.meta != null ? `+${m.groups.meta}` : "");
			return {
				success: true,
				value: canonical
			};
		},
		format(value) {
			return value;
		},
		*suggest(prefix) {
			for (const s of suggestions) if (s.startsWith(prefix)) yield {
				kind: "literal",
				text: s
			};
		}
	};
}
/**
* Creates a {@link ValueParser} for parsing JSON-encoded strings from the
* command line.
*
* The parser accepts any well-formed JSON string by default.  Use the
* `rootType` option to restrict which JSON root type is accepted and to
* narrow the TypeScript return type accordingly.
*
* @example
* ```typescript
* // Accept any JSON value
* const anyJson = json();
*
* // Accept only JSON objects (return type narrowed)
* const objJson = json({ rootType: "object" });
* ```
*
* @param options Optional configuration for the parser.
* @returns A {@link ValueParser} that converts JSON-encoded strings to the
*   appropriate JavaScript type.
* @throws {TypeError} If `options.metavar` is provided but is an empty string.
* @throws {TypeError} If `options.rootType` is provided but is not one of the
*   six allowed values.
* @throws {TypeError} If `options.placeholder` or any value nested within
*   it is a non-finite number (`Infinity`, `-Infinity`, or `NaN`).
* @throws {TypeError} If `options.placeholder` is provided with a `rootType`
*   but its JSON type does not match the `rootType`.
* @throws {TypeError} If the returned parser's `format()` method is called
*   with a value that contains a non-finite number anywhere in its structure.
* @since 1.1.0
*/
function json(options) {
	const metavar$1 = options?.metavar ?? "JSON";
	require_nonempty.ensureNonEmptyString(metavar$1);
	checkEnumOption(options, "rootType", [
		"string",
		"number",
		"boolean",
		"null",
		"object",
		"array"
	]);
	const rootType = options?.rootType;
	const invalidJsonError = options?.errors?.invalidJson;
	const invalidRootTypeError = options?.errors?.invalidRootType;
	if (options?.placeholder !== void 0) {
		const p = options.placeholder;
		const nonFinitePlaceholder = findNonFiniteNumber(p);
		if (nonFinitePlaceholder !== void 0) throw new TypeError(`Expected placeholder to contain only finite numbers, but found ${String(nonFinitePlaceholder)}.`);
		if (rootType != null && jsonTypeOf(p) !== rootType) throw new TypeError(`Expected placeholder to be a JSON ${rootType}, but got ${jsonTypeOf(p)}.`);
	}
	const defaultPlaceholder = rootType === "string" ? "" : rootType === "number" ? 0 : rootType === "boolean" ? false : rootType === "object" ? {} : rootType === "array" ? [] : null;
	const placeholder = options?.placeholder ?? defaultPlaceholder;
	return {
		mode: "sync",
		metavar: metavar$1,
		placeholder,
		parse(input) {
			let value;
			try {
				value = JSON.parse(input);
			} catch (e) {
				const err = e instanceof Error ? e : new Error(String(e));
				const error = invalidJsonError instanceof Function ? invalidJsonError(input) : invalidJsonError ?? [require_message.text(`Not a valid JSON: ${err.message}`)];
				return {
					success: false,
					error
				};
			}
			const nonFinite = findNonFiniteNumber(value);
			if (nonFinite !== void 0) {
				const error = invalidJsonError instanceof Function ? invalidJsonError(input) : invalidJsonError ?? [require_message.text(`Not a valid JSON: number out of range.`)];
				return {
					success: false,
					error
				};
			}
			if (rootType != null) {
				const actual = jsonTypeOf(value);
				if (actual !== rootType) {
					const error = invalidRootTypeError instanceof Function ? invalidRootTypeError(value, rootType) : invalidRootTypeError ?? [require_message.text(`Expected JSON ${rootType}, but got ${actual}.`)];
					return {
						success: false,
						error
					};
				}
			}
			return {
				success: true,
				value
			};
		},
		format(value) {
			const nonFinite = findNonFiniteNumber(value);
			if (nonFinite !== void 0) throw new TypeError(`Expected a finite JSON number, but got ${String(nonFinite)}.`);
			return JSON.stringify(value);
		}
	};
}
function jsonTypeOf(value) {
	if (value === null) return "null";
	if (Array.isArray(value)) return "array";
	return typeof value;
}
/**
* Finds the first non-finite number (`NaN`, `Infinity`, or `-Infinity`)
* anywhere in a JSON structure.
*
* Uses an explicit stack rather than recursion to avoid call-stack overflows
* on deeply nested inputs.  Tracks visited objects to handle circular
* references without looping forever.
*
* @returns The first non-finite number found, or `undefined` if all numbers
*   in the structure are finite.
*/
function findNonFiniteNumber(root) {
	const stack = [root];
	const seen = /* @__PURE__ */ new Set();
	while (stack.length > 0) {
		const value = stack.pop();
		if (typeof value === "number") {
			if (!Number.isFinite(value)) return value;
		} else if (Array.isArray(value)) {
			if (seen.has(value)) continue;
			seen.add(value);
			for (const item of value) stack.push(item);
		} else if (value !== null && typeof value === "object") {
			if (seen.has(value)) continue;
			seen.add(value);
			for (const item of Object.values(value)) stack.push(item);
		}
	}
	return void 0;
}
/**
* Implementation of the {@link firstOf} combinator.
*/
function firstOf(...rawArgs) {
	const args = rawArgs.length > 0 && rawArgs.at(-1) === void 0 ? rawArgs.slice(0, -1) : rawArgs;
	let parsers;
	let options;
	if (args.length > 0 && Array.isArray(args[0])) {
		parsers = [...args[0]];
		options = args[1] ?? {};
	} else {
		const last = args.at(-1);
		if (args.length > 0 && typeof last === "object" && last != null && !isValueParser(last)) {
			options = last;
			parsers = args.slice(0, -1);
		} else {
			options = {};
			parsers = args;
		}
	}
	if (parsers.length < 2) throw new TypeError("firstOf() requires at least two value parsers.");
	for (const parser of parsers) {
		if (!isValueParser(parser)) throw new TypeError("Every firstOf() constituent must be a value parser.");
		if (parser.mode !== "sync") throw new TypeError("firstOf() only supports sync value parsers, but an async one was given.");
		if (require_internal_dependency.isDerivedValueParser(parser)) throw new TypeError("firstOf() does not support dependency-derived value parsers (created via deriveFrom() or dependency().derive()); pass the derived parser directly to option() or argument() instead.");
	}
	const metavar$1 = options.metavar ?? parsers.map((parser) => parser.metavar).join("|");
	require_nonempty.ensureNonEmptyString(metavar$1);
	const mergedChoices = parsers.every((parser) => parser.choices != null) ? Object.freeze(parsers.flatMap((parser) => [...parser.choices]).filter((value, index, all) => all.findIndex((other) => Object.is(other, value)) === index)) : void 0;
	function findOwner(value) {
		let canonicalizing;
		for (const parser of parsers) {
			if (typeof parser.validate === "function") {
				let result$1;
				try {
					result$1 = parser.validate(value);
				} catch {
					continue;
				}
				if (result$1.success) return {
					parser,
					result: result$1
				};
				continue;
			}
			let formatted;
			try {
				formatted = parser.format(value);
			} catch {
				continue;
			}
			if (typeof formatted !== "string") continue;
			const result = parser.parse(formatted);
			if (!result.success) continue;
			if (ownsRoundTrippedValue(parser, result.value, value)) return {
				parser,
				result
			};
			if (canonicalizing == null && value !== null && typeof value !== "object" && typeof result.value === typeof value) canonicalizing = {
				parser,
				result
			};
		}
		return canonicalizing;
	}
	const hasNormalize = parsers.some((parser) => typeof parser.normalize === "function");
	const hasSuggest = parsers.some((parser) => typeof parser.suggest === "function");
	return {
		mode: "sync",
		metavar: metavar$1,
		placeholder: parsers[0].placeholder,
		...mergedChoices != null ? { choices: mergedChoices } : {},
		parse(input) {
			const errors = [];
			for (const parser of parsers) {
				const result = parser.parse(input);
				if (result.success) return result;
				errors.push(result.error);
			}
			const noMatch = options.errors?.noMatch;
			return {
				success: false,
				error: noMatch == null ? firstOfNoMatchError(errors) : typeof noMatch === "function" ? noMatch(input, errors) : noMatch
			};
		},
		format(value) {
			let canonicalizing;
			let fallback;
			let firstError;
			let hasError = false;
			for (const parser of parsers) {
				let formatted;
				try {
					formatted = parser.format(value);
				} catch (e) {
					if (!hasError) {
						firstError = e;
						hasError = true;
					}
					continue;
				}
				if (typeof formatted !== "string") continue;
				if (typeof parser.validate === "function") {
					let owned = false;
					try {
						owned = parser.validate(value).success;
					} catch {}
					if (owned) return formatted;
				} else {
					const result = parser.parse(formatted);
					if (result.success) {
						if (ownsRoundTrippedValue(parser, result.value, value)) return formatted;
						if (canonicalizing == null && value !== null && typeof value !== "object" && typeof result.value === typeof value) canonicalizing = formatted;
					}
				}
				fallback ??= formatted;
			}
			if (canonicalizing !== void 0) return canonicalizing;
			if (fallback !== void 0) return fallback;
			if (hasError) throw firstError;
			throw new TypeError("No constituent parser could format the given value.");
		},
		validate(value) {
			const owner = findOwner(value);
			if (owner == null) return {
				success: false,
				error: require_message.message`Expected a value matching ${require_message.metavar(metavar$1)}.`
			};
			return owner.result;
		},
		...hasNormalize ? { normalize(value) {
			const owner = findOwner(value);
			if (owner == null || typeof owner.parser.normalize !== "function") return value;
			return owner.parser.normalize(value);
		} } : {},
		...hasSuggest ? { *suggest(prefix) {
			const collected = [];
			for (const parser of parsers) {
				if (typeof parser.suggest !== "function") continue;
				for (const suggestion of parser.suggest(prefix)) collected.push(suggestion);
			}
			yield* require_suggestion.deduplicateSuggestions(collected);
		} } : {}
	};
}
/**
* Builds the default error message for {@link firstOf} when every
* constituent parser fails: a header followed by each constituent's error
* on its own line.
*/
function firstOfNoMatchError(errors) {
	const terms = [require_message.text("Expected one of the following:")];
	for (const error of errors) terms.push(require_message.lineBreak(), require_message.text("- "), ...require_message.cloneMessage(error));
	return terms;
}
/**
* Determines whether a {@link firstOf} constituent owns a value, given the
* result of round-tripping the value through the constituent's `format()`
* and `parse()`.  The constituent owns the value when the round-trip
* preserves it exactly, when it yields the constituent's own normalization
* of it (e.g. a MAC address parser canonicalizing separators and case), or
* when the round-trip is a recognized parse-level canonicalization: a
* case-insensitive string match (e.g. a case-insensitive `choice()`
* folding `"INFO"` to `"info"`) or numeric equality (a number parser
* folding `-0` to `0`)—the same acceptance the constituent alone would
* grant a fallback value through round-trip validation.  Parsers with
* richer parse-level canonicalization should expose it via `normalize()`.
* A merely *successful* round-trip is not enough otherwise: an arbitrary
* same-type change means data loss rather than canonicalization (e.g. a
* clamping `format()` folding 15 into 10), a round-trip that changes the
* primitive type means the string form belongs to a different branch of
* the union (e.g. `choice(["1"])` capturing the integer 1), and for
* object values structural equality is the only way to tell
* canonicalization from data loss (e.g. a lossy `color()` `format()`
* dropping fields that belong to a later, more faithful constituent).
*/
function ownsRoundTrippedValue(parser, roundTripped, value) {
	if (valuesEqual(roundTripped, value)) return true;
	if (typeof parser.normalize === "function") {
		let normalized;
		try {
			normalized = parser.normalize(value);
		} catch {
			return false;
		}
		if (valuesEqual(roundTripped, normalized)) return true;
	}
	if (typeof value === "string" && typeof roundTripped === "string") return value.toLowerCase() === roundTripped.toLowerCase();
	if (typeof value === "number" && typeof roundTripped === "number") return value === roundTripped;
	return false;
}
/**
* Structural equality for parsed CLI values: primitives compare with
* `Object.is` (so `NaN` equals itself and `0` differs from `-0`, matching
* the distinction `choice()` makes), `Date` and `URL` instances compare by
* time value and href respectively, arrays and plain objects compare
* recursively, and other class instances compare by their overridden
* `toString()` serialization.  Recursion is bounded by the structure of
* the second argument, which in {@link firstOf} is always a freshly
* parsed (acyclic) value.
*/
function valuesEqual(a, b) {
	if (Object.is(a, b)) return true;
	if (typeof a !== "object" || typeof b !== "object" || a == null || b == null) return false;
	if (Array.isArray(a) || Array.isArray(b)) return Array.isArray(a) && Array.isArray(b) && a.length === b.length && a.every((item, i) => valuesEqual(item, b[i]));
	if (a instanceof Date || b instanceof Date) return a instanceof Date && b instanceof Date && a.getTime() === b.getTime();
	if (a instanceof URL || b instanceof URL) return a instanceof URL && b instanceof URL && a.href === b.href;
	const prototypeA = Object.getPrototypeOf(a);
	const prototypeB = Object.getPrototypeOf(b);
	const plainA = prototypeA === Object.prototype || prototypeA === null;
	const plainB = prototypeB === Object.prototype || prototypeB === null;
	if (plainA !== plainB) return false;
	if (!plainA) {
		if (prototypeA !== prototypeB) return false;
		const overridesA = hasCustomToString(a);
		const overridesB = hasCustomToString(b);
		if (overridesA !== overridesB) return false;
		const hasKeys = Object.keys(a).length > 0 || Object.keys(b).length > 0;
		if (hasKeys && !plainObjectsEqual(a, b)) return false;
		if (!overridesA) return hasKeys;
		try {
			return String(a) === String(b);
		} catch {
			return false;
		}
	}
	return plainObjectsEqual(a, b);
}
/**
* Checks whether an object provides a `toString()` implementation other
* than the generic `Object.prototype.toString`.
*/
function hasCustomToString(value) {
	try {
		const toString = value.toString;
		return typeof toString === "function" && toString !== Object.prototype.toString;
	} catch {
		return false;
	}
}
/**
* Compares two objects by their enumerable own keys, recursing into the
* values with {@link valuesEqual}.
*/
function plainObjectsEqual(a, b) {
	const aKeys = Object.keys(a);
	const bKeys = Object.keys(b);
	if (aKeys.length !== bKeys.length) return false;
	const bRecord = b;
	return aKeys.every((key) => Object.hasOwn(b, key) && valuesEqual(a[key], bRecord[key]));
}

//#endregion
exports.checkBooleanOption = checkBooleanOption;
exports.checkEnumOption = checkEnumOption;
exports.choice = choice;
exports.cidr = cidr;
exports.color = color;
exports.domain = domain;
exports.email = email;
exports.ensureNonEmptyString = require_nonempty.ensureNonEmptyString;
exports.fileSize = fileSize;
exports.firstOf = firstOf;
exports.float = float;
exports.hostname = hostname;
exports.integer = integer;
exports.ip = ip;
exports.ipv4 = ipv4;
exports.ipv6 = ipv6;
exports.isNonEmptyString = require_nonempty.isNonEmptyString;
exports.isValueParser = isValueParser;
exports.json = json;
exports.keyValue = keyValue;
exports.locale = locale;
exports.macAddress = macAddress;
exports.port = port;
exports.portRange = portRange;
exports.semVer = semVer;
exports.socketAddress = socketAddress;
exports.string = string;
exports.url = url;
exports.uuid = uuid;
